{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f5d384f0-503d-4709-abdd-c5ecab30a2ae",
   "metadata": {},
   "source": [
    "# Todo : adding PADOG, Wilcoxon test, KS test method to the notebook\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3716dad",
   "metadata": {},
   "source": [
    "# Enrichment Analysis\n",
    "\n",
    "Enrichment analysis (EA) is a technique used to derive biological insight from lists of significantly altered genes. The list of genes can be obtained from Differential Expression (DE) analysis or users' interest. The EA methods rely on the knowledge databases (e.g. KEGG and GO) to identify biological pathways or terms that are enriched in a gene list more than would be expected by chance. The outcome of the EA would be the in-depth and contextualized findings to help understand the mechanisms of disease, genes and proteins associated with the etiology of a specific disease or drug target.\n",
    "\n",
    "Over more than a decade, there are over 50 methods have been developed for EA. In this module, we will focus on pathway analysis using three popular methods including Over Representation Analysis (ORA), Fast Gene Set Enrichment Analysis (FGSEA), and Gene Set Analysis (GSA).\n",
    "\n",
    "## Learning Objectives:\n",
    "1. Data preparation\n",
    "2. Perform enrichment analysis using ORA, FGSEA and GSA\n",
    "3. Visualize and interpret the outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "37ff2142-8b48-4250-bb3b-b47175b79f9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<iframe src=\"./Quizzes/Quiz_Submodule4-1.html\" width=100% height=250></iframe>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "IRdisplay::display_html('<iframe src=\"./Quizzes/Quiz_Submodule4-1.html\" width=100% height=250></iframe>')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4731471e",
   "metadata": {},
   "source": [
    "## Gene Mapping\n",
    "In this module, we will use the DE genes with statistic generated in the [**submodule 02** ](./Module02-DE_Analysis.ipynb) using `limma` R package. We can use the following command to load the DE analysis result."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5c0da12c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>PROBEID</th><th scope=col>logFC</th><th scope=col>AveExpr</th><th scope=col>t</th><th scope=col>P.Value</th><th scope=col>adj.P.Val</th><th scope=col>B</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>222178_s_at</th><td>222178_s_at</td><td>-0.4473780</td><td>0.3653633</td><td>-73.01106</td><td>4.019112e-55</td><td>2.197450e-50</td><td>81.05538</td></tr>\n",
       "\t<tr><th scope=row>224687_at</th><td>224687_at  </td><td>-4.1431631</td><td>2.2572347</td><td>-48.92036</td><td>5.489424e-46</td><td>1.500671e-41</td><td>74.18354</td></tr>\n",
       "\t<tr><th scope=row>207488_at</th><td>207488_at  </td><td>-0.4017530</td><td>0.6243144</td><td>-39.29933</td><td>4.735684e-41</td><td>8.630785e-37</td><td>68.83493</td></tr>\n",
       "\t<tr><th scope=row>239226_at</th><td>239226_at  </td><td> 0.4483302</td><td>1.3245501</td><td> 28.36499</td><td>7.626558e-34</td><td>1.042455e-29</td><td>58.84908</td></tr>\n",
       "\t<tr><th scope=row>234109_x_at</th><td>234109_x_at</td><td>-0.2289726</td><td>0.7508256</td><td>-27.67010</td><td>2.635337e-33</td><td>2.881741e-29</td><td>58.00438</td></tr>\n",
       "\t<tr><th scope=row>212833_at</th><td>212833_at  </td><td>-2.5976592</td><td>1.7308944</td><td>-24.01172</td><td>2.914912e-30</td><td>2.656213e-26</td><td>52.99567</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & PROBEID & logFC & AveExpr & t & P.Value & adj.P.Val & B\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t222178\\_s\\_at & 222178\\_s\\_at & -0.4473780 & 0.3653633 & -73.01106 & 4.019112e-55 & 2.197450e-50 & 81.05538\\\\\n",
       "\t224687\\_at & 224687\\_at   & -4.1431631 & 2.2572347 & -48.92036 & 5.489424e-46 & 1.500671e-41 & 74.18354\\\\\n",
       "\t207488\\_at & 207488\\_at   & -0.4017530 & 0.6243144 & -39.29933 & 4.735684e-41 & 8.630785e-37 & 68.83493\\\\\n",
       "\t239226\\_at & 239226\\_at   &  0.4483302 & 1.3245501 &  28.36499 & 7.626558e-34 & 1.042455e-29 & 58.84908\\\\\n",
       "\t234109\\_x\\_at & 234109\\_x\\_at & -0.2289726 & 0.7508256 & -27.67010 & 2.635337e-33 & 2.881741e-29 & 58.00438\\\\\n",
       "\t212833\\_at & 212833\\_at   & -2.5976592 & 1.7308944 & -24.01172 & 2.914912e-30 & 2.656213e-26 & 52.99567\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 7\n",
       "\n",
       "| <!--/--> | PROBEID &lt;chr&gt; | logFC &lt;dbl&gt; | AveExpr &lt;dbl&gt; | t &lt;dbl&gt; | P.Value &lt;dbl&gt; | adj.P.Val &lt;dbl&gt; | B &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 222178_s_at | 222178_s_at | -0.4473780 | 0.3653633 | -73.01106 | 4.019112e-55 | 2.197450e-50 | 81.05538 |\n",
       "| 224687_at | 224687_at   | -4.1431631 | 2.2572347 | -48.92036 | 5.489424e-46 | 1.500671e-41 | 74.18354 |\n",
       "| 207488_at | 207488_at   | -0.4017530 | 0.6243144 | -39.29933 | 4.735684e-41 | 8.630785e-37 | 68.83493 |\n",
       "| 239226_at | 239226_at   |  0.4483302 | 1.3245501 |  28.36499 | 7.626558e-34 | 1.042455e-29 | 58.84908 |\n",
       "| 234109_x_at | 234109_x_at | -0.2289726 | 0.7508256 | -27.67010 | 2.635337e-33 | 2.881741e-29 | 58.00438 |\n",
       "| 212833_at | 212833_at   | -2.5976592 | 1.7308944 | -24.01172 | 2.914912e-30 | 2.656213e-26 | 52.99567 |\n",
       "\n"
      ],
      "text/plain": [
       "            PROBEID     logFC      AveExpr   t         P.Value     \n",
       "222178_s_at 222178_s_at -0.4473780 0.3653633 -73.01106 4.019112e-55\n",
       "224687_at   224687_at   -4.1431631 2.2572347 -48.92036 5.489424e-46\n",
       "207488_at   207488_at   -0.4017530 0.6243144 -39.29933 4.735684e-41\n",
       "239226_at   239226_at    0.4483302 1.3245501  28.36499 7.626558e-34\n",
       "234109_x_at 234109_x_at -0.2289726 0.7508256 -27.67010 2.635337e-33\n",
       "212833_at   212833_at   -2.5976592 1.7308944 -24.01172 2.914912e-30\n",
       "            adj.P.Val    B       \n",
       "222178_s_at 2.197450e-50 81.05538\n",
       "224687_at   1.500671e-41 74.18354\n",
       "207488_at   8.630785e-37 68.83493\n",
       "239226_at   1.042455e-29 58.84908\n",
       "234109_x_at 2.881741e-29 58.00438\n",
       "212833_at   2.656213e-26 52.99567"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load previously saved data\n",
    "DE.df <- readRDS(\"./data/DE_genes.rds\")\n",
    "rownames(DE.df) <- DE.df$PROBEID\n",
    "head(DE.df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "722ce8e2",
   "metadata": {},
   "source": [
    "We can see that the genes are saved with probe IDs, we need to convert them into gene symbols so that they can be analyzed using the enrichment analysis method later in this module. We will use the same approach presented in the [**submodule 01** ](./Module01-GEO_Data_Processing.ipynb) with `hgu133plus2.db` and `AnnotationDbi` databases."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0cb1e8f5-4142-49e5-89cc-1df2da15e1e7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<iframe src=\"./Quizzes/Quiz_Submodule4.html\" width=100% height=250></iframe>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "IRdisplay::display_html('<iframe src=\"./Quizzes/Quiz_Submodule4.html\" width=100% height=250></iframe>')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "84f28dcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the databases\n",
    "suppressMessages({\n",
    "    library(hgu133plus2.db)\n",
    "    library(AnnotationDbi)\n",
    "})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f51c9e14",
   "metadata": {},
   "source": [
    "Then, we can retrieve vector of probe IDs to perform gene symbol mapping using the following command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f04cf4fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>PROBEID</th><th scope=col>GENENAME</th><th scope=col>SYMBOL</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>222178_s_at</td><td>NA                                        </td><td>NA      </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>224687_at  </td><td>ankyrin repeat and IBR domain containing 1</td><td>ANKIB1  </td></tr>\n",
       "\t<tr><th scope=row>3</th><td>207488_at  </td><td>NA                                        </td><td>NA      </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>239226_at  </td><td>NA                                        </td><td>NA      </td></tr>\n",
       "\t<tr><th scope=row>5</th><td>234109_x_at</td><td>one cut homeobox 3                        </td><td>ONECUT3 </td></tr>\n",
       "\t<tr><th scope=row>6</th><td>212833_at  </td><td>solute carrier family 25 member 46        </td><td>SLC25A46</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & PROBEID & GENENAME & SYMBOL\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & 222178\\_s\\_at & NA                                         & NA      \\\\\n",
       "\t2 & 224687\\_at   & ankyrin repeat and IBR domain containing 1 & ANKIB1  \\\\\n",
       "\t3 & 207488\\_at   & NA                                         & NA      \\\\\n",
       "\t4 & 239226\\_at   & NA                                         & NA      \\\\\n",
       "\t5 & 234109\\_x\\_at & one cut homeobox 3                         & ONECUT3 \\\\\n",
       "\t6 & 212833\\_at   & solute carrier family 25 member 46         & SLC25A46\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 3\n",
       "\n",
       "| <!--/--> | PROBEID &lt;chr&gt; | GENENAME &lt;chr&gt; | SYMBOL &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | 222178_s_at | NA                                         | NA       |\n",
       "| 2 | 224687_at   | ankyrin repeat and IBR domain containing 1 | ANKIB1   |\n",
       "| 3 | 207488_at   | NA                                         | NA       |\n",
       "| 4 | 239226_at   | NA                                         | NA       |\n",
       "| 5 | 234109_x_at | one cut homeobox 3                         | ONECUT3  |\n",
       "| 6 | 212833_at   | solute carrier family 25 member 46         | SLC25A46 |\n",
       "\n"
      ],
      "text/plain": [
       "  PROBEID     GENENAME                                   SYMBOL  \n",
       "1 222178_s_at NA                                         NA      \n",
       "2 224687_at   ankyrin repeat and IBR domain containing 1 ANKIB1  \n",
       "3 207488_at   NA                                         NA      \n",
       "4 239226_at   NA                                         NA      \n",
       "5 234109_x_at one cut homeobox 3                         ONECUT3 \n",
       "6 212833_at   solute carrier family 25 member 46         SLC25A46"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "probeIDs <- DE.df$PROBEID\n",
    "# Map probe IDs to their symbols\n",
    "suppressMessages({\n",
    "    annotLookup <- AnnotationDbi::select(hgu133plus2.db, keys = probeIDs, columns = c('PROBEID', 'GENENAME', 'SYMBOL'))\n",
    "})\n",
    "# View the first few genes in the mapping table\n",
    "head(annotLookup)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ed42b60",
   "metadata": {},
   "source": [
    "Now, we can merge DE data frame with annotation table by the `PROBEID` and remove NA gene symbols using the following command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5c5ac1b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "DE.df <- merge(annotLookup, DE.df, by = \"PROBEID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3bef4b3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>PROBEID</th><th scope=col>GENENAME</th><th scope=col>SYMBOL</th><th scope=col>logFC</th><th scope=col>AveExpr</th><th scope=col>t</th><th scope=col>P.Value</th><th scope=col>adj.P.Val</th><th scope=col>B</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1007_s_at</td><td>discoidin domain receptor tyrosine kinase 1 </td><td>DDR1  </td><td>-0.23858465</td><td>1.0967772</td><td>-2.08257585</td><td>0.04210211</td><td>0.2531265</td><td>-4.284228</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>1053_at  </td><td>replication factor C subunit 2              </td><td>RFC2  </td><td> 0.05830367</td><td>0.9657758</td><td> 1.42311356</td><td>0.16052793</td><td>0.4674726</td><td>-5.359907</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>117_at   </td><td>heat shock protein family A (Hsp70) member 6</td><td>HSPA6 </td><td>-0.01138494</td><td>1.1501582</td><td>-0.08100885</td><td>0.93573821</td><td>0.9861536</td><td>-6.337785</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>121_at   </td><td>paired box 8                                </td><td>PAX8  </td><td> 0.02301084</td><td>1.0581847</td><td> 0.87166663</td><td>0.38729816</td><td>0.7236528</td><td>-5.968512</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>1255_g_at</td><td>guanylate cyclase activator 1A              </td><td>GUCA1A</td><td> 0.36999950</td><td>1.4080351</td><td> 1.78615022</td><td>0.07976123</td><td>0.3351480</td><td>-4.812134</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>1294_at  </td><td>ubiquitin like modifier activating enzyme 7 </td><td>UBA7  </td><td>-0.11221434</td><td>1.0859983</td><td>-1.61593670</td><td>0.11200995</td><td>0.3925252</td><td>-5.082923</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & PROBEID & GENENAME & SYMBOL & logFC & AveExpr & t & P.Value & adj.P.Val & B\\\\\n",
       "  & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & 1007\\_s\\_at & discoidin domain receptor tyrosine kinase 1  & DDR1   & -0.23858465 & 1.0967772 & -2.08257585 & 0.04210211 & 0.2531265 & -4.284228\\\\\n",
       "\t2 & 1053\\_at   & replication factor C subunit 2               & RFC2   &  0.05830367 & 0.9657758 &  1.42311356 & 0.16052793 & 0.4674726 & -5.359907\\\\\n",
       "\t3 & 117\\_at    & heat shock protein family A (Hsp70) member 6 & HSPA6  & -0.01138494 & 1.1501582 & -0.08100885 & 0.93573821 & 0.9861536 & -6.337785\\\\\n",
       "\t4 & 121\\_at    & paired box 8                                 & PAX8   &  0.02301084 & 1.0581847 &  0.87166663 & 0.38729816 & 0.7236528 & -5.968512\\\\\n",
       "\t5 & 1255\\_g\\_at & guanylate cyclase activator 1A               & GUCA1A &  0.36999950 & 1.4080351 &  1.78615022 & 0.07976123 & 0.3351480 & -4.812134\\\\\n",
       "\t6 & 1294\\_at   & ubiquitin like modifier activating enzyme 7  & UBA7   & -0.11221434 & 1.0859983 & -1.61593670 & 0.11200995 & 0.3925252 & -5.082923\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 9\n",
       "\n",
       "| <!--/--> | PROBEID &lt;chr&gt; | GENENAME &lt;chr&gt; | SYMBOL &lt;chr&gt; | logFC &lt;dbl&gt; | AveExpr &lt;dbl&gt; | t &lt;dbl&gt; | P.Value &lt;dbl&gt; | adj.P.Val &lt;dbl&gt; | B &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 1007_s_at | discoidin domain receptor tyrosine kinase 1  | DDR1   | -0.23858465 | 1.0967772 | -2.08257585 | 0.04210211 | 0.2531265 | -4.284228 |\n",
       "| 2 | 1053_at   | replication factor C subunit 2               | RFC2   |  0.05830367 | 0.9657758 |  1.42311356 | 0.16052793 | 0.4674726 | -5.359907 |\n",
       "| 3 | 117_at    | heat shock protein family A (Hsp70) member 6 | HSPA6  | -0.01138494 | 1.1501582 | -0.08100885 | 0.93573821 | 0.9861536 | -6.337785 |\n",
       "| 4 | 121_at    | paired box 8                                 | PAX8   |  0.02301084 | 1.0581847 |  0.87166663 | 0.38729816 | 0.7236528 | -5.968512 |\n",
       "| 5 | 1255_g_at | guanylate cyclase activator 1A               | GUCA1A |  0.36999950 | 1.4080351 |  1.78615022 | 0.07976123 | 0.3351480 | -4.812134 |\n",
       "| 6 | 1294_at   | ubiquitin like modifier activating enzyme 7  | UBA7   | -0.11221434 | 1.0859983 | -1.61593670 | 0.11200995 | 0.3925252 | -5.082923 |\n",
       "\n"
      ],
      "text/plain": [
       "  PROBEID   GENENAME                                     SYMBOL logFC      \n",
       "1 1007_s_at discoidin domain receptor tyrosine kinase 1  DDR1   -0.23858465\n",
       "2 1053_at   replication factor C subunit 2               RFC2    0.05830367\n",
       "3 117_at    heat shock protein family A (Hsp70) member 6 HSPA6  -0.01138494\n",
       "4 121_at    paired box 8                                 PAX8    0.02301084\n",
       "5 1255_g_at guanylate cyclase activator 1A               GUCA1A  0.36999950\n",
       "6 1294_at   ubiquitin like modifier activating enzyme 7  UBA7   -0.11221434\n",
       "  AveExpr   t           P.Value    adj.P.Val B        \n",
       "1 1.0967772 -2.08257585 0.04210211 0.2531265 -4.284228\n",
       "2 0.9657758  1.42311356 0.16052793 0.4674726 -5.359907\n",
       "3 1.1501582 -0.08100885 0.93573821 0.9861536 -6.337785\n",
       "4 1.0581847  0.87166663 0.38729816 0.7236528 -5.968512\n",
       "5 1.4080351  1.78615022 0.07976123 0.3351480 -4.812134\n",
       "6 1.0859983 -1.61593670 0.11200995 0.3925252 -5.082923"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Remove NA gene symbol\n",
    "DE.df <-  DE.df[!is.na(DE.df$SYMBOL), ]\n",
    "# Remove duplicated genes\n",
    "DE.df <-  DE.df[!duplicated(DE.df$SYMBOL, fromLast = FALSE), ]\n",
    "head(DE.df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adf678d9",
   "metadata": {},
   "source": [
    "As a result, we obtained the new DE table with two more columns containing gene names and gene symbols. The use of this DE table can be varied based on the selected enrichment analysis tools."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8538d33d",
   "metadata": {},
   "source": [
    "## Enrichment Analysis Using Over-Representation Analysis\n",
    "Over-representation analysis (ORA) is a statistical method that determines whether genes from a pre-defined gene set of a specific GO term or KEGG pathway are presented more than would be expected (over-represented) in a subset of your data. In our learning module, this subset refers to the list of DE genes generated from the `limma` method. For each gene set, an enrichment *p-value* is calculated using the Binomial distribution, Hypergeometric distribution, the Fisher exact test, or the Chi-square test.  The hypergeometric distribution is a popular approach used to calculate enrichment p-value. The formula can be presented as follows:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eca231f2",
   "metadata": {},
   "source": [
    "$$\n",
    "     P(X\\geq x) = 1 - P(X \\leq x-1) = 1 - \\sum\\limits_{i=0}^{x-1}\\frac{\\hphantom{}{M \\choose i }{N - M \\choose n-i}}{N \\choose n}\n",
    "$$"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c18d887",
   "metadata": {},
   "source": [
    "where *N* is the number of background genes (all genes presented in the expression matrix), *n* is the number of “interesting” genes (DE genes), *M* is the number of genes that are annotated to a particular gene set *S* (list of genes in a specific KEGG pathway or GO term), and *x* is the number of “interesting” genes that are annotated to *S* (genes presented in DE genes list and a specific KEGG pathway or GO term)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccade46c",
   "metadata": {},
   "source": [
    "For example, suppose we have an expression matrix with 20,000 genes, of which 500 are differently expressed. Also, suppose that 100 of the 20,000 genes are annotated to a particular gene set S. Of these 100 genes, 20 are members of the DE genes list. The probability that 20 or more (up to 100) genes annotated to S are in the DE genes list by chance is given by"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed50cd53",
   "metadata": {},
   "source": [
    "$$\n",
    "P(X\\geq 20) = 1 - P(X \\leq 19) = 1-\\sum \\limits_{i=0}^{19}\\frac{\\hphantom{}{100 \\choose i}{20000 - 100 \\choose 500-i}}{20000 \\choose 500} = 5.26 \\times 10^{-13}\n",
    "$$"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2e2d72e",
   "metadata": {},
   "source": [
    "The *p-value* indicates that it is very rare to observe 20 of the 100 genes from this set are in the DE genes list by chance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2b0be768-1749-4752-b31f-bf5818b33be5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<iframe src=\"./Quizzes/Quiz_Submodule4-2.html\" width=100% height=250></iframe>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "IRdisplay::display_html('<iframe src=\"./Quizzes/Quiz_Submodule4-2.html\" width=100% height=250></iframe>')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e332d07b",
   "metadata": {},
   "source": [
    "### Data preparation\n",
    "To conduct enrichment analysis using ORA, there are several input data that we need to prepare. First, we need to select a set of genes that\n",
    "are significantly altered (*p-value < 0.05*) in the DE genes generated from the `limma` method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b1b28353",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Selecting a list of significant DE genes\n",
    "DEGenes <- DE.df[DE.df$adj.P.Val <= 0.05, ]\n",
    "# Select genes with symbol\n",
    "DEGenes <- DEGenes$SYMBOL"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0d9f75f",
   "metadata": {},
   "source": [
    "Next, we need to define a list of background genes. In this analysis, they are all the genes generated from the DE analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "303d3696",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Defining background genes\n",
    "backgroundSet <- DE.df$SYMBOL"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f19e685",
   "metadata": {},
   "source": [
    "Then, we need to obtain a list of geneset from knowledge databases such as GO and KEGG. In this learning module, the geneset will be retrieved from\n",
    "the *.gmt* files that were processed from the [**submodule 03**](./Module03-Gene_Set_and_Pathway.ipynb). To load the geneset, we will use the `gmt2geneset` function below:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0042697a",
   "metadata": {},
   "source": [
    "Here, we also write a function to perform over-representation analysis based on the hyper-geometric testing formula presented above. The `ORA` method will perform\n",
    "hyper-geometric testing for each geneset obtained from GO or KEGG using the function `phyper` available for the `stats` R base package. The output of the `ORA` function is a table that contains a column of terms or pathway names and a column of *p-value*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "38686455",
   "metadata": {},
   "outputs": [],
   "source": [
    "# A customized function to parse a gmt file to a list of genesets\n",
    "gmt2geneset <- function(path) {\n",
    "    # Read the GMT file and create a list of gene sets\n",
    "    gmt_content <- read.table(path, header = FALSE, fill = TRUE, sep = \"\\t\")\n",
    "    \n",
    "    genesets <- apply(gmt_content, MARGIN = 1, function(r) {\n",
    "        # Extract unique genes from the current row (excluding the first two elements)\n",
    "        genes = unique(r[-(1:2)]) \n",
    "        # Create a list for each gene set with ID, description, and non-NA genes\n",
    "        list(\n",
    "            id = r[1],\n",
    "            description = r[2],\n",
    "            genes = genes[!is.na(genes)]\n",
    "        )\n",
    "    })\n",
    "    # Extract the gene sets from the list and convert gene vectors to character\n",
    "    gs <- lapply(genesets, function(g) {\n",
    "        as.character(g$genes)\n",
    "    })\n",
    "    # Set names for the gene sets based on their descriptions\n",
    "    names(gs) <- lapply(genesets, function(g) g$description)\n",
    "    # Return the resulting list of gene sets\n",
    "    return (gs)\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d809f4b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Over-Representation Analysis (ORA) Function\n",
    "ORA <- function(geneset, DEGenes, backgroundSet, DE.df) {\n",
    "    # Initialize an empty vector to store the results of the analysis\n",
    "    res <- sapply(geneset, function(gs) {\n",
    "        # Count the number of differentially expressed genes in the current gene set (gs)\n",
    "        wBallDraw <- length(intersect(gs, DEGenes)) - 1\n",
    "        # Check for a potential issue with the intersection count being negative\n",
    "        if (wBallDraw < 0) return(1)\n",
    "        # Total number of differentially expressed genes\n",
    "        wBall <- length(DEGenes)\n",
    "        # Remaining genes in the background set\n",
    "        bBall <- nrow(DE.df) - length(DEGenes)\n",
    "        # Number of genes in the current gene set that are in the background set\n",
    "        ballDraw <- length(intersect(gs, backgroundSet))\n",
    "        # Calculate the hypergeometric probability using the phyper function\n",
    "        hypergeo_prob <- phyper(wBallDraw, wBall, bBall, ballDraw)\n",
    "        # Calculate the complementary probability (1 - hypergeometric probability)\n",
    "        return(1 - hypergeo_prob)\n",
    "    })\n",
    "    # Return the vector of results (p-values)\n",
    "    return(res)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "050d0119-5c7d-4b09-96d3-a4cb76e8ee2e",
   "metadata": {},
   "source": [
    "To use the `ORA` function, we need to pass the following arguments:\n",
    "\n",
    "- `geneset`: A list of gene sets to be analyzed.\n",
    "- `DEGenes`: A vector of differentially expressed genes.\n",
    "- `backgroundSet`: A vector containing a set of genes representing the background.\n",
    "- `DE.df`: A dataframe containing information about differentially expressed genes.\n",
    "\n",
    "The `ORA` function returns a vector of the results (p-values)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dedfa2a3",
   "metadata": {},
   "source": [
    "### Enrichment analysis using ORA and GO terms\n",
    "In this section, we will perform ORA using genesets obtained from the GO database and the function `ORA` defined above. The detailed code is presented below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "975e07c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>GOterms</th><th scope=col>pvalue</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>601</th><td>mitochondrial electron transport, cytochrome c to oxygen</td><td>0.0001704211</td></tr>\n",
       "\t<tr><th scope=row>1854</th><td>energy derivation by oxidation of organic compounds     </td><td>0.0004553500</td></tr>\n",
       "\t<tr><th scope=row>2204</th><td>respiratory electron transport chain                    </td><td>0.0004962063</td></tr>\n",
       "\t<tr><th scope=row>2203</th><td>electron transport chain                                </td><td>0.0009369785</td></tr>\n",
       "\t<tr><th scope=row>598</th><td>oxidative phosphorylation                               </td><td>0.0013190484</td></tr>\n",
       "\t<tr><th scope=row>1282</th><td>aerobic respiration                                     </td><td>0.0016870236</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & GOterms & pvalue\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t601 & mitochondrial electron transport, cytochrome c to oxygen & 0.0001704211\\\\\n",
       "\t1854 & energy derivation by oxidation of organic compounds      & 0.0004553500\\\\\n",
       "\t2204 & respiratory electron transport chain                     & 0.0004962063\\\\\n",
       "\t2203 & electron transport chain                                 & 0.0009369785\\\\\n",
       "\t598 & oxidative phosphorylation                                & 0.0013190484\\\\\n",
       "\t1282 & aerobic respiration                                      & 0.0016870236\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | GOterms &lt;chr&gt; | pvalue &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| 601 | mitochondrial electron transport, cytochrome c to oxygen | 0.0001704211 |\n",
       "| 1854 | energy derivation by oxidation of organic compounds      | 0.0004553500 |\n",
       "| 2204 | respiratory electron transport chain                     | 0.0004962063 |\n",
       "| 2203 | electron transport chain                                 | 0.0009369785 |\n",
       "| 598 | oxidative phosphorylation                                | 0.0013190484 |\n",
       "| 1282 | aerobic respiration                                      | 0.0016870236 |\n",
       "\n"
      ],
      "text/plain": [
       "     GOterms                                                  pvalue      \n",
       "601  mitochondrial electron transport, cytochrome c to oxygen 0.0001704211\n",
       "1854 energy derivation by oxidation of organic compounds      0.0004553500\n",
       "2204 respiratory electron transport chain                     0.0004962063\n",
       "2203 electron transport chain                                 0.0009369785\n",
       "598  oxidative phosphorylation                                0.0013190484\n",
       "1282 aerobic respiration                                      0.0016870236"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Loading geneset from GO database\n",
    "suppressWarnings({suppressMessages({geneset <- gmt2geneset(\"./data/GO_terms.gmt\")})})\n",
    "# Perform ORA\n",
    "res.ORA <- ORA(geneset = geneset, DEGenes = DEGenes, backgroundSet = backgroundSet, DE.df = DE.df)\n",
    "# Save the result to a table where the first column is GO term name and the second column is the p-value\n",
    "res.df <- data.frame(\n",
    "    GOterms = names(res.ORA),\n",
    "    pvalue = res.ORA\n",
    ")\n",
    "# Order the table based on the p-value\n",
    "res.df <- res.df[order(res.df$pvalue), ]\n",
    "# View the most significant GO terms\n",
    "head(res.df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae0d8278",
   "metadata": {},
   "source": [
    "From the result table, we might conclude that the pathways shown are key components of the disease in study."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfa3dbd7",
   "metadata": {},
   "source": [
    "### Enrichment analysis using ORA and KEGG pathways"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69dc9176-9f21-4198-8873-c391fd58b1d2",
   "metadata": {},
   "source": [
    "We will perform the same steps as we did above for the GO terms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7c31dbde",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading the KEGG pathways\n",
    "suppressWarnings({suppressMessages({geneset <- gmt2geneset(\"./data/KEGG_pathways.gmt\")})})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ee2d6ff9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>pathways</th><th scope=col>pvalue</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>622</th><td>UQCRH</td><td>0.0008616251</td></tr>\n",
       "\t<tr><th scope=row>680</th><td>COX2 </td><td>0.0008753422</td></tr>\n",
       "\t<tr><th scope=row>670</th><td>CYC1 </td><td>0.0057312227</td></tr>\n",
       "\t<tr><th scope=row>17</th><td>COX5A</td><td>0.0060597917</td></tr>\n",
       "\t<tr><th scope=row>472</th><td>SDHC </td><td>0.0060597917</td></tr>\n",
       "\t<tr><th scope=row>612</th><td>SDHC </td><td>0.0060597917</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & pathways & pvalue\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t622 & UQCRH & 0.0008616251\\\\\n",
       "\t680 & COX2  & 0.0008753422\\\\\n",
       "\t670 & CYC1  & 0.0057312227\\\\\n",
       "\t17 & COX5A & 0.0060597917\\\\\n",
       "\t472 & SDHC  & 0.0060597917\\\\\n",
       "\t612 & SDHC  & 0.0060597917\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | pathways &lt;chr&gt; | pvalue &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| 622 | UQCRH | 0.0008616251 |\n",
       "| 680 | COX2  | 0.0008753422 |\n",
       "| 670 | CYC1  | 0.0057312227 |\n",
       "| 17 | COX5A | 0.0060597917 |\n",
       "| 472 | SDHC  | 0.0060597917 |\n",
       "| 612 | SDHC  | 0.0060597917 |\n",
       "\n"
      ],
      "text/plain": [
       "    pathways pvalue      \n",
       "622 UQCRH    0.0008616251\n",
       "680 COX2     0.0008753422\n",
       "670 CYC1     0.0057312227\n",
       "17  COX5A    0.0060597917\n",
       "472 SDHC     0.0060597917\n",
       "612 SDHC     0.0060597917"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Perform Hyper-geometric testing\n",
    "res.ORA <- ORA(geneset = geneset, DEGenes = DEGenes, \n",
    "               backgroundSet = backgroundSet, DE.df = DE.df)\n",
    "# Save the results to a data frame\n",
    "res.df <- data.frame(\n",
    "    pathways = names(res.ORA),\n",
    "    pvalue = res.ORA\n",
    ")\n",
    "\n",
    "res.df <- res.df[order(res.df$pvalue), ]\n",
    "head(res.df)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63818b96-1b67-49be-81d3-c45dae212ecc",
   "metadata": {},
   "source": [
    "### Enrichment analysis using ORA and REACTOME pathways\n",
    "\n",
    "We will perform basically the same steps as we did above for the GO terms and KEEG pathways, the only thing that is different is we will map the PROBEID to ENTREZID instead of SYMBOL since REACTOME use the ENTREZID for gene id."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d3a72d64-a93e-4680-9c9c-a969923f79a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>PROBEID</th><th scope=col>ENTREZID</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1007_s_at</td><td>780 </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>1053_at  </td><td>5982</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>117_at   </td><td>3310</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>121_at   </td><td>7849</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>1255_g_at</td><td>2978</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>1294_at  </td><td>7318</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & PROBEID & ENTREZID\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & 1007\\_s\\_at & 780 \\\\\n",
       "\t2 & 1053\\_at   & 5982\\\\\n",
       "\t3 & 117\\_at    & 3310\\\\\n",
       "\t4 & 121\\_at    & 7849\\\\\n",
       "\t5 & 1255\\_g\\_at & 2978\\\\\n",
       "\t6 & 1294\\_at   & 7318\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | PROBEID &lt;chr&gt; | ENTREZID &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 1 | 1007_s_at | 780  |\n",
       "| 2 | 1053_at   | 5982 |\n",
       "| 3 | 117_at    | 3310 |\n",
       "| 4 | 121_at    | 7849 |\n",
       "| 5 | 1255_g_at | 2978 |\n",
       "| 6 | 1294_at   | 7318 |\n",
       "\n"
      ],
      "text/plain": [
       "  PROBEID   ENTREZID\n",
       "1 1007_s_at 780     \n",
       "2 1053_at   5982    \n",
       "3 117_at    3310    \n",
       "4 121_at    7849    \n",
       "5 1255_g_at 2978    \n",
       "6 1294_at   7318    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>PROBEID</th><th scope=col>ENTREZID</th><th scope=col>GENENAME</th><th scope=col>SYMBOL</th><th scope=col>logFC</th><th scope=col>AveExpr</th><th scope=col>t</th><th scope=col>P.Value</th><th scope=col>adj.P.Val</th><th scope=col>B</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1007_s_at</td><td>780 </td><td>discoidin domain receptor tyrosine kinase 1 </td><td>DDR1  </td><td>-0.23858465</td><td>1.0967772</td><td>-2.08257585</td><td>0.04210211</td><td>0.2531265</td><td>-4.284228</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>1053_at  </td><td>5982</td><td>replication factor C subunit 2              </td><td>RFC2  </td><td> 0.05830367</td><td>0.9657758</td><td> 1.42311356</td><td>0.16052793</td><td>0.4674726</td><td>-5.359907</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>117_at   </td><td>3310</td><td>heat shock protein family A (Hsp70) member 6</td><td>HSPA6 </td><td>-0.01138494</td><td>1.1501582</td><td>-0.08100885</td><td>0.93573821</td><td>0.9861536</td><td>-6.337785</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>121_at   </td><td>7849</td><td>paired box 8                                </td><td>PAX8  </td><td> 0.02301084</td><td>1.0581847</td><td> 0.87166663</td><td>0.38729816</td><td>0.7236528</td><td>-5.968512</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>1255_g_at</td><td>2978</td><td>guanylate cyclase activator 1A              </td><td>GUCA1A</td><td> 0.36999950</td><td>1.4080351</td><td> 1.78615022</td><td>0.07976123</td><td>0.3351480</td><td>-4.812134</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>1294_at  </td><td>7318</td><td>ubiquitin like modifier activating enzyme 7 </td><td>UBA7  </td><td>-0.11221434</td><td>1.0859983</td><td>-1.61593670</td><td>0.11200995</td><td>0.3925252</td><td>-5.082923</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 10\n",
       "\\begin{tabular}{r|llllllllll}\n",
       "  & PROBEID & ENTREZID & GENENAME & SYMBOL & logFC & AveExpr & t & P.Value & adj.P.Val & B\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & 1007\\_s\\_at & 780  & discoidin domain receptor tyrosine kinase 1  & DDR1   & -0.23858465 & 1.0967772 & -2.08257585 & 0.04210211 & 0.2531265 & -4.284228\\\\\n",
       "\t2 & 1053\\_at   & 5982 & replication factor C subunit 2               & RFC2   &  0.05830367 & 0.9657758 &  1.42311356 & 0.16052793 & 0.4674726 & -5.359907\\\\\n",
       "\t3 & 117\\_at    & 3310 & heat shock protein family A (Hsp70) member 6 & HSPA6  & -0.01138494 & 1.1501582 & -0.08100885 & 0.93573821 & 0.9861536 & -6.337785\\\\\n",
       "\t4 & 121\\_at    & 7849 & paired box 8                                 & PAX8   &  0.02301084 & 1.0581847 &  0.87166663 & 0.38729816 & 0.7236528 & -5.968512\\\\\n",
       "\t5 & 1255\\_g\\_at & 2978 & guanylate cyclase activator 1A               & GUCA1A &  0.36999950 & 1.4080351 &  1.78615022 & 0.07976123 & 0.3351480 & -4.812134\\\\\n",
       "\t6 & 1294\\_at   & 7318 & ubiquitin like modifier activating enzyme 7  & UBA7   & -0.11221434 & 1.0859983 & -1.61593670 & 0.11200995 & 0.3925252 & -5.082923\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 10\n",
       "\n",
       "| <!--/--> | PROBEID &lt;chr&gt; | ENTREZID &lt;chr&gt; | GENENAME &lt;chr&gt; | SYMBOL &lt;chr&gt; | logFC &lt;dbl&gt; | AveExpr &lt;dbl&gt; | t &lt;dbl&gt; | P.Value &lt;dbl&gt; | adj.P.Val &lt;dbl&gt; | B &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 1007_s_at | 780  | discoidin domain receptor tyrosine kinase 1  | DDR1   | -0.23858465 | 1.0967772 | -2.08257585 | 0.04210211 | 0.2531265 | -4.284228 |\n",
       "| 2 | 1053_at   | 5982 | replication factor C subunit 2               | RFC2   |  0.05830367 | 0.9657758 |  1.42311356 | 0.16052793 | 0.4674726 | -5.359907 |\n",
       "| 3 | 117_at    | 3310 | heat shock protein family A (Hsp70) member 6 | HSPA6  | -0.01138494 | 1.1501582 | -0.08100885 | 0.93573821 | 0.9861536 | -6.337785 |\n",
       "| 4 | 121_at    | 7849 | paired box 8                                 | PAX8   |  0.02301084 | 1.0581847 |  0.87166663 | 0.38729816 | 0.7236528 | -5.968512 |\n",
       "| 5 | 1255_g_at | 2978 | guanylate cyclase activator 1A               | GUCA1A |  0.36999950 | 1.4080351 |  1.78615022 | 0.07976123 | 0.3351480 | -4.812134 |\n",
       "| 6 | 1294_at   | 7318 | ubiquitin like modifier activating enzyme 7  | UBA7   | -0.11221434 | 1.0859983 | -1.61593670 | 0.11200995 | 0.3925252 | -5.082923 |\n",
       "\n"
      ],
      "text/plain": [
       "  PROBEID   ENTREZID GENENAME                                     SYMBOL\n",
       "1 1007_s_at 780      discoidin domain receptor tyrosine kinase 1  DDR1  \n",
       "2 1053_at   5982     replication factor C subunit 2               RFC2  \n",
       "3 117_at    3310     heat shock protein family A (Hsp70) member 6 HSPA6 \n",
       "4 121_at    7849     paired box 8                                 PAX8  \n",
       "5 1255_g_at 2978     guanylate cyclase activator 1A               GUCA1A\n",
       "6 1294_at   7318     ubiquitin like modifier activating enzyme 7  UBA7  \n",
       "  logFC       AveExpr   t           P.Value    adj.P.Val B        \n",
       "1 -0.23858465 1.0967772 -2.08257585 0.04210211 0.2531265 -4.284228\n",
       "2  0.05830367 0.9657758  1.42311356 0.16052793 0.4674726 -5.359907\n",
       "3 -0.01138494 1.1501582 -0.08100885 0.93573821 0.9861536 -6.337785\n",
       "4  0.02301084 1.0581847  0.87166663 0.38729816 0.7236528 -5.968512\n",
       "5  0.36999950 1.4080351  1.78615022 0.07976123 0.3351480 -4.812134\n",
       "6 -0.11221434 1.0859983 -1.61593670 0.11200995 0.3925252 -5.082923"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "probeIDs <- DE.df$PROBEID\n",
    "\n",
    "suppressMessages({\n",
    "    annotLookup <- AnnotationDbi::select(hgu133plus2.db, keys = probeIDs, columns = c('PROBEID', 'ENTREZID'))\n",
    "})\n",
    "head(annotLookup)\n",
    "\n",
    "DE.df <- merge(annotLookup, DE.df, by = \"PROBEID\")\n",
    "\n",
    "# Remove NA gene symbol\n",
    "DE.df <-  DE.df[!is.na(DE.df$ENTREZID), ]\n",
    "# Remove duplicated genes\n",
    "DE.df <-  DE.df[!duplicated(DE.df$ENTREZID, fromLast = FALSE), ]\n",
    "head(DE.df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6805c48b-7775-4ad1-a5f3-ca19d9f74b5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Selecting a list of significant DE genes\n",
    "DEGenes <- DE.df[DE.df$adj.P.Val <= 0.05, ]\n",
    "# Select genes with symbol\n",
    "DEGenes <- DEGenes$ENTREZID\n",
    "#Defining background genes\n",
    "backgroundSet <- DE.df$ENTREZID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7a7baa26-8eb4-4762-bcd0-3ebaf18d8cc0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>pathways</th><th scope=col>pvalue</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1278</th><td>5264                                       </td><td>0.0007996188</td></tr>\n",
       "\t<tr><th scope=row>2879</th><td>Homo sapiens: Type I hemidesmosome assembly</td><td>0.0054458977</td></tr>\n",
       "\t<tr><th scope=row>1651</th><td>Homo sapiens: P2Y receptors                </td><td>0.0066064555</td></tr>\n",
       "\t<tr><th scope=row>2202</th><td>Homo sapiens: Ribavirin ADME               </td><td>0.0066064555</td></tr>\n",
       "\t<tr><th scope=row>1863</th><td>Homo sapiens: Protein folding              </td><td>0.0091140070</td></tr>\n",
       "\t<tr><th scope=row>1971</th><td>84271                                      </td><td>0.0091140070</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & pathways & pvalue\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1278 & 5264                                        & 0.0007996188\\\\\n",
       "\t2879 & Homo sapiens: Type I hemidesmosome assembly & 0.0054458977\\\\\n",
       "\t1651 & Homo sapiens: P2Y receptors                 & 0.0066064555\\\\\n",
       "\t2202 & Homo sapiens: Ribavirin ADME                & 0.0066064555\\\\\n",
       "\t1863 & Homo sapiens: Protein folding               & 0.0091140070\\\\\n",
       "\t1971 & 84271                                       & 0.0091140070\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | pathways &lt;chr&gt; | pvalue &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| 1278 | 5264                                        | 0.0007996188 |\n",
       "| 2879 | Homo sapiens: Type I hemidesmosome assembly | 0.0054458977 |\n",
       "| 1651 | Homo sapiens: P2Y receptors                 | 0.0066064555 |\n",
       "| 2202 | Homo sapiens: Ribavirin ADME                | 0.0066064555 |\n",
       "| 1863 | Homo sapiens: Protein folding               | 0.0091140070 |\n",
       "| 1971 | 84271                                       | 0.0091140070 |\n",
       "\n"
      ],
      "text/plain": [
       "     pathways                                    pvalue      \n",
       "1278 5264                                        0.0007996188\n",
       "2879 Homo sapiens: Type I hemidesmosome assembly 0.0054458977\n",
       "1651 Homo sapiens: P2Y receptors                 0.0066064555\n",
       "2202 Homo sapiens: Ribavirin ADME                0.0066064555\n",
       "1863 Homo sapiens: Protein folding               0.0091140070\n",
       "1971 84271                                       0.0091140070"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Loading geneset from GO database\n",
    "suppressWarnings({suppressMessages({geneset <- gmt2geneset(\"./data/REACTOME_pathways.gmt\")})})\n",
    "# Perform ORA\n",
    "res.ORA <- ORA(geneset = geneset, DEGenes = DEGenes, backgroundSet = backgroundSet, DE.df = DE.df)\n",
    "# Save the result to a table where the first column is GO term name and the second column is the p-value\n",
    "res.df <- data.frame(\n",
    "    pathways = names(res.ORA),\n",
    "    pvalue = res.ORA\n",
    ")\n",
    "# Order the table based on the p-value\n",
    "res.df <- res.df[order(res.df$pvalue), ]\n",
    "# View the most significant GO terms\n",
    "head(res.df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72a138b4-a0aa-461b-b529-67e76c7767d6",
   "metadata": {},
   "source": [
    "# Enrichment analysis using KS and Wilcoxon test\n",
    "\n",
    "The Kolmogorov-Smirnov (KS) test is a non-parametric test that compares the cumulative distribution functions of two datasets. In the context of enrichment analysis, it can be employed to assess whether the distribution of DE genes is significantly different from what would be expected by chance for a given gene set. The test produces a KS statistic and a corresponding p-value, indicating the degree of over-representation.\n",
    "\n",
    "The Wilcoxon test, also known as the Mann-Whitney U test, is another non-parametric test used in enrichment analysis. It assesses whether there is a significant difference in the ranks of genes between the DE gene set and the rest of the genes."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2a98cfa-2b91-46a5-92ea-38ee1deb2eac",
   "metadata": {},
   "source": [
    "### Data preparation\n",
    "To conduct enrichment analysis using KS or Wilcoxon test, we will use the limma DE result as input. First, we can use the DE.df variable above and select a set of genes that\n",
    "are significantly altered (*p-value < 0.05*)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1135da3e-3277-408f-a4b2-7ec34d09a659",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>PROBEID</th><th scope=col>ENTREZID</th><th scope=col>GENENAME</th><th scope=col>SYMBOL</th><th scope=col>logFC</th><th scope=col>AveExpr</th><th scope=col>t</th><th scope=col>P.Value</th><th scope=col>adj.P.Val</th><th scope=col>B</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>146</th><td>1552467_at</td><td>259234</td><td>Down syndrome critical region 10                    </td><td>DSCR10</td><td> 0.07632117</td><td>1.1024122</td><td> 3.721644</td><td>4.778227e-04</td><td>4.325324e-02</td><td>-0.28303673</td></tr>\n",
       "\t<tr><th scope=row>149</th><td>1552473_at</td><td>2593  </td><td>guanidinoacetate N-methyltransferase                </td><td>GAMT  </td><td> 0.12655694</td><td>1.0378989</td><td> 3.829094</td><td>3.402464e-04</td><td>3.481091e-02</td><td> 0.03057276</td></tr>\n",
       "\t<tr><th scope=row>175</th><td>1552514_at</td><td>164684</td><td>WBP2 N-terminal like                                </td><td>WBP2NL</td><td>-0.43701990</td><td>1.2072571</td><td>-3.705169</td><td>5.031544e-04</td><td>4.458666e-02</td><td>-0.33067421</td></tr>\n",
       "\t<tr><th scope=row>181</th><td>1552524_at</td><td>116969</td><td>ADP-ribosyltransferase 5                            </td><td>ART5  </td><td>-0.15248106</td><td>1.2214668</td><td>-5.601889</td><td>7.601467e-07</td><td>3.172597e-04</td><td> 5.74686253</td></tr>\n",
       "\t<tr><th scope=row>220</th><td>1552596_at</td><td>246176</td><td>growth arrest specific 2 like 2                     </td><td>GAS2L2</td><td> 0.09704702</td><td>0.9937790</td><td> 8.231539</td><td>4.578441e-11</td><td>4.490998e-08</td><td>14.90292753</td></tr>\n",
       "\t<tr><th scope=row>381</th><td>1552768_at</td><td>84254 </td><td>calcium/calmodulin dependent protein kinase kinase 1</td><td>CAMKK1</td><td> 0.07873062</td><td>0.8284583</td><td> 3.985720</td><td>2.057635e-04</td><td>2.541044e-02</td><td> 0.49642261</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 10\n",
       "\\begin{tabular}{r|llllllllll}\n",
       "  & PROBEID & ENTREZID & GENENAME & SYMBOL & logFC & AveExpr & t & P.Value & adj.P.Val & B\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t146 & 1552467\\_at & 259234 & Down syndrome critical region 10                     & DSCR10 &  0.07632117 & 1.1024122 &  3.721644 & 4.778227e-04 & 4.325324e-02 & -0.28303673\\\\\n",
       "\t149 & 1552473\\_at & 2593   & guanidinoacetate N-methyltransferase                 & GAMT   &  0.12655694 & 1.0378989 &  3.829094 & 3.402464e-04 & 3.481091e-02 &  0.03057276\\\\\n",
       "\t175 & 1552514\\_at & 164684 & WBP2 N-terminal like                                 & WBP2NL & -0.43701990 & 1.2072571 & -3.705169 & 5.031544e-04 & 4.458666e-02 & -0.33067421\\\\\n",
       "\t181 & 1552524\\_at & 116969 & ADP-ribosyltransferase 5                             & ART5   & -0.15248106 & 1.2214668 & -5.601889 & 7.601467e-07 & 3.172597e-04 &  5.74686253\\\\\n",
       "\t220 & 1552596\\_at & 246176 & growth arrest specific 2 like 2                      & GAS2L2 &  0.09704702 & 0.9937790 &  8.231539 & 4.578441e-11 & 4.490998e-08 & 14.90292753\\\\\n",
       "\t381 & 1552768\\_at & 84254  & calcium/calmodulin dependent protein kinase kinase 1 & CAMKK1 &  0.07873062 & 0.8284583 &  3.985720 & 2.057635e-04 & 2.541044e-02 &  0.49642261\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 10\n",
       "\n",
       "| <!--/--> | PROBEID &lt;chr&gt; | ENTREZID &lt;chr&gt; | GENENAME &lt;chr&gt; | SYMBOL &lt;chr&gt; | logFC &lt;dbl&gt; | AveExpr &lt;dbl&gt; | t &lt;dbl&gt; | P.Value &lt;dbl&gt; | adj.P.Val &lt;dbl&gt; | B &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 146 | 1552467_at | 259234 | Down syndrome critical region 10                     | DSCR10 |  0.07632117 | 1.1024122 |  3.721644 | 4.778227e-04 | 4.325324e-02 | -0.28303673 |\n",
       "| 149 | 1552473_at | 2593   | guanidinoacetate N-methyltransferase                 | GAMT   |  0.12655694 | 1.0378989 |  3.829094 | 3.402464e-04 | 3.481091e-02 |  0.03057276 |\n",
       "| 175 | 1552514_at | 164684 | WBP2 N-terminal like                                 | WBP2NL | -0.43701990 | 1.2072571 | -3.705169 | 5.031544e-04 | 4.458666e-02 | -0.33067421 |\n",
       "| 181 | 1552524_at | 116969 | ADP-ribosyltransferase 5                             | ART5   | -0.15248106 | 1.2214668 | -5.601889 | 7.601467e-07 | 3.172597e-04 |  5.74686253 |\n",
       "| 220 | 1552596_at | 246176 | growth arrest specific 2 like 2                      | GAS2L2 |  0.09704702 | 0.9937790 |  8.231539 | 4.578441e-11 | 4.490998e-08 | 14.90292753 |\n",
       "| 381 | 1552768_at | 84254  | calcium/calmodulin dependent protein kinase kinase 1 | CAMKK1 |  0.07873062 | 0.8284583 |  3.985720 | 2.057635e-04 | 2.541044e-02 |  0.49642261 |\n",
       "\n"
      ],
      "text/plain": [
       "    PROBEID    ENTREZID GENENAME                                            \n",
       "146 1552467_at 259234   Down syndrome critical region 10                    \n",
       "149 1552473_at 2593     guanidinoacetate N-methyltransferase                \n",
       "175 1552514_at 164684   WBP2 N-terminal like                                \n",
       "181 1552524_at 116969   ADP-ribosyltransferase 5                            \n",
       "220 1552596_at 246176   growth arrest specific 2 like 2                     \n",
       "381 1552768_at 84254    calcium/calmodulin dependent protein kinase kinase 1\n",
       "    SYMBOL logFC       AveExpr   t         P.Value      adj.P.Val   \n",
       "146 DSCR10  0.07632117 1.1024122  3.721644 4.778227e-04 4.325324e-02\n",
       "149 GAMT    0.12655694 1.0378989  3.829094 3.402464e-04 3.481091e-02\n",
       "175 WBP2NL -0.43701990 1.2072571 -3.705169 5.031544e-04 4.458666e-02\n",
       "181 ART5   -0.15248106 1.2214668 -5.601889 7.601467e-07 3.172597e-04\n",
       "220 GAS2L2  0.09704702 0.9937790  8.231539 4.578441e-11 4.490998e-08\n",
       "381 CAMKK1  0.07873062 0.8284583  3.985720 2.057635e-04 2.541044e-02\n",
       "    B          \n",
       "146 -0.28303673\n",
       "149  0.03057276\n",
       "175 -0.33067421\n",
       "181  5.74686253\n",
       "220 14.90292753\n",
       "381  0.49642261"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Select genes that are significantly altered\n",
    "kw_de <- DE.df[DE.df$adj.P.Val <= 0.05, ]\n",
    "head(kw_de)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fe00efa-e665-476b-8872-5482f4dafde1",
   "metadata": {},
   "source": [
    "Next, we will remove rows containing NA, dupicated genes symbol."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "47e8c745-2bc5-4991-979f-89e2d0d643e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>SYMBOL</th><th scope=col>logFC</th><th scope=col>AveExpr</th><th scope=col>t</th><th scope=col>P.Value</th><th scope=col>adj.P.Val</th><th scope=col>B</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>DSCR10</th><td>DSCR10</td><td> 0.07632117</td><td>1.1024122</td><td> 3.721644</td><td>4.778227e-04</td><td>4.325324e-02</td><td>-0.28303673</td></tr>\n",
       "\t<tr><th scope=row>GAMT</th><td>GAMT  </td><td> 0.12655694</td><td>1.0378989</td><td> 3.829094</td><td>3.402464e-04</td><td>3.481091e-02</td><td> 0.03057276</td></tr>\n",
       "\t<tr><th scope=row>WBP2NL</th><td>WBP2NL</td><td>-0.43701990</td><td>1.2072571</td><td>-3.705169</td><td>5.031544e-04</td><td>4.458666e-02</td><td>-0.33067421</td></tr>\n",
       "\t<tr><th scope=row>ART5</th><td>ART5  </td><td>-0.15248106</td><td>1.2214668</td><td>-5.601889</td><td>7.601467e-07</td><td>3.172597e-04</td><td> 5.74686253</td></tr>\n",
       "\t<tr><th scope=row>GAS2L2</th><td>GAS2L2</td><td> 0.09704702</td><td>0.9937790</td><td> 8.231539</td><td>4.578441e-11</td><td>4.490998e-08</td><td>14.90292753</td></tr>\n",
       "\t<tr><th scope=row>CAMKK1</th><td>CAMKK1</td><td> 0.07873062</td><td>0.8284583</td><td> 3.985720</td><td>2.057635e-04</td><td>2.541044e-02</td><td> 0.49642261</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & SYMBOL & logFC & AveExpr & t & P.Value & adj.P.Val & B\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\tDSCR10 & DSCR10 &  0.07632117 & 1.1024122 &  3.721644 & 4.778227e-04 & 4.325324e-02 & -0.28303673\\\\\n",
       "\tGAMT & GAMT   &  0.12655694 & 1.0378989 &  3.829094 & 3.402464e-04 & 3.481091e-02 &  0.03057276\\\\\n",
       "\tWBP2NL & WBP2NL & -0.43701990 & 1.2072571 & -3.705169 & 5.031544e-04 & 4.458666e-02 & -0.33067421\\\\\n",
       "\tART5 & ART5   & -0.15248106 & 1.2214668 & -5.601889 & 7.601467e-07 & 3.172597e-04 &  5.74686253\\\\\n",
       "\tGAS2L2 & GAS2L2 &  0.09704702 & 0.9937790 &  8.231539 & 4.578441e-11 & 4.490998e-08 & 14.90292753\\\\\n",
       "\tCAMKK1 & CAMKK1 &  0.07873062 & 0.8284583 &  3.985720 & 2.057635e-04 & 2.541044e-02 &  0.49642261\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 7\n",
       "\n",
       "| <!--/--> | SYMBOL &lt;chr&gt; | logFC &lt;dbl&gt; | AveExpr &lt;dbl&gt; | t &lt;dbl&gt; | P.Value &lt;dbl&gt; | adj.P.Val &lt;dbl&gt; | B &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| DSCR10 | DSCR10 |  0.07632117 | 1.1024122 |  3.721644 | 4.778227e-04 | 4.325324e-02 | -0.28303673 |\n",
       "| GAMT | GAMT   |  0.12655694 | 1.0378989 |  3.829094 | 3.402464e-04 | 3.481091e-02 |  0.03057276 |\n",
       "| WBP2NL | WBP2NL | -0.43701990 | 1.2072571 | -3.705169 | 5.031544e-04 | 4.458666e-02 | -0.33067421 |\n",
       "| ART5 | ART5   | -0.15248106 | 1.2214668 | -5.601889 | 7.601467e-07 | 3.172597e-04 |  5.74686253 |\n",
       "| GAS2L2 | GAS2L2 |  0.09704702 | 0.9937790 |  8.231539 | 4.578441e-11 | 4.490998e-08 | 14.90292753 |\n",
       "| CAMKK1 | CAMKK1 |  0.07873062 | 0.8284583 |  3.985720 | 2.057635e-04 | 2.541044e-02 |  0.49642261 |\n",
       "\n"
      ],
      "text/plain": [
       "       SYMBOL logFC       AveExpr   t         P.Value      adj.P.Val   \n",
       "DSCR10 DSCR10  0.07632117 1.1024122  3.721644 4.778227e-04 4.325324e-02\n",
       "GAMT   GAMT    0.12655694 1.0378989  3.829094 3.402464e-04 3.481091e-02\n",
       "WBP2NL WBP2NL -0.43701990 1.2072571 -3.705169 5.031544e-04 4.458666e-02\n",
       "ART5   ART5   -0.15248106 1.2214668 -5.601889 7.601467e-07 3.172597e-04\n",
       "GAS2L2 GAS2L2  0.09704702 0.9937790  8.231539 4.578441e-11 4.490998e-08\n",
       "CAMKK1 CAMKK1  0.07873062 0.8284583  3.985720 2.057635e-04 2.541044e-02\n",
       "       B          \n",
       "DSCR10 -0.28303673\n",
       "GAMT    0.03057276\n",
       "WBP2NL -0.33067421\n",
       "ART5    5.74686253\n",
       "GAS2L2 14.90292753\n",
       "CAMKK1  0.49642261"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Remove NA value\n",
    "kw_de <- kw_de[!is.na(kw_de$SYMBOL), ]\n",
    "# Remove duplicated genes symbol\n",
    "kw_de <- kw_de[!duplicated(kw_de$SYMBOL, fromLast = FALSE), ]\n",
    "rownames(kw_de) <- kw_de$SYMBOL\n",
    "# remove the first 3 columns from the DE dataframe  \n",
    "kw_de <- kw_de[, -c(1:3)]\n",
    "head(kw_de)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f7c106eb-b3bf-40d8-9b57-acd92bb7adad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>241</li><li>7</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 241\n",
       "\\item 7\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 241\n",
       "2. 7\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 241   7"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(kw_de)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "610c0f93-6a3b-4221-bf0a-b93d65c4419c",
   "metadata": {},
   "source": [
    "To perform KS or Wilcoxon test, we will use the `.runKsWilcox` function bellow."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "14a9698b-6464-451e-aa95-a7a13cbe2858",
   "metadata": {},
   "outputs": [],
   "source": [
    ".runKsWilcox <- function(DE.df, statCol, genesets, sTest) {\n",
    "    # remove rows that contain NA value in adj.P.Val column\n",
    "    DE_data <- DE.df[!is.na(DE.df[[statCol]]), ]\n",
    "    # get the adjusted p-values from DE.df to perform the test\n",
    "    ranks <- DE_data[[statCol]]\n",
    "    # set names of the DE genes\n",
    "    names(ranks) <- rownames(DE_data)\n",
    "\n",
    "    # choose the test to perform based on the sTest argument\n",
    "    test <- if (sTest == \"ks\") ks.test else wilcox.test\n",
    "    \n",
    "    background.genes <- rownames(DE_data)\n",
    "\n",
    "    DEhit <- sapply(genesets, function(gs) ranks[background.genes[background.genes %in% gs]])\n",
    "    DEmiss <- sapply(genesets, function(gs) ranks[background.genes[!background.genes %in% gs]])\n",
    "\n",
    "    # Initialize an empty list to store results\n",
    "    PA_res_list <- lapply(1:length(DEhit), function(i) {\n",
    "        cur.DEhit <- DEhit[[i]]\n",
    "        cur.DEmiss <- DEmiss[[i]]\n",
    "    \n",
    "        # Check if either list is empty\n",
    "        if (length(cur.DEhit) == 0 | length(cur.DEmiss) == 0) return(1)\n",
    "    \n",
    "        # Perform the test and extract the p-value\n",
    "        test_result <- test(cur.DEhit, cur.DEmiss)\n",
    "        test_result$p.value\n",
    "    })\n",
    "    \n",
    "    # Combine the results into a data frame\n",
    "    PA_res <- data.frame(\n",
    "        ID = names(genesets),\n",
    "        p.value = unlist(PA_res_list),\n",
    "        stringsAsFactors = FALSE\n",
    "    )\n",
    "    \n",
    "    # Remove rows with missing values\n",
    "    PA_res <- PA_res[complete.cases(PA_res), ]\n",
    "    # Ensure row names are unique\n",
    "    PA_res$ID <- make.unique(PA_res$ID)\n",
    "    # Set row names\n",
    "    rownames(PA_res) <- PA_res$ID\n",
    "\n",
    "    return(PA_res)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0245522a-5d4e-4d40-8bc8-b04af61a2358",
   "metadata": {},
   "source": [
    "To use the `.runKsWilcox` function, we need to pass the following arguments:\n",
    "\n",
    "- `DE.df`: A dataframe containing information about differentially expressed genes.\n",
    "- `statCol`: A character string specifying the column in the DE.df data frame that contains the statistical values (e.g., p-value, adjusted p-values) associated with each gene.\n",
    "- `geneset`: A list of gene sets to be analyzed.\n",
    "- `sTest`: A character string specifying the testing method to be used; it can take either \"ks\" for the Kolmogorov-Smirnov test or \"wilcox\" for the Wilcoxon test.\r\n",
    "\n",
    "The `.runKsWilcox` function returns a dataframe of the analysis result.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1461242-eab0-40cc-8168-e99c577f08f5",
   "metadata": {},
   "source": [
    "### Enrichment analysis using KS and GO terms"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7eb5e19-f357-4656-9632-02118bb129e9",
   "metadata": {},
   "source": [
    "We will need a gene set to perform enrichment analysis using KS/Wilcoxon test. Here, we will use the GO terms saved in the \"GO_terms.gmt\" file in module 3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b6aca88a-249c-4d46-bd99-ae3ad29415db",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ID</th><th scope=col>p.value</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>cellular respiration</th><td>cellular respiration                </td><td>0.0001154964</td></tr>\n",
       "\t<tr><th scope=row>aerobic respiration</th><td>aerobic respiration                 </td><td>0.0003372217</td></tr>\n",
       "\t<tr><th scope=row>oxidative phosphorylation</th><td>oxidative phosphorylation           </td><td>0.0010481317</td></tr>\n",
       "\t<tr><th scope=row>respiratory electron transport chain</th><td>respiratory electron transport chain</td><td>0.0010481317</td></tr>\n",
       "\t<tr><th scope=row>leukocyte apoptotic process</th><td>leukocyte apoptotic process         </td><td>0.0024896266</td></tr>\n",
       "\t<tr><th scope=row>signal transduction</th><td>signal transduction                 </td><td>0.0042530813</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & ID & p.value\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\tcellular respiration & cellular respiration                 & 0.0001154964\\\\\n",
       "\taerobic respiration & aerobic respiration                  & 0.0003372217\\\\\n",
       "\toxidative phosphorylation & oxidative phosphorylation            & 0.0010481317\\\\\n",
       "\trespiratory electron transport chain & respiratory electron transport chain & 0.0010481317\\\\\n",
       "\tleukocyte apoptotic process & leukocyte apoptotic process          & 0.0024896266\\\\\n",
       "\tsignal transduction & signal transduction                  & 0.0042530813\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | ID &lt;chr&gt; | p.value &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| cellular respiration | cellular respiration                 | 0.0001154964 |\n",
       "| aerobic respiration | aerobic respiration                  | 0.0003372217 |\n",
       "| oxidative phosphorylation | oxidative phosphorylation            | 0.0010481317 |\n",
       "| respiratory electron transport chain | respiratory electron transport chain | 0.0010481317 |\n",
       "| leukocyte apoptotic process | leukocyte apoptotic process          | 0.0024896266 |\n",
       "| signal transduction | signal transduction                  | 0.0042530813 |\n",
       "\n"
      ],
      "text/plain": [
       "                                     ID                                  \n",
       "cellular respiration                 cellular respiration                \n",
       "aerobic respiration                  aerobic respiration                 \n",
       "oxidative phosphorylation            oxidative phosphorylation           \n",
       "respiratory electron transport chain respiratory electron transport chain\n",
       "leukocyte apoptotic process          leukocyte apoptotic process         \n",
       "signal transduction                  signal transduction                 \n",
       "                                     p.value     \n",
       "cellular respiration                 0.0001154964\n",
       "aerobic respiration                  0.0003372217\n",
       "oxidative phosphorylation            0.0010481317\n",
       "respiratory electron transport chain 0.0010481317\n",
       "leukocyte apoptotic process          0.0024896266\n",
       "signal transduction                  0.0042530813"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load the GO gene sets from saved gmt file\n",
    "suppressWarnings({suppressMessages({geneset <- gmt2geneset(\"./data/GO_terms.gmt\")})})\n",
    "# perform enrichment analysis using KS test\n",
    "suppressWarnings(k_results <- .runKsWilcox(DE.df = kw_de, statCol = \"adj.P.Val\", geneset = geneset, sTest = \"ks\"))\n",
    "\n",
    "# order the results \n",
    "k_results <- k_results[order(k_results$p.value), ]\n",
    "# show some first rows of the result table\n",
    "head(k_results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "214d9285-55fa-4b53-a974-ea18096ffea8",
   "metadata": {},
   "source": [
    "### Enrichment analysis using KS and KEGG pathways\n",
    "To do the analysis with KEEG pathways, we simply replace the GO gene sets with the KEGG pathways."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6b3d0d05-3bae-4b41-8ade-479d1985cf40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ID</th><th scope=col>p.value</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>Complement and coagulation cascades - Homo sapiens (human)</th><td>Complement and coagulation cascades - Homo sapiens (human)</td><td>0.01317731</td></tr>\n",
       "\t<tr><th scope=row>COX5A</th><td>COX5A                                                     </td><td>0.01447322</td></tr>\n",
       "\t<tr><th scope=row>SDHC</th><td>SDHC                                                      </td><td>0.01447322</td></tr>\n",
       "\t<tr><th scope=row>SDHC.1</th><td>SDHC.1                                                    </td><td>0.01447322</td></tr>\n",
       "\t<tr><th scope=row>COX2</th><td>COX2                                                      </td><td>0.01447322</td></tr>\n",
       "\t<tr><th scope=row>NDUFA4L2</th><td>NDUFA4L2                                                  </td><td>0.01447322</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & ID & p.value\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\tComplement and coagulation cascades - Homo sapiens (human) & Complement and coagulation cascades - Homo sapiens (human) & 0.01317731\\\\\n",
       "\tCOX5A & COX5A                                                      & 0.01447322\\\\\n",
       "\tSDHC & SDHC                                                       & 0.01447322\\\\\n",
       "\tSDHC.1 & SDHC.1                                                     & 0.01447322\\\\\n",
       "\tCOX2 & COX2                                                       & 0.01447322\\\\\n",
       "\tNDUFA4L2 & NDUFA4L2                                                   & 0.01447322\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | ID &lt;chr&gt; | p.value &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| Complement and coagulation cascades - Homo sapiens (human) | Complement and coagulation cascades - Homo sapiens (human) | 0.01317731 |\n",
       "| COX5A | COX5A                                                      | 0.01447322 |\n",
       "| SDHC | SDHC                                                       | 0.01447322 |\n",
       "| SDHC.1 | SDHC.1                                                     | 0.01447322 |\n",
       "| COX2 | COX2                                                       | 0.01447322 |\n",
       "| NDUFA4L2 | NDUFA4L2                                                   | 0.01447322 |\n",
       "\n"
      ],
      "text/plain": [
       "                                                           ID                                                        \n",
       "Complement and coagulation cascades - Homo sapiens (human) Complement and coagulation cascades - Homo sapiens (human)\n",
       "COX5A                                                      COX5A                                                     \n",
       "SDHC                                                       SDHC                                                      \n",
       "SDHC.1                                                     SDHC.1                                                    \n",
       "COX2                                                       COX2                                                      \n",
       "NDUFA4L2                                                   NDUFA4L2                                                  \n",
       "                                                           p.value   \n",
       "Complement and coagulation cascades - Homo sapiens (human) 0.01317731\n",
       "COX5A                                                      0.01447322\n",
       "SDHC                                                       0.01447322\n",
       "SDHC.1                                                     0.01447322\n",
       "COX2                                                       0.01447322\n",
       "NDUFA4L2                                                   0.01447322"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Loading the KEGG pathways from saved gmt file\n",
    "suppressWarnings({suppressMessages({geneset <- gmt2geneset(\"./data/KEGG_pathways.gmt\")})})\n",
    "# perform enrichment analysis using KS test\n",
    "k_results <- .runKsWilcox(DE.df = kw_de, statCol = \"adj.P.Val\", geneset = geneset, sTest = \"ks\")\n",
    "\n",
    "# order the results \n",
    "k_results <- k_results[order(k_results$p.value), ]\n",
    "# show some first rows of the result table\n",
    "head(k_results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "036d1f5a-8292-442c-9e81-d6cdfd286ae5",
   "metadata": {},
   "source": [
    "### Enrichment analysis using KS and REACTOME pathways\n",
    "To do the analysis with REACTOME pathways, we will use the same steps as for GO terms and KEEG pathways. But instead of select the SYMBOL column to set for row names, we will use the ENTREZID column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "905fba8b-ad78-4242-a32c-878205b91a9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>PROBEID</th><th scope=col>ENTREZID</th><th scope=col>GENENAME</th><th scope=col>SYMBOL</th><th scope=col>logFC</th><th scope=col>AveExpr</th><th scope=col>t</th><th scope=col>P.Value</th><th scope=col>adj.P.Val</th><th scope=col>B</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>146</th><td>1552467_at</td><td>259234</td><td>Down syndrome critical region 10                    </td><td>DSCR10</td><td> 0.07632117</td><td>1.1024122</td><td> 3.721644</td><td>4.778227e-04</td><td>4.325324e-02</td><td>-0.28303673</td></tr>\n",
       "\t<tr><th scope=row>149</th><td>1552473_at</td><td>2593  </td><td>guanidinoacetate N-methyltransferase                </td><td>GAMT  </td><td> 0.12655694</td><td>1.0378989</td><td> 3.829094</td><td>3.402464e-04</td><td>3.481091e-02</td><td> 0.03057276</td></tr>\n",
       "\t<tr><th scope=row>175</th><td>1552514_at</td><td>164684</td><td>WBP2 N-terminal like                                </td><td>WBP2NL</td><td>-0.43701990</td><td>1.2072571</td><td>-3.705169</td><td>5.031544e-04</td><td>4.458666e-02</td><td>-0.33067421</td></tr>\n",
       "\t<tr><th scope=row>181</th><td>1552524_at</td><td>116969</td><td>ADP-ribosyltransferase 5                            </td><td>ART5  </td><td>-0.15248106</td><td>1.2214668</td><td>-5.601889</td><td>7.601467e-07</td><td>3.172597e-04</td><td> 5.74686253</td></tr>\n",
       "\t<tr><th scope=row>220</th><td>1552596_at</td><td>246176</td><td>growth arrest specific 2 like 2                     </td><td>GAS2L2</td><td> 0.09704702</td><td>0.9937790</td><td> 8.231539</td><td>4.578441e-11</td><td>4.490998e-08</td><td>14.90292753</td></tr>\n",
       "\t<tr><th scope=row>381</th><td>1552768_at</td><td>84254 </td><td>calcium/calmodulin dependent protein kinase kinase 1</td><td>CAMKK1</td><td> 0.07873062</td><td>0.8284583</td><td> 3.985720</td><td>2.057635e-04</td><td>2.541044e-02</td><td> 0.49642261</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 10\n",
       "\\begin{tabular}{r|llllllllll}\n",
       "  & PROBEID & ENTREZID & GENENAME & SYMBOL & logFC & AveExpr & t & P.Value & adj.P.Val & B\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t146 & 1552467\\_at & 259234 & Down syndrome critical region 10                     & DSCR10 &  0.07632117 & 1.1024122 &  3.721644 & 4.778227e-04 & 4.325324e-02 & -0.28303673\\\\\n",
       "\t149 & 1552473\\_at & 2593   & guanidinoacetate N-methyltransferase                 & GAMT   &  0.12655694 & 1.0378989 &  3.829094 & 3.402464e-04 & 3.481091e-02 &  0.03057276\\\\\n",
       "\t175 & 1552514\\_at & 164684 & WBP2 N-terminal like                                 & WBP2NL & -0.43701990 & 1.2072571 & -3.705169 & 5.031544e-04 & 4.458666e-02 & -0.33067421\\\\\n",
       "\t181 & 1552524\\_at & 116969 & ADP-ribosyltransferase 5                             & ART5   & -0.15248106 & 1.2214668 & -5.601889 & 7.601467e-07 & 3.172597e-04 &  5.74686253\\\\\n",
       "\t220 & 1552596\\_at & 246176 & growth arrest specific 2 like 2                      & GAS2L2 &  0.09704702 & 0.9937790 &  8.231539 & 4.578441e-11 & 4.490998e-08 & 14.90292753\\\\\n",
       "\t381 & 1552768\\_at & 84254  & calcium/calmodulin dependent protein kinase kinase 1 & CAMKK1 &  0.07873062 & 0.8284583 &  3.985720 & 2.057635e-04 & 2.541044e-02 &  0.49642261\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 10\n",
       "\n",
       "| <!--/--> | PROBEID &lt;chr&gt; | ENTREZID &lt;chr&gt; | GENENAME &lt;chr&gt; | SYMBOL &lt;chr&gt; | logFC &lt;dbl&gt; | AveExpr &lt;dbl&gt; | t &lt;dbl&gt; | P.Value &lt;dbl&gt; | adj.P.Val &lt;dbl&gt; | B &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 146 | 1552467_at | 259234 | Down syndrome critical region 10                     | DSCR10 |  0.07632117 | 1.1024122 |  3.721644 | 4.778227e-04 | 4.325324e-02 | -0.28303673 |\n",
       "| 149 | 1552473_at | 2593   | guanidinoacetate N-methyltransferase                 | GAMT   |  0.12655694 | 1.0378989 |  3.829094 | 3.402464e-04 | 3.481091e-02 |  0.03057276 |\n",
       "| 175 | 1552514_at | 164684 | WBP2 N-terminal like                                 | WBP2NL | -0.43701990 | 1.2072571 | -3.705169 | 5.031544e-04 | 4.458666e-02 | -0.33067421 |\n",
       "| 181 | 1552524_at | 116969 | ADP-ribosyltransferase 5                             | ART5   | -0.15248106 | 1.2214668 | -5.601889 | 7.601467e-07 | 3.172597e-04 |  5.74686253 |\n",
       "| 220 | 1552596_at | 246176 | growth arrest specific 2 like 2                      | GAS2L2 |  0.09704702 | 0.9937790 |  8.231539 | 4.578441e-11 | 4.490998e-08 | 14.90292753 |\n",
       "| 381 | 1552768_at | 84254  | calcium/calmodulin dependent protein kinase kinase 1 | CAMKK1 |  0.07873062 | 0.8284583 |  3.985720 | 2.057635e-04 | 2.541044e-02 |  0.49642261 |\n",
       "\n"
      ],
      "text/plain": [
       "    PROBEID    ENTREZID GENENAME                                            \n",
       "146 1552467_at 259234   Down syndrome critical region 10                    \n",
       "149 1552473_at 2593     guanidinoacetate N-methyltransferase                \n",
       "175 1552514_at 164684   WBP2 N-terminal like                                \n",
       "181 1552524_at 116969   ADP-ribosyltransferase 5                            \n",
       "220 1552596_at 246176   growth arrest specific 2 like 2                     \n",
       "381 1552768_at 84254    calcium/calmodulin dependent protein kinase kinase 1\n",
       "    SYMBOL logFC       AveExpr   t         P.Value      adj.P.Val   \n",
       "146 DSCR10  0.07632117 1.1024122  3.721644 4.778227e-04 4.325324e-02\n",
       "149 GAMT    0.12655694 1.0378989  3.829094 3.402464e-04 3.481091e-02\n",
       "175 WBP2NL -0.43701990 1.2072571 -3.705169 5.031544e-04 4.458666e-02\n",
       "181 ART5   -0.15248106 1.2214668 -5.601889 7.601467e-07 3.172597e-04\n",
       "220 GAS2L2  0.09704702 0.9937790  8.231539 4.578441e-11 4.490998e-08\n",
       "381 CAMKK1  0.07873062 0.8284583  3.985720 2.057635e-04 2.541044e-02\n",
       "    B          \n",
       "146 -0.28303673\n",
       "149  0.03057276\n",
       "175 -0.33067421\n",
       "181  5.74686253\n",
       "220 14.90292753\n",
       "381  0.49642261"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>SYMBOL</th><th scope=col>logFC</th><th scope=col>AveExpr</th><th scope=col>t</th><th scope=col>P.Value</th><th scope=col>adj.P.Val</th><th scope=col>B</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>259234</th><td>DSCR10</td><td> 0.07632117</td><td>1.1024122</td><td> 3.721644</td><td>4.778227e-04</td><td>4.325324e-02</td><td>-0.28303673</td></tr>\n",
       "\t<tr><th scope=row>2593</th><td>GAMT  </td><td> 0.12655694</td><td>1.0378989</td><td> 3.829094</td><td>3.402464e-04</td><td>3.481091e-02</td><td> 0.03057276</td></tr>\n",
       "\t<tr><th scope=row>164684</th><td>WBP2NL</td><td>-0.43701990</td><td>1.2072571</td><td>-3.705169</td><td>5.031544e-04</td><td>4.458666e-02</td><td>-0.33067421</td></tr>\n",
       "\t<tr><th scope=row>116969</th><td>ART5  </td><td>-0.15248106</td><td>1.2214668</td><td>-5.601889</td><td>7.601467e-07</td><td>3.172597e-04</td><td> 5.74686253</td></tr>\n",
       "\t<tr><th scope=row>246176</th><td>GAS2L2</td><td> 0.09704702</td><td>0.9937790</td><td> 8.231539</td><td>4.578441e-11</td><td>4.490998e-08</td><td>14.90292753</td></tr>\n",
       "\t<tr><th scope=row>84254</th><td>CAMKK1</td><td> 0.07873062</td><td>0.8284583</td><td> 3.985720</td><td>2.057635e-04</td><td>2.541044e-02</td><td> 0.49642261</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & SYMBOL & logFC & AveExpr & t & P.Value & adj.P.Val & B\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t259234 & DSCR10 &  0.07632117 & 1.1024122 &  3.721644 & 4.778227e-04 & 4.325324e-02 & -0.28303673\\\\\n",
       "\t2593 & GAMT   &  0.12655694 & 1.0378989 &  3.829094 & 3.402464e-04 & 3.481091e-02 &  0.03057276\\\\\n",
       "\t164684 & WBP2NL & -0.43701990 & 1.2072571 & -3.705169 & 5.031544e-04 & 4.458666e-02 & -0.33067421\\\\\n",
       "\t116969 & ART5   & -0.15248106 & 1.2214668 & -5.601889 & 7.601467e-07 & 3.172597e-04 &  5.74686253\\\\\n",
       "\t246176 & GAS2L2 &  0.09704702 & 0.9937790 &  8.231539 & 4.578441e-11 & 4.490998e-08 & 14.90292753\\\\\n",
       "\t84254 & CAMKK1 &  0.07873062 & 0.8284583 &  3.985720 & 2.057635e-04 & 2.541044e-02 &  0.49642261\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 7\n",
       "\n",
       "| <!--/--> | SYMBOL &lt;chr&gt; | logFC &lt;dbl&gt; | AveExpr &lt;dbl&gt; | t &lt;dbl&gt; | P.Value &lt;dbl&gt; | adj.P.Val &lt;dbl&gt; | B &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 259234 | DSCR10 |  0.07632117 | 1.1024122 |  3.721644 | 4.778227e-04 | 4.325324e-02 | -0.28303673 |\n",
       "| 2593 | GAMT   |  0.12655694 | 1.0378989 |  3.829094 | 3.402464e-04 | 3.481091e-02 |  0.03057276 |\n",
       "| 164684 | WBP2NL | -0.43701990 | 1.2072571 | -3.705169 | 5.031544e-04 | 4.458666e-02 | -0.33067421 |\n",
       "| 116969 | ART5   | -0.15248106 | 1.2214668 | -5.601889 | 7.601467e-07 | 3.172597e-04 |  5.74686253 |\n",
       "| 246176 | GAS2L2 |  0.09704702 | 0.9937790 |  8.231539 | 4.578441e-11 | 4.490998e-08 | 14.90292753 |\n",
       "| 84254 | CAMKK1 |  0.07873062 | 0.8284583 |  3.985720 | 2.057635e-04 | 2.541044e-02 |  0.49642261 |\n",
       "\n"
      ],
      "text/plain": [
       "       SYMBOL logFC       AveExpr   t         P.Value      adj.P.Val   \n",
       "259234 DSCR10  0.07632117 1.1024122  3.721644 4.778227e-04 4.325324e-02\n",
       "2593   GAMT    0.12655694 1.0378989  3.829094 3.402464e-04 3.481091e-02\n",
       "164684 WBP2NL -0.43701990 1.2072571 -3.705169 5.031544e-04 4.458666e-02\n",
       "116969 ART5   -0.15248106 1.2214668 -5.601889 7.601467e-07 3.172597e-04\n",
       "246176 GAS2L2  0.09704702 0.9937790  8.231539 4.578441e-11 4.490998e-08\n",
       "84254  CAMKK1  0.07873062 0.8284583  3.985720 2.057635e-04 2.541044e-02\n",
       "       B          \n",
       "259234 -0.28303673\n",
       "2593    0.03057276\n",
       "164684 -0.33067421\n",
       "116969  5.74686253\n",
       "246176 14.90292753\n",
       "84254   0.49642261"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Select genes that are significantly altered\n",
    "kwR_de <- DE.df[DE.df$adj.P.Val <= 0.05, ]\n",
    "head(kwR_de)\n",
    "\n",
    "# Remove NA value\n",
    "kwR_de <- kwR_de[!is.na(kwR_de$ENTREZID), ]\n",
    "# Remove duplicated genes symbol\n",
    "kwR_de <- kwR_de[!duplicated(kwR_de$ENTREZID, fromLast = FALSE), ]\n",
    "rownames(kwR_de) <- kwR_de$ENTREZID\n",
    "# remove the first 3 columns from the DE dataframe  \n",
    "kwR_de <- kwR_de[, -c(1:3)]\n",
    "head(kwR_de)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "3a67eacd-72f8-4ac8-99b5-90048bcad9ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ID</th><th scope=col>p.value</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>Homo sapiens: Activation of HOX genes during differentiation</th><td>Homo sapiens: Activation of HOX genes during differentiation                                                      </td><td>0.007936508</td></tr>\n",
       "\t<tr><th scope=row>Homo sapiens: Activation of anterior HOX genes in hindbrain development during early embryogenesis</th><td>Homo sapiens: Activation of anterior HOX genes in hindbrain development during early embryogenesis                </td><td>0.007936508</td></tr>\n",
       "\t<tr><th scope=row>Homo sapiens: Regulation of gene expression in early pancreatic precursor cells</th><td>Homo sapiens: Regulation of gene expression in early pancreatic precursor cells                                   </td><td>0.007936508</td></tr>\n",
       "\t<tr><th scope=row>Homo sapiens: Regulation of gene expression in late stage (branching morphogenesis) pancreatic bud precursor cells</th><td>Homo sapiens: Regulation of gene expression in late stage (branching morphogenesis) pancreatic bud precursor cells</td><td>0.007936508</td></tr>\n",
       "\t<tr><th scope=row>Homo sapiens: RNA polymerase II transcribes snRNA genes</th><td>Homo sapiens: RNA polymerase II transcribes snRNA genes                                                           </td><td>0.015873016</td></tr>\n",
       "\t<tr><th scope=row>Homo sapiens: AKT phosphorylates targets in the nucleus</th><td>Homo sapiens: AKT phosphorylates targets in the nucleus                                                           </td><td>0.023809524</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & ID & p.value\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\tHomo sapiens: Activation of HOX genes during differentiation & Homo sapiens: Activation of HOX genes during differentiation                                                       & 0.007936508\\\\\n",
       "\tHomo sapiens: Activation of anterior HOX genes in hindbrain development during early embryogenesis & Homo sapiens: Activation of anterior HOX genes in hindbrain development during early embryogenesis                 & 0.007936508\\\\\n",
       "\tHomo sapiens: Regulation of gene expression in early pancreatic precursor cells & Homo sapiens: Regulation of gene expression in early pancreatic precursor cells                                    & 0.007936508\\\\\n",
       "\tHomo sapiens: Regulation of gene expression in late stage (branching morphogenesis) pancreatic bud precursor cells & Homo sapiens: Regulation of gene expression in late stage (branching morphogenesis) pancreatic bud precursor cells & 0.007936508\\\\\n",
       "\tHomo sapiens: RNA polymerase II transcribes snRNA genes & Homo sapiens: RNA polymerase II transcribes snRNA genes                                                            & 0.015873016\\\\\n",
       "\tHomo sapiens: AKT phosphorylates targets in the nucleus & Homo sapiens: AKT phosphorylates targets in the nucleus                                                            & 0.023809524\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | ID &lt;chr&gt; | p.value &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| Homo sapiens: Activation of HOX genes during differentiation | Homo sapiens: Activation of HOX genes during differentiation                                                       | 0.007936508 |\n",
       "| Homo sapiens: Activation of anterior HOX genes in hindbrain development during early embryogenesis | Homo sapiens: Activation of anterior HOX genes in hindbrain development during early embryogenesis                 | 0.007936508 |\n",
       "| Homo sapiens: Regulation of gene expression in early pancreatic precursor cells | Homo sapiens: Regulation of gene expression in early pancreatic precursor cells                                    | 0.007936508 |\n",
       "| Homo sapiens: Regulation of gene expression in late stage (branching morphogenesis) pancreatic bud precursor cells | Homo sapiens: Regulation of gene expression in late stage (branching morphogenesis) pancreatic bud precursor cells | 0.007936508 |\n",
       "| Homo sapiens: RNA polymerase II transcribes snRNA genes | Homo sapiens: RNA polymerase II transcribes snRNA genes                                                            | 0.015873016 |\n",
       "| Homo sapiens: AKT phosphorylates targets in the nucleus | Homo sapiens: AKT phosphorylates targets in the nucleus                                                            | 0.023809524 |\n",
       "\n"
      ],
      "text/plain": [
       "                                                                                                                   ID                                                                                                                \n",
       "Homo sapiens: Activation of HOX genes during differentiation                                                       Homo sapiens: Activation of HOX genes during differentiation                                                      \n",
       "Homo sapiens: Activation of anterior HOX genes in hindbrain development during early embryogenesis                 Homo sapiens: Activation of anterior HOX genes in hindbrain development during early embryogenesis                \n",
       "Homo sapiens: Regulation of gene expression in early pancreatic precursor cells                                    Homo sapiens: Regulation of gene expression in early pancreatic precursor cells                                   \n",
       "Homo sapiens: Regulation of gene expression in late stage (branching morphogenesis) pancreatic bud precursor cells Homo sapiens: Regulation of gene expression in late stage (branching morphogenesis) pancreatic bud precursor cells\n",
       "Homo sapiens: RNA polymerase II transcribes snRNA genes                                                            Homo sapiens: RNA polymerase II transcribes snRNA genes                                                           \n",
       "Homo sapiens: AKT phosphorylates targets in the nucleus                                                            Homo sapiens: AKT phosphorylates targets in the nucleus                                                           \n",
       "                                                                                                                   p.value    \n",
       "Homo sapiens: Activation of HOX genes during differentiation                                                       0.007936508\n",
       "Homo sapiens: Activation of anterior HOX genes in hindbrain development during early embryogenesis                 0.007936508\n",
       "Homo sapiens: Regulation of gene expression in early pancreatic precursor cells                                    0.007936508\n",
       "Homo sapiens: Regulation of gene expression in late stage (branching morphogenesis) pancreatic bud precursor cells 0.007936508\n",
       "Homo sapiens: RNA polymerase II transcribes snRNA genes                                                            0.015873016\n",
       "Homo sapiens: AKT phosphorylates targets in the nucleus                                                            0.023809524"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Loading the KEGG pathways from saved gmt file\n",
    "suppressWarnings({suppressMessages({geneset <- gmt2geneset(\"./data/REACTOME_pathways.gmt\")})})\n",
    "# perform enrichment analysis using KS test\n",
    "k_results <- .runKsWilcox(DE.df = kwR_de, statCol = \"adj.P.Val\", geneset = geneset, sTest = \"ks\")\n",
    "\n",
    "# order the results \n",
    "k_results <- k_results[order(k_results$p.value), ]\n",
    "# show some first rows of the result table\n",
    "head(k_results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9a069b0-0c79-455d-90b4-cfeaedcb4fb7",
   "metadata": {},
   "source": [
    "### Enrichment analysis using WILCOXON and GO terms\n",
    "\n",
    "For enrichment analysis with Wilcoxon test, we simply pass the value \"wilcox\" to the *sTest* argument in the `.runKsWilcox` function. We will do the same steps as KS test."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "7c3f6872-3317-43ac-bfe7-f57f991b7d56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ID</th><th scope=col>p.value</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>cellular respiration</th><td>cellular respiration                               </td><td>0.0006071947</td></tr>\n",
       "\t<tr><th scope=row>aerobic respiration</th><td>aerobic respiration                                </td><td>0.0022047165</td></tr>\n",
       "\t<tr><th scope=row>energy derivation by oxidation of organic compounds</th><td>energy derivation by oxidation of organic compounds</td><td>0.0036405380</td></tr>\n",
       "\t<tr><th scope=row>oxidative phosphorylation</th><td>oxidative phosphorylation                          </td><td>0.0041849457</td></tr>\n",
       "\t<tr><th scope=row>MAP10</th><td>MAP10                                              </td><td>0.0042461484</td></tr>\n",
       "\t<tr><th scope=row>generation of precursor metabolites and energy</th><td>generation of precursor metabolites and energy     </td><td>0.0043314488</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & ID & p.value\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\tcellular respiration & cellular respiration                                & 0.0006071947\\\\\n",
       "\taerobic respiration & aerobic respiration                                 & 0.0022047165\\\\\n",
       "\tenergy derivation by oxidation of organic compounds & energy derivation by oxidation of organic compounds & 0.0036405380\\\\\n",
       "\toxidative phosphorylation & oxidative phosphorylation                           & 0.0041849457\\\\\n",
       "\tMAP10 & MAP10                                               & 0.0042461484\\\\\n",
       "\tgeneration of precursor metabolites and energy & generation of precursor metabolites and energy      & 0.0043314488\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | ID &lt;chr&gt; | p.value &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| cellular respiration | cellular respiration                                | 0.0006071947 |\n",
       "| aerobic respiration | aerobic respiration                                 | 0.0022047165 |\n",
       "| energy derivation by oxidation of organic compounds | energy derivation by oxidation of organic compounds | 0.0036405380 |\n",
       "| oxidative phosphorylation | oxidative phosphorylation                           | 0.0041849457 |\n",
       "| MAP10 | MAP10                                               | 0.0042461484 |\n",
       "| generation of precursor metabolites and energy | generation of precursor metabolites and energy      | 0.0043314488 |\n",
       "\n"
      ],
      "text/plain": [
       "                                                    ID                                                 \n",
       "cellular respiration                                cellular respiration                               \n",
       "aerobic respiration                                 aerobic respiration                                \n",
       "energy derivation by oxidation of organic compounds energy derivation by oxidation of organic compounds\n",
       "oxidative phosphorylation                           oxidative phosphorylation                          \n",
       "MAP10                                               MAP10                                              \n",
       "generation of precursor metabolites and energy      generation of precursor metabolites and energy     \n",
       "                                                    p.value     \n",
       "cellular respiration                                0.0006071947\n",
       "aerobic respiration                                 0.0022047165\n",
       "energy derivation by oxidation of organic compounds 0.0036405380\n",
       "oxidative phosphorylation                           0.0041849457\n",
       "MAP10                                               0.0042461484\n",
       "generation of precursor metabolites and energy      0.0043314488"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load the GO gene sets from saved gmt file\n",
    "suppressWarnings({suppressMessages({geneset <- gmt2geneset(\"./data/GO_terms.gmt\")})})\n",
    "# perform enrichment analysis with Wilcoxon test \n",
    "w_results <- .runKsWilcox(DE.df = kw_de, statCol = \"adj.P.Val\", geneset = geneset, sTest = \"wilcox\")\n",
    "\n",
    "# order the result  \n",
    "w_results <- w_results[order(w_results$p.value), ]\n",
    "# show some first rows of the result\n",
    "head(w_results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b13fabf-227e-47d1-ad29-1cd5fbf81de7",
   "metadata": {},
   "source": [
    "### Enrichment analysis using WILCOXON and KEGG pathways\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "bc304aaf-85f9-412a-8fb4-c0e6b6a391ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ID</th><th scope=col>p.value</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>COX5A</th><td>COX5A                                                     </td><td>0.02472462</td></tr>\n",
       "\t<tr><th scope=row>SDHC</th><td>SDHC                                                      </td><td>0.02472462</td></tr>\n",
       "\t<tr><th scope=row>SDHC.1</th><td>SDHC.1                                                    </td><td>0.02472462</td></tr>\n",
       "\t<tr><th scope=row>COX2</th><td>COX2                                                      </td><td>0.02472462</td></tr>\n",
       "\t<tr><th scope=row>CYC1</th><td>CYC1                                                      </td><td>0.02472462</td></tr>\n",
       "\t<tr><th scope=row>Complement and coagulation cascades - Homo sapiens (human)</th><td>Complement and coagulation cascades - Homo sapiens (human)</td><td>0.02692431</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & ID & p.value\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\tCOX5A & COX5A                                                      & 0.02472462\\\\\n",
       "\tSDHC & SDHC                                                       & 0.02472462\\\\\n",
       "\tSDHC.1 & SDHC.1                                                     & 0.02472462\\\\\n",
       "\tCOX2 & COX2                                                       & 0.02472462\\\\\n",
       "\tCYC1 & CYC1                                                       & 0.02472462\\\\\n",
       "\tComplement and coagulation cascades - Homo sapiens (human) & Complement and coagulation cascades - Homo sapiens (human) & 0.02692431\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | ID &lt;chr&gt; | p.value &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| COX5A | COX5A                                                      | 0.02472462 |\n",
       "| SDHC | SDHC                                                       | 0.02472462 |\n",
       "| SDHC.1 | SDHC.1                                                     | 0.02472462 |\n",
       "| COX2 | COX2                                                       | 0.02472462 |\n",
       "| CYC1 | CYC1                                                       | 0.02472462 |\n",
       "| Complement and coagulation cascades - Homo sapiens (human) | Complement and coagulation cascades - Homo sapiens (human) | 0.02692431 |\n",
       "\n"
      ],
      "text/plain": [
       "                                                           ID                                                        \n",
       "COX5A                                                      COX5A                                                     \n",
       "SDHC                                                       SDHC                                                      \n",
       "SDHC.1                                                     SDHC.1                                                    \n",
       "COX2                                                       COX2                                                      \n",
       "CYC1                                                       CYC1                                                      \n",
       "Complement and coagulation cascades - Homo sapiens (human) Complement and coagulation cascades - Homo sapiens (human)\n",
       "                                                           p.value   \n",
       "COX5A                                                      0.02472462\n",
       "SDHC                                                       0.02472462\n",
       "SDHC.1                                                     0.02472462\n",
       "COX2                                                       0.02472462\n",
       "CYC1                                                       0.02472462\n",
       "Complement and coagulation cascades - Homo sapiens (human) 0.02692431"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Loading the KEGG pathways from saved gmt file\n",
    "suppressWarnings({suppressMessages({geneset <- gmt2geneset(\"./data/KEGG_pathways.gmt\")})})\n",
    "# perform enrichment analysis with Wilcoxon test \n",
    "w_results <- .runKsWilcox(DE.df = kw_de, statCol = \"adj.P.Val\", geneset = geneset, sTest = \"wilcox\")\n",
    "\n",
    "# order the result  \n",
    "w_results <- w_results[order(w_results$p.value), ]\n",
    "# show some first rows of the result\n",
    "head(w_results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "402f89b4-83d0-4dbb-af54-265cfb4cf314",
   "metadata": {},
   "source": [
    "### Enrichment analysis using WILCOXON and REACTOME pathways\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "b5641e20-85d2-4ca8-9e66-4a1aabeb216b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ID</th><th scope=col>p.value</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1329</th><td>1329                                                                                              </td><td>0.03413802</td></tr>\n",
       "\t<tr><th scope=row>84271</th><td>84271                                                                                             </td><td>0.06002829</td></tr>\n",
       "\t<tr><th scope=row>Homo sapiens: Respiratory electron transport</th><td>Homo sapiens: Respiratory electron transport                                                      </td><td>0.06636071</td></tr>\n",
       "\t<tr><th scope=row>Homo sapiens: Complex I biogenesis</th><td>Homo sapiens: Complex I biogenesis                                                                </td><td>0.08040866</td></tr>\n",
       "\t<tr><th scope=row>Homo sapiens: Activation of HOX genes during differentiation</th><td>Homo sapiens: Activation of HOX genes during differentiation                                      </td><td>0.08573066</td></tr>\n",
       "\t<tr><th scope=row>Homo sapiens: Activation of anterior HOX genes in hindbrain development during early embryogenesis</th><td>Homo sapiens: Activation of anterior HOX genes in hindbrain development during early embryogenesis</td><td>0.08573066</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & ID & p.value\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1329 & 1329                                                                                               & 0.03413802\\\\\n",
       "\t84271 & 84271                                                                                              & 0.06002829\\\\\n",
       "\tHomo sapiens: Respiratory electron transport & Homo sapiens: Respiratory electron transport                                                       & 0.06636071\\\\\n",
       "\tHomo sapiens: Complex I biogenesis & Homo sapiens: Complex I biogenesis                                                                 & 0.08040866\\\\\n",
       "\tHomo sapiens: Activation of HOX genes during differentiation & Homo sapiens: Activation of HOX genes during differentiation                                       & 0.08573066\\\\\n",
       "\tHomo sapiens: Activation of anterior HOX genes in hindbrain development during early embryogenesis & Homo sapiens: Activation of anterior HOX genes in hindbrain development during early embryogenesis & 0.08573066\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | ID &lt;chr&gt; | p.value &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| 1329 | 1329                                                                                               | 0.03413802 |\n",
       "| 84271 | 84271                                                                                              | 0.06002829 |\n",
       "| Homo sapiens: Respiratory electron transport | Homo sapiens: Respiratory electron transport                                                       | 0.06636071 |\n",
       "| Homo sapiens: Complex I biogenesis | Homo sapiens: Complex I biogenesis                                                                 | 0.08040866 |\n",
       "| Homo sapiens: Activation of HOX genes during differentiation | Homo sapiens: Activation of HOX genes during differentiation                                       | 0.08573066 |\n",
       "| Homo sapiens: Activation of anterior HOX genes in hindbrain development during early embryogenesis | Homo sapiens: Activation of anterior HOX genes in hindbrain development during early embryogenesis | 0.08573066 |\n",
       "\n"
      ],
      "text/plain": [
       "                                                                                                   ID                                                                                                \n",
       "1329                                                                                               1329                                                                                              \n",
       "84271                                                                                              84271                                                                                             \n",
       "Homo sapiens: Respiratory electron transport                                                       Homo sapiens: Respiratory electron transport                                                      \n",
       "Homo sapiens: Complex I biogenesis                                                                 Homo sapiens: Complex I biogenesis                                                                \n",
       "Homo sapiens: Activation of HOX genes during differentiation                                       Homo sapiens: Activation of HOX genes during differentiation                                      \n",
       "Homo sapiens: Activation of anterior HOX genes in hindbrain development during early embryogenesis Homo sapiens: Activation of anterior HOX genes in hindbrain development during early embryogenesis\n",
       "                                                                                                   p.value   \n",
       "1329                                                                                               0.03413802\n",
       "84271                                                                                              0.06002829\n",
       "Homo sapiens: Respiratory electron transport                                                       0.06636071\n",
       "Homo sapiens: Complex I biogenesis                                                                 0.08040866\n",
       "Homo sapiens: Activation of HOX genes during differentiation                                       0.08573066\n",
       "Homo sapiens: Activation of anterior HOX genes in hindbrain development during early embryogenesis 0.08573066"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Loading the KEGG pathways from saved gmt file\n",
    "suppressWarnings({suppressMessages({geneset <- gmt2geneset(\"./data/REACTOME_pathways.gmt\")})})\n",
    "# perform enrichment analysis using KS test\n",
    "k_results <- .runKsWilcox(DE.df = kwR_de, statCol = \"adj.P.Val\", geneset = geneset, sTest = \"wilcox\")\n",
    "\n",
    "# order the results \n",
    "k_results <- k_results[order(k_results$p.value), ]\n",
    "# show some first rows of the result table\n",
    "head(k_results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9aa0e1c7",
   "metadata": {},
   "source": [
    "## Enrichment Analysis using FGSEA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47b46d66",
   "metadata": {},
   "source": [
    "This submodule describes `FGSEA`, one of the methods for evaluating pathway enrichment in transcriptional data and it stands for Fast preranked Gene Set Enrichment Analysis (GSEA). `FGSEA` quickly calculates arbitrarily low GSEA P-values for a collection of gene sets and is based on an algorithm that allows  it to make more permutations and get accurate *p-values*.\n",
    "It extends the GSEA algorithm which calculates the enrichment score (the degree to which a set S is over-represented at the top or bottom of the ranked list L), estimates the p-value of the Enrichment Score using a permutation test and adjusts the estimated significance level to account for multiple hypotheses in addition to calculating the q-values for FDR control. Typically, GSEA requires inputs of a list of gene sets (GO term or pathway with a set of genes), a vector DE genes with statistic. In addition, users can limit the number of terms or pathways of interest by adjusting the `minSize` and `maxSize` parameters. The code to install and load the FGSEA R package is shown below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "b9702a89",
   "metadata": {},
   "outputs": [],
   "source": [
    "# To perform enrichment analysis using FGSEA, we first install the fgsea package\n",
    "suppressMessages({\n",
    "    if (!require(\"BiocManager\", quietly = TRUE)) {\n",
    "        install.packages(\"BiocManager\")\n",
    "        suppressWarnings(BiocManager::install(\"fgsea\", update = F))\n",
    "    }\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "52f7aae1-0bd8-4ff4-86f2-041eb6e41db4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading the package\n",
    "suppressPackageStartupMessages({\n",
    "    library(\"fgsea\")\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "cdb41931-b689-4fb9-9c5d-0e1c443bfe00",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<iframe src=\"./Quizzes/Quiz_Submodule4-3.html\" width=100% height=250></iframe>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Run the following command to take the quiz\n",
    "IRdisplay::display_html('<iframe src=\"./Quizzes/Quiz_Submodule4-3.html\" width=100% height=250></iframe>')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb6f9407",
   "metadata": {},
   "source": [
    "Next, we need to prepare a vector of DE genes with its statistic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "38d25313",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>DDR1</dt><dd>0.25312654919025</dd><dt>RFC2</dt><dd>0.467472580609517</dd><dt>HSPA6</dt><dd>0.986153555292704</dd><dt>PAX8</dt><dd>0.723652752171903</dd><dt>GUCA1A</dt><dd>0.335147977832318</dd><dt>UBA7</dt><dd>0.392525231374342</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[DDR1] 0.25312654919025\n",
       "\\item[RFC2] 0.467472580609517\n",
       "\\item[HSPA6] 0.986153555292704\n",
       "\\item[PAX8] 0.723652752171903\n",
       "\\item[GUCA1A] 0.335147977832318\n",
       "\\item[UBA7] 0.392525231374342\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "DDR1\n",
       ":   0.25312654919025RFC2\n",
       ":   0.467472580609517HSPA6\n",
       ":   0.986153555292704PAX8\n",
       ":   0.723652752171903GUCA1A\n",
       ":   0.335147977832318UBA7\n",
       ":   0.392525231374342\n",
       "\n"
      ],
      "text/plain": [
       "     DDR1      RFC2     HSPA6      PAX8    GUCA1A      UBA7 \n",
       "0.2531265 0.4674726 0.9861536 0.7236528 0.3351480 0.3925252 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Get the gene list and their statistic from DE results\n",
    "stats = DE.df$adj.P.Val\n",
    "names(stats) = DE.df$SYMBOL\n",
    "head(stats)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9950a6f6",
   "metadata": {},
   "source": [
    "### Enrichment analysis using FGSEA and GO terms"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5093c0c",
   "metadata": {},
   "source": [
    "The package `fgsea` has its own function named `gmtPathways` to load the genesets from the `gmt` file. It is recommended to use this function for compatibility. We can load the GO terms geneset and view the first five GO terms with associated genes using the following command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "b2d3d20a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$`GO:0000002`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'AKT3'</li><li>'DNA2'</li><li>'DNAJA3'</li><li>'ENDOG'</li><li>'FLCN'</li><li>'LIG3'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$`GO:0000003`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'A1CF'</li><li>'A2M'</li><li>'AAAS'</li><li>'ABAT'</li><li>'ABCC8'</li><li>'ABHD2'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$`GO:0000012`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'APLF'</li><li>'APTX'</li><li>'ERCC6'</li><li>'ERCC8'</li><li>'LIG4'</li><li>'PARP1'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$`GO:0000018`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ABL1'</li><li>'ACTB'</li><li>'ACTL6A'</li><li>'ACTR2'</li><li>'ALYREF'</li><li>'ANKLE1'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$`GO:0000022`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'AURKB'</li><li>'AURKC'</li><li>'BIRC5'</li><li>'CDCA8'</li><li>'INCENP'</li><li>'KIF23'</li></ol>\n",
       "</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$`GO:0000002`] \\begin{enumerate*}\n",
       "\\item 'AKT3'\n",
       "\\item 'DNA2'\n",
       "\\item 'DNAJA3'\n",
       "\\item 'ENDOG'\n",
       "\\item 'FLCN'\n",
       "\\item 'LIG3'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$`GO:0000003`] \\begin{enumerate*}\n",
       "\\item 'A1CF'\n",
       "\\item 'A2M'\n",
       "\\item 'AAAS'\n",
       "\\item 'ABAT'\n",
       "\\item 'ABCC8'\n",
       "\\item 'ABHD2'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$`GO:0000012`] \\begin{enumerate*}\n",
       "\\item 'APLF'\n",
       "\\item 'APTX'\n",
       "\\item 'ERCC6'\n",
       "\\item 'ERCC8'\n",
       "\\item 'LIG4'\n",
       "\\item 'PARP1'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$`GO:0000018`] \\begin{enumerate*}\n",
       "\\item 'ABL1'\n",
       "\\item 'ACTB'\n",
       "\\item 'ACTL6A'\n",
       "\\item 'ACTR2'\n",
       "\\item 'ALYREF'\n",
       "\\item 'ANKLE1'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$`GO:0000022`] \\begin{enumerate*}\n",
       "\\item 'AURKB'\n",
       "\\item 'AURKC'\n",
       "\\item 'BIRC5'\n",
       "\\item 'CDCA8'\n",
       "\\item 'INCENP'\n",
       "\\item 'KIF23'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$`GO:0000002`\n",
       ":   1. 'AKT3'\n",
       "2. 'DNA2'\n",
       "3. 'DNAJA3'\n",
       "4. 'ENDOG'\n",
       "5. 'FLCN'\n",
       "6. 'LIG3'\n",
       "\n",
       "\n",
       "\n",
       "$`GO:0000003`\n",
       ":   1. 'A1CF'\n",
       "2. 'A2M'\n",
       "3. 'AAAS'\n",
       "4. 'ABAT'\n",
       "5. 'ABCC8'\n",
       "6. 'ABHD2'\n",
       "\n",
       "\n",
       "\n",
       "$`GO:0000012`\n",
       ":   1. 'APLF'\n",
       "2. 'APTX'\n",
       "3. 'ERCC6'\n",
       "4. 'ERCC8'\n",
       "5. 'LIG4'\n",
       "6. 'PARP1'\n",
       "\n",
       "\n",
       "\n",
       "$`GO:0000018`\n",
       ":   1. 'ABL1'\n",
       "2. 'ACTB'\n",
       "3. 'ACTL6A'\n",
       "4. 'ACTR2'\n",
       "5. 'ALYREF'\n",
       "6. 'ANKLE1'\n",
       "\n",
       "\n",
       "\n",
       "$`GO:0000022`\n",
       ":   1. 'AURKB'\n",
       "2. 'AURKC'\n",
       "3. 'BIRC5'\n",
       "4. 'CDCA8'\n",
       "5. 'INCENP'\n",
       "6. 'KIF23'\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$`GO:0000002`\n",
       "[1] \"AKT3\"   \"DNA2\"   \"DNAJA3\" \"ENDOG\"  \"FLCN\"   \"LIG3\"  \n",
       "\n",
       "$`GO:0000003`\n",
       "[1] \"A1CF\"  \"A2M\"   \"AAAS\"  \"ABAT\"  \"ABCC8\" \"ABHD2\"\n",
       "\n",
       "$`GO:0000012`\n",
       "[1] \"APLF\"  \"APTX\"  \"ERCC6\" \"ERCC8\" \"LIG4\"  \"PARP1\"\n",
       "\n",
       "$`GO:0000018`\n",
       "[1] \"ABL1\"   \"ACTB\"   \"ACTL6A\" \"ACTR2\"  \"ALYREF\" \"ANKLE1\"\n",
       "\n",
       "$`GO:0000022`\n",
       "[1] \"AURKB\"  \"AURKC\"  \"BIRC5\"  \"CDCA8\"  \"INCENP\" \"KIF23\" \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load the pathways into a named list\n",
    "GO_term_hallmark <- gmtPathways(\"./data/GO_terms.gmt\")\n",
    "# Show the first few GO terms, and within those, show only the first few genes.\n",
    "tmp = lapply(GO_term_hallmark, head)\n",
    "tmp[1:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d06027d9",
   "metadata": {},
   "source": [
    "Running the FGSEA can be done by calling the `fgsea` with two required inputs that we created from the two previous steps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6efd9f83",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Running fgsea\n",
    "suppressWarnings(\n",
    "    fgseaRes <- fgsea(pathways = GO_term_hallmark, stats = stats)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb9dba30-e8a2-449c-ba72-4fc3f1db04bb",
   "metadata": {},
   "source": [
    "To use the `fgsea` function, we need to pass the following arguments:\n",
    "\n",
    "- `pathways`: A list of gene sets to check.\n",
    "- `stats`: A named vector of gene-level stats. Names should be the same as in 'pathways'.\n",
    "\n",
    "The `fgsea` function returns a dataframe with GSEA results. Each row corresponds to a tested pathway.\r\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad26fa54",
   "metadata": {},
   "source": [
    "The output of the FGSEA method is a table where rows are GO terms IDs or pathway IDs. The columns are *p-values*, *adjusted p-values*, *Enrichment Score*, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "f90ac106",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>pathway</th><th scope=col>pval</th><th scope=col>padj</th><th scope=col>log2err</th><th scope=col>ES</th><th scope=col>NES</th><th scope=col>size</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>GO:0006122</td><td>8.754388e-06</td><td>0.06000257</td><td>0.5933255</td><td>-0.7106070</td><td>-2.620884</td><td> 11</td></tr>\n",
       "\t<tr><td>GO:0060390</td><td>1.829102e-05</td><td>0.06268331</td><td>0.5756103</td><td> 0.4911948</td><td> 1.862244</td><td> 46</td></tr>\n",
       "\t<tr><td>GO:0050907</td><td>4.798773e-05</td><td>0.10963597</td><td>0.5573322</td><td> 0.3756389</td><td> 1.551850</td><td>128</td></tr>\n",
       "\t<tr><td>GO:0030850</td><td>1.099860e-04</td><td>0.18846105</td><td>0.5384341</td><td> 0.4547892</td><td> 1.743769</td><td> 51</td></tr>\n",
       "\t<tr><td>GO:0007606</td><td>1.802479e-04</td><td>0.22144289</td><td>0.5188481</td><td> 0.3392928</td><td> 1.432718</td><td>178</td></tr>\n",
       "\t<tr><td>GO:0009593</td><td>1.938514e-04</td><td>0.22144289</td><td>0.5188481</td><td> 0.3497537</td><td> 1.463237</td><td>159</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 7\n",
       "\\begin{tabular}{lllllll}\n",
       " pathway & pval & padj & log2err & ES & NES & size\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <int>\\\\\n",
       "\\hline\n",
       "\t GO:0006122 & 8.754388e-06 & 0.06000257 & 0.5933255 & -0.7106070 & -2.620884 &  11\\\\\n",
       "\t GO:0060390 & 1.829102e-05 & 0.06268331 & 0.5756103 &  0.4911948 &  1.862244 &  46\\\\\n",
       "\t GO:0050907 & 4.798773e-05 & 0.10963597 & 0.5573322 &  0.3756389 &  1.551850 & 128\\\\\n",
       "\t GO:0030850 & 1.099860e-04 & 0.18846105 & 0.5384341 &  0.4547892 &  1.743769 &  51\\\\\n",
       "\t GO:0007606 & 1.802479e-04 & 0.22144289 & 0.5188481 &  0.3392928 &  1.432718 & 178\\\\\n",
       "\t GO:0009593 & 1.938514e-04 & 0.22144289 & 0.5188481 &  0.3497537 &  1.463237 & 159\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 7\n",
       "\n",
       "| pathway &lt;chr&gt; | pval &lt;dbl&gt; | padj &lt;dbl&gt; | log2err &lt;dbl&gt; | ES &lt;dbl&gt; | NES &lt;dbl&gt; | size &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| GO:0006122 | 8.754388e-06 | 0.06000257 | 0.5933255 | -0.7106070 | -2.620884 |  11 |\n",
       "| GO:0060390 | 1.829102e-05 | 0.06268331 | 0.5756103 |  0.4911948 |  1.862244 |  46 |\n",
       "| GO:0050907 | 4.798773e-05 | 0.10963597 | 0.5573322 |  0.3756389 |  1.551850 | 128 |\n",
       "| GO:0030850 | 1.099860e-04 | 0.18846105 | 0.5384341 |  0.4547892 |  1.743769 |  51 |\n",
       "| GO:0007606 | 1.802479e-04 | 0.22144289 | 0.5188481 |  0.3392928 |  1.432718 | 178 |\n",
       "| GO:0009593 | 1.938514e-04 | 0.22144289 | 0.5188481 |  0.3497537 |  1.463237 | 159 |\n",
       "\n"
      ],
      "text/plain": [
       "  pathway    pval         padj       log2err   ES         NES       size\n",
       "1 GO:0006122 8.754388e-06 0.06000257 0.5933255 -0.7106070 -2.620884  11 \n",
       "2 GO:0060390 1.829102e-05 0.06268331 0.5756103  0.4911948  1.862244  46 \n",
       "3 GO:0050907 4.798773e-05 0.10963597 0.5573322  0.3756389  1.551850 128 \n",
       "4 GO:0030850 1.099860e-04 0.18846105 0.5384341  0.4547892  1.743769  51 \n",
       "5 GO:0007606 1.802479e-04 0.22144289 0.5188481  0.3392928  1.432718 178 \n",
       "6 GO:0009593 1.938514e-04 0.22144289 0.5188481  0.3497537  1.463237 159 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(fgseaRes[order(pval), ][, -8])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7866448",
   "metadata": {},
   "source": [
    "From the result table, we can select top five up regulated GO terms and top five down regulated GO terms. Then we can plot them using the built-in function `plotGseaTable`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "f1b4ed14",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAM1BMVEUAAAAzMzNNTU1oaGh8\nfHyMjIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enw8PD////p0TvgAAAACXBIWXMAABJ0\nAAASdAHeZh94AAAgAElEQVR4nO2di9ayKtdA2R7SMg/3f7VbwAMgmvWuHivnHP//PoawtAVT\nkfzGVh0A/DPq6BMA+AUQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEA\nBEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAA\nRAIQAJFWUYbk0gTlld15wBm5tNe8P7v8KhTund9Hqco5iJrQRW2RKpVJfYu3fQ8ddjv00ePh\ngxn7O/FNSpXdecg5TVRJ/Oxe5b0iJc5BPJHa4WskrdixhAItwyLSiwyXzEwVkeKDRaqUutz7\nv/d8HKT/yHtFUuV8EO9IF5X1F4ImzPG/HEso0LNhEWmVIXltMFQ/QqRknC31Q1FkXvRekVLV\nTAfxjqSUuRW1YodHpM9jTJ75W/VPJEnRDTMTU1iopOzu6qLrVHZgX9R9qtmq1DQf/0pyU/m4\n2RT6ztRdU5Vc7dk2uT4xr3D8Im2qG07fxalsv2yhbx1V1j+1VJ0cStX2fGMiPR9MJ97Jb9q7\n6H6j189yCNy5GepzkvS3S6Z2/4B7RyrtVL5wRNLP+v3dIDG1LnZu0leda+bKDPHbMK+RZAjt\nFmgyc9bmsaP0CscvlOsTm8/QqWy+bKErX+1uued/HftiTngpUqEWazkPg5XDt8rMba7pN71v\n9PpZjoHdeP1RlMkbIr2MTZydvyt100rMA6HPeNuPubRP+q0bn6e1M3PNyt6sLkpmOWB5ajOV\nPpv+aa5yTswtHFtlbRd8l7Gy/mg86q8Mtd4teBvtY9vbh7/YYPb14zQt7pvNw2BJ3dVJ/x3s\nFarsv9+id147yzGwG+82lCLS60yrdu1c0s0iDZdYfUXsJ3i5Hn7Z5IyplJoHgDfM7MY+nUZk\nbo7U6gnUdGJu4djqHoSYK5uJTWZ3SE7rxkNd9S0uIlJXXXSK9x/Snl5lvqrObKrmPd2/iTQF\nduLZe3+FSP+A/ztSU5WZJ9L4b9aP16J/CiitU07Nq75k3t8ws1uI5AzO+cSCEeuMhMh36e9N\ng2dFP5Opa/Gz1VeVlaF+L5NwqvogmP2j7/WNnYAF3+jls5zmIUE8RPoHvMRl05gMRap6U5K0\nS8dZ3lxzeLqSn9l5z0jPixT7LvqSMdw7S/3cJPTz1HxcvSqzOtTr/bdtZ2Tra1RhMhF+o5fP\ncviziIdI/4CbuItKr1UTFamfYdz7y2Khl8Rar2ZfVmnB5LnZx6/pNJ1Tdd2If6Hod+lvR7V5\nNNBURSr8jNRp+evFUJ+2949/d7z3169En+fiG71+lsOtLoyHSP/AosdXRCrUpRem6v+9+DX7\nC21Wv2Nm5/6O1NrHIWdFYfw3Dx923DG4FEnfPJOwrgjDbEmli6GeD4uD4W91m8HsQ4vOdaHM\n7Hn5jV47yymwE88m8Y5I/4Av0r2rx1lz03ki9VnWtyJln1admvrBIHnHzM48/ebmzYbCrBea\ntaX+mSx3T8wp9L6Q913G8uE5ptT/3ORX7TTlcvLVp+7ap+6e7V9tHxfXdK4bm/nlN3rtLKfA\nTryKVbt/xk1cMTxu3PVA00PXvS/ZeVBmr6pOTTPe3zGz67mP79rZSZ6d1OsHG+fE5kLvCzln\nGIhU63F5c85fhjGRyTDUnWe38VyyrfZBMPvLjvmQ2oaLb/TaWU6B3T40P8ZdEOkf8BLXpzK7\nm7XRexqKVJqVo3J4X2yuaeZdb5nZaW55PzCzctDk2vt9CW6Vc6H/heYzDB+oSvPig36zQdCj\n6bjVUqSuvuhvcXsqWN4/wtgPt2H2Gn6jF89yCuz2YcmbDcdTqffM7E6M5MPbc4ER6TiemPvD\nPg4TqUGko3hq7g/7OEqkQjk/OcSai54MuCTzghlIcZRIuflfTm00Fz0ZgJOCSAACIBKAAIgE\nIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAAC\nIBKAAIgEIAAiAQiASAACIBKAAIh0MMN/osH8//CfUxj+oxHOv/Dx0FEHs0ckp0qHYZ8J/XEw\nr4ukxsbwAdARB4NIvwEdcTBvFYne/TNI9cH8gUj08R9Akg/mj0Wiv98EiT2Yg0Si34UhoQeD\nSL8BCT0YRPoNSOjBfJRIivHwKiTuYD5PJIbEK5C1g/lkkRgc+yFXB4NIvwG5OphvEOl7B4lS\nb3qLahH3e3P0IyDSO3nXmTsd8eYjwU6+TqRvGjFvO1dE+ji+V6QvGDpvm9nZ6Ksf4O9BpDey\nvHGIR49+gL/nJ0RSYaVP4q3PSW8/ymfTXvN+DObX6XORKpUWrVepSFQyFDmb9UWpSxOUtrqw\nnppl1SLCKr8jktoaTVdvVxtP6NNtl1XGnc6cbvWkwiO7vbgaeQochD2lSFWiLIlN4234qCqn\nUmZK0mCzsu1av9TGq+e6ZdBsnR8TaeV8au9Em2TOvpPQFdbbLqsEKd+c2i2O7PTiw8iLqGcU\nqc/g5d7/vecqGT4Xfb80hWvSXSV1Vyfq7m12Sb/Z5qrwSgt10f/knb6AZa2+tNV+s3VOIVKf\nBXfHRefPZs1J6AobbRdVwpRvTjnDIzu9uC+yxxlFSiZfLqqf3bWTP5VzfSpM4U3fW5zNm8l7\nqwV0ShOlm5mUZybVja7mVNjgDCL1VxfvRMfJkfISGmej7bLKIuUbq3aLIzu9+DCyWqwHnlCk\n23zRaYp+2JfuRUk/NpkM5UpPHmpd19m8zHd+p9RiumTs5yxSIcovirSY9hT++EyG80q8hK58\nhfW2yyr7Um5ZObKn9BORTyhSHky28jmjd5MoNY6L4Y+zmaquTNSl9SsYHAkXzTY4g0h1kIVy\nmJ6VXkLjbLRdVtmXckv8yLYXX4h8QpHCZLif5+24EUrl5gm1C1N7U/a+lpor1x2Rwshe2VU/\n1yf2wjMldP1brLRdVnlGpNiRx158IfJ5RVLjPPc5kfQT6kVfD/3UXvPEXCRLlbddnSFSGNkr\nK6elTSeh699ipW28txYNVqMujzz24guREelJkfQ0sNHLoIvUmpULu4aaI1IY2S276ut+a/Ll\nJHT9W6y0/VeRokd+NfIJRXKfkXRisvkZyXmYTObMOZsqWmqxC0B9HydlvEKUM4qUmvWxNno5\nip3VSttllX0p95qEdcMFxL2RTyjSzfkNQifGW7WbbvR2maaZV+2GzamZU+oEs9S6n5cVYpxR\nJOVmcbl7b9tl6b6Ud0Pd+JFfjHxCkZzfkVqdpvjvSKUprLRki81Gr247pfYXCDNLsJtXnXCn\nwgZnFMle3s2130no+rdYabussi/lnVPXObLTiy9EPqNIVf8QY95sKOyijfNmw22qFH+zoU+z\neXHhtnyzoc317Nps3tOgAngyFEq/DlfoUekk9Om2yyrPpHxxZKcXX4h8RpG6+/iu3TDJq8aP\nNqk2d6kpyYLNMlqaTJut3cyDCqdnlMH+zaIJfbrtssozKXeObJsn/xL5lCL1z0l5n7WsHN98\nbEv99nc5zOts7uxbxl2w2VVZrLTfTO2VrLn0GlWLCmfHl6GLJvT5tosqT6V8PvIcOQ1+oNod\n+aQiAciCSAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAAC\nIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAi\nAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIh0IEr/J0qHf+x/q9RuWew++A7oqgPZI9JUMu7t\nhs1jTx0C6I8D+TeR5nuWLacvj4TkH4i0SGMBc8K/h4wfyBtFMjco+w9d/BeQ5QP5O5EUt6k3\nQ3IP5K9F4lnqfZDUAzlQJG5PwpDPAzlaJNV1jAAhSOOBfIxIjIJ/hhQeyKeJpBgOL0PmDuQj\nRVIsR7wCKTuQTxaJgfEc5OtAEOl3IF8H8kUifenj05i5d4Te/Ah/yZeJ9K4h+Ubed8ZhMr4v\nNz/EV4r0TdO+N3rEHemDQKQ388aZHSJ9EF8v0qePnukLviv26kf4SxDpL0Ck99Fe835Q5dfp\nc5EqlRatV6lIVDIUzZvjIF9WyKqgmVJu5Ri/J9Lqd716e5zEPMqR5u7ubp20L8OPO52YG7Gv\n4b5FwSLyFBiRuq5Khu5LGvP5No7dyqmUmZLU36ydfl9UKP3SMWiyeh7nEan2VHGy6CZ0jTZx\ndjeJ23NheKdPptNZj12Hx10UrEZGpJ5Kqcu9/3vP7SDvPxd9vzSFa9JdJXVXJ+rubdYqj1S4\nqqzt2ouqvdLxYPdujdOI1CfEFymPbK6Su40vquj/LdQlFj7MvvPdHp9VpGA9MiL1JJMvF9XP\n7trJn0ol05ShMIU3fZtxNq/2thNUyEx+G93FTqmhTTYGyllE6i803gh1suhsrnHzbljjxMop\nmsOH2XdneA/PalmwFRmR+mxMY7spegNKc40zFFos20m50pMHc8F0Nq9qeq5ySsfOzbzSoVYw\nm3f5VZHC4j7BgUjXyOYKjT+8h2meO2Gew4fZ3yQ8q2XBM5FPKFIeTLZyPSWz3E2i1DTShmE1\nb+aquvTPnN1qheCCWc+SRjiLSLV/B3Gz6GyukKnGbVwOUzvnRlaHaV+/C22d1bLgmcgnFClM\nRmTi0HmZ80QyZH6F1Fyu7hGRNm9IpxEpSLKbRWczTqluwe1MrzYkwX3sFZFiFZctEWmVKSfD\n+HxGJNV3a9eaGaBTWqq87epsKVLtPRMvz+SkInlZnDajuFNnS2m+R+kHlRbp+ciI9JRIllYv\ng7qlZkk2X9YtvAX15ZmcVCQni4vNgDRp/cZXPbVrL/PDrBsekf4S9xlJJyabn5Gch8lkzlyy\nSGJY2ndsUkbqJtvpPbdIYUJjXMyVyN2bmslyaJ6f738X6fnIJxTp5sy37MTMWbWbnmHtMk0z\nr9o1zoqNbrYorXXneqWPlo8QaWv38M2mmYNXM2gwfFz21Db7RXoY+YQiOb8jtTpN8d+RSlNY\nacmczcRcD00+F6XXoPTx2u5ZRXKy6GxGM7QQyd4a2uB9kWG/l/0d7BfpYeQzilT1zzPmzYbC\nvr/jvNlwmyrF32wozAzdPPp4pf097p7q1t4v4M7C+tnxRqiTRWdzX+O+QTs0W9ZYvn/wxFlF\nC3ZHPqNI3X18126Y5FXjR+uRzV1qSjJ/s7Uti2hpHjQb5vPQzSPUPlLOWXQT+qix/ZM5CV6E\nT2M7d56Vt/F05FOK1D8n5X0HZuX45mNb6re/y2HY29zZt4y7yGZ6DUubS69RFZbuf+j9ffwh\nG2QxffB2g/9c5CR4Eb6N7dx7Vt2WSI8i09UAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEI\ngEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACI\nBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIdBzk/oegM49D\njdmf/xOpyhYPf+id74GuOo6dIpnPC6nouM+C/jiOfxVJzd1HNx4NPXAcbxSJbv1ryPhxINIP\nQcaPA5F+CDJ+HH8uEp39PsjtcRwpEv0uDAk9DkT6IUjocXyOSPz0+8+QwOP4LJEYCf8E6TuO\nDxVJrbaHdcjYcXy6SIyNJyBZx/EVIjFA9kGejuOLRPriYfKuU1/2BxzEN4r0deNFvemMw7hf\nl5gf4itFetfAfBPvWo3kf9byQXyxSLEz+kTU+0Y4In0MXy/SG4epEIh0BhDp3Uxf5U2x1z+e\nhPaa9yMiv06fi1SptGi9SkWikqFo3mwvSl3qsEI7bypLGCHGmUS6uhWVcrPUdfftKNdw96Jg\nLhlTbqI7X2BX5L5xVj2IPJ02InVdlQwdmTTm823sWDeLmSlJg03bsvZLm2SKVjtDxGkW5RdF\nWhlPtbdGMaY7sR/bZHMU+m1jBXOJn3Jf14eRbeNyT+QOkXqq/qZy7//ec9uV/eeid6ApXJPu\nKqm7OlF3b7NQF/1P7le4qKIb9tVmXxghznlE6rOw3FGNmck3VwIXbZfBppJYyjc88gNdVdbq\nGUe9LzIi9XeV0ZeL6md37eRPpZJpIlaYwpu+QDmbidIVTHKd0vFmr3RvlJEIcX5VpMjEK1u6\n0ibDJee2uaS+aLsMNpfEUr4aOwyUGUsac1HcERmRbvNNoyn63JVz6gotlh3HudLTPnODcTYt\n5kbmlA5TE118VdOD16JZyGlE6hO8dCVX9qLVxCzbaLsMNpc8TPlWoPFymL0U+YQi5cFkK59v\n5neTKDuOndE9bxoc3eyfcpjalTpYdekfSrtu2WzBaUSqI1mox8tXppotkRZtl8HqMNf7fjQO\nA0UaPxH5hCItXu5Qse1VkfqZyMKTq15tSLReuX2+zbpduT+JSF0kC+MNqVS3RwN/sXtZ/xWR\nwoqpuencw9sdIq0x5WRY0nlSpGuemJmyW1pO6z2qHxdda+5ZiOSWB4W1XpnphsnSqyJFeyvW\nYFfkUuVtV2fufOSJyIj0rEjdsEbhrTAU5hem+XcpvU6KSG55UFgMCzxp0n6KSPanjRyR9uI+\nI+nEZPMzkvMwmcyZS8IktnpZwSlNzTSldX5liDYLObNIQ3Iuxiexqd3DlG8G6i+FSRk23hv5\nhCLdhlmFRifGW7WbljftMk0zr9o1zoqNbuaULi9XQYU4JxZpvGJN7zhsjcP9Ij1M+cNA/Zml\nsSo7OvN8zL8jtTpN8d+RSlNYacmcTfs7UqOT7ZXaYMlcIfcqxPllkR6M/fFXAmGRHqZ8K5Dt\numvgyt7IZxSp6mfC5s2Gwr6l4rzZcJsqbbzZ0OZ6GHil+vWuQmfZ/NOaJ4CHbzaciWDs5977\nA2JTu2dT7gUyfXtPnTHwTOQzitTdx3fthkleNX60ObS5S6d1bHcziZZm02ZrKxRBhdMTrA+k\nqo3s3Nd2Q6RnU+5FHrouj1d5FPmUIvXPSXmftaxsho9tqd/+LofOtbmzb3R3waZ+CTi9xkqd\nupEKZyeQwVdBTqQnU+5Hbi69RuHb33sjn1QkAFkQCUAARAIQAJEABEAkAAEQCUAARAIQAJEA\nBEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAA\nRAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEORKn+\n/1Rn/jEb45ZhKIKvgK46kD0izSXDdjdUPPrkwYP+OJB/E2nYMbaiLw+F5B+ItEhOFPr1jyHh\nB/JWkca7FE79CWT5QP5OpDECvAuSeyB/LRLPUu+DpB7IYSKpsQlIQTIP5GiRhm0QgDQeyMeI\nxOPTP0MGD+STRJpawEuQuQP5SJF4enoJUnYgnywSA+M5yNeBfLxICqP2QpoO5GtEGgN/I+86\nb7X5Ef6SLxPpK+9P77oAhHG/LjG/xFeK9F02vWvhZBH3i3Lye3yvSN8ybN5oPSJ9Dl8vkuo+\nfAQh0in4KZE+cSS9U3RE6mmveT8y8uv0uUiVSovWq1QkKhmKYpvz2OypL0pdmrCuUxrl90Ra\nHU7XxZ67LWl1jur1HC3atk6Cl1XGneN33Rrh1zQItCiIRB57HJF6qmQYVokd5bdxmFVOpcyU\npKubY5tEB7RbrV/XKY3ziyLFB1StwvI2sSW2K7ZM8ts2idtzYRUn+/O3XolbhN2zKFiP3CFS\nZwb45d7/vefGAv256PulKVyT7iqpuzpR97XNMZbeTPrSNleFX2EuXeE0IvUJCctzW1Koi/4n\nX++toO3F5NM0W1ZZdE+3PsJrdWn1DeeyWrAdGZH6i+Doy0X1s7t28qdyLkeFKbypcm3T0Ca5\n+aw7t9VaOhWc0hXOItJVZaFIt+FOkah2OM4KYdtxYqViVcLumc4rQh4GWhRsR0ak23z9a4r+\nClPO94xCi2VTmSs9eah13fimITfj4DJPTZwKl80JS9f9rkjhoOoTHKjS+HpsXWyCtsOM0G0x\nVwm7ZwcLh73z2h/5hCLl3p1ff54G/N0kahyp3fAnvqmprYOp6srETAzcCk7pCmcRqV6M10w1\nTom9fsUJ25bD1K6MVYncrh7Qqmyj4InIJxQpTIZ3CVLB1rZI9obUf9SLgOYi6dWdStfP5BQi\nLZJeqttc0s/y1h8jl22verUhuUarPC/S1VtgihQg0ipTTqbhudjXeZlbFakeHkyVfhJtL/oq\n6dWdStfP5JQimRnSVHLNk40chW17CTWlv+NVkZokjxY8HxmR/kGkYrh62eXbRi+OenWn0vUz\nOaVIadL6JZeNuV3Q9qpvX+1lfph1qzwrUpuEE7uhAJF24D4j6cRk8zOS8zCZzJmLb8513Cw7\nFR736klFupjrj1uytbQZSmhm021wefI7ZLdIWXiRWxTsjnxCkW7OLwU6Md6q3TTHsMs0zbxU\nF2462jkLp17dqXSFk4qk5rOJ7d5su3J5Gj563fOQJs2a7YInIp9QJOd3pFanKf47UmkKKy1Z\nfFPPMq5u3Uav9yzqNuGykMMPixSOqzWR7O9IW/PfQBp7awhvYUMVr3seUYU9syh4JvIZRaqU\nys2bDcX0gs/4ZsNtqrTjzYZp3bwfB62et9+8Ck4pLG85tsS8otDm+5+RCqXfhiuCET1Uib3Z\nsMbiChe/5O2NfEaRuvv4rt0wyavGj3bM29ylpiRb3xym65oyWsEpPT2jDLMU08PHwyT5bbNY\ng7FKuj/lF2d+af5xC56PfEqR+uekvO/ArBxnxG2p3/4uBy9s7uxbxt36pnulrLJYhbn09KyK\npN+qTrfuR4u2ToIXVdrYzrWogUjLJ7dnIp9UJABZEAlAAEQCEACRAARAJAABEAlAAEQCEACR\nAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlA\nAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRDkSZ\n/3Cp/Wf8j5mq6b9l2oX/OVP4YOiqA9kjkv/fDO4Lu/k/HwyfA/1xIP8m0vyf4Lbl9OWRkPwD\nkRbJjQJ/Cxk/kDeKZG5Q9h+6+C8gywfydyKxdPFuSO6B/LVIPEu9D5J6IAeKxO1JGPJ5IEeL\npLqOESAEaTyQzxDJljAQ/g3ydyCfJpJiOLwMmTuQTxRpaApPQsoO5JNFYmA8B/k6kI8XidGx\nG1J1IN8jkvrex6d3nbfa/Ah/yZeJpL5xsLzrpMO435ibn+ErRfqu56d33UkXcb8oJ7/HN4v0\nHQPnjdYj0ufw7SKpj787IdIp+BGRPncMvfPWiUg97TXvB0x+nT4XqVJp0XqVikQlQ9HqZlaZ\nrXmsu6XtRalLvXEaPyfS+pP9dbHnPpTUOkvNRpbCtnOCI1XG7hm/69YIv6ZTV64URCJP/YxI\nXVclw+BMbAfexsHq9k9mStJHm2W/VTsizaWdPcqGSecRqV7saRNbUtmOaJdtVto6CV5Wcbqn\n869vS4rwyIuC9cgdInWm7y73/u89V8nwueiNagrXpLtK6q5O1H1t86qyVt90ap3tfGzmlBbq\nov/JlycwchqR+oyFe/KhJOkT2uaqWE1S0NZJ8LKK0z1ukuNx1aXV4S6rBduREam/U4y+XFQ/\nu2snfyrnclSYwpu+9MU3M5PURg+B63yBdEoTpYOtz3bOI1I/+MM9t+HUb0ah1l7R9rR1Erys\n4nSPm+QouQpOeVGwHRmRbvNNoin6XinnTim0WDaVudLTPnOviW+OU+VM53t62nJKh4LVIfK7\nIoWDqk9wIFIzjtDL1tQ31naRYLeK0z17WbjvFjwR+YQi5d6dX3+e+vJuEqWmkTaMlUebuaou\n/ZNo5zczFLNjS84iUr0Yr5lqbEmqujIxc6oVwrZhgr0qkZ0PaF0jFwVPRD6hSGEyvEuQCrbW\n7UnNNepuRTLoHnBKOzOBWZ/8n0ekRdJLdRujKZO8jdt20NZPcHjgp0W6qmq7AJFWmXIyDMQX\nRSpV3nZ1Zktvegld33yc0p5rngTrS/6ZnFQkZ2qs9EN8e9nKkt/WTXC0t8IGmzRJHi14PjIi\nvSqSXd3O5xatWRwNSy8bc7uzipQm7SySnlc37rrydls3wf8sUpuEE7uhAJF24D4j6cRk8zOS\n8zCZzJmLb+rraFK6qY2WbixInVWki5k9KeW22BqH/s5F2p0qyY5oLlno76Jgd+QTinRzfinQ\nifFW7aYphl2maealunDTUjvX0jnL8dIFJxVJOd8psua82XagDm5hQ5VF92zSpFmzXfBE5BOK\n5PyO1Oo0xX9HKk1hpSWLb9qfia46tXazmTe90vVZCyINCW3CtbPVtm6Cl1Wc7nlMFR50UfBM\n5DOKVPVzbPNmQ2GXi5w3G25TpYdvNpgXF+6pblLo/LbmNzuvtN9s86317/OwvKvYkv46Y15U\nuC2bxNs6CV5Wib3ZsMZC3rjNeyOfUaTuPr5rN0zyqvGj7R2bu9SUZKubrQ2Sz5uFXzq8a7d1\nrT0PowyLp/jycZK8tm6Cl1XS/Sm/eLf6oOD5yKcUqX9Oyvv+yMpxRtyW+u3vcpjX2dy15nXf\nbn2z6TOfV1Npeg1Lu7n09KyK1FXZmNB9bd0EL6o43fP4jAKR1KZIjyKfVCQAWRAJQABEAhAA\nkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJ\nQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAE\nQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQKTfhz7+A0jy76M2P4IIZPX3iYlEvwtD\nQqFG0FUAABtRSURBVH+fLZHofyFI5O/zQCSGgARk8fd5QiSGw6uQud/nOZEYES9B2n4fRPoD\nSNvv828iMUJ2QZp+HwGRvnWYqJ53hd78CD+IpEjfNV7eeBEIBf2uxMArINI7IiPS6XiPSF8z\nct50ooh0Os4sEs9IIMb7RfroUcQd6X2017y/VOXX6XORKpUWrVepSFQyFPmbWaU31MhQ/T5s\n1BelLo2NOjc7kk8R6epXmzK5nahrutx3Dw84hR4jOR2zcnJh/60cKow8tUCkrquSIYeJGfDd\nbcxp5VTKTEka3Sy7uSMSW71NbCorW9h3R5O4xziQPxVpfUDV/ixrzuRmooopnzNjrpehnZ5a\nnrtP0H8rh1qLjEidGeuXe//3ntss9p+LvhObwjXprpK6qxN19zavKmu79qLqOdbdbuRDxpO+\nbpurousu+p++dy5/983i/K1IayOqT6C7y8nkVqJqdWl1ZW9fHhxkCu30VOTcY0z9Fz/UamRE\n6kkmXy6qn921kz+VczkqTOFNXzKdzcxksjE9r2mT3G7chjv+zexqtaHjFODwFH+ESL043i4n\nk1uJyiP7xlwvQzs9NZ7Ndvqn/osfaiMyIvXZmJLXFH1vlpMWfbr0Y5PJXa70TKPWdZ3Nsc+z\noUGurHnNmHHnZjXMP5y5w0F8hEh9kr1dTiZ3JMpt2gRKOqGdntrH2H8rh3oi8glFyp37s/08\njf27SZTJnXOZjG9q6tHBTDW2LFVdmZg5wiBo4VzEDuIjRKq7uEhqT6La6crVOblehn52ElDP\n19DooZ6IfEKRFi93RIdB3J7UXJjuY7Xxglaq21BLqXx6gr3qh+jk2h3NR4gU7nIz+TBRV+fh\ndc71MvSzIkVuSFdvxQmRNphyMix+PiNSqfK2q8epRT08mDqzPqUfSvtnaH11Lad1qWP5SJHc\nTD5KVOM8yXgz7FhvbZ+DR71c3xgO9XxkRHpKpM6s1I6rRsVw9UqTdhZJzxMbvU561fOG1qxn\nHMtHiuRk8lGi2sSZbTm5/meRiuDuMx8KkXbgPiPpxGTzM5LzMJnMmXM2dX8n5ZjPYcfF9Mco\n0lQ3NfOG1v3t4RiOFMn9zdPfNWfyUaIyZ4eba+8w5k/yaLj7hD9H+Yd6LvIJRbo5N/T5Wdfg\nPPDaZZpmXrVrZslq2+ejdtNv5MpbQ2X5u9sSyWAyuZ2oJs2cX2rdXHuHMX8WPbXJYg3OP9Rz\nkQ/v5QOYf0dqdZrivyOVprDSkjmbibl4Xm0+r8NkxO1cW7eZV3Vblr+ju5xMbiaqclfRHonk\n9NQOrsFkMjjUc5HPKFLVz83Nmw2FXV5z3my4TZXibzaYn9/vqa3nrJt3Y8b7pyPzg/1N19Xv\nbRU7e/UEeIPfyeRWoprY4F6d2i3fP9jC77/tQz2MfEaRuvv4rt0wyavGj9aj8RchTeZvtral\nvcGn3uLpdOmammXzJnTz8LcPm04mNxJ1cW5AixWAZej0mZTP/Weau4d6PvIpReqfk/K+H7Ny\nnBG3pX77uxzyanNnX0nugs2mT3deufVGxk9VNtbt5mbQBSJ5mVxPlHpOpPaZlAfxNieNDyOf\nVCQAWRAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABE\nAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQA\nARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQKTfhz7+A0jy76M2\nP4IIZPX3iYlEvwtDQn+fLZHofyFI5O/zQCSGgARk8fd5QiSGw6uQud/nOZEYES9B2n4fRPoD\nSNvv828iMUJ2QZp+HwGRvnWYqJ53hd78CD+IpEjfNV7eeBEIBf2uxMArINI7IiPS6TivSIY3\nnTIinY4zi8QzEojxfpE+ehRxR3of7TXvL1X5dfpcpEqlRetVKhKVDEXOZtdd1bJCz30ovqax\nukfyKSL5yWj9/GbVWis3n7rdRalLvRZ6DOos162cnBpZP8VY5KkFInVdlQw5TBrz+Tbm1O3M\nzJSkwWbX1WPqvdKuTWxxYQO3ft1D+VOR1r+wn4wmmXvAZrKMtvLyqbHtfJNW+mRzsWHs82T1\nFLciI1JP1V/S7v3fe26z2H8u+v5sCteku0rqrk7U3dvs9F8VVtDktrhWl1Zfxi5e3WP5W5HW\nvnGQjIsqOq3JRWcra/WNJrzNmFZuPruxRaHyaOigT3as2lVT3eUpbkVGpE5f00ZfLqqf3bWT\nP5Vz5StM4U1fJ51N3elKhRU6e1PTf/Phvq+8usfyESKFyVBzojIzPBtjVoibT0Oi2vAoa33S\ndQ9/kG0Tx8hlf21ERqQ+G1PymqLvwnLuwUKLZXspV3rSUeu6zmbX1x1S65b2oyAyTOa6x/IR\nIoXJGGbCek4wOpXFW8aiehOylT7ZQ66cp69lfz0R+RM6+o/J3bu5+TzNKe5WFtV17hXT2exq\n/+P4J1ON2wetGRT1+rD6Uz5CpDAZ5TC1K8OExmgDyewFbxn6caTgpLy74LK/noj8CR39xyyu\nbtFhsCJSF01tqW5emOs4WUSkbm3XVa8aJFqI1Fzt7xstr94yUD+NDmeBL4rk3ZDiLRFplSkn\nw+Lnv4tk7vhOmGaaeSNSt7arnJbqSpW3Xb3xPNkk3ozqmif2WSXaW9vn4FHPaxjBKT4f+RM6\n+o+RFylNWrdpm2RBlGP5SJGu+q7SmtUeu6Kdr7Z08jlymR9m3dBPilSo8McrRHoC9xnJLhpN\nz0jOw2QyZy7xkzj8nUsvpj/mHGfzrw2nF8n9zdPblZpZVWt+mel9Ssr1XDn5HGm91YZIn+wi\nWZvIvRD5Ezr6j7k5N3SdGG/VblretMs0zbxq14ySees4unT6jdzsaNKs8eIfzkeKtLjE12rp\ni8bL5/phzJ+gpx4QWYNbFelh5E/o6L9m/h2p1WmK/45UmsJKS+ZsaobUzqWeSJW3wHR6kbxa\n7i57iTd3FvvT0DU+SKtgwc5WbgLrFn2yh6u39rc8xacif0JH/zVVPyE3bzYU9gUR582G21Rp\n7c2GKbWLX9Ht/cjv948Q6VPwklEo/UpcoUemeVnhnjrJn2nCX5dM5TYPHFjpk23y5bsUqyI9\njHzKjr6P79oNk7xq/Gi70uYuNSVZsNnNyfZLh+KL/yokIjn4Scmm9LW2O6I3JDeftl3ip90P\nncZ2rpFOi99TN62K9DDySTv6lvf9kZXj5Lst9dvf5ZBXmzv7dnIXbHZzav3SoVgh0ipBUub0\nNb0uefztb7UQSbdLwznZSp/sO59dIj2KTEcDCIBIAAIgEoAAiAQgACIBCIBIAAIgEoAAiAQg\nACIBCIBIAAIgEoAAiAQgACIBCIBIAAIgEoAAiAQgACIBCIBIAAIgEoAAiAQgACIBCIBIAAIg\nEoAAiAQgACIBCIBIAAIgEoAAiAQgACIBCIBIAAIgEoAAiAQgACIBCIBIAAIgEoAAiPTZ0D9f\nAh312dA/XwId9dnQP18CHfXZ0D9fAh312dA/XwId9dnQP18CHfXZ0D9fAh312dA/XwId9dnQ\nP18CHfXZ/E3/MAr+GVL42fyZSMociuHwKmTus/lzkdTfHfWnIGWfDSJ9CaTss/kAkRgieyBL\nnw0i/ROq512hNz/Ch/FRIn3dYHnjPDUU9OtyczI+TqSvGjDvE0lxR/ouEOmfedMZI9JXgUj/\nBs9IYPgakT53IHFHeh/tNe8vVfl1+lykSqVF61UqEpUMRdOmGgkqdN11yGR7UepSd0Hpy/yS\nSEEy+vRlldlqvUzuaLeoPFUZdzrLdVtndnd2en27GnmqgEhdVyVDypLGfL6NKaycSpkpSf3N\nsWLiV+i6esy/DV0HpS/zbSJtfOEgGTZ9Zb/VJG5n7GuXxqsEOx8tNrRJRKRkT2RE6qn6e8a9\n/3vPbdL6z0XfiU3hmnRXSd3Vibp7m2OAsLT/azNZqIv+J++80tf5HZGCZFxV1urbd3/Juaii\nGzL3uN2iM9wq4c5HIuXL0632RUakTt80Rl8uqp/dtZM/lUqmKUNhCm/6kulsGtok9yvoUTEk\nPFF2Bth5pa/zMyKFycjMmGy0Q+NcKdo0bBd2hldlsXP7B9nbcq/t2x2REanPxpSrpuh7szQX\nREOhxbI9mis906h1XWfTkBtZ3NI+gtcl9k4Xlr7Cz4i0SNFgT6YvPnYziTRbtAs7w6sS2blB\nE7nQ2b59IfIJRcq9m7f+PK0N3K0WqvMuk8EVs7biuaW1P4Ksj2HpS/yMSGEynPSVw9SufLLd\nssrmvW1BppqwZj1fVJ+MfEKRFi93qNj2ukjDRSsonVv2E4YiiPIPJ/uP7XcfRVSkeCUvGam5\nxN9N2VWvNiTXaKOwXXxEvyJSqW6Lmv4N6ZnI5xVpWut8UqR6eCheFemaJ9PF9awiRWt5paXK\n2662c6tyWsCLh14TKdpbq4cPMNO0oObYty9ERqRnRSqGpYlVkbphEWNR+tLJ/mP73Uf5a5Hs\n7wRm2eyq7+DtlLPtdmIipUm7qDn2LSLtwX1G0onJ5mck52EymTOXeEkcf3rwS/0Ut+OD8+lF\ncn/i9JPRm5OUpiw186k2+G3ICe22S6Ij2u+bPVm/GGeCmsmy4d7IJxTp5vxgMT/rGpwHXrtM\n08yrdo2VbHLNKw1THB87L/C7Ihlqbc+jq723I0i7XyW+cyXo8kWG2KLc3sgnFMn5HanVaYr/\njlSawkpL5mzqech1UUEzXbl0hGa8up5eJK+Kf2fRibrqkWmv9m18+TtsF6TdrxLfuRJ0KdI1\nMrvcG/mMIlX93Ny82VDY90GcNxtuU6W1NxumxfLgt+7pCaq/3bW52DPSL+ElwyTqnuqMF0q/\nHVesDlKvXezNhqlKfOfeU3J/CHk68ik7+j6+azdM8qrxo/XI5i41JVmwOUzow9K5S5JoKXTB\nfLe1iTJTpczL2Xa7IO1BlejOx6c0/EkXi9/7I5+0o295349ZOb4m2Zb67e9ySKPNnX0luQs2\nHTXcUqe8L02vi1IIhWguvUbDnNrL5IN2bazyWCW68/EpbTyl7Y1MRwMIgEgAAiASgACIBCAA\nIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiAS\ngACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEI\ngEgAAiASgACIBCAAIgEIgEgAAiASgACI9MvQu38Gqf5l6N0/g1T/MvTun0Gqfxl6988g1b/M\ndu/S94KQzF9mr0iMgn+GFP4yT4vEcHgVMvfLINKfQeZ+GUT6M8jcL4NIfwaZ+2WkRFJfPFDe\ndeJq8yP8FLIifeVYUW866zDuVyYHdvIGkb5swKg3nfAi7pflBZ7ivSJ9wdh5430UkU7Eu0VS\n6m1TJxkQCST4I5HUw0MdxDtPDJF62muulMqv0+ciVSotWq9SkahkKHI2u+465qy+KHVpggqt\nU/eaOs2O4JNEWiRjSpQaiR56sSuW1alPxqCmzfZDXau7r948VCTyVAORuq5KhpwlVoPb2JOV\nUykzJWmw2eszprqyIVqvQpPMgYu5wkH8vUirR1wkY87ZmP0k0qxejO5YVqc+8XoqbsaE7anZ\npOWh1iMvv+kZReoNuNz7v/fc9l3/uegHflO4Jt1VUnd1ou7eZqf/DjlL+tI2V4VX4aI/9519\n0X1wafUV7fLnX3DiEJGiMRbJ8JKqqeZNr13+IFDn9Mki6PpXsn1UOOEXh9qOjEj9pWj05aL6\n2V07+VM5F7rCFN5U6W32PZgNqb0ZZVrtolNhvO/3f/J58yg+R6RFMtykatpkOYw7ne/SL4hk\nde6TMOjGV+pHQevHWRxqOzIi3eYrT1P0V5jSGGEotFg2ubnSszNzlXI2u77ukNrLPCtwKiRD\nP8/TFERyA0073aTaz9Ep8FVdY8VeVuc+CYM+xumoyKGeiHxCkfJgDpHPRtytLKrr3FuLs9nV\n08dUdWVi5hluhXKY2k1XrlZl7/sqj/g0kZxkuEnt9BAtoi1yVV36h/yNQG6fBEEfUzjyRA71\nROQTirR4uUPFtldEcvfk4xOyW+Gqn2GTuXuu3grGH/NpIjnJCJK6ckPqcrsAEF6Mwqy+JtJN\nufrGD4VIq0w5GdZoXhZJLzZcggej/pakmW5ITXzq/0d8mEhuMvyk1mtLMkrd9E8KwaxrCBTt\nra7bLdI1T9yHHudQz0dGpNdF0jPCRq+IOhWu+hrXXsaOb5MDJ3afJpKXDD+pxfZ9u3XXnedA\n/ypSNyw3LQ+FSDtwn5F0YrL5Gcl5mEzmzCUrIsUqpGaKMnV85g2AP+d4kdyfZrxk+ElNHoxD\nfwAvs+qHeUKkdvHrVdB4b+QTinRzphHz+oDBWSSwyzTNvGrXjJJN6zjTJ6eCZ1mTZs1bv8oj\nPkmkIBleUh+us7kDOJZVb22teWLVbqlGXKSHkU8okvM7UqvTFP8dqTSFlZbM2dQMqbWljX44\ndSrYK5e9zlVHLtgZjhdpIkyGl9S1Re7x1x53AEez6vVJtbICGI+chgW+K3sjn1GkSqncvNlQ\nqGHAT2823KZKa282TKntu6DVT0M3r0Kh9JtehU54c7hHH8QiGV5SnV8gAkwmW+cRKp7VoU9i\nbzasYd5saPNZ4cWhnop8RpG6+/iu3TDJq8aP1iObu3ReDHU2u/neX0YrZNPmZYz6d1/sc3GT\nscjv8GQZo7VdZW4Epl08q+NHv6e2Sea6prl7qOcjn7SXb3mftawc59ptqd/+LofetLmzr3F3\nwWY3p7arsliFaVNFu/ykuMlY5HdrdUBXS69zrXhWx49+Tz0giOwe6vnI9DKAAIgEIAAiAQiA\nSAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgE\nIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAAC\nIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAi\nAQiASBCFgfEc5AuiqOgmrEGSIMqPiKTUe85ehYG/OUnwRpYifeNQUd17zltN/3glACE/IdLb\nThmRYCcPRfqGkfPec0QkeMyPiPSuZyQbfe0DwMgTIn3uGFLvekbqFnE/NwnvpL3m/aUqv06f\ni1SptGi9SkWikqEovllflLo0Q+3rmMlrOlVwNr8OcZEWyXDSt5Go1kl4vMAEGM9h3GlX1ZaP\nMi5e/7mNtyJPy3VfeIMWp0qUJbFpvA0fVeVUykxJur5Z2RCDVOMUophLC7fCt/GiSKszqUUy\nql2JahK3o2IFmin7TvdMZ7Z2RlV41KDxZuRF1DOK1Gfwcu//3nOVDJ+Lvl+awjXprpK6qxN1\nX9vskn6zzVWha/dFNpO1urT6MnbxNr+Q10WKDqllMub0bSXqYvJbzPsWBZ2Tfbd7pjNbG+Ju\n/8UaP47sckaRksmXi+pnd+3kT+VcnwpTeFPl2ubNdEFrXLyqbEh4Ptz3lbf5hfyDSLEvvEiG\nk76tRKlw36LAzb7TPWPt1fS7/RdvvBHZTmC8E105zA9zU/m42RT9FaacL0qFFst2Uq705KHW\ndeObF1VPIfsIQVpVbPOb+BeR1r/xvMtNX7jPIRm8SVYL3Ow73fOQxQlEGj8R+Tt7+Z/IvZu3\n/jxl9G4SZdd65itffDNVXZmYSUmf4GAYtCqLbH4VbxHJSYabvnCfQznM5MrVAjf7kdvVKosT\niDR+IvIJRQqT4X6OTiHWNnPzsBqLep0ftq7eCsb38BaRnGQE6VtN1FUvLiTXjYLhsMs/2yxO\nIN4YkVaZcjLMc18WST+sXuZZ8xylSfLI5nfxDpHcZATpW01UaXqpjBasPTntFMk/Aa/x85ER\n6XWR9IywGVdEnShtkkU2v4w3iOQlI0jfWqKueibXmkWhZcE/iuSdACI9jfuMpBOTzc9IzsNk\nMmcuvhmk1klxlsY2vwwhkdzlLS8ZQfrWEpUq/RDTzsN9UeCGSZ4SKagbb7w38glFujm/QejE\neKt2043eLtM081LdcnOK4P7tmjRrlptfh7hIQTK89K0nas/y9/zR6Z6HLBbd4433Rj6hSM7v\nSK1OU/x3pNIUVlqyrc1mXGsaO6SaF5+qL12wM0hP7cJkuOnbSJS9E7Th8nfrLn93U/ad7nlI\n0H9rjfdGPqNIlVK5ebOhsIs2zpsNt6nSwzcb+tl1qyfrQ5Mh4U7HNN/skTSLZDjp20pUofTL\nb4U7aQgKDEP2Yy8nrJ+R139rjfdGPqNI3X18126Y5FXjR5tUm7vUlGTrm+W8OTXqLmMobxPc\nZNiMzOnbTFQ2Z9lWcApmxrZpbOcKZRg52nhv5JP28i3vXcrKcWLelvrt73KY19nc2beMu/XN\nrsqmzXAt0K5GINKEm4whI1P6thM1J3yo4PSAE97+bWM715j7b9Hjz0emlwEEQCQAARAJQABE\nAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQA\nARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAA\nkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARDp\nOP57f4s/OMQLLX4RRDoORPohEOk4EOmHQKTjQKQfApGOA5F+CEQCEACRAARAJAABEAlAAEQC\nEACRjiLp2VcrcWqHf8PqXazSVqO5xb7DPIq972v9Hoh0EMn0z+Nqc+3wb1g72aocazSM+t2H\neRR759f6PRDpIN4hUuKU7xvkyXhH2nsYRFoBkQ5i34hL3L87xuvTdyS/cO+ND5EWINJB7BRp\nfHYZar9LpOcOg0hLEOkgnrgjPTFe/+WOtPswW8ohEvwtT4y494v03GEQKQIiHcT3ivTCPewE\nINJBfO3ULnnpxH4eRDqI/SJtzaTWoz4t0s7DJH4zRBpApKPY/2bD1t+w+tONnmyRPHoHgjcb\nAOBlEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQC\nEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAAB\nEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAgP8BcD/7KrIlmv0AAAAASUVORK5C\nYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "topGOUp <- fgseaRes[ES > 0][head(order(pval), n = 5), pathway]\n",
    "topGODown <- fgseaRes[ES < 0][head(order(pval), n = 5), pathway]\n",
    "topGO <- c(topGOUp, rev(topGODown))\n",
    "# create the plot\n",
    "plotGseaTable(pathways = GO_term_hallmark[topGO], stats = stats, \n",
    "              fgseaRes = fgseaRes, gseaParam = 0.5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86cfabed-97a6-4e5e-8cfa-5ffa01c57906",
   "metadata": {},
   "source": [
    "To use the `plotGseaTable` function, we need to pass the following arguments:\n",
    "\n",
    "- `pathways`: A list of pathways to plot table.\n",
    "- `stats`: A dataframe specifying gene-level stats.\n",
    "- `fgseaRes`: A dataframe of fgsea results.\n",
    "- `gseaParam`: A numeric value to adjust displayed statistic values, values closer to 0 flatten plots. Default = 1, value of 0.5 is a good choice too."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ef50627",
   "metadata": {},
   "source": [
    "### Enrichment analysis using FGSEA and KEGG pathways\n",
    "We can perform enrichment analysis using FGSEA with KEGG pathway using the same procedure mentioned above. The only thing we need to change is the\n",
    "list of genesets that are available in the KEGG database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "93b11008",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$hsa00010</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'HK3'</li><li>'HK1'</li><li>'HK2'</li><li>'HKDC1'</li><li>'GCK'</li><li>'GPI'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$hsa00020</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'CS'</li><li>'ACLY'</li><li>'ACO2'</li><li>'ACO1'</li><li>'IDH1'</li><li>'IDH2'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$hsa00030</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'GPI'</li><li>'G6PD'</li><li>'PGLS'</li><li>'H6PD'</li><li>'PGD'</li><li>'RPE'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$hsa00040</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'GUSB'</li><li>'KL'</li><li>'UGT2A1'</li><li>'UGT2A3'</li><li>'UGT2B17'</li><li>'UGT2B11'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$hsa00051</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'MPI'</li><li>'PMM2'</li><li>'PMM1'</li><li>'GMPPB'</li><li>'GMPPA'</li><li>'GMDS'</li></ol>\n",
       "</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$hsa00010] \\begin{enumerate*}\n",
       "\\item 'HK3'\n",
       "\\item 'HK1'\n",
       "\\item 'HK2'\n",
       "\\item 'HKDC1'\n",
       "\\item 'GCK'\n",
       "\\item 'GPI'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$hsa00020] \\begin{enumerate*}\n",
       "\\item 'CS'\n",
       "\\item 'ACLY'\n",
       "\\item 'ACO2'\n",
       "\\item 'ACO1'\n",
       "\\item 'IDH1'\n",
       "\\item 'IDH2'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$hsa00030] \\begin{enumerate*}\n",
       "\\item 'GPI'\n",
       "\\item 'G6PD'\n",
       "\\item 'PGLS'\n",
       "\\item 'H6PD'\n",
       "\\item 'PGD'\n",
       "\\item 'RPE'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$hsa00040] \\begin{enumerate*}\n",
       "\\item 'GUSB'\n",
       "\\item 'KL'\n",
       "\\item 'UGT2A1'\n",
       "\\item 'UGT2A3'\n",
       "\\item 'UGT2B17'\n",
       "\\item 'UGT2B11'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$hsa00051] \\begin{enumerate*}\n",
       "\\item 'MPI'\n",
       "\\item 'PMM2'\n",
       "\\item 'PMM1'\n",
       "\\item 'GMPPB'\n",
       "\\item 'GMPPA'\n",
       "\\item 'GMDS'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$hsa00010\n",
       ":   1. 'HK3'\n",
       "2. 'HK1'\n",
       "3. 'HK2'\n",
       "4. 'HKDC1'\n",
       "5. 'GCK'\n",
       "6. 'GPI'\n",
       "\n",
       "\n",
       "\n",
       "$hsa00020\n",
       ":   1. 'CS'\n",
       "2. 'ACLY'\n",
       "3. 'ACO2'\n",
       "4. 'ACO1'\n",
       "5. 'IDH1'\n",
       "6. 'IDH2'\n",
       "\n",
       "\n",
       "\n",
       "$hsa00030\n",
       ":   1. 'GPI'\n",
       "2. 'G6PD'\n",
       "3. 'PGLS'\n",
       "4. 'H6PD'\n",
       "5. 'PGD'\n",
       "6. 'RPE'\n",
       "\n",
       "\n",
       "\n",
       "$hsa00040\n",
       ":   1. 'GUSB'\n",
       "2. 'KL'\n",
       "3. 'UGT2A1'\n",
       "4. 'UGT2A3'\n",
       "5. 'UGT2B17'\n",
       "6. 'UGT2B11'\n",
       "\n",
       "\n",
       "\n",
       "$hsa00051\n",
       ":   1. 'MPI'\n",
       "2. 'PMM2'\n",
       "3. 'PMM1'\n",
       "4. 'GMPPB'\n",
       "5. 'GMPPA'\n",
       "6. 'GMDS'\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$hsa00010\n",
       "[1] \"HK3\"   \"HK1\"   \"HK2\"   \"HKDC1\" \"GCK\"   \"GPI\"  \n",
       "\n",
       "$hsa00020\n",
       "[1] \"CS\"   \"ACLY\" \"ACO2\" \"ACO1\" \"IDH1\" \"IDH2\"\n",
       "\n",
       "$hsa00030\n",
       "[1] \"GPI\"  \"G6PD\" \"PGLS\" \"H6PD\" \"PGD\"  \"RPE\" \n",
       "\n",
       "$hsa00040\n",
       "[1] \"GUSB\"    \"KL\"      \"UGT2A1\"  \"UGT2A3\"  \"UGT2B17\" \"UGT2B11\"\n",
       "\n",
       "$hsa00051\n",
       "[1] \"MPI\"   \"PMM2\"  \"PMM1\"  \"GMPPB\" \"GMPPA\" \"GMDS\" \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load the pathways into a named list\n",
    "KEGG_hallmark <- gmtPathways(\"./data/KEGG_pathways.gmt\")\n",
    "# Show the first few GO terms, and within those, show only the first few genes.\n",
    "tmp = lapply(KEGG_hallmark, head)\n",
    "tmp[1:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "9eb1ede2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Running fgsea\n",
    "suppressWarnings(\n",
    "    fgseaRes <- fgsea(pathways = KEGG_hallmark, stats = stats)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "cbe55bb6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>pathway</th><th scope=col>pval</th><th scope=col>padj</th><th scope=col>log2err</th><th scope=col>ES</th><th scope=col>NES</th><th scope=col>size</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>hsa04060</td><td>4.044034e-06</td><td>0.001374972</td><td>0.6105269</td><td>0.3328477</td><td>1.436012</td><td>290</td></tr>\n",
       "\t<tr><td>hsa04740</td><td>1.058477e-03</td><td>0.179941158</td><td>0.4550599</td><td>0.3505170</td><td>1.429992</td><td>120</td></tr>\n",
       "\t<tr><td>hsa04061</td><td>5.614319e-03</td><td>0.636289434</td><td>0.4070179</td><td>0.3461236</td><td>1.391701</td><td> 96</td></tr>\n",
       "\t<tr><td>hsa04080</td><td>8.145521e-03</td><td>0.692369245</td><td>0.3807304</td><td>0.2779229</td><td>1.210259</td><td>356</td></tr>\n",
       "\t<tr><td>hsa05340</td><td>1.279142e-02</td><td>0.869816875</td><td>0.3807304</td><td>0.4011639</td><td>1.445081</td><td> 37</td></tr>\n",
       "\t<tr><td>hsa03460</td><td>2.330367e-02</td><td>1.000000000</td><td>0.3524879</td><td>0.3637500</td><td>1.374373</td><td> 52</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 7\n",
       "\\begin{tabular}{lllllll}\n",
       " pathway & pval & padj & log2err & ES & NES & size\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <int>\\\\\n",
       "\\hline\n",
       "\t hsa04060 & 4.044034e-06 & 0.001374972 & 0.6105269 & 0.3328477 & 1.436012 & 290\\\\\n",
       "\t hsa04740 & 1.058477e-03 & 0.179941158 & 0.4550599 & 0.3505170 & 1.429992 & 120\\\\\n",
       "\t hsa04061 & 5.614319e-03 & 0.636289434 & 0.4070179 & 0.3461236 & 1.391701 &  96\\\\\n",
       "\t hsa04080 & 8.145521e-03 & 0.692369245 & 0.3807304 & 0.2779229 & 1.210259 & 356\\\\\n",
       "\t hsa05340 & 1.279142e-02 & 0.869816875 & 0.3807304 & 0.4011639 & 1.445081 &  37\\\\\n",
       "\t hsa03460 & 2.330367e-02 & 1.000000000 & 0.3524879 & 0.3637500 & 1.374373 &  52\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 7\n",
       "\n",
       "| pathway &lt;chr&gt; | pval &lt;dbl&gt; | padj &lt;dbl&gt; | log2err &lt;dbl&gt; | ES &lt;dbl&gt; | NES &lt;dbl&gt; | size &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| hsa04060 | 4.044034e-06 | 0.001374972 | 0.6105269 | 0.3328477 | 1.436012 | 290 |\n",
       "| hsa04740 | 1.058477e-03 | 0.179941158 | 0.4550599 | 0.3505170 | 1.429992 | 120 |\n",
       "| hsa04061 | 5.614319e-03 | 0.636289434 | 0.4070179 | 0.3461236 | 1.391701 |  96 |\n",
       "| hsa04080 | 8.145521e-03 | 0.692369245 | 0.3807304 | 0.2779229 | 1.210259 | 356 |\n",
       "| hsa05340 | 1.279142e-02 | 0.869816875 | 0.3807304 | 0.4011639 | 1.445081 |  37 |\n",
       "| hsa03460 | 2.330367e-02 | 1.000000000 | 0.3524879 | 0.3637500 | 1.374373 |  52 |\n",
       "\n"
      ],
      "text/plain": [
       "  pathway  pval         padj        log2err   ES        NES      size\n",
       "1 hsa04060 4.044034e-06 0.001374972 0.6105269 0.3328477 1.436012 290 \n",
       "2 hsa04740 1.058477e-03 0.179941158 0.4550599 0.3505170 1.429992 120 \n",
       "3 hsa04061 5.614319e-03 0.636289434 0.4070179 0.3461236 1.391701  96 \n",
       "4 hsa04080 8.145521e-03 0.692369245 0.3807304 0.2779229 1.210259 356 \n",
       "5 hsa05340 1.279142e-02 0.869816875 0.3807304 0.4011639 1.445081  37 \n",
       "6 hsa03460 2.330367e-02 1.000000000 0.3524879 0.3637500 1.374373  52 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(fgseaRes[order(pval), ][, -8])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "a63408eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAM1BMVEUAAAAzMzNNTU1oaGh8\nfHyMjIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enw8PD////p0TvgAAAACXBIWXMAABJ0\nAAASdAHeZh94AAAgAElEQVR4nO2di7qyKtdA2VqZmYf7v9ot4AEQT63Zq9UY+/lahDi1CUOR\n+v9XNQDwZ9TRJwDwDSASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEI\ngEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACI\nBCAAIs2iDMmtCuoLu/GAM3Kp82t7dtdcKNw7P49ShXMQNaCr6ixV6iL1Kd72OXTY5dBHj4cT\n0/d34puUKrvxkHMaKJL42b3Ke0VKnIN4ItXdx0hqsWMJBZqGRaQX6S6ZF5VFqg8WqVDq9mz/\nPq/9IP0j7xVJ3ceDeEe6qUt7IajCHP/lWEKB9oZFpFm65NXBUD2FSEk/W2qHosi86L0ipaoa\nDuIdSSlzK6rFDo9I56NPnvlbtE8kSdZ0MxNTmank3jzVTbcp7MC+qefQslap2b3/K8lDXfti\nlek7U5OnKsnt2VZXfWJeZf9B6lTvOHwWp7H9sJm+dRSX9qmlaORQqrTnGxNpfzCdeCe/aeui\n+4leP8sucONmqM1J0t4umdr9AfeOdLdT+cwRST/rt3eDxLS62blJ23RseVVmiD+6eY0kXWi3\nQnMxZ20eO+5eZf+BrvrExjN0GpsPm+nGud0s9/yvY9/MCU9FytRkLWc12L37VBdzm6vaoveJ\nXj/LPrAbrz2KMnlDpJexibPzd6UeWolxILQZr9sxl7ZJfzT987R2ZmxZ2JvVTcksB0xPbaTQ\nZ9M+zRXOibmV/V6Xugk+S99YvzUetVeGUm8WvI22se3tw19sMNvacZpmz8Xdw2BJ2ZRJ+xns\nFerefr5J77x2ln1gN96jq0Wk1xlW7eqxphlF6i6x+orYTvCuevhdBmdMo9Q8ALxhZtf36TAi\nr+ZItZ5ADSfmVvZ7PYMQY2MzsbnYDZLTuv5Qub7FRURqiptO8fZD2tMrzEfVmU3VuKX5m0hD\nYCeevfcXiPQH/O+RquJ+8UTqXy/teM3ap4C7dcppmetL5vMNM7uJSM7gHE8sGLHOSIh8lvbe\n1HmWtTOZshQ/W31VmRnqz3sSTlVXgtk/+l5f2QlY8IlePsthHhLEQ6Q/4CXuMozJUKSiNSVJ\nm7Sf5Y0tu6cr+Zmd94y0X6TYZ9GXjO7eedfPTUJfT43H1asys0O93H7bdka2vkZlJhPhJ3r5\nLLs/k3iI9AfcxN1UmhdVVKR2hvFsL4uZXhKrvZZtXaEFk+dhH7+G03RO1XUj/oGin6W9HZXm\n0UBTZKnwM1Kj5S8nQ30obx//7nhvr1+JPs/JJ3r9LLtbXRgPkf7ApMdnRMrUrRWmaF9vfsv2\nQnsp3zGzc79Hqu3jkLOi0L9ew4cddwxORdI3zyRsK0I3W1LpZKhfu8XB8Lu6xWD2oUXnOlNm\n9jz9RK+d5RDYiWeT+ESkP+CL9GzKftZcNZ5IbZb1rUjZp1WnpX4wSN4xszNPv1fzy4bMrBea\ntaX2mezqnphT6X0g77P09d1zzF2/PORX7TT36eSrTV3epu552b7a3i+u6VxXNvPTT/TaWQ6B\nnXgFq3Z/xk1c1j1uPPVA00PXvS/ZedDFXlWdlma8v2Nm1/Lsf2tnJ3l2Uq8fbJwTGyu9D+Sc\nYSBSqcflwzl/GfpEJt1Qd57d+nO5LO0fBLPf7Jg3qd1x8oleO8shsNuH5su4GyL9AS9xbSov\nT7M2+kxDke5m5eje/V5sbGnmXW+Z2Wke13ZgXu6dJnnr9y24VY6V/gcazzB8oLqbHz7oXzYI\nejQct5iK1JQ3/Skeu4Jd20cY++bRzV7DT/TiWQ6B3T6888uG4ynUe2Z2P4zkw9u+wIh0HDvm\n/rCNw0SqEOkods39YRtHiZQp5yuH2O6iJwMuybhgBlIcJdLV/F9OLewuejIAPwoiAQiASAAC\nIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAi\nAQiASAACIBKAAIgEIAAiAQiASAACINJxKIfxnwZr3Lq3/X9xA2HoqOPYJJKVyW85/v+gh7NA\ndxzHH0Vyb1t9BDgKkn8c8iKN7+jXfwwJP463itT/W0T08L+BNB/HvxKpj3H05/1qyO5x/HuR\nFPeod0FSj+M4kYYGIAXZPI7DRbL/gQTk8TjOIlLXGP4CGTyOk4nU7wSvQOaO44wi9RtgJ6Ts\nOE4skm3P4NgOuTqO84s07AtrkKTj+ByR1MdO9951GZhcXz4zPd/BB4r0acNFNe8552kyPi0z\n38RHimTP5tjEbeZt54lIZ+KDReoPcHLee4qIdBI+XiQ17nxO1NuekSaBz5uE7+d7ROrCnw71\nrmckG332DaxP/7NEJVndv3uqSW15U+pWRdpGjvVVIqmlwZRPtj1VrBjF3x7L6hC+3zh85MUh\nvnhW8chdsiaBEclnTaSLGUJp965OVFhbmGJST9rGjvV1Is3ZVE7y2qfOL0bxt8eyOoQPNq6I\ntHhWi5FZbFhjRaSnSsqmTNTTvr3a5m5t0hbrq8ombWPH+kqRIhlssxDWXseaa2yXeNN4Vofw\n4cZlkZbPajEyIq2x0qeZKtrXh7qbdw87At3ah1aoqVUSto0e60dEytUlrO1T5xej+NsjWR3D\nTzaqhdjLZ7USeRIYkXza9GQqMekq2iReClNqL1SJFqS9Yumnn1Jd9ZuqT7NTe1NlH8prGz/W\nl4oUDqr24hIMu2ocxNV0PM81ncnqGH415ZvPandkRPJR6qqHRq6vRqor3W0pa4Yblv1zUZUt\nOLWpau6JutVh2/ixvlWk8EFjkoUhdV4xSrA9ktUyrF2ZVmw5q92REcmnvQnVrUPtQ2Wi7y0P\nXVLqYe76jZ/Pe1s9EakTMWm25P5XRJpkYUydW4wSbo9n9RWRls5qf2RE8lHmcdIOjCLY0nj5\nNLf5iEh6seGm59K/LNLkM3s1TuqcYpTJdjerY72ESO6h9kdGJJ/RjKy9uZTdA09V3C+hSGlS\nR0XSu1T2RuZEjB/rR0VyUucUo0y2v08k91CI9FccM9pnnfb+op8xL8PITIZ83swNyzYfa92E\nJ6u5/1GRnNS5WYwx3R7Pavd2NeXbzuqVyIjk491iiizVt5abSvPCPobaxZuqnQQ4w86pbYtD\nGKd27li/KVL0s6ylqK+JZ7XbvprybWf1SmRE8vFEapy7jBXpbi5bhVkWHfM+1nbFSl282rlj\nIdJukeJZ7bavpnzbWb0SGZF8RpFSu1ZnHnaeTWmfkSZfnZvmTm37dFTrxYbHhl82/BKLaq1N\nxdzt8ax2LfamfFn4XZERyWcU6WEvT0+z7NAXW700F7+5W3sfi0Hbn6YfoP69flpc2tn+iWa1\nj7Az5YtntS8yIvk4UzvzywZzBbrpQmHmx7X5EXDQ3Kttd+uKQdufRk6kaFb7CDtTvkOktciI\nBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgA\nAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAA\nIgEIgEgAAiASgACIBCAAIgEIgEgAAiDSgSjzj9Lbl+4fr9X/mbeNWv2nVeFE0FcHskUku83U\nmMquy5DsZNAfB/I3keyGbgfFDexYSP6ByIukhn3g30LGD+SdIqku9IEf76cg0Qfyj0RSfXN4\nH2T3QP61SNyi3gd5PZCjRMIneUjogRwt0vgCf4U0HshJROLpSQByeCCnEqkPCy9B4g7knCLx\nBPUKpOxAziwSA2Mf5OtATi8STm2GJB3Ix4j0saNEqbc99anFt/Av+SyR3jcm38b4weRDI9J5\n+EiR1AfdoN4n0iQJH5OTb+RzRfqkYfOmc0Wk8/DpIjmneVZ4RvoFEOlfwB3pCFZHRZaoJKv7\nd89+iHWD2a0N20YO9jsi5ZNGXZLqm1K3ct++02BDTZ9ypzcWTs8PVMf6K4w8BEakRdZGxcUM\n2rR7VyeeSIlXG7SNHexnRConjfokJebkl0ya7DsNNtQEKR8/2IbIlT2VpNoSGZFWWBkVT5WU\nTZmop317dZsXQW3QNnawXxGpzULYqEtSpm765bpj32mwoSZM+bJIQaCbyvoTWo+MSCusjIpM\nFe3rQ93Nu4f7KFsn3Wjoa/220YP9iEi5uoSN+iQlqm4Wsz7ZdxpsrJmkXKn50GGgfsqmYk2m\nnYlIi7R5y1Ri0lW0SbwUptRePhN9uWqvo/rGX9oraOX1xFXZ2fVQ67WNH+xLRQor2kt90MhP\nXdNPimM5CvedBhtr1lO+EKiba7rnsiMyIvkoddWjMtdXI9WV7raUNcFV66KqsSdKs92tjVzh\nwoP9iEjlJAte6tpLVz6bo8m+02BjzXrKFwLdu6ndPdZkNTIi+bQ3obp1KNUXqFLfy1Nd9zBz\nkcbP572tHhPb35DGWkTy8uq+81L3UP01aC5Nkzvc9IgviDRpmOvVhiSPNkGknSjzOGnHZBFs\nabx8mtv8kNiye0Z1an9ZpLmR3ifLTV1+TRaeI2MZ7CvGDSIi2ZnH/bXIiOQzOpC1s7yyW5at\nivslFClNaiexWaedU4tIk7x2+Klr9ILZ/Nzun4mU6xtjbU8Fkf6K48Dd3On1M+alG6rjA6lq\n+75onMR2G9zaZLVXf1SkIHUt9cJqw4JIk5r1lC8ESs3kvA6++NsaGZF8vJtJkaU6rTeV5oV9\nOLaLN1U7NVEDumW/nOPWjm1nD/abIvmpm2yepmmzSOspXwgUv+lsjYxIPp5IXcm8syLdzcW0\nMMui7mjIu7mJWzu2nT0YItnvkaql33/sEGk95QuB7E0nvDlujYxIPqNIqV2rM6t2z6a0z0iT\nr867/F29H7nY2vVfNvwQc2Pf/JCgvr72jDSp2ZtyL1Cm9O/sssCVrZERyWcU6WEvmk+z7NAX\nW700F795P78OaoO2P02fKP9e3/S/tVtM0mTfWZH2ptyPfIntvDUyIvk4UzvzywZzBbrpQmHm\nx/YXwkHzsGu7d0Hbn2ZWJP2r6nTpfrRLpJ0pDyLHdt4aGZEABEAkAAEQCUAARAIQAJEABEAk\nAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQ\nAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQ\nCUAARDoUZf6JUv1q/vavqvsf3fMx0FOHsi6S6v95aFOwze2uh500RKA7DuXPIqkuhtnjmM8A\nGpJ/KPIiKZQ6BFJ+KG8WyWz5xx/pRyHNh/JPROp0oqvfCdk9lH8s0ngQEIakHsphIvEYJQz5\nPJSjRaL7pSCTh3IOkRgEf4ccHsqZRGLZ/C+QukM5pUiMiRcgaYdyWpEYFzshYYdybpH6c4N1\nyNKhfIhIHyuTet91QC2+hX/Lp4n0YcNl+BRvCI1IZwKR3sr7RJr8PuTDMvNtfKpI6oNGzptO\nFJHOxGeL1DU68xjiGek3+BaRzjyMuCMdweoFLEtUktX9u6ea1JY3pW5Vtz1fCfczIvmJqHWO\nSltuU3cpFjMULL15CQ7D9x3h7DN7brFAz7B1GHkIjEiLrIl0Md2adu/qRIW1hSkmnVRr4X5F\npCARifk45Zi6+8KOgUhegsPwfveMnyJGLFDfoauREWmFlZH/VEnZlIl62rdX29ytTdpifVWZ\nrm+rflCkyEcOEpGpm365NvqCf6n1DaqcSVBpWrm4CQ7DB92zLFIkUN+h65ERaYWVkZ8pPQt5\ndFfQRzfKndqH6ZhaJY0ZJIjURBKRqLo/hYsZmpU/nL1dg5uVm+BJeL97Gm+GFxIJNHTohsiI\ntEybt0wlJl1Fm0Q7ey/aC1Viuvqq9Jy6u0xWfZqdWvfa2vYUIs0nwgzh/pCXmQTlKvcrIjev\nMbzXPStEAlXhlW9HZETyUeqqh2yuu1B1pbstZc3Y7+bPRVX94B1qU9XcE3UzE+9y9Qb3rSIF\nnzqaiMwo4iU0wlUVt+4iZnETPAm/Fs0lEmjo0BciI5KP0nP2XD9UJvqK9dAlpR7mrt/4+by3\n1ROROhGTIdza4X5BpEgiHvbC1A5nfaV/LohkGG9YQYL98HtEmgYaO/SFyIjko8yc3Q6eItjS\nePk0t/mISPoR9tbPpX9VpMnHDmvya2JydFfXuinnnyXNRazOxgmem+Bxr5dE8nrK69AXIiOS\nz2hG1l6yym4aXRX3SyhSmtRRkfQuVb9OikheXj1uxg6zEh4uloXUzoq2k+A/iuT1lNehiPRn\nHDPuuo8TPfO4DN9lJEM+b+aGZZuPtWHCEcnLq4ddMGvvCcl9Q56CUrBD9zaJblwMObR1O/SF\nyIjk491iiizVF6ybSvPCPobaxZuqnQSoAbe2LTphECnI61xd6X7TGc9TXwoS7G93OmKVMJDb\noS9ERiQfT6TGuctYke7mslWYZdEx72NtV6z6p2NE8vLaYb9HquySji7mswO0bzpsDxLsh3c6\nYpUw0LJIq5ERyWcUKbVrdWbV7tk/D0++OjfNndp2dJhv6h/udvATYX7ZUF/1M5IpPtM+XxMy\nPXTrbFz4CRLsh5/+/mCexUD7I9PRPqNID3t5epplh77Y6uWtxnZpdmrvXgNE6vDnusmQpNoW\nZ2dM3fZs3NdPcBA+jW2cwQk0eQTbH5mO9nGmduaXDeYKdNOFwvR2bX4EHDT3atvdxgaI1BE8\nNLb5Su2KdtUm97rw6+96bNo/uV68b2i98EH3LDMG2iDSWmQ6GkAARAIQAJEABEAkAAEQCUAA\nRAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAk\nAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQ\nAJEABECk4xj/PXr/77iF7vkY6Knj2CSS8mmsXYp/5flk0B/H8TeRJpvoyiMh+8chLZJbd/Rn\n+znI+HG8VST3Fd4PWT6OfyXSGB/eBsk9jn8uEjq9D9J6HIeJNIQBMcjmcRwt0vAe/g5pPI7z\niIRPf4b8Hce5RBrewCuQueM4o0j9IY/OzcdBxo7jzCLxM6SdkKvjOL9IYwtYgSQdx8eI9Lm3\npzdeBtTiW/iHfJhI3dZPQg0v8qER6TR8pEjd/z6D94k0uaR8Skq+kY8W6XPuTm86TUQ6DZ8v\n0vlvT+87PUQ6Dd8i0rDHOeGOdASr4yFLVJLV/bunCmtrp4HfNnKsrxJJLSUv9zflqZOZfCnn\ndSSJ0x2Gmr61Y/Zs9PKm1K1yKtqdL8VK5CEwIi2yJtLFDJm0e1cnKqitElNMqknb2LGC4f8F\nIs3kr/TrM5ukOrbNx83nTDC3Jkj54mJD4Z7EuPN9U2REWmFFpKdKyqZM1NO+vdrmTu1NZY0e\nKLdJ29ixguH/FSLFEthmwa0u1a3W1/pbZFuAk8+ZYG5NmPJFkZK2bX018Q25utRNfVPllsiI\ntMKKSJnSt/5Hd9l6dMPGqe3v+ypsGz3Wl4oU5rAdol5er2OSJtvi/eE0me4w1kxSruJiN6aV\nVqhWSV9xMZZUo1mLkRFpmTZvmUpMuoo2iXbKXLR3nsTk96r0HKNUV/2m6tPs1HZzPd09Xtv4\nsX5EpHZsxsazPXR8W4+Tz9lgY81qyh38W08zOnt5KTIi+Sh11WMi11cj1ZXutpQ1wQXyoipb\ncGrv3VTkHruYTo4VDP9vFamMZqE2Qza+bcDJ52ywMsz1yrTCkqrmnpg55tAdk513REYkH6Un\nyrl+qEz0FeuhS0o9zDSu8fN5b6snIjW5fjpO8gaRgs86qcpVMbvNaTXkcynYCyJ1l8zxVpea\nm84zvN0h0ksoM1G2I6UItjRePs1tfiqSvX3dG0SaZs+jSq6z2xzGfE6DjTu+JJJebLiNoe/q\nWjflxe3nHZERyWc0I2svWWU3ja6Ku5dg/SdN6ohIuZ6KtN2TI1Ikrw51cpnd5uDkcxrsjyLp\nzq2cFW2z0n5FJBEcM+5mUjF+IaTrkiGfN3PDss3H2nZ+oCfdtZ0bOhHjx/pakebmXiOXdH6b\ng5PPhWB9zWrKp7s4bVtfk3u489bIiOTj3WKKLNU9eFNpXth1Bbt4U7WzOneUjbVu9zi1c8f6\nWZGq9FLNbYvt5zeZFWk15Q7XuBpl8A361siI5OOJ1PRPS43OoX0C0vehwiyLjqNsrO2uXObb\nCad27li/KlLhrjEviuTkc2mHrmY15Q62bTWeSWJufnngytbIiOQzipTatTqzavfsH0InX52b\n5k5tpvQvvTKd8NVfNvwS3tivPI8WRXLyubRDV7Mn5e3Tkfkhw2M8lP45SjpW7IqMSD6jSA97\nmX32vwyz63mpKYXPyU7tZSwGbX+afuybvzfnZtgsi+Tmc2g3K9KulN+DyLX9WV8wedsaGZF8\nnKmd+WWDuQLddKEwKbY/Rg6ae7VjMWj703jSqB0iOfncINKulLcd7EWu2n6+hr/+3hoZkQAE\nQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABE\nAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQA\nARAJQABEAhAAkQAEQCQAARAJQABEOhBl/pvUqSZSD+eG/jqQjSL1bxTddV7omQP5u0iq+x/9\neDR0wIG8QSTdRjEx/PeQ8QNBpO+BjB/IvxKJyd/7Ib0HcoxIdPk7IKsHcqxIdL0kZPNAEOl7\nIJsHchKRGAMCkMQDQaTvgSQeCCJ9DyTxQBDpeyCJB/IBIjE+NkKiDuSDRPrQcaJ0Mt4UevEt\n/EsQ6c34H1w2NCKdh88U6W3XeHneJ9Kk4z4nKV8IIv0L3nS6iHQeEOlfgEhfDyL9A1hsOITV\nUZIlKslq21T1i0L1TalbObR5qrDtzMF+R6R8slOXJCeL2/YNch006VM+xJyPXAfdo1TsXMLI\nc4ERyWdtlFxMqtO2VDppT0yp7906UUHbuYN9o0jRreVkpy5J5bpIwb5+roMmk5TPB65soKQa\nmnYkL0VGJJ+VUfJUSdmUiXrqFF/72kzd9Ev//mqDOG3nDvYrIrVZCKu7JDlZnCHYN8i132SS\n8oVTvamsD+dSeP01HxmRllkZJZkq2teHuut7/r2vTVTt7ProLrBO27mD/YhIubqEO/VJyhfS\nE93Xz3XQJEz50t2un6H5m+vEVXQ+MiKt0OYtU4lJV9Em8VKYUnv5TPT1q72O6pmAuYzmKg92\ntXOCqk++03buYD8iUnvtD3YakjTJ4vq+pjaJNllP+Ug3//Zncm0E95l2R2RE8lHqqi9iue5i\n1ZXutpQ13mXsqopbp5ch64bERVW2VfyS5x3sW0UKWpSTnYYkhVmcMN23GXMdNllP+ci9m9p5\nd8RSeaeyIzIi+bQ3obp1KNVXrFLfy1Nd9zBzkSYQyXCx+z1U1wX3tjEiTWu9qjFJfhZnIk6n\nhaF5L4jU5Hq1IfFviP4NaU9kRPJR5nHSjGUzK3a3NF4+jV51f23Mr4m5tpmbPyJNa90qL0lu\nFmciBuH6XDsbXhHJTjSCG9ItPCQivcboQNbO8spulbUq7pdQJEs9roje9GhIkxqR1kRykmSp\nl74kiB3Q5PpvIuX6tlbfPIWz/uKJSH/FceCe9F8zXIbVn2SST9eppO3gYqibtp0c7DdFcpO0\nJWZkYx2sEXRN1lM+kppZnK9wMnviq5ERyce7mRRZqvN8U2le2Idju3hTOYs3vlPDt+Mq1nZy\nsN8UyU3SlpixjUGdt7a2lPLJLm6g2KLc1siI5OOJ1JXMOyvS3VxMCz0rsF9omNT2xdQbI07b\nuYMhkpPF+YhuuDHXkSbrKXcD6Vfv1hZbi98aGZF8RpFSu1ZnVu2eTWmfkZwvuDMzxzazavP1\neH0dusEGWf9lww8xtc/WOFnctm+Ya6/JnpS3R6674/dcVfgbvu2REclnFOlhL5pGmaHY6tWv\n1tbJ8O1S9/uvix/EaQu9DOFTvJvFbfsmsaz2Tfak/DK2HTps+gvjrZERyceZ2plfNpgr0E0X\nCjP9sD8ZNk11Me0ujE5xCOK2/XnmRPKyuG3f2A59k10pH9surMltjYxIAAIgEoAAiAQgACIB\nCIBIAAIgEoAAiAQgACIBCIBIAAIgEoAAiAQgACIBCIBIAAIgEoAAiAQgACIBCIBIAAIgEoAA\niAQgACIBCIBIAAIgEoAAiAQgACIBCIBIAAIgEoAAiAQgACIBCIBIAAIgEoAAiAQgACIBCIBI\nAAIgEoAAiAQgACIBCIBIp4cu+gTopdOz0kUq+o9xwz+GPjg9m0WiLw+E5J+ev4hE9/4ryPTp\nQaRPgEyfHkT6BMj06XmPSHS8LOTz9CDSJ0A+T8/bRWIMCEASTw8ifQIk8fT8Y5GU+Q92QspO\nzxEiMSz2QsZOz5EiMTy2QqZODyL9hbd8AqUm9+zPz9TXg0h/4C2TVDW8BHVwZk4m0kc9P71n\n3QSRPhJE+guIBB1nF+ncQwiRoAOR/gIiHcPqxCVLVJLVflH1tOV6bDBW/u2U1jZ/iki53zKW\nydjhpknMJ4ccavpIwz4rZzd/TvOxEWkDa+P+Yro19Yu9R0nTVIktVU1TIpJH6e8ZzWRsp0kS\ny8kpDDXTSMtnN39OS7ERaZ2Vcf9USdmUiXp6RUuhizeVtcVM3XQPXGVOaW3zZ4jUpsptuZRJ\nd69pEoNAbk0k0uLZzZ/TYmxEWmdFpEwV7etD3b2ioU6u4/76Tz5s+uMprW0+m0jRg+bq4u25\nkEl/t7A2DOTWRCItZW/hnJZj84XsKm2CMpWYTBZtCi+FKV3byZq+0zRX1c7Z7GXSKTZ2k54+\nJyH87UoAABaFSURBVJ1Iie6DXOaU1jZ/hEjtndrbcyGTLtMkhoHcmqVI+85pd2xE8lHqqi83\nue5C1ZXutpQ13g3HKWpKs71ta6d2d5354tb597dTWtv8ESKVwZ7zmfSYJjEM5NYsRdp3Trtj\nI5JPexOqW4dSfW8p9Z081XUPXVLNUvfbG1Lrn15tSHJTY7j8+ZTWNn+ESOGem0WKJHHa8jWR\nFs5pd2xE8lHmSdN2bxFsaRa6v9TLCxp7+7IT6YdeDf/zBO+nRYomcfjAKqgRFGlvbETysXnS\nr1k7yytLW1sV98uySFmnXa6ndvVt7Pg6vqy755TWNn+zSJYgiYh0fkaRmnvSfSHUfYOg65Ix\nm4mX2O5dk5oZntvxO3p15pTWNp9ZJPdbIG/P2UzGz1bNv3NqNkRaChzffVtsRPJxRGqaIku1\nETeV5kVl6uzSTTWuNVV2FWdYzJletxApKtJcJmfOdptIGyItBY7vvi02Ivl4IjXO1MOKdDcz\nuELP35yis0prr1u1Xv5OzM1pT6/OnNLa5jOLNLvnXCYDokmcFWkp0r5z2h0bkXxGkVK7VmdW\n7Z5NaZ+R5r6Pv6ruaSpT+tdYmU63eamzYM3ih/EG7cZfNkSTOCvSUqR957Q7NiL5jCI97Kzk\naZYd+mKr17Aa6xS7RyPNZaitk+HrJ9D4U7yZTAa4SRzvk3MiLUXaeU57YyOSjzO1M79sMNef\nmy4UZnphf9xtmjpFt2v9BqnMrxu+An/QzmQyxEniBpGWIu07p72xEQlAAEQCEACRAARAJAAB\nEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACR\nAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlA\nAEQCEACRAARAJAABEAlAAEQCEACRAARAJAABEAlAAET6HejrN0Jyfwf6+o2Q3N+Bvn4jJPd3\noK/fCMn9HejrN0Jyfwf6+o2Q3N9hT18zLnZCwn4HRHojJOx3+EmR3vJBVMu/OA6ckl8UaTLg\nRYIOL0Ed/AQyIn3UiFFvOV1E+m1+UKT3nC4i/TaIJBkTkX4WRJKMiUiLrD6cZolKstovqp62\nXMcanINjRZoko624FJPi5HBDZmcDteQq2Djss3K6ub99U2xE2sCaSBfTralf7D1KmqZKbKny\n256CQ0WaJMNW3INiSDkRKZbVsm8w3bh8uqXf3xtjI9I6KyI9VVI2ZaKeXtFS6OJNZW0xU7dI\ng6M5UqRJMnJ1qZv6pkqvOKFU15VAulHSdVtk4+LpDjvui41I66yIlCk9A3noq6dTNNTJddxf\n/wkbHM6RIk2ScTHDtdLXHac4IQ/TF8lqK2KX9sjGpdMdd9wZmy9kV2kTlKnEZLJoU2hn7sW1\nnayZbr4qPWczl0mn2NhNevqcdCIl0waHc6RIk2T0V5yLV5yQq3wlULtj1ofYmfJxx7/HRiQf\npa76cpPrLlRd6W5LWePdcJyiprQX1Hs3tbtPGhzPkSJNkjGfSY+rKm7dRWwmUJv6sHZrysug\n6V9in6aXT4LS8/VcP1Imes7+0CWlHrqkmqXutzek1j+92pDkDSL5e4XJSM0F/qkrnOKEq72I\njTereFZfEyls+pfYp+nlk6DMfN2KUgRbmgWRSr28oLkPK1CI5O4VJuOurnVTmgcQpxjZ76G/\nUhgneG6g6P2tkRFpb+zT9PJJsHmyawXqWnYLSVVxvyyLlHXa5XpqV990xyOSu9ckGeZ7gqup\ncIpx6nFJGpE+g1Gk5p74XwjZLh9aJV5iu3ftNEXP8Go7N3QbnIAjROozN01Ge7VJ7rbCKc6G\n6UvxrPq98KpIf4l9ml4+CY5ITVNkqTbiptK8qEydXbqpxlW7yq7iDIs5znXLa3AGjhRpJhnl\neKspl766HodvPJC3srYn5ZFVu9diI5KPJ1LjTOKsSHczgyv0/M0pOqu09rpV6+Vvr8EZOHJq\nN0lGYm7duR6XTnGC3eYM33hWuw7bn3JPpL/ERiSfUaTUrtWldgGiexae+2XDtf9aPlP611iZ\nTvfpftlwJJNk2B9/mCQ7xQkmk3U2LvzEs9oN9v0p90T6S2xE8hlFethZydMsO/TFVq9hNdYp\ndo9Gmku8wc/jJMM+F9kfJV794oRuWzbsF89q78PulPc7/jk2Ivk4UzvzywZz/bnpQmE73fwE\n2DR1iu6FLd7g53GSYZNVtVm92huNU4zul+bOftGs9vnfnXJfpD/ERiQAARAJQABEAhAAkQAE\nQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABE\nAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQA\nARAJQABEAhAAkQAEQCQAARAJQABE+jbU8AL/EDL+bSDSIZDxbwORDoGMfxuIdAhk/NtApEMg\n498GIh0CGf82EOkQyPi3gUiHQMa/jb+JxHh4ERL3bSDSIZC4bwORDoHEfRuI5PGWD6Ra/sVx\n4EAQyWUy4EWCDi9BHXwRiOSg3vKBEOkX+Ccifc4aOyLBayCSByLBayCSByIdw+rDaZaoJKuD\norKY2jwdGrQ8/3mCzytS7u/gZHKyzWeSxEnFsHsfdOiOlZNcOqe52Ii0gTWRLsaY1C+WjkiZ\nKSVdZ9QJIvWUfm6dTE62+UySOKkYdveCjmf60jnNx0akdVZEeqqkbMpEPb1iqa59g1Ldan0R\nu9m317csvy5yVpHaVLk7OOmbbAuYJDGsGHb3gjpn+so5LcRGpHVWBn6mivb1oe5eMdcvlqty\nwzymX9y9nZOKlKuLlwsnfZNtPpMkhhXj7m5Q90xfOKfF2Hwhu0qboEwlJpNFm8JLYUrtJTDJ\ndOmqqqa7AznFXOXTMC3V4gB5EycVSWX+RcpJ32SbxySJk4pxdzfoH89pb2xE8lHqqi83ubZD\ndaW7LWXNcKfRf5ziVRW3zjRLrS76z0VViNRRBnd7J32TbR6TJE4qSj/a9p8zLJ3T3tiI5NPe\nhOrWofaRMlGlvpOnuu5h5hPNvEiGyxAlN9OAe7sfIjk7LQza2TxNkhjN6msirZzTrtiI5KPM\nk6YVpQi2NHMiGdPqbJjgVYm+/5tZACI5O70g0iSJbsVYLy/S3tiI5GPzpF+zdpZXlra2Ku6X\nBZEsdb9uWifm3pQmNSJ5O70g0iSJbgUinZdRpOae6C+E9BPmZfiaKBmzmUwSO3zhYIS6mTva\nr4vkfFPt5yJI30yeJkmcyWpXMe2T9fObP6ddsRHJxxGpaYos1XeZm0rzwj7g2qWbaly1q5xV\nnG6pLr1U9l3PP/4Ew8vLewseZFakIH0zWZokcSar3spatXXVbuWcdsVGJB9PpMaZxFmR7uaC\nWOgVPKeYqLrpc1z0aw6INNnJ3cFJ32Sbu8sukYKgfzynXbERyWcUKbVrdaldgCjtM1L8lw2Z\nzm5tvrGrnLW7MR5ovFwEvyJYzNPcTG5SEftpwuvntCs2/ewzivSwF75n//M5u56XjgvdY7FO\nhi+abpMr6DGf44z4Uzwnk802kSbP/9PQftA/ntOu2PSzjzO1M79sMNefmy4UZuZWm58Am6ZB\nMc3tnog0hz9onfQ1YiL5Qf94Trti088AAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgA\nAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAA\nIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEgAAiASgACIBCAAIgEIgEifhGro\nsZNCt3wSMyLh1/GQ/09io0iKfv3nkPBPYodIyPRvIdmfBCKdFpL9SQiIRIe/B/L6SSDSaSGv\nnwQinRby+km8TSSGwV8hg5/Eu0ViNLwMqfskEOm0kLpP4t+JxOr5TsjWJ4FIp4VsfRLHiPTR\nY+QtJ6+UCuN+dJJ+jhOI9GEDZjLgRYIOL0EdfAjnEOmDxox6y8ki0qdzIpE+ZeAgEkxBpN0g\nEkw5q0hveRKRAZGOYXVIZIlKsjooKkvXJO+HpVcrcnbNmUVa+6S53yBP+/TVN6VuZXyn6bY2\n7ZdiLnTfJ0PiV04qOCend+djI9IG1sb9xbiR+sXSU6bsCiUieZR+KjKTnEQPzMQU4yZNttm0\n3+Ohne6JnfzaOU13j8ZGpHVWxv1TJWVTJurpFUt1HZu0Vb1I15kor59d87kiDXnp3qpbra/2\nN62UeYlma7ItV5da36Zc7YbQTp9ET37lnCK7R2Mj0jorImVKTyoe+oLoFHPnAtl2dBcjDy6b\nEmfXfKxIY14sVzXsmai6mcv8ZNvFDPNKZbHQTp9ET37lnCK7x2PzhewqbYIylZhMFm0K7WS8\nuLZzENNzV1U13b3GKeYqHwNkzSBSHkb/89k1HyvSmBe/ejxKsrDzuK1/9LnEQjt9soXwnCK7\nb46NSD5KXfXlJtceqK50t6WsGbtRecWrKm6daU3ZjJkfa6XOrvlYkca8uNSDENnCZcfd5qR9\nGjqycc85/SU2IvkoPQXP9SNloqfhD11S6qFLZrjOiGS4DDHMn6BW5OyajxWpiY7BXNn1t4dS\ns1ccf1tqbgzPINZrIoVN47sj0isoMwW3ohTBlqaZEcmYVg/XzWGLVytyds2HiBQ/9KS2SrqJ\nUn5NZp8o/W13da2b8uL2RtO8QaS9sRHJpxsSyizPXstubagq7pcFkSx1v27q5boOVlP/dnbN\nV4lUJ87t+rZwxXG3mfXwKyKdnFGk5q67LNETicvwhVAyZjOZJLYvqsggEzq75tNEcr9Km2Ti\n4l5j6oXVBndbfdOLQfHp17RP1vCaxnffFhuRfByRmqbIUn0/uak0LypTZ5duqnHVrnJWcRBp\nepQFkar0Uk2DxAm3lcF93ltZq3Z8gRdZtQt33xYbkXw8kRpnEmdFupsHp0I//TpF+2XHkOPh\nEubVipxd82kiTVv1FMMyTJ+o2Bx4ss1W5PHB7vTJRrxziu++LTYi+YwipXatzqzaPfvH2/gv\nGzKd3TrrVye6zAe14A3aalzONL9eqK/RZ6TJNlPxNJ0zDR37acKOc4rvvi02IvmMIj3srOTZ\n/yrMruel45L2WKyT4YumMUZYC94U76bGnyImC98TONtM2y6rwW2+D50uRFo/p/ju22Ijko8z\ntTO/bDDXn5suFKbzavMTYNM0KKa5HyOsBW/QKkekZilR4zbbtmp74xre5d2U7/oO3Bcpuvu2\n2IgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiA\nSAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgE\nIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAiAQiASAACIBKAAIgEIAAi/RR097sg\nsz+Fcl5BEnL6UyDSuyCnP0VMJMUoEIAU/hSI9C5I4U+xKhLj4UVI3E+xXSQGxj7I10+BSO+C\nfP0UvyaSavk3Yb8jX7CRHxNJDS9vD/sV+YKtINK7wn5FvmArPyaSAZFAHEQSi4hIv8wPivSG\nD4JI66wu8mSJSrK6f/ccmud9KU+HBk7xJBwqUu7HLG9K3ar4tpFatyrdCr8Hgt37jcO62srn\nCI67KTYibWBNpIvuIZV27+qkb172O2amQVL7xbPwfpHm9yv93BZudsrZvCem2WhSZSuSym00\n7O53z+LpxI472T0eG5HWWRHpqZKyKRP1tG+vffO2ypZKdav1RezmFU/DgSINKepI2kzWV5XF\nto1kOn2Zug4VN7NH5mV12D3onnWP/ONOdp+JjUjrrIiUqaJ9fai7effo5w+5unSlqxrCOMXT\ncJxIY4osDyNErZLINodEmbmUc4LTrI67+91jJ3gL+Q+PG+4+G5svZFdpE5SpxGSyaFN4KUyp\nvfMk5tJ5VXpKUdorZDUkuR0UfmanPX8OjhMpTNHNma5N0jfZORmK3WTaqXF297pnw7kGx43s\nvjn2mXr5DCh11ZebXF+LVFe621LWBNfDi6r6CV3gS60ukeIJOE6kMEWpau6JmfxOt4Vkuhc6\n7t3UzrlrlEG/bL52hceN7L45NiL5tDehunUo1Ze+Ut/JU133MNO4xs/mva0e0+olODfTgLB4\nAo5cbAgHrbliJbFtPg97DevJ9WpDkvttXhNpek6x3RHpFZR50tTZUoEBoUjmJh8XqUqukeIZ\nOJNIerHh1t9Zlp5krol7/7Hzg7u/k7xIe2Mjko/Nk37N2mtm2c3jq+J+CUVKk3pGpDq5RIqn\n4Ewi6dxW/Vrz8ti/jXO7XN+ealuBSOdlFElP4bvvKy7D6k8yZPNmblhRkS5prHgK/rlI7rrZ\n0jRqeezX49pCapbx6uDLnm735G8ixXffFhuRfByRmqbIUt1hN5XmhV1XsEs3VTurUwPujnpj\neqmmxZNwHpGue0SavT8EtWP3bD5d/5xiu2+LjUg+nkhN/7TU6Ayau5S5DxVmUTQuUjGu0hWn\nWrAznGdqZzNZ9SmaE8l+j1SNNyB7Y6jd5e9x97F7Np/u9JzC3bfFRiSfUaTUrtWldgGitM9I\n4XffkytlNcpTnc+jY/EGbetGrZ91HpFtDuYnDPV1fEbKlP65WxYf7LGfJuw4p/ju22Ijks8o\n0sPeb579b+bsel5qShe/uVO6jfepm3/LAv/mffcyOZulZGxm21z8HvB3T2Mbt59TdPdtsell\nH2dqZ37ZYK4/N10ozOzY/vY4aO6UnAmfQiSfYBbc5nfM5HyW2oSnudvG7wF/9zq2cfs5RXff\nFpteBhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAE\nQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABE\nAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQAARAJQABEAhAAkQAEQCQA\nARAJQABEAhAAkQAEQKTj+O/9e/yDQ7ywxzeCSMeBSF8EIh0HIn0RiHQciPRFINJxINIXgUgA\nAiASgACIBCAAIgEIgEgAAiDSUSQt21olTuvwb9i8iTVa2mncY9th1mJv+1jfByIdRDK8rDcb\nW4d/w9bJUuPYTt2o33yYtdgbP9b3gUgH8Q6REqd+2yBP+jvS1sMg0gyIdBDbRlzi/t0wXnff\nkfzKrTc+RJqASAexUaT+2aVr/S6R9h0GkaYg0kHsuCPtGK9/uSNtPsyScogE/5YdI+79Iu07\nDCJFQKSD+FyRXriH/QCIdBAfO7VLXjqxrweRDmK7SEszqfmou0XaeJjE3w2ROhDpKLb/smHp\nb9h8904790jWfgPBLxsA4GUQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQ\nAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQ\nCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUAARAIQAJEABEAkAAEQCUCA/wG7W3eo\nrsX6bAAAAABJRU5ErkJggg==",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# select the top five up-regulated pathways \n",
    "topPathwaysUp <- fgseaRes[ES > 0][head(order(pval), n = 5), pathway]\n",
    "# select the top five down-regulated pathways \n",
    "topPathwaysDown <- fgseaRes[ES < 0][head(order(pval), n = 5), pathway]\n",
    "# concatenate them for visualization\n",
    "topPathways <- c(topPathwaysUp, rev(topPathwaysDown))\n",
    "#Viewing the 5 most significantly up-regulated and down-regulated pathways each with the FGSEA internal plot function\n",
    "plotGseaTable(pathways = KEGG_hallmark[topPathways], stats = stats, \n",
    "              fgseaRes = fgseaRes, gseaParam = 0.5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29cbb971-2f26-42fa-92d1-3f7b8e6347ef",
   "metadata": {},
   "source": [
    "### Enrichment analysis using FGSEA and REACTOME pathways\n",
    "We can perform enrichment analysis using FGSEA with REACTOME pathways using the same procedure mentioned above. The only thing we need to change is the\n",
    "list of genesets that are available in the REACTOME database and the gene id (from SYMBOL to ENTREZID)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "b48340bd-3c93-46fa-88eb-6d48239849f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>780</dt><dd>0.25312654919025</dd><dt>5982</dt><dd>0.467472580609517</dd><dt>3310</dt><dd>0.986153555292704</dd><dt>7849</dt><dd>0.723652752171903</dd><dt>2978</dt><dd>0.335147977832318</dd><dt>7318</dt><dd>0.392525231374342</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[780] 0.25312654919025\n",
       "\\item[5982] 0.467472580609517\n",
       "\\item[3310] 0.986153555292704\n",
       "\\item[7849] 0.723652752171903\n",
       "\\item[2978] 0.335147977832318\n",
       "\\item[7318] 0.392525231374342\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "780\n",
       ":   0.253126549190255982\n",
       ":   0.4674725806095173310\n",
       ":   0.9861535552927047849\n",
       ":   0.7236527521719032978\n",
       ":   0.3351479778323187318\n",
       ":   0.392525231374342\n",
       "\n"
      ],
      "text/plain": [
       "      780      5982      3310      7849      2978      7318 \n",
       "0.2531265 0.4674726 0.9861536 0.7236528 0.3351480 0.3925252 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Get the gene list and their statistic from DE results\n",
    "stats = DE.df$adj.P.Val\n",
    "names(stats) = DE.df$ENTREZID\n",
    "head(stats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "ab574490-afc7-428d-8bd9-fc0a2514777e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$`R-HSA-164843`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'8815'</li><li>'155348'</li><li>'155030'</li><li>'11168'</li><li>'7518'</li><li>'3981'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$`R-HSA-73843`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'5634'</li><li>'5631'</li><li>'221823'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$`R-HSA-1971475`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'221914'</li><li>'27087'</li><li>'10675'</li><li>'1464'</li><li>'3339'</li><li>'633'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$`R-HSA-5619084`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'122706'</li><li>'5716'</li><li>'26154'</li><li>'5826'</li><li>'5682'</li><li>'8647'</li></ol>\n",
       "</dd>\n",
       "\t<dt>$`R-HSA-1369062`</dt>\n",
       "\t\t<dd><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'20'</li><li>'64241'</li><li>'225'</li><li>'64240'</li><li>'10347'</li><li>'21'</li></ol>\n",
       "</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$`R-HSA-164843`] \\begin{enumerate*}\n",
       "\\item '8815'\n",
       "\\item '155348'\n",
       "\\item '155030'\n",
       "\\item '11168'\n",
       "\\item '7518'\n",
       "\\item '3981'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$`R-HSA-73843`] \\begin{enumerate*}\n",
       "\\item '5634'\n",
       "\\item '5631'\n",
       "\\item '221823'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$`R-HSA-1971475`] \\begin{enumerate*}\n",
       "\\item '221914'\n",
       "\\item '27087'\n",
       "\\item '10675'\n",
       "\\item '1464'\n",
       "\\item '3339'\n",
       "\\item '633'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$`R-HSA-5619084`] \\begin{enumerate*}\n",
       "\\item '122706'\n",
       "\\item '5716'\n",
       "\\item '26154'\n",
       "\\item '5826'\n",
       "\\item '5682'\n",
       "\\item '8647'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item[\\$`R-HSA-1369062`] \\begin{enumerate*}\n",
       "\\item '20'\n",
       "\\item '64241'\n",
       "\\item '225'\n",
       "\\item '64240'\n",
       "\\item '10347'\n",
       "\\item '21'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$`R-HSA-164843`\n",
       ":   1. '8815'\n",
       "2. '155348'\n",
       "3. '155030'\n",
       "4. '11168'\n",
       "5. '7518'\n",
       "6. '3981'\n",
       "\n",
       "\n",
       "\n",
       "$`R-HSA-73843`\n",
       ":   1. '5634'\n",
       "2. '5631'\n",
       "3. '221823'\n",
       "\n",
       "\n",
       "\n",
       "$`R-HSA-1971475`\n",
       ":   1. '221914'\n",
       "2. '27087'\n",
       "3. '10675'\n",
       "4. '1464'\n",
       "5. '3339'\n",
       "6. '633'\n",
       "\n",
       "\n",
       "\n",
       "$`R-HSA-5619084`\n",
       ":   1. '122706'\n",
       "2. '5716'\n",
       "3. '26154'\n",
       "4. '5826'\n",
       "5. '5682'\n",
       "6. '8647'\n",
       "\n",
       "\n",
       "\n",
       "$`R-HSA-1369062`\n",
       ":   1. '20'\n",
       "2. '64241'\n",
       "3. '225'\n",
       "4. '64240'\n",
       "5. '10347'\n",
       "6. '21'\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$`R-HSA-164843`\n",
       "[1] \"8815\"   \"155348\" \"155030\" \"11168\"  \"7518\"   \"3981\"  \n",
       "\n",
       "$`R-HSA-73843`\n",
       "[1] \"5634\"   \"5631\"   \"221823\"\n",
       "\n",
       "$`R-HSA-1971475`\n",
       "[1] \"221914\" \"27087\"  \"10675\"  \"1464\"   \"3339\"   \"633\"   \n",
       "\n",
       "$`R-HSA-5619084`\n",
       "[1] \"122706\" \"5716\"   \"26154\"  \"5826\"   \"5682\"   \"8647\"  \n",
       "\n",
       "$`R-HSA-1369062`\n",
       "[1] \"20\"    \"64241\" \"225\"   \"64240\" \"10347\" \"21\"   \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load the pathways into a named list\n",
    "REACTOME_hallmark <- gmtPathways(\"./data/REACTOME_pathways.gmt\")\n",
    "# Show the first few GO terms, and within those, show only the first few genes.\n",
    "tmp = lapply(REACTOME_hallmark, head)\n",
    "tmp[1:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "8c812668-2971-4e80-b10d-7986b290574c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Running fgsea\n",
    "suppressWarnings(\n",
    "    fgseaRes <- fgsea(pathways = REACTOME_hallmark, stats = stats)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "2d20e0c2-f9ec-46a5-98b3-d7e4aad75cd2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>pathway</th><th scope=col>pval</th><th scope=col>padj</th><th scope=col>log2err</th><th scope=col>ES</th><th scope=col>NES</th><th scope=col>size</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>R-HSA-6805567</td><td>4.182288e-06</td><td>0.01056446</td><td>0.6105269</td><td>0.3884187</td><td>1.607719</td><td>140</td></tr>\n",
       "\t<tr><td>R-HSA-381753 </td><td>1.530411e-04</td><td>0.15176876</td><td>0.5188481</td><td>0.3841968</td><td>1.549905</td><td>101</td></tr>\n",
       "\t<tr><td>R-HSA-9752946</td><td>1.802479e-04</td><td>0.15176876</td><td>0.5188481</td><td>0.3910022</td><td>1.569605</td><td> 94</td></tr>\n",
       "\t<tr><td>R-HSA-500792 </td><td>2.529114e-04</td><td>0.15971358</td><td>0.4984931</td><td>0.2906890</td><td>1.282994</td><td>453</td></tr>\n",
       "\t<tr><td>R-HSA-6809371</td><td>5.174794e-04</td><td>0.21544769</td><td>0.4772708</td><td>0.3667173</td><td>1.482304</td><td>105</td></tr>\n",
       "\t<tr><td>R-HSA-375276 </td><td>5.724316e-04</td><td>0.21544769</td><td>0.4772708</td><td>0.3243598</td><td>1.374254</td><td>195</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 7\n",
       "\\begin{tabular}{lllllll}\n",
       " pathway & pval & padj & log2err & ES & NES & size\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <int>\\\\\n",
       "\\hline\n",
       "\t R-HSA-6805567 & 4.182288e-06 & 0.01056446 & 0.6105269 & 0.3884187 & 1.607719 & 140\\\\\n",
       "\t R-HSA-381753  & 1.530411e-04 & 0.15176876 & 0.5188481 & 0.3841968 & 1.549905 & 101\\\\\n",
       "\t R-HSA-9752946 & 1.802479e-04 & 0.15176876 & 0.5188481 & 0.3910022 & 1.569605 &  94\\\\\n",
       "\t R-HSA-500792  & 2.529114e-04 & 0.15971358 & 0.4984931 & 0.2906890 & 1.282994 & 453\\\\\n",
       "\t R-HSA-6809371 & 5.174794e-04 & 0.21544769 & 0.4772708 & 0.3667173 & 1.482304 & 105\\\\\n",
       "\t R-HSA-375276  & 5.724316e-04 & 0.21544769 & 0.4772708 & 0.3243598 & 1.374254 & 195\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 7\n",
       "\n",
       "| pathway &lt;chr&gt; | pval &lt;dbl&gt; | padj &lt;dbl&gt; | log2err &lt;dbl&gt; | ES &lt;dbl&gt; | NES &lt;dbl&gt; | size &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| R-HSA-6805567 | 4.182288e-06 | 0.01056446 | 0.6105269 | 0.3884187 | 1.607719 | 140 |\n",
       "| R-HSA-381753  | 1.530411e-04 | 0.15176876 | 0.5188481 | 0.3841968 | 1.549905 | 101 |\n",
       "| R-HSA-9752946 | 1.802479e-04 | 0.15176876 | 0.5188481 | 0.3910022 | 1.569605 |  94 |\n",
       "| R-HSA-500792  | 2.529114e-04 | 0.15971358 | 0.4984931 | 0.2906890 | 1.282994 | 453 |\n",
       "| R-HSA-6809371 | 5.174794e-04 | 0.21544769 | 0.4772708 | 0.3667173 | 1.482304 | 105 |\n",
       "| R-HSA-375276  | 5.724316e-04 | 0.21544769 | 0.4772708 | 0.3243598 | 1.374254 | 195 |\n",
       "\n"
      ],
      "text/plain": [
       "  pathway       pval         padj       log2err   ES        NES      size\n",
       "1 R-HSA-6805567 4.182288e-06 0.01056446 0.6105269 0.3884187 1.607719 140 \n",
       "2 R-HSA-381753  1.530411e-04 0.15176876 0.5188481 0.3841968 1.549905 101 \n",
       "3 R-HSA-9752946 1.802479e-04 0.15176876 0.5188481 0.3910022 1.569605  94 \n",
       "4 R-HSA-500792  2.529114e-04 0.15971358 0.4984931 0.2906890 1.282994 453 \n",
       "5 R-HSA-6809371 5.174794e-04 0.21544769 0.4772708 0.3667173 1.482304 105 \n",
       "6 R-HSA-375276  5.724316e-04 0.21544769 0.4772708 0.3243598 1.374254 195 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(fgseaRes[order(pval), ][, -8])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "cb80919c-714f-4788-b266-6ee71146d0cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAM1BMVEUAAAAzMzNNTU1oaGh8\nfHyMjIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enw8PD////p0TvgAAAACXBIWXMAABJ0\nAAASdAHeZh94AAAgAElEQVR4nO2dibqqurKFs0Mv0rz/0+60UGlAdJYLlfGf+zljSIpYyUhT\ncPcSMwDgz4izGwDALwAhAcAAhAQAAxASAAxASAAwACEBwACEBAADEBIADEBIADAAIQHAAIQE\nAAMQEgAMQEgAMAAhAcAAhAQAAxASAAxASAAwACEBwACEBAADEBIADEBIADAAIQHAAIQEAAMQ\nEgAMQEgAMAAhAcAAhAQAAxDSJsIg6zHK7+3FE1pEmbpKta7qmMy98/cI0ZObiAWdNTWFECXX\nr3jb79Bm902fPR4+GN/fMlRSIezFU9q00Mt8617lvUKS5CaBkCb3M+TEdi8mQ6lZCOlF3JRZ\niiaTfbKQeiHqu/p7r/wg/SPvFZJo15sEd6pFqSaCMfbxX+7FZOhZsxDSJs55UzRUP0JI0u+W\n1FBk2Re9V0iFGJebBHcSwixFE9vtIaTPwzvP/O3ViUQ2s9uZmMxGyHa+i1qX6e3ArsV9KTmJ\nwlT3fzm5iconx0avTHNXCNnZ1o6VbliQ6X/IVOiKy28hhe2PbfTS0Zfq1NLPfAgx2PbmhPS8\nMe144t9CaZH+otdb6QzP1EPKJ1Itl9ja/QG6IrV2K98QIemzvloNpClV272JKrqWrIQZ4je3\nr+HEmaYZmtK02hw72iDT/6BKN2xtISlsfmyjC3f2Mt/5X9uuTYNTITUiieU8NNa6X1WaZW5U\nyeAXvd5Kb5jaU3cRxm8Q0stYx9n9uxA3LYl1ICiPT2rMFcrpt9mfp7Vm1pK9XaxqwRMOSJu2\n0uvWqNNcTxpGM32tcpqj3+IL669GR2pmGPRlxmVU2bbLRxhsMNfUOC2a+2712Jgc5kGq32Bn\nqFb9vqR3XmulN0zt3VwuhPQ6S9RuWnPmVUhuitUzotrgVXr4lYtmTKHCHADesLPzfbqMyMrc\nadIbqKVhNNPXukcm1sJmY1PaC5zbOn+rTi9xGSHNfa1dfPyWtnm9+anas4VYr8x/E9JimNiz\na38PIf2B8DnS2LdlICT/Warx2qhTQGs1RUp2esq8v2FnlwiJDM61YdGIJSMh81vU2uR01qid\nzDCwt1bPKhtD/d7KeKv6wJj9o9f60W7Aol/0ciuXfUhkD0L6A4HjymVMxkLqlVJkMRd+l7eW\ndKcr/p1dcEZ6Xki536KnDLd2tvrcxPR4ar2vjspsDvXh+LJNRraeoxrjifgXvdxK9yexByH9\nAeq4WhRdP2aFpHYYdzUtNjokNgUlVV6vBcbPzR6/lmaSplJt5H9Q9reo5WgwRwNN3xTMZ6RZ\ni39IhvqSPj7+6XhX85fU7Ux+0eutdEtdbA9C+gNJj28IqRG1EkyvPuuwpJpoy+EdOzv6HGmy\nxyESUfCfVXzYoWMwFZJePGVclgW3WxJFMtQrFxyMn9XtGrOHFu3rRpjdc/qLXmvlYpjYs068\nQ0h/IBTSfR78rnmcAyEpL+ulSNjTKimpDwbyHTs7c/qtzJsNjYkXmtiSOpNVtGEkM/hBwW/x\n+e4c0+qPG3/UTtOmmy/luk657l4ej7b74Jr29Wg9n/6i11q5GCb2ekTt/gx1XOOOG3c90PTQ\npeuS3QeVdlYlJc14f8fOTnH379rZTZ7d1OuDDWnYmhn8INLCSEiDHpc30n4evCOlG+rk7Obb\nUu7Vj4zZJzvmS2ErJr/otVYuhmkfmodxNYT0BwLHKVeWdxMbvRexkFoTOWrd+2JrSbPvesvO\nTnOr1MAsWyeTTum7jpbKNTP8QWsL4wNVa1580G82MOpouW+fCmkeav0rbk8Zq9QRxn65ud1r\n/ItebOVimPZhizcbzqcX79nZXRjOw9tzhiGk83hi7w+OcZqQRgjpLJ7a+4NjnCWkRpBHDrnq\nrI0BFLkGzAAXZwmpMv+fUzvVWRsDwEWBkABgAEICgAEICQAGICQAGICQAGAAQgKAAQgJAAYg\nJAAYgJAAYABCAoABCAkABiAkABiAkABgAEICgAEICQAGICQAGICQAGAAQgKAAQgJAAYgJAAY\ngJAAYABCAoABCOk8xOz/fQYx2/+2tJjt/7kPgf75GtBR53FMSPa/N28/lgtv+88kghdBh5zH\n34TkPt01SOtk4P7zYBeSMyLe99/1BVvA4+fxbiE5u+BfAEefxz8SkrnFjK5+L/DueZwhpIf/\n8hx4DTj1PM4Tkv2KvmcEzjyPTxAS5MQE3HgeHyIknw/+Ahx4HhDSDwEHnseHCklgv/cC8Nh5\nfLKQrEFwGHjrPL5ASBgeR4GnzuN7hPS9e713LazJiv2tDvoFvlFIX7bhe9eqmgZovssvv8X3\nCulbRs3b2gkhfRLfLCRf77N5bxMhpA/hF4QkPnoIeQ+/y/rWF/BP+Rkhfezi9NbIo9j59jO4\nsVXeg7wgMTWFKtD5q1LIyMZQC1GPtqi62kwmd00u4zdIrtWEoNm5Rl5ISF1YZPVtei3lHhTo\niqUzUvO+exaH7A7x5M5pU2LLS4dGJX9bSELcaR5NTNIWkLZLepXsAxP9cnm0RaXu+NIkC5Ua\nVpmQJKnmmxALdG3QdYQ0hEVWJ6XXUlRXkW8NrRubX7tnXn7VtvXkzmlTtiwnVn9WSOZPI8ok\nzyVqUSphjKVoTF4tGlEHJqQc5qnSl2tTxly/C5U7SK3PQVS+KEmSapaeajlq5GWEpDwWFKFO\niq+lVLTAIOpJrxO0sxYTpHvm5VdtWk/unDZl3zLlt4UUdHEoJCHMnDb5hVrOwbw330w3T3o9\n8UWEVlNvrrW6L1tfliRJNcMkq3mLywipE2VQhDopvpZyE7RAtXZGap50j2/utvHkzmlTNi2v\nO5DlTru/4Ws5IiRaXndtI24kpxaDT8pFa6ob9f7OrECdWM5XJEmqGSoR7eZpI68iJOXcoAh1\nUnwtYcwqLejYxQTpnsckd06b8oTl3xbSztauWQ+7s94B39XqTQrPhZhbabYRc+u2dm2wOFWi\nr9XxU38lSVJNM6x7vEwjryKkIZq3qJPiawmlGNMCE+2rIeiX+ZHBrVZlmvKE5Z8VkmPI5Flv\nqHmuaNyW1+4yJF09hKjMoVanOx1tkN0cCcmgO5QkabV5f0H6YSHtryCxk3bHfav2CWmBLooM\nvSKkXMG0JoRkhvaQ5i2bW7WKqM40HXJza86NltYH4tpsiltTqZ0Dfwpdemr0ri5IrtXMyXiv\nkdcVEnXS7rg3m6mkwGhPnms+t5Cet/yzQlIfheyTvCAx31sbhynM50Cim7NdzEad1WmZqX7v\nMv6c1jomSarN/oi62cjrCok6aXfcF/ZBQpg5yTKqCCG9CfOD70KsD0azQnLiGZelaoxL6z+F\n2Z8ZncjEn1EydLjcde+FhRTk7Yz72sxEcYGyiMuF/n7D1u6h5V8WkjqhVPOGkIIv7SKkdilN\nwqyk323wZiTBm0hIQXT2QfjoskKKQtg74z7ajRvGohzTcs5w3D37HBfSQ8s/LaQhCjaQROUi\n1ibMUAjbMyPZ27VmLhx1BMHORqakze31Xs+GJoxrSZJUC8Li2UZeVUiBk54VUh8EV0PzpHsO\ncVxIDy3/tJDckhTm2YTa9nVq9N9LNdbXhaNchadENemD0U2fdPTbW412InnAbb5P5hREkqSa\nvn34UOnCBCM0cNKBnViwHuV05EvE7x881ap8U45a/m0hTXRJCs9IjZvpShoSoDNO6y/796xM\nsliS7mW9JkzSau5wBeZ4G0eddFxI5k+d7vWIiSIw/GSrsk05avm3haQ0kjnNuH1frYZ/qSdF\nub5WSpJzX7qHrO7NX5OawmTRxUla7fCh9/eJhixx0pNCyhyaiAnSPS+0ak9IjyyjqwFgAEIC\ngAEICQAGICQAGICQAGAAQgKAAQgJAAYgJAAYgJAAYABCAoABCAkABiAkABiAkABgAEICgAEI\nCQAGICQAGICQAGAAQgKAAQgJAAYgJAAYgJAAYABCAoABCAkABiAkABiAkABgAEICgAEICQAG\nICQAGICQAGAAQgKAAQgJAAYgJAAYgJAAYABCAoABCAkABiCkExHmn0S1H8s/Z2r+ldTZ/A+d\n8z2gr07kiJDsNXtVmO6ySXTcZ4H+OJG/CclecEaMmbN+B4DzT4VdSK6gwHr1z4HHT+SdQrL/\nQ//+K+DoE/l3QhI4Vb0ZePdE/rWQfD7gB149kROFhH5nBg49kbOFBEnxATeeyGcICU9+OYAL\nT+TzhITh8Crw3Il8ppAwJF4BXjuRTxYSBsZzwF8n8vFCwug4DFx1Il8jpG8eJe9qu9j9Cv4l\nXyakrwzuvavRsd1v9M3P8JVC+q5ouXjTCE/sfpFPfg8I6d2I941wCOlzgJDeDYR0CSCkN7P+\nnPfY3v76gwhLeQ/ygsTUFKpA569KIUMTUy1EPdhko642k7vQ+RHmWNNRtY2mXUdIXVCHOIkk\nn66bKeK7xzmR/AYmy0uJqwpJiDvNo4lJ2gLS6qNXyT4wYa9rSYw2KUeTP3ifOqTJWnpjrbbV\ntMsIaQjqECdRfz1bN1OkNNcK3+wHtl+3PF9SSOZPI8okzyVqUSphjKVoTF4tGlFTC+ZrIyp7\nzWcoJ8vA672W6mCKxdW2mnYVIUWeIk4a9vzzoG5a5C7koL8Fu483Wb6qkILuD4UkhFmKJr++\nSLWWUAvSXHdF12qdKKnNSVYms81U22raRYQUeYo6iSSfrpsWacxW4hYU2WzsHy1DSDRN1eG4\nqUWnEbfUjj44SbGmVTlaszK66USXqbbVtB8VUnw98hR1UuKvxNZ23bRIJfSe+/Eq9wbLlxHS\nztZO7cHG9Vqplu87LTy7+trNrdva6YlpCCQ42I1hJfpaHUqjaltNu4iQQk8FTor9lbBTNy0S\n/nkAr+UrCMkxZPKsY9T6XTRu9zu5lWcKjNyEO0B1OoIgu8XOUsIuSOqPoYyqbTTtIkKa53jI\nLk4K/bVhbqNuWuQpIfFavoyQyiHN8wN0VlORkocJ1d3cmhPu7bpK2u1xayq1ix1fYHDxCaEr\nTn4dWqptNO2iQiJOCv21YW6jLm03g5D+ZvkKQlIfheyTvCAx31sbkinM50ADnZZae7jTMptq\n3+9r9SaImE9r9XpnjFxVSBbipCl195G6vEL6m+WLCOkuzGHRL0IZITnxjMtSNS6lLWbPV5gN\n3NLv69UwzjdH1baadmkh0bxdkw/qkq9S5C7+E8sXEZLaAFfzhpCCL+0ipDYSEi27jHp/NQ7n\nHBsjENLe5aN1yVcbWxuPRe14LV9FSEMUbCCJym2+3JJjA3gj2WzYyIPJsRPTssosdpbQqS9b\nBdW2mnZRIWWctDf0t+qmRVqzwe53IzxvsnwVIbklKcyzCbXt65QL76USw7qylKvwzCsKk9Fb\nI/SLXI3352Kn8qXNpcmcmEg1EAzZwElL8um6aZHcmw2vtOoFy5cR0kSXpPCM1PjAHo0Z0MlH\n+sv+lasytlP4cLl7b6+Jql2eYDNMnET99WzdTJHiOZczWr6MkIK33qJgw1ArH5Y64C3XyABJ\n6jd/i25Jrg/tMkGLKSxbYD0yhKdK4qTpsZM266ZFpqB7Xm/V85Z/X0gA/AMgJAAYgJAAYABC\nAoABCAkABiAkABiAkABgAEICgAEICQAGICQAGICQAGAAQgKAAQgJAAYgJAAYgJAAYABCAoAB\nCAkABiAkABiAkABgAEICgAEICQAGICQAGICQAGAAQgKAAQgJAAYgJAAYgJAAYABCAoABCAkA\nBiAkABiAkABgAEICgAEICQAGICQAGICQAGAAQjoRof+JUmH/ndLlnzPVWWI2/0PnfA/oqxM5\nIiTzRdirwnSXTaLjPgv0x4n8TUj2gv1izZzzK4AGzj8RdiG5ggLr1T8HHj+R9wrJlEb//iPg\n6BP5d0KCpN4N3Hsi/1pI3jDgB149kROFhH5nBg49kbOFBEnxATeeyGcICU9+OYALT+TzhITh\n8Crw3Il8ppAwJF4BXjuRTxYSBsZzwF8nAiH9DvDXiUBI/4B3/RCx+xX8S75MSF8Z3HtXo2O7\n3+ibn+ErhfRd0XLxphGe2P0in/weENK7Ee8b4RDS5wAhvRsI6RJASG9m/Tnvsb399VcQlvIe\n5AWJqSlUgc5flUJGNhopyp4Y82PUD+u5K4RsprCs4e7uNNRC1ONOK68jpC6ok/hz12RYd1K+\n9m5Pi/iLzonz/gjfatWO5aXEpYQkxJ3m0cQkbQFpu6RXyT4wUZrLLTGmlDYQZzdr/bXsbE0v\nNpcbZFt5GSENQR3iReLbY3VH229yzBaxHVEsDd7X6GarHlqeryMk86cRZZLnErUoVV+MpWhM\nXi0aUVMLnSinearF4DN6LcpBVP77IOpJF6uTspW7k5TDPFXuBtlWXkVIg4yGbBVe7+mEt1+3\nNv4MO2spchfK5epbsBF5uVX7li8lpKDPQyEJYVaKyS/UcpaBL0rjs3GRwSS1m7tl2VFyWYyF\nZW9ueN/M92lnsr2KkNREE9QhXjRY3x6rKwT9ExVpzK7iFpjfbOyjVj2wDCH5AUrL60HfiFtq\nYVnSKiO8TnRzCBGBLTt655PVbKuVPyqk+LpybjRkQy9a327YiurKZeLLFKmE3vHllpYDltO+\nfcbybwtpZ2vXBGEAvabcaeF45hvsclOJvlZnzrXYpCsFZUsx2kQh5laa/d9mKy8ipCGatyIv\nDju736Ru67Z2ba5IZrk6bjnp26cs/6yQHEMmzw94UTRuy2v3X5JOjIWZgu7ec27SrGz9VXGd\nXvJp2Vata354Vw+O0VcR0jzHQzbw4t6ClNSdOx1tkF22yFNCetCqJy3/tpDKIc1bAjNq/lE9\nYkJ1NzfNkb1dK6ppHvwWeXCHW6GLTM2yCRjN7p6UNYu/H9462FCnO++1RRcVUujFIYzyPKir\nnK1pwwscQqKtet7yzwpJfRSyT/KCxHxvbRymMJ8DjW7OJszqA3BNEBuffMFJllHZQk6rkLSM\nx8Bo1MqLCsnivdhEzx3263Z6zlPT02vD/WirICSH+cFqrzXOfiXKCsmJZ1yWqnEprTtLtr5s\nGNFbLJReUL5sbQaFH97J3eJWXlpIG759ULcw+8Apmp5CU38RUpx31PIvC0nteqt5Q0jBl3YR\nUrsKyeAWqThaYwuMRUkfC+qydPdYPexVCGk+EmOLNmA5e+6rja2Nx6J2u6162vJPC2mIgg0k\nUbkNugkz2GBBuA2TLtxtPLdERm2u9We/nEvXslRIrVmdxvj0Slt5USFRL+aizvt19Wf8dM4V\nsS7v96KAB1v1tOWfFpJbksI8m1Dbvk757V6qblznxHIVnnl2fi9s/KHy+Y3ZoZtdPVFIUHa5\nk5Kled+BPp26LMGQJV4kvj1cd3IW0iK5Nxtea9XTln9bSBNdksIzUuNWjpKedsmM497FsxIr\nfIDW5epS9br2BGVnMou5G4B1wdV/iReJb4/VdS+95YPUythTLt9u1dOWf1tISiNVkuf3fbVy\nXKmXC7luE0hyVEqp+qiqefe46GzeIqSgLCnel9EDvusSimH14pHIQFjXvoa9UWTKXXylVU9b\n/lEhAfBvgZAAYABCAoABCAkABiAkABiAkABgAEICgAEICQAGICQAGICQAGAAQgKAAQgJAAYg\nJAAYgJAAYABCAoABCAkABiAkABiAkABgAEICgAEICQAGICQAGICQAGAAQgKAAQgJAAYgJAAY\ngJAAYABCAoABCAkABiAkABiAkABgAEICgAEICQAGICQAGICQAGAAQgKAAQjpPNw/QWv+T7jP\n9Z+mXf+FWvAFoKPO45CQEmZX4ezWgwD0x3n8TUjZS+As4PzzYBcSRHUecPh5vFVI0NS/BV4+\nj38lJF8Rff1G4Nzz+OdC8vXO/uG/CJx6HqcJyd8L8AFvnsfZQloMgr8DN57HpwgJemIA/juP\nzxIS5PQn4Lnz+EQh6dgeBPUC8Nh5fLKQ8BrSk8BX5/H5QlpuCR4BJ53H1wjJVf5O3tVusfsV\n/EOi8f3pQrLXznbas7xrAojtfp1jfohofH+HkL7s8PQu6Sd2v8clv0c0vr9JSNbY2Q58jMDW\n7gJE4/v7hLTU+FggpCsQje8vFpJr2MchZgiJDdfd5T3ICxJTU6gCnb8qhcyaMOlGXW2mMEkG\nuckte1uvK5ay2y37ISGJbTV14ZXVMxNxaL5m5MTFDVnz3ppr2Lw/wqNWDbUQ9bhvebn3VYUk\nxJ3m0cQkbQFpu6tXyZ5aGEjflSZVhEl/B7nktrpAQ61utuy3hLQxnobwwuqZ0fpejvl6qRNp\nZ6TmSffMG5rbblUf3WrH8nxJIZk/jSiTPJeoRan6cSxFY/Jq0YiaWhhE5ZN3IYd5kFqUJGnp\ndbIT5TRPtRh0tXrSGYGtsGXBkPwVIWWGlHITzSWeqY3Tm00nJU4knZGaT/pk3hvhUatmqSpP\nlRsFDy1fVUgzdVooJCHsJsOv2VJt7qiFzi4wmsasVTedQZKGSeoeLo2rR90blUjuG7csGJI/\nI6TkF6vZJcgjnhEPnJQ4kXRGaj7uE1N1y/1xq25GQhPd1+9ZhpDmVEi0vPZnI24kpxPL8akS\neg9iZkWSdJem1RZZ/iCkWXk05wWdJ5fJaxcqpC65uJiP+2TfZtQqs4142fJlhLSztWuCE6Ze\nU+60sHJiX6uDJqmYmUsHERewTIGpqGXBkPwhIUWjashOJ8YzrdvaJctMWjTpjNT8o/Vtt1WF\nmFtpdpKvWL6CkBxDJk/4CELRuN2vXdqlIP6sbMly3hOSXZBUb2hN3ukMGgQu0pZdQEhzdgRa\nz3Q62iCTVSZXNO6M1PxTQooLCmFsJxHbY5YvI6RySPOWLleTnPKg6aybmyJvtPRNh2n1nmJT\nSIM7DreimuZh3XyPcnubEY3vXxJSMqzSHOeZ1hTfX5CoE0lnEKM8QtLBhtq05XnLVxCS+ihk\nn+QFifne2pBMYT4HGui0TDprU0iNnzNNOLfydie5ubG7uJCcZzo9b6nRu7ckZZw4uccOkfk/\nCknPteOLli8iJLXXGudl6OaE5MQzLkNnXEqTslLkkvO8xvnUoJDtYrdM9Bi0LBiS1xKS80xh\ndsRTOm+lRffshb3wqpD+YvkiQlJb62reEFLwpV2GTpsTkg3ejGvUbrRxnCic4xa0sSg3HzTO\nlxbS4pnHa0jeifnhHvTJAQIr+ccVRy1fRUhDFGwgicrtK0yYwQYL/ApvsZEH48TW7OB6vR8h\nSRKUtWU74/B+O2Dnbh4MyQsJafWMnemn7fB37ETSGan5oE8OELTKVh6j+x21fBUhuSUpzLMJ\nte3rVOfcSyWGdWUpV+E1Zh9vTkFbbzZUvrR5Rn8v9Ik47pJLEwxZ4hnl28l5OEviRNIZqfnc\nmw1PtKowL6XcckUeWr6MkCa6JIVnpMZNrSWJGdDJx72LZ74XvmSQdFv9tayWY73M5mDxuPlL\nPVMSL2agRc0H7YzUfLFvbbdVblufD6w/tPz7vew90YgqyfP7vlr1TqlnIrluMUhSv6FcdEvS\nPQ4kSTK1jarrq97mQUgLwZANPEO8mK0XCYl2Rmp+2re22yo1eZZp5aOW0csAMAAhAcAAhAQA\nAxASAAxASAAwACEBwACEBAADEBIADEBIADAAIQHAAIQEAAMQEgAMQEgAMAAhAcAAhAQAAxAS\nAAxASAAwACEBwACEBAADEBIADEBIADAAIQHAAIQEAAMQEgAMQEgAMAAhAcAAhAQAAxASAAxA\nSAAwACEBwACEBAADEBIADEBIADAAIQHAAIQEAAMQ0nmI9Z9UNV+FyfFXlhT4BtBR53FMSMJe\nFzO9gH/h+dNAh5zH34Qklmv4t9M/APj/PNiF5C+gW/898Ph5vFtIazXwduDo8/hHQvLf0dXv\nBN49jzOEhEDFm4BTz+M8IYl5uQHgAc48j08QEuTEBNx4Hh8iJJ8P/gIceB4Q0g8BB57HhwpJ\nYL/3AvDYeXy0kPCuxHPAW+cBIf0Q8NZ5fI+QvnmUvKvtYvcr+Id8o5C+bp16V4Nju9/ml1/i\ne4X0PaPmXXGTxO73uOT3+GYhie9Ym8T7RjiE9DH8gpDeOFI5gJCuwM8I6WMXp+Vnv8n29tdf\nwY4tUd6DvCAxNYUq0PmrUsjIxlALUY+2qLraTCY3TJa9SU266EBu7O/V7fn3SkJKHXG3OdS3\nWSdRd1pTRVphMe+tmTrxL4vbFBs6YnlpzKWEJMSd5tHEJG0BaR3Xq2QfmOiXy6MtKrWoSpMs\n5iXZ6qQtMJAbW1UOu0PsQkJKHaH8r/9Q325UjITU0G6LzZPumXMipCSGnrA8X0dI5k8jyiTP\nJWpRqs4bS9GYvFo0og5MSDnMU6Uv16aMuX4XKneQWp+dKCe9FA3uUiOqpW5vBawKQkgbjqhs\nDvFtvibxqsuoJ+18WmExT7on0/YHhp60fCkhBV0cCkkIM/NMfqGWswx8cTM9POmlxRcRusd7\nc63VU5R26qiLSWNrvdckTe8rqUFIG464rT+a/MlVbcOMKq2wmifdk2n7A0NPWoaQ/Oil5bVs\nGnEjOWapschFa8rZeg9ipkk/BJY1bz1kVVakyiaElHfE6Eco8W2WTnTZ/KB1i3nSPUdJ2v6i\n5d8W0s7WrnGBBIteXu608FyIuZVmsZ9bt/1ogwk0nkubpcsHt10cdmZaU/UqQkodUYrR5hDf\nZqlEX6tDfpw90b5azT9a31ICQ3+x/LNCcgyZPOsNNSUWjdvymh2c26EtpaslaNDpE7Hs5sCf\nhZmj7i5HbVWW3q5WOxcVUvqrwyKtWvxdzurbLJX9cdFoVwtVGBl6WUixoZct/7aQyiHNW+I4\naqpTXWjcdnPz4o2W1sGG2kyV7RKfI/5sRTXNg99Dd5X0s+pAjqsQUs4RdGu8+jbvI90nUxNt\n8EZ7CF2NviokZ4jB8s8KSX0Usk/ygsR8b20cpjCfQxDdNIvZqLM6LTOlqS70pwncVqux2vV2\nQ+Y0CCnniEJOPof4doeJ9owO5pSR0ReF5A0xWP5lIalt1zjPyzjLCMmJZ1yWqjEubTdxeqtm\nOk0bKNgAAB6TSURBVFMSf6ruly0xNrkTM43+QUgZR9RmprE5xLd7hHcok9LuOu2eIySGXrf8\ny0JSG+xq3hBS8KVdhNQupUkwlGjKBm/GNXhDVzFbLojsQEgZR9At9sE1hF4fizJ5fOuuJ92z\nS2LoL5Z/WkhDFGwgicptJswyYuMGbiPnaM20Oepjrp2NTEmb268PjzrtWpt0tYOALYSUcQQV\nEvFtFu9a+rA7jjys5kn3PCYx9CfLPy0ktySFeTahtn2d6qJ7qYb9uoaUq/CULMyLCzd96NEv\nYDXaieQBt33TobAFVHJy2qyy4r06qSNsDvFtFnNlIsfOMTfavfncmw1bJIb+ZvlHO9r9/Iku\nSeEZqfGBPRodoDNOu8ZdyzVZLEn3sp7RoFwLuF1/fM+r4x2RnOKJb3M4LzdLjTqIvMbmi31r\nFGqIwfKPdrT/+fQFuCjYMNSqj0od8JbrvoIk575cngSaN39NalqTo/J81S8Fii66TfLlymwK\nifg2y7S61u+Dd4b79MBaUCUS0t8so6MBYABCAoABCAkABiAkABiAkABgAEICgAEICQAGICQA\nGICQAGAAQgKAAQgJAAYgJAAYgJAAYABCAoABCAkABiAkABiAkABgAEICgAEICQAGICQAGICQ\nAGAAQgKAAQgJAAYgJAAYgJAAYABCAoABCAkABiAkABiAkABgAEICgAEICQAGICQAGICQAGAA\nQgKAAQgJAAYgJAAYgJAAYABCAoABCAkABiAkABiAkC4AOvn9wMcXAJ38fuDjC4BOfj/w8QXY\n7GT0Phtw5QU4ICSMgz8CB14ACOn9wIEXAEJ6P3DgBYCQ3g8ceAEgpPcDB16A6wpJKN5levcr\n+EUuKySxfPCbhpCuB6uQvmnEvE9IAivSBXmDkL5o3LypqRDS9biwkHBGAnxcWEgzViQuhKW8\nB3lBYmoKVaDzV6WQoYmpFqIebLqRouyXpGwmX+gu6O0yBc7jXwnpwezfBZeJo6Y9RxGHL9zj\nGy2mvSUSrttolBC0q4LKe5aXGlcVkhB3mkcTk7QFpPVhr5Jhz9nrRkmlSbZrsnBllBH9ZyC9\nExY4kX8npD0zQ3CVOGq0/pVjrhZx+ILzdcZ05PK9YIMfFuukmeuvLcuXFJL504gyyXOJWpSq\nD8dSNCavFo2oqQXztRHVrOenctIL1KBnRTnMg/T6rIQTUuWrRQVO5COEpDwRCmlxVG0cHznd\nQRy+UkX3WUzHLn8ctevX7sn116bl6wop2HeEQhLCLEWTX7PlHM540lw3V0vjyVH3fGOWrZub\nLG9uEHXr5BkWOJNPEJJSRHCVOMr7PVeZOHzhFt1nNZ24/NED2Ukucs71145lCClIOyHR8jfV\naY24pXbkWlSvbpXQWxE3sY7e451YjlpBgVP510LKXVRejYREzqS2hIzrzIHDPWMkSWL6aZdX\nYiLppPITli8jpJ2tndpUkO25ngPvtPDs6ndzMHkG82gpRpuoRF+rQ+k8z7sT7b/lE4Q0RJ4g\njmrd1i63dGecuPg6Nf2sywe60mUqP2H5/F5+N0uwYcjkLVGBonG738lMjFKEwRu1mzAuL8zE\ndI+F1KoVzAvJxghnCClzmXwhjpo7HW2QXa4Ocbhj9XVq+lmX0wVpozKE5PHh7yHNW3bQanJU\nPWlCdTc3PYZ7u66SZsJsRTXNQxkJyaz49rvQFSezfEFIyeXgy+Io5dU0MucgDrcQXxODrwlp\nCOIbtPLzls/v5Xdjfnsh+yQvSMz31oZkCvM5pFHr2vS5idRWkZAKOYU+nnR1CCm5nGQZR3V6\n7ppqkV2SVodbqK//KqQmeMwBIe1jfrvaG4yzX4myQnLiGZelapzJkjX7PZ/qb9maanJxbW36\nI5pvaYGzOVFI1Ic5T5h5yGywpvwTt9XhhoyvieknXR4HZ3OVj1o+v5ffjT+7VPOGkIIv7SKk\nNhIScaKRnI3jjMputE/0ZdcCZ/PhQnq8jqwbhIyvSd3nXB7F4PKVj1q+ipCGKNhAEpXbU5gl\nxx5utcvW+dFGHkyOTXban62ZG3sTIV0715cNCpzNJ27tiKPsbD9lw9/E4f4WO0J6zuVduJnM\nVz5q+SpCcktSmGcTatvXqd66l8qv6yRVrsIzz9wnozeTvBf6oJw8RRe2bGPO0P0nvdnwIQSD\nnzhKJSeXkUAcvmWJZDzn8ip8XyJf+ajlywhpoktSeEZqfGCPHj/p5COXSK17Lc+orVjjt6tN\nV6DJFLg8wRaPOqrcdhR1eBIBSE0/5fJiCX7b6tnKRy1fRkjuZbkwz+/7atVdpZ705Lq9IEn9\n5m9htwFjrXrVis2+sxzbnNayUYHLE56ViKPmHUcRhx8Q0lMuj+1lKx+1/PtCAuAfACEBwACE\nBAADEBIADEBIADAAIQHAAIQEAAMQEgAMQEgAMAAhAcAAhAQAAxASAAxASAAwACEBwACEBAAD\nEBIADEBIADAAIQHAAIQEAAMQEgAMQEgAMAAhAcAAhAQAAxASAAxASAAwACEBwACEBAADEBIA\nDEBIADAAIQHAAIQEAAMQEgAMQEgAMAAhAcAAhAQAAxDSuYiNNPgy0HnnAiH9COi8c4GQfgR0\n3rlASD8COu9cIKQfAZ13Lv9MSOjo9wL/nguE9CPAv+cCIf0I8O+5QEg/Avx7Lp8hJIyCPwMX\nnsvnCAkj4U/AfefycUISGBIvAa+dy6cKaf0Ah4CvzuWzhYTRcRi46ly+QUhfPEaEeNtkIHa/\ngn/MFwlJfN9oeeMGNRbot7nm1/hKIX3NoHmfkJKYzNf45Ef5ZiF9ydh5UzMhpI/iu4X0+dEI\nnJEuwk8ISXx0gA8rEg/CUt6DvCAxNYUq0PmrUsjIRlcI2Uz+i6/cSJ/r7hHl0uRm6zbS7PwT\nIe3/gi64HPms7DdqxdemWoh62DLtXU7CdRuNEoK2IKi8Z3mpcVUhCXGneTQxSVtAWh/2Khn2\nakMvD75yaXILm7V0yZpLk9ut20izc76QhmDVSn3WZmsl12xnhUrK9Mna0AdCWifNXH9tWb6k\nkMyfRpRJnkvUohzneSxFY/Jq0YiaWhhEPempyWQO0lW+Cznob3ddoPJlSS5J7rRuI83O6UJa\nHOe+rj7rRDnphSZeZrLXTN80a+XAdOzyx+ru1+7J9dem5esKKdjFh0ISwqw1kx8WUs161EIl\n1rKqYxdl6mXrpifLbp0xSS5J7rRuI83O2UJaHee/L34pzfAc3TwWklyTprMiU5k+cU18EGyY\n5KrITH/tWIaQgrQTEi1/U53WiNuGIXXRFa+EWsbsxNqJ5XxFcklyp3UbaXb+qZAyN1sdZyE+\n81bKuM72teAUm+2TY1RiIumk8hOWLyOkna2d2i2M6zU9B94znTqZvGHt2/VPJfpanUSj3LDc\nVus20uycLaQh8kTeZ2nLsteaVYXzRp8cYqCrYKbyE5avICTHkMlbDrtF43a/k5nspEiCbZ2P\nQOSEZAODYS6EFN8lFNLis8LM9veso3LXbiLZBb4oJLogbVSGkDw+/D2kecsOWk2OQkgjlJvp\npHRvNy676dS1QpeezDQJIW3fLPAE8Vkrqmkeyqyjcte6Stqzypr7mpCGIKhEKz9v+QpCUh+F\n7JO8IDHfWxuSKcznEEetJ1lGlRPXTroOhLR9s4wnjM9sRLvKOyp/rTZ7u78KqQkec0BI+5jf\nrvYG4+xXoqyQnHjGZakaZ7JkzWURGpx9ZI9Y0EmSmxbItW4jzc5JQqI+zHnC5E3qtNRuOCp/\nbYqemW/2yS5hcDZf+ajliwhJ7YareUNIwZd2EVJLBsFYlGNk0MVxRhLHscelJTctkGvdRpqd\nTxaSIdkEENJr+eF+yOXEalAuX/mo5asIaYiCDSRRuRCQmeXs4Va7jHRdHwTxXOXW7At6faSy\noQnjZZJLkjut20iz82lbO+Izm+yygzS55uuFykr75ABdEPvbqHzU8lWE5JakMM8m1LavU51z\nL5Vf10mqXIU3hsFwV5k86260fyez4372zYYrEQiJ+My8rHAvcs/u0msmY6pCDWT65ABV+C5F\nvvJRy5cR0kSXpPCM1PjAHj1+ksmnDiJ8S+Viid+6l/WaMJcmwbw6zp6LVp+5ZHbXRK/Z+jLn\n1bRPDlAswW9bPVv5qOXLCCl4PysKNgy16p1ST3pyPcOuyTBUvlSezPvAS7Locrny2C7jEoQO\nJD4b1VRVbbz9Ta65+mu91PRTLo9Dc9nKRy3/vpAA+AdASAAwACEBwACEBAADEBIADEBIADAA\nIQHAAIQEAAMQEgAMQEgAMAAhAcAAhAQAAxASAAxASAAwACEBwACEBAADEBIADEBIADAAIQHA\nAIQEAAMQEgAMQEgAMAAhAcAAhAQAAxASAAxASAAwACEBwACEBAADEBIADEBIADAAIQHAAIQE\nAAMQEgAMQEgAMAAhAcAAhAQAAxASAAxASN8Keu6jQHd8K7s9h27918Dj3wqE9FHA498KhPRR\nwOPfygtCEuJhRfAi8Oq38lchoedZgTu/FR4hof+ZgCO/FUYhCYyCPwMXfivvERLGw4vAcd8K\nhPRRwHHfCoT0UcBx3wqE9FHAcd/KvxXSl44TId4WSBG7X8HXcI6Qvmu8vPEBdCzQ73IMWIGQ\nHvM+IQmsSL8ChHSQNzUZQvoRPkBI9gDy0UMIZyTwgA8T0ue+HYEViQdhKe9BXpCYmkIV6PxV\nKWRoYqqFqAdizFZrpCj7JHeeO2990PVG9l/kmv78xfcKSTwaTV1ceXHZ4smUSV2TzURycoUX\n0740CddtNCrqNFp5z/JS46pCEuJO82hikraAtD7sVTLsKHt9IMa00kqTaqNcJR9vvadW+X/X\n8xffL6S9Rg3B1YEM49WTCaP1vVyno1zhxbS9WAStfiCkddKMKu9avqSQzJ9GlEmeS9SiVP00\nlqIxebVoRE0tmK+NqJaMXouyE+Wk16ohyFWul966lMM8Vc4qOx8rpA0Tq1/s19WfiScJtXEf\n6ZBc4cX0XSiXq29+znwctevX+TWuvGv5ukIKZstQSEKYRWPya7ZUSxC1IM31tf4k9RgojVPH\nRSc2V3e0K3kzl6Z4n8jFlwlp9Yv/viwqsSfTJpMWZgqvphuzlbitph89kHWdlq+8axlCCtJO\nSLS8Hv+NuKV2FkFUVFjLQmdzZ1XbXclPsWx8mZBWv1g6sZxJY09S5DK5bRdeTVdCbwHJYvcI\n12lblZ+wfBkh7WztmiAioKe8e9qpzdLxg50No7nS5c7DklWIuZWiftMR6duENMyhkCrR1+r0\nTm6fXTtat7Uja0xSeIhzD0cQB7qw/c3yFYTkGDJ5y2G3aNzu127FpAjH/02sLnezWGHmqLt3\nLZnbFqdX4WGWly8T0jzHQjLoCSvyZEinow1yWb02Cr8oJLogbVSGkDxOMOWQ5i07aDU5qt4y\nobqbmwLDvV1XST8pDu7g24pqmge/hx5IfGJxug421Plo1N/5ciEJ7eHJrPOhJyPaKEhHC681\nXhPSEASVaOXnLV9BSOqjkH2SFyTme2tDMoX5HKIo6KyPPHZabHxs3ARmq+AwGhi1a+CYWuLh\ny4VkmYx3Ak+GdHpem+r1QEUL/1VITfCYA0Lax/x2tRUYZ78SZYXkxDMuS9U4kyVrXsNvS0RP\nda9svQUa53txn/Ek3yEk6sOcJ0xe4MmQwmy+JjIbZQuHvXDU5XFwNlf5qOWLCEnthqt5Q0jB\nl3YRUhsJyZWLAjdu6QpjPbZOBSEdFJIhswlYr8dV48KLy/V0OR6M2kVdma981PJVhDREwQaS\nqNy+wSw59iwb7shs5MHlLGFbm9uJKsgl1luzcRizUV0GvkNI6e0d3qlV5MkIuxKQh3H5woHL\n+4PPwINO26p81PJVhOSWpDDPJtS2r1Odcy+VX9dJqlyFZ56rT05vlc83uffCRiWqjEyV8swz\n+PSR1EUJhNSYw485pQSejFDFJlfYZ+QKO9O5lxO2qcIHffnKRy1fRkgTXZLCM1LjA3v0+Ekn\nH7lEav2mfV7e0KvCXGq9JdXA6hd7LrLua+bIkzHl6kRaLyrsTRfPuHztNFs9W/mo5csIKXhZ\nLgo2DLXqnVLPcXJ96kOS+s3foouqzmOt+rOPc+mXvnSPHIEmPCtNq1OpJxPMW9ekXrawNz2t\npY+3h7YoqXzU8u8LCYB/AIQEAAMQEgAMQEgAMAAhAcAAhAQAAxASAAxASAAwACEBwACEBAAD\nEBIADEBIADAAIQHAAIQEAAMQEgAMQEgAMAAhAcAAhAQAAxASAAxASAAwACEBwACEBAADEBIA\nDEBIADAAIQHAAIQEAAMQEgAMQEgAMAAhAcAAhAQAAxASAAxASAAwACEBwACEBAADEBIADEBI\nADAAIQHAAIQEAAMQEgAMQEgAMAAhAcAAhAQAAxASAAxASAAwACEBwACEBAADEBIADEBIYAMM\njWeAt8AGGBrPAG+BDcTml+9BKN5levcrAJ4fEJJYPvhNQ0jgGBDSrmUICRxjV0hfNG7e1FQI\nCRzjmJA+fADhjATO5ikhffA4worEg7CU9yAvSExNoQp0/qoUMjVzd1W6QshmMsmhFqIeTbKR\nS+5adgpzvw12IXW0kPAEyQzEy7TihmnvclJks2mx5Wx/xZYXw1cVkhB3mkcTk7QFpPVhr5J9\nbEWVMX+btWS/JkuTLMKyozUrx9jWl8AtpEHkhCSDZEpPu8ZaSYW0mA47Yj/YEFvO9teW5UsK\nyfxpRJnkuUQtSuW8sRSNyatFI+rYSmWrDKKe9Cylr0s5zFOlK92FSg7SS9WVrY29jK0vgVlI\nyj9poX6d3no6062sXnZmRLVtOuqIfSHFlnP9tWn5ukIKAv+hkIQws9Lk12w5y9gtNzcHVmKp\ndjN+n/Q82pgV7CbaoKxYy34lvELqRJl6YpJVJkkhXvZ22m3TYUfMuw9kE8uZ/tqxDCEFaSck\nWl47uBG3wMYYjQL9pRaD/1oJvR1wc+VSVi66/E54haS8mo7pSkyZJIV42dKJLi6zmg464gGJ\n5Ux/PWH5MkLa2do1wZGzVMv3nRY2eSMdBZO+XIi5lWanF85lS9nWbRWSGfRL4BXSkFmbB7pj\na+YcxMuWSvS1OvXnTT+zCUgsZ/rrCctXEJJjyOT5sS+Kxu1+7Vovg/mxVQtUuOL32kblT8jU\ny6Rsp0+vMplAvwX2qF0yCh8vSNTLvqANwmZNPyOkxHK+vyAkjw9/D2nesoNWk5xyoQnV3dy8\nRPZ2ZkEnLhzNdl7o4+dU6wmMeJmWbc0NvnVBer+QhvVcP2yFZIiXfcZNB6rNBm81+JKQIsu0\nv563fAUhqY9C9klekJjvrQ3JFOZzoIHOQk605CRLW1VLc9QFiZdJ2U5LUvXUty5JbxdSsz5l\naNIHDr7O4mXKZDL+KKTIMu0vCCnF/Pa7MIdFvwhlhOTEMy5L1ehL16aT15JlQavqP3JJ0rKF\n2a1M8Rj4GliERB/6xKOQxEaTMOlaP1s1znBf5VNCisrm++uo5YsISW2t7XYsI6TgS7sIqfWl\nw23gWJQ2MkEi4TakM6pb0LIIf8+7QiIhsO1oWPWUkNaOeExiOd9fRy1/ay8fx3piiIINJFG5\nzZcJMxTCyoQs+YGQ+uWY25rFZ9TfbbI3wdK1rJ3EJoS/l6JhIRLIzsS0HcTLFhsGikf0cir1\nHfGYjGX9GffXUctXEZJbksI8m1Dbvk51zr1UvblOjWX0mMGtR6vfldQmvaW+ZR6om7KN0O9t\nNYd69RJEQqrog7ghKW0hXrYYf07xmcqZTt8/2CZnOdNfRy1fRkgTXZLi50hLSHXtoHjysVVq\nsji1axy2CEOyznyZC9RemGiLV6wR72Ij+D0HXjb13IuR0ezkTRdPuDy2nO+vo5YvIySlkSrJ\n8/u+WvVOqacmua7rMrfEB8elvvRPBu17w+ktZfTo8NpEQiIL1N45cvWyLaV9XcQ7QW9gesbl\nseVsfx21/PtCAuAfACEBwACEBAADEBIADEBIADAAIQHAAIQEAAMQEgAMQEgAMAAhAcAAhAQA\nAxASAAxASAAwACEBwACEBAADEBIADEBIADAAIQHAAIQEAAMQEgAMQEgAMAAhAcAAhAQAAxAS\nAAxASAAwACEBwACEBAADEBIADEBIADAAIQHAAIQEAAMQEgAMQEgAMAAhAcAAhAQAAxASAAxA\nSAAwACEBwACEBAADEBIADEBIADAAIQHAAIQEAAMQEgAMQEgAMAAhAcAAhAQAAxASAAxASAAw\nACGBH0Yo3mV69ysAP4RYPvhNQ0jgMrxPSAIrErgYbxriEBJgJJmZPwyckcBX8OlCmrEicSEs\n5T3ICxJTU6gCnb8qhUysdOJBcr6LJDnUQtTjXxr/8TwjJOot3y26A0gyZdI+HNKKG6Yb1XnN\nFITrthqYGsr1V2x5qXJVIQlxp3k0MUlbQE4mr1fJPjIyrA7PJ7WRONlTqz/KE0IaRE5IMkim\n2L5ZlTTkhLSYLs21Ym3dvDnEU0O5/tqyfEkhmT+NKJM8l6hFqaahsRSNyatFI+rQxiCXGvnk\nPFdr2ielHOapclZ/lFRIWyMq8JanX6e3ns50C6YrGlGtZkg6MX0XyuXqm7e0L6TYUKa/Ni1f\nV0hB4D8UkhBmEpr8mi1nGbqlE6WvkU/O822d23zyZrpkyk+0v8JhIQXe8kyyyiQp0vQNqdmJ\ndtt0Y7YSt7XIzgPZxFCmv3YsQ0hB2gmJltf+bMQtMNGsVrLJeVyHyZKsBd3a/yiHhUS9tVCJ\nKZPMVF4Hdye65OpiuhL6hJNbtFISQ5n+esLyZYS0s7VrghNmqZbvOy2s/LfWyCdVpXHdTLtk\nIeZWivqnj0jHhTTEE5bOW7dRw94OuCFjvhJ9rU79edPhn30SQ5n+esLyFYTkGDJ51jFqCSka\nt/u1S7uM50fiwUyyVSuYS69JIarNI/TPcPyMlBmFxxYktVUmw71yQdis6eeEFBnK9xeE5PHh\n70wIddlBq7lJedCE6uxOOdrbPRCSWfFtmiSFPp9Odbqn/yX+IqRhjekMcXiH0FWSHk10z0x2\njVoNviAkashlrP31vOUrCEl9FLJP8oLEfG9tSKYwnwMNdIYFM8lCTj5NknYNHGNLv8VfhNSs\nTxma5IFDQB2fZybj1T8JiRpytdb+gpBSzG+/C3NYXB8B0osOI55xWarGmSxZu0KqzShwkfQl\n+UKvfiGPhLThQwOJjcp9J6WhTxHfhpp5xuVbknnW8k/3ssH+9srEW7JCCr60i5Daw0Ii+0S6\nZawgpF0hkRDYwzhb4sT8cLextfFQ1C4xlO+vo5Z/upcN/uwSBhtIonL7BjPtFcIG8OId2QtC\nas3qNMbn4t/iD1s7En/OxLQdNupDesNnhCN6ifRol/eHnoEnhvL9ddTyVYTklqQwzybUtq9T\nPr2XqjfXqbEMnyrsR+2y2ar3J314jcIW1yUSUrV6uNp85GbebJiqVWiNHstTfKZyptP3D7ZJ\nDOX766jlywhpoktS/BxpiYSufo0mnxeE5LaJP70gPUW0xSvWiHexHfyWqxNNPfdiZLQyeNPF\ncZdTQ7Z6tr+OWr6MkIIXtqJgw1Arp5Z6JpLroVYG59tXhDT3ZfTo8NpEQtrwY0wjRdGRUtOa\nkZqezDvax5ozxZaz/XXU8u8LCYB/AIQEAAMQEgAMQEgAMAAhAcAAhAQAAxASAAxASAAwACEB\nwACEBAADEBIADEBIADAAIQHAAIQEAAMQEgAMQEgAMAAhAcAAhAQAAxASAAxASAAwACEBwACE\nBAADEBIADEBIADAAIQHAAIQEAAMQEgAMQEgAMAAhAcAAhAQAAxASAAxASAAwACEBwACEBAAD\nEBIADEBIADAAIQHAAIQEAAMQEgAMQEgAMAAhAcAAhAQAAxASAAxASAAwACEBwACEBAADENJ5\n/Pf+Gv/gFi/U+EUgpPOAkH4ICOk8IKQfAkI6Dwjph4CQzgNC+iEgJAAYgJAAYABCAoABCAkA\nBiAkABiAkM5CKo6VkqR0/DcuPucK7VVaaxy7zSPbx37W7wEhnYRcPh4XW0vHf+PScq9wrpIb\n9Ydv88j2wZ/1e0BIJ/EOIUmSf2yQS78iHb0NhLQBhHQSx0acpH8PjNenV6Qw8+jCByElQEgn\ncVBI/uziSr9LSM/dBkJKgZBO4okV6Ynx+pcV6fBt9iQHIYF/yxMj7v1Ceu42EFIGCOkkvldI\nL6xhFwBCOomv3drJlxr280BIJ3FcSHs7qW2rTwvp4G1kWA1CckBIZ3H8zYa9v3Hxpys9WUM+\negcCbzYAAF4GQgKAAQgJAAYgJAAYgJAAYABCAoABCAkABiAkABiAkABgAEICgAEICQAGICQA\nGICQAGAAQgKAAQgJAAYgJAAYgJAAYABCAoABCAkABiAkABiAkABgAEICgAEICQAGICQAGICQ\nAGAAQgKAAQgJAAYgJAAYgJAAYABCAoABCAkABiAkABiAkABgAEICgAEICQAGICQAGICQAGAA\nQgKAgf8BG98SaCANT+sAAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# select the top five up-regulated pathways \n",
    "topPathwaysUp <- fgseaRes[ES > 0][head(order(pval), n = 5), pathway]\n",
    "# select the top five down-regulated pathways \n",
    "topPathwaysDown <- fgseaRes[ES < 0][head(order(pval), n = 5), pathway]\n",
    "# concatenate them for visualization\n",
    "topPathways <- c(topPathwaysUp, rev(topPathwaysDown))\n",
    "#Viewing the 5 most significantly up-regulated and down-regulated pathways each with the FGSEA internal plot function\n",
    "plotGseaTable(pathways = REACTOME_hallmark[topPathways], stats = stats, \n",
    "              fgseaRes = fgseaRes, gseaParam = 0.5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "181989b1",
   "metadata": {},
   "source": [
    "## Gene Set Enrichment Analysis using GSA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2f471b1",
   "metadata": {},
   "source": [
    "Gene Set Analysis (GSA), an Enrichment Analysis, is a method that is commonly used to summarize high-dimensional gene expression data sets into sets according to their biological relevance. GSA takes the ranked gene lists from the initial stage of a gene expression analysis and aggregates the genes into sets based on shared biological or functional properties as specified by a reference knowledge base. Such databases often contain phenotype associations, molecular interactions and regulation and are referenced in the analysis of the resultant gene sets to find the relevance of the gene properties to the phenotype of interest."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "443e636c-265e-4829-85f3-1999eeff4385",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<iframe src=\"./Quizzes/Quiz_Submodule4-4.html\" width=100% height=250></iframe>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Run the following command to take the quiz\n",
    "IRdisplay::display_html('<iframe src=\"./Quizzes/Quiz_Submodule4-4.html\" width=100% height=250></iframe>')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f0c9233",
   "metadata": {},
   "source": [
    "### Data preparation\n",
    "The GSA method is freely available as standalone package in CRAN repository. We can use the following code to install the package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "c3efb930",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install GSA from CRAN\n",
    "suppressMessages({\n",
    "    if (!require(\"GSA\")) suppressWarnings(install.packages(\"GSA\"))\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "ee423bc8-1c31-46b5-8955-5d89f6f1a745",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressMessages({\n",
    "    library(GSA)\n",
    "})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4a2ed98",
   "metadata": {},
   "source": [
    "The GSA method requires an expression matrix, a numeric vector containing the class of each sample and a vector of the genes inputs. We can easily get those inputs by loading the data that we processed in the [**Module 01** ](./Module01-GEO_Data_Processing.ipynb)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "3982c128",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading expression data with groups\n",
    "data <- readRDS(\"./data/GSE48350.rds\")\n",
    "expression_data <- data$expression_data\n",
    "norm_expression_data <- data$norm_expression_data\n",
    "groups <- data$groups"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad173974",
   "metadata": {},
   "source": [
    "We can also use the sample approach available in the [**Module 01** ](./Module01-GEO_Data_Processing.ipynb) to map the probe IDs to gene symbols. The step-by-step coding instruction is shown below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "e0d1acf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the probe IDs\n",
    "expression_data$PROBEID <- rownames(expression_data)\n",
    "probeIDs <- rownames(expression_data)\n",
    "# Perform gene mapping\n",
    "suppressMessages({\n",
    "    annotLookup <- AnnotationDbi::select(hgu133plus2.db, keys = probeIDs, columns = c('PROBEID', 'GENENAME', 'SYMBOL'))\n",
    "})\n",
    "# Merge DE result  data frame with annotation table\n",
    "new_expression_data = merge(annotLookup, expression_data, by = \"PROBEID\")\n",
    "# Remove NA value\n",
    "new_expression_data <- new_expression_data[!is.na(new_expression_data$SYMBOL), ]\n",
    "# Remove duplicated genes symbol\n",
    "new_expression_data <- new_expression_data[!duplicated(new_expression_data$SYMBOL, fromLast = FALSE), ]\n",
    "rownames(new_expression_data) <- new_expression_data$SYMBOL\n",
    "# Drop PROBEID, GENENAME, and SYMBOL columns\n",
    "new_expression_data <- new_expression_data[, -c(1:3)]\n",
    "genenames <- rownames(new_expression_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8acf4f4",
   "metadata": {},
   "source": [
    "### Enrichment Analysis using GSA and GO Terms\n",
    "Using data obtained from the previous step, we can run the GSA method by calling the function `GSA`. We can reuse `GO_term_hallmark` and `KEGG_hallmark` loaded in FGSEA to perform analysis. The code details are shown below:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0648ecb6-f0ec-41aa-b3e6-e812ba18d85c",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-danger\">\n",
    "    <i class=\"fa fa-exclamation-circle\" aria-hidden=\"true\"></i>\n",
    "    <b>Alert: </b>  Please note that the outputs generated from GSA can be varied depending on R environments and software versions. Especially, when the users run the scripts using User-managed Notebooks instances which have a preinstalled suite of packages.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "e026aa03",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "perm= 10 / 100 \n",
      "perm= 20 / 100 \n",
      "perm= 30 / 100 \n",
      "perm= 40 / 100 \n",
      "perm= 50 / 100 \n",
      "perm= 60 / 100 \n",
      "perm= 70 / 100 \n",
      "perm= 80 / 100 \n",
      "perm= 90 / 100 \n",
      "perm= 100 / 100 \n"
     ]
    }
   ],
   "source": [
    "# Getting \n",
    "genesets <- GO_term_hallmark\n",
    "GSA.obj <- GSA(x = as.matrix(new_expression_data), y = as.numeric(groups$groups), \n",
    "               genenames = genenames, genesets = genesets, \n",
    "               resp.type = \"Two class unpaired\", nperms = 100, random.seed = 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6abece1d-95a4-46a5-9240-4db9d62a170b",
   "metadata": {},
   "source": [
    "To use the `GSA` function, we need to pass the following arguments:\n",
    "\n",
    "- `x`: Data x: p by n matrix of features (expression values), one observation per column (missing values allowed); y: n-vector of outcome measurements.\n",
    "- `y`: A vector of response values: 1,2 for two class problem, or 1,2,3 ... for multiclass problem, or real numbers for quantitative or survival problems\n",
    "- `genesets`: A list containing the gene set collection.\n",
    "- `genenames`: A vector of genenames in the expression dataset.\n",
    "- `resp.type`: A character string specifying the problem type: \"quantitative\" for a continuous parameter; \"Two class unpaired\"; \"Survival\" for censored survival outcome; \"Multiclass\": more than 2 groups, coded 1,2,3...; \"Two class paired\" for paired outcomes, coded -1,1 (first pair), -2,2 (second pair), etc.\n",
    "- `nperms`: An integer specifying the number of permutations used to estimate false discovery rates.\n",
    "- `random.seed`: An integer specifying the optional initial seed for random number generator.\n",
    "\n",
    "The `GSA` function returns a list containing the analysis results. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "ebba9b67",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List the results from a GSA analysis\n",
    "res <- GSA.listsets(GSA.obj = GSA.obj, geneset.names = names(genesets), FDRcut = .5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f280f564-9698-4896-9eb9-c22a5ac00399",
   "metadata": {},
   "source": [
    "To use the `GSA.listsets` function, we need to pass the following arguments:\n",
    "\n",
    "- `GSA.obj`: A GSA object returned by GSA function.\n",
    "- `geneset.names`: An optional vector of names for the gene sets.\n",
    "- `FDRcut`: A numeric value specifying the false discovery rate cutpoint for listed sets. A value of 1 will cause all sets to be listed.\n",
    "\n",
    "The `GSA.listsets` function returns a list of the results from the GSA call."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df55e53c",
   "metadata": {},
   "source": [
    "A table of the negative gene sets. “Negative” means that lower expression of most genes in the gene set correlates with higher values of the phenotype y. E.g., for two classes coded 1,2, lower expression correlates with class 2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "ae97649e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 × 5 of type chr</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Gene_set</th><th scope=col>Gene_set_name</th><th scope=col>Score</th><th scope=col>p-value</th><th scope=col>FDR</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>57 </td><td>GO:0000422</td><td>-0.4838</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>641</td><td>GO:0006123</td><td>-1.3695</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>893</td><td>GO:0006851</td><td>-0.802 </td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>925</td><td>GO:0006906</td><td>-0.4592</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>976</td><td>GO:0007007</td><td>-1.2636</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>987</td><td>GO:0007029</td><td>-0.5442</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 5 of type chr\n",
       "\\begin{tabular}{lllll}\n",
       " Gene\\_set & Gene\\_set\\_name & Score & p-value & FDR\\\\\n",
       "\\hline\n",
       "\t 57  & GO:0000422 & -0.4838 & 0 & 0\\\\\n",
       "\t 641 & GO:0006123 & -1.3695 & 0 & 0\\\\\n",
       "\t 893 & GO:0006851 & -0.802  & 0 & 0\\\\\n",
       "\t 925 & GO:0006906 & -0.4592 & 0 & 0\\\\\n",
       "\t 976 & GO:0007007 & -1.2636 & 0 & 0\\\\\n",
       "\t 987 & GO:0007029 & -0.5442 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 5 of type chr\n",
       "\n",
       "| Gene_set | Gene_set_name | Score | p-value | FDR |\n",
       "|---|---|---|---|---|\n",
       "| 57  | GO:0000422 | -0.4838 | 0 | 0 |\n",
       "| 641 | GO:0006123 | -1.3695 | 0 | 0 |\n",
       "| 893 | GO:0006851 | -0.802  | 0 | 0 |\n",
       "| 925 | GO:0006906 | -0.4592 | 0 | 0 |\n",
       "| 976 | GO:0007007 | -1.2636 | 0 | 0 |\n",
       "| 987 | GO:0007029 | -0.5442 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "     Gene_set Gene_set_name Score   p-value FDR\n",
       "[1,] 57       GO:0000422    -0.4838 0       0  \n",
       "[2,] 641      GO:0006123    -1.3695 0       0  \n",
       "[3,] 893      GO:0006851    -0.802  0       0  \n",
       "[4,] 925      GO:0006906    -0.4592 0       0  \n",
       "[5,] 976      GO:0007007    -1.2636 0       0  \n",
       "[6,] 987      GO:0007029    -0.5442 0       0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "neg.table <- res$negative\n",
    "head(neg.table)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c237dea8",
   "metadata": {},
   "source": [
    "A table of the positive gene sets. \"Positive\" means that higher expression of most genes in the gene set correlates with higher values of the phenotype y."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "e640ff47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 × 5 of type chr</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Gene_set</th><th scope=col>Gene_set_name</th><th scope=col>Score</th><th scope=col>p-value</th><th scope=col>FDR</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1681</td><td>GO:0014037</td><td>0.7351</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>1682</td><td>GO:0014044</td><td>0.9529</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>1713</td><td>GO:0014855</td><td>0.3543</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>1729</td><td>GO:0014912</td><td>0.6334</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>1877</td><td>GO:0017145</td><td>0.5308</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>2009</td><td>GO:0021533</td><td>0.596 </td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 5 of type chr\n",
       "\\begin{tabular}{lllll}\n",
       " Gene\\_set & Gene\\_set\\_name & Score & p-value & FDR\\\\\n",
       "\\hline\n",
       "\t 1681 & GO:0014037 & 0.7351 & 0 & 0\\\\\n",
       "\t 1682 & GO:0014044 & 0.9529 & 0 & 0\\\\\n",
       "\t 1713 & GO:0014855 & 0.3543 & 0 & 0\\\\\n",
       "\t 1729 & GO:0014912 & 0.6334 & 0 & 0\\\\\n",
       "\t 1877 & GO:0017145 & 0.5308 & 0 & 0\\\\\n",
       "\t 2009 & GO:0021533 & 0.596  & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 5 of type chr\n",
       "\n",
       "| Gene_set | Gene_set_name | Score | p-value | FDR |\n",
       "|---|---|---|---|---|\n",
       "| 1681 | GO:0014037 | 0.7351 | 0 | 0 |\n",
       "| 1682 | GO:0014044 | 0.9529 | 0 | 0 |\n",
       "| 1713 | GO:0014855 | 0.3543 | 0 | 0 |\n",
       "| 1729 | GO:0014912 | 0.6334 | 0 | 0 |\n",
       "| 1877 | GO:0017145 | 0.5308 | 0 | 0 |\n",
       "| 2009 | GO:0021533 | 0.596  | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "     Gene_set Gene_set_name Score  p-value FDR\n",
       "[1,] 1681     GO:0014037    0.7351 0       0  \n",
       "[2,] 1682     GO:0014044    0.9529 0       0  \n",
       "[3,] 1713     GO:0014855    0.3543 0       0  \n",
       "[4,] 1729     GO:0014912    0.6334 0       0  \n",
       "[5,] 1877     GO:0017145    0.5308 0       0  \n",
       "[6,] 2009     GO:0021533    0.596  0       0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pos.table <- res$positive\n",
    "head(pos.table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "79687f5d",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 114 × 2 of type chr</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Gene</th><th scope=col>Score</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>SH3GLB1  </td><td>2.729</td></tr>\n",
       "\t<tr><td>IFT88    </td><td>2.491</td></tr>\n",
       "\t<tr><td>MTMR3    </td><td>2.271</td></tr>\n",
       "\t<tr><td>TP53INP1 </td><td>2.134</td></tr>\n",
       "\t<tr><td>ATG16L2  </td><td>1.973</td></tr>\n",
       "\t<tr><td>ATG2B    </td><td>1.944</td></tr>\n",
       "\t<tr><td>ATG14    </td><td>1.935</td></tr>\n",
       "\t<tr><td>ATG7     </td><td>1.597</td></tr>\n",
       "\t<tr><td>STING1   </td><td>1.457</td></tr>\n",
       "\t<tr><td>MAP1LC3B2</td><td>1.385</td></tr>\n",
       "\t<tr><td>ATG9B    </td><td>1.105</td></tr>\n",
       "\t<tr><td>TBC1D12  </td><td>1.061</td></tr>\n",
       "\t<tr><td>FEZ1     </td><td>1    </td></tr>\n",
       "\t<tr><td>EPHB2    </td><td>0.888</td></tr>\n",
       "\t<tr><td>NUPR1    </td><td>0.822</td></tr>\n",
       "\t<tr><td>ATG4C    </td><td>0.784</td></tr>\n",
       "\t<tr><td>WDR45B   </td><td>0.775</td></tr>\n",
       "\t<tr><td>TMEM39A  </td><td>0.751</td></tr>\n",
       "\t<tr><td>PIP4K2A  </td><td>0.659</td></tr>\n",
       "\t<tr><td>ATG2A    </td><td>0.622</td></tr>\n",
       "\t<tr><td>RAB7A    </td><td>0.588</td></tr>\n",
       "\t<tr><td>EFNB1    </td><td>0.571</td></tr>\n",
       "\t<tr><td>RNF186   </td><td>0.551</td></tr>\n",
       "\t<tr><td>UBXN2A   </td><td>0.516</td></tr>\n",
       "\t<tr><td>WIPI2    </td><td>0.464</td></tr>\n",
       "\t<tr><td>BECN1    </td><td>0.392</td></tr>\n",
       "\t<tr><td>IFT20    </td><td>0.389</td></tr>\n",
       "\t<tr><td>TBC1D14  </td><td>0.381</td></tr>\n",
       "\t<tr><td>WIPI1    </td><td>0.352</td></tr>\n",
       "\t<tr><td>SNX18    </td><td>0.35 </td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>SCFD1    </td><td>-1.858</td></tr>\n",
       "\t<tr><td>LRSAM1   </td><td>-1.899</td></tr>\n",
       "\t<tr><td>UBXN2B   </td><td>-1.922</td></tr>\n",
       "\t<tr><td>RAB3GAP1 </td><td>-1.924</td></tr>\n",
       "\t<tr><td>GABARAPL3</td><td>-1.988</td></tr>\n",
       "\t<tr><td>MAP1LC3B </td><td>-2.017</td></tr>\n",
       "\t<tr><td>MTOR     </td><td>-2.019</td></tr>\n",
       "\t<tr><td>RNF5     </td><td>-2.026</td></tr>\n",
       "\t<tr><td>VMP1     </td><td>-2.122</td></tr>\n",
       "\t<tr><td>SMURF1   </td><td>-2.124</td></tr>\n",
       "\t<tr><td>RAB1A    </td><td>-2.16 </td></tr>\n",
       "\t<tr><td>TRAF6    </td><td>-2.232</td></tr>\n",
       "\t<tr><td>RAB3GAP2 </td><td>-2.265</td></tr>\n",
       "\t<tr><td>PHF23    </td><td>-2.39 </td></tr>\n",
       "\t<tr><td>ATP2A2   </td><td>-2.517</td></tr>\n",
       "\t<tr><td>ATG5     </td><td>-2.518</td></tr>\n",
       "\t<tr><td>RALB     </td><td>-2.528</td></tr>\n",
       "\t<tr><td>STX12    </td><td>-2.608</td></tr>\n",
       "\t<tr><td>MOAP1    </td><td>-2.609</td></tr>\n",
       "\t<tr><td>ATG4D    </td><td>-2.669</td></tr>\n",
       "\t<tr><td>PIP4K2C  </td><td>-2.684</td></tr>\n",
       "\t<tr><td>MAP1LC3A </td><td>-2.716</td></tr>\n",
       "\t<tr><td>PINK1    </td><td>-2.797</td></tr>\n",
       "\t<tr><td>UBQLN1   </td><td>-2.802</td></tr>\n",
       "\t<tr><td>EMC6     </td><td>-2.957</td></tr>\n",
       "\t<tr><td>ATG9A    </td><td>-2.995</td></tr>\n",
       "\t<tr><td>ARFIP2   </td><td>-3.055</td></tr>\n",
       "\t<tr><td>SMC2     </td><td>-3.079</td></tr>\n",
       "\t<tr><td>SEC22B   </td><td>-3.088</td></tr>\n",
       "\t<tr><td>RB1CC1   </td><td>-4.001</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 114 × 2 of type chr\n",
       "\\begin{tabular}{ll}\n",
       " Gene & Score\\\\\n",
       "\\hline\n",
       "\t SH3GLB1   & 2.729\\\\\n",
       "\t IFT88     & 2.491\\\\\n",
       "\t MTMR3     & 2.271\\\\\n",
       "\t TP53INP1  & 2.134\\\\\n",
       "\t ATG16L2   & 1.973\\\\\n",
       "\t ATG2B     & 1.944\\\\\n",
       "\t ATG14     & 1.935\\\\\n",
       "\t ATG7      & 1.597\\\\\n",
       "\t STING1    & 1.457\\\\\n",
       "\t MAP1LC3B2 & 1.385\\\\\n",
       "\t ATG9B     & 1.105\\\\\n",
       "\t TBC1D12   & 1.061\\\\\n",
       "\t FEZ1      & 1    \\\\\n",
       "\t EPHB2     & 0.888\\\\\n",
       "\t NUPR1     & 0.822\\\\\n",
       "\t ATG4C     & 0.784\\\\\n",
       "\t WDR45B    & 0.775\\\\\n",
       "\t TMEM39A   & 0.751\\\\\n",
       "\t PIP4K2A   & 0.659\\\\\n",
       "\t ATG2A     & 0.622\\\\\n",
       "\t RAB7A     & 0.588\\\\\n",
       "\t EFNB1     & 0.571\\\\\n",
       "\t RNF186    & 0.551\\\\\n",
       "\t UBXN2A    & 0.516\\\\\n",
       "\t WIPI2     & 0.464\\\\\n",
       "\t BECN1     & 0.392\\\\\n",
       "\t IFT20     & 0.389\\\\\n",
       "\t TBC1D14   & 0.381\\\\\n",
       "\t WIPI1     & 0.352\\\\\n",
       "\t SNX18     & 0.35 \\\\\n",
       "\t ⋮ & ⋮\\\\\n",
       "\t SCFD1     & -1.858\\\\\n",
       "\t LRSAM1    & -1.899\\\\\n",
       "\t UBXN2B    & -1.922\\\\\n",
       "\t RAB3GAP1  & -1.924\\\\\n",
       "\t GABARAPL3 & -1.988\\\\\n",
       "\t MAP1LC3B  & -2.017\\\\\n",
       "\t MTOR      & -2.019\\\\\n",
       "\t RNF5      & -2.026\\\\\n",
       "\t VMP1      & -2.122\\\\\n",
       "\t SMURF1    & -2.124\\\\\n",
       "\t RAB1A     & -2.16 \\\\\n",
       "\t TRAF6     & -2.232\\\\\n",
       "\t RAB3GAP2  & -2.265\\\\\n",
       "\t PHF23     & -2.39 \\\\\n",
       "\t ATP2A2    & -2.517\\\\\n",
       "\t ATG5      & -2.518\\\\\n",
       "\t RALB      & -2.528\\\\\n",
       "\t STX12     & -2.608\\\\\n",
       "\t MOAP1     & -2.609\\\\\n",
       "\t ATG4D     & -2.669\\\\\n",
       "\t PIP4K2C   & -2.684\\\\\n",
       "\t MAP1LC3A  & -2.716\\\\\n",
       "\t PINK1     & -2.797\\\\\n",
       "\t UBQLN1    & -2.802\\\\\n",
       "\t EMC6      & -2.957\\\\\n",
       "\t ATG9A     & -2.995\\\\\n",
       "\t ARFIP2    & -3.055\\\\\n",
       "\t SMC2      & -3.079\\\\\n",
       "\t SEC22B    & -3.088\\\\\n",
       "\t RB1CC1    & -4.001\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 114 × 2 of type chr\n",
       "\n",
       "| Gene | Score |\n",
       "|---|---|\n",
       "| SH3GLB1   | 2.729 |\n",
       "| IFT88     | 2.491 |\n",
       "| MTMR3     | 2.271 |\n",
       "| TP53INP1  | 2.134 |\n",
       "| ATG16L2   | 1.973 |\n",
       "| ATG2B     | 1.944 |\n",
       "| ATG14     | 1.935 |\n",
       "| ATG7      | 1.597 |\n",
       "| STING1    | 1.457 |\n",
       "| MAP1LC3B2 | 1.385 |\n",
       "| ATG9B     | 1.105 |\n",
       "| TBC1D12   | 1.061 |\n",
       "| FEZ1      | 1     |\n",
       "| EPHB2     | 0.888 |\n",
       "| NUPR1     | 0.822 |\n",
       "| ATG4C     | 0.784 |\n",
       "| WDR45B    | 0.775 |\n",
       "| TMEM39A   | 0.751 |\n",
       "| PIP4K2A   | 0.659 |\n",
       "| ATG2A     | 0.622 |\n",
       "| RAB7A     | 0.588 |\n",
       "| EFNB1     | 0.571 |\n",
       "| RNF186    | 0.551 |\n",
       "| UBXN2A    | 0.516 |\n",
       "| WIPI2     | 0.464 |\n",
       "| BECN1     | 0.392 |\n",
       "| IFT20     | 0.389 |\n",
       "| TBC1D14   | 0.381 |\n",
       "| WIPI1     | 0.352 |\n",
       "| SNX18     | 0.35  |\n",
       "| ⋮ | ⋮ |\n",
       "| SCFD1     | -1.858 |\n",
       "| LRSAM1    | -1.899 |\n",
       "| UBXN2B    | -1.922 |\n",
       "| RAB3GAP1  | -1.924 |\n",
       "| GABARAPL3 | -1.988 |\n",
       "| MAP1LC3B  | -2.017 |\n",
       "| MTOR      | -2.019 |\n",
       "| RNF5      | -2.026 |\n",
       "| VMP1      | -2.122 |\n",
       "| SMURF1    | -2.124 |\n",
       "| RAB1A     | -2.16  |\n",
       "| TRAF6     | -2.232 |\n",
       "| RAB3GAP2  | -2.265 |\n",
       "| PHF23     | -2.39  |\n",
       "| ATP2A2    | -2.517 |\n",
       "| ATG5      | -2.518 |\n",
       "| RALB      | -2.528 |\n",
       "| STX12     | -2.608 |\n",
       "| MOAP1     | -2.609 |\n",
       "| ATG4D     | -2.669 |\n",
       "| PIP4K2C   | -2.684 |\n",
       "| MAP1LC3A  | -2.716 |\n",
       "| PINK1     | -2.797 |\n",
       "| UBQLN1    | -2.802 |\n",
       "| EMC6      | -2.957 |\n",
       "| ATG9A     | -2.995 |\n",
       "| ARFIP2    | -3.055 |\n",
       "| SMC2      | -3.079 |\n",
       "| SEC22B    | -3.088 |\n",
       "| RB1CC1    | -4.001 |\n",
       "\n"
      ],
      "text/plain": [
       "      Gene      Score \n",
       " [1,] SH3GLB1   2.729 \n",
       " [2,] IFT88     2.491 \n",
       " [3,] MTMR3     2.271 \n",
       " [4,] TP53INP1  2.134 \n",
       " [5,] ATG16L2   1.973 \n",
       " [6,] ATG2B     1.944 \n",
       " [7,] ATG14     1.935 \n",
       " [8,] ATG7      1.597 \n",
       " [9,] STING1    1.457 \n",
       "[10,] MAP1LC3B2 1.385 \n",
       "[11,] ATG9B     1.105 \n",
       "[12,] TBC1D12   1.061 \n",
       "[13,] FEZ1      1     \n",
       "[14,] EPHB2     0.888 \n",
       "[15,] NUPR1     0.822 \n",
       "[16,] ATG4C     0.784 \n",
       "[17,] WDR45B    0.775 \n",
       "[18,] TMEM39A   0.751 \n",
       "[19,] PIP4K2A   0.659 \n",
       "[20,] ATG2A     0.622 \n",
       "[21,] RAB7A     0.588 \n",
       "[22,] EFNB1     0.571 \n",
       "[23,] RNF186    0.551 \n",
       "[24,] UBXN2A    0.516 \n",
       "[25,] WIPI2     0.464 \n",
       "[26,] BECN1     0.392 \n",
       "[27,] IFT20     0.389 \n",
       "[28,] TBC1D14   0.381 \n",
       "[29,] WIPI1     0.352 \n",
       "[30,] SNX18     0.35  \n",
       "[31,] ⋮         ⋮     \n",
       "[32,] SCFD1     -1.858\n",
       "[33,] LRSAM1    -1.899\n",
       "[34,] UBXN2B    -1.922\n",
       "[35,] RAB3GAP1  -1.924\n",
       "[36,] GABARAPL3 -1.988\n",
       "[37,] MAP1LC3B  -2.017\n",
       "[38,] MTOR      -2.019\n",
       "[39,] RNF5      -2.026\n",
       "[40,] VMP1      -2.122\n",
       "[41,] SMURF1    -2.124\n",
       "[42,] RAB1A     -2.16 \n",
       "[43,] TRAF6     -2.232\n",
       "[44,] RAB3GAP2  -2.265\n",
       "[45,] PHF23     -2.39 \n",
       "[46,] ATP2A2    -2.517\n",
       "[47,] ATG5      -2.518\n",
       "[48,] RALB      -2.528\n",
       "[49,] STX12     -2.608\n",
       "[50,] MOAP1     -2.609\n",
       "[51,] ATG4D     -2.669\n",
       "[52,] PIP4K2C   -2.684\n",
       "[53,] MAP1LC3A  -2.716\n",
       "[54,] PINK1     -2.797\n",
       "[55,] UBQLN1    -2.802\n",
       "[56,] EMC6      -2.957\n",
       "[57,] ATG9A     -2.995\n",
       "[58,] ARFIP2    -3.055\n",
       "[59,] SMC2      -3.079\n",
       "[60,] SEC22B    -3.088\n",
       "[61,] RB1CC1    -4.001"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Individual gene scores from a gene set analysis\n",
    "# look at 10th gene set\n",
    "GSA.genescores(geneset.number = 10, genesets = genesets, \n",
    "               GSA.obj = GSA.obj, genenames = genenames)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c77eafab-f0e9-4522-8306-496d91dec8d5",
   "metadata": {},
   "source": [
    "To use the `GSA.genescores` function, we need to pass the following arguments:\n",
    "\n",
    "- `geneset.number`: A number specifying which gene set is to be examined.\n",
    "- `genesets`: A list containing the gene set collection.\n",
    "- `GSA.obj`: A GSA object returned by the GSA function.\n",
    "- `genenames`: A vector containing gene names for genes in the expression dataset.\n",
    "\n",
    "The `GSA.genescores` returns a matrix of gene names and gene scores (eg t-statistics) for each gene in the gene set.\r\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "1cfde1f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAANlBMVEUAAAAil+ZNTU1h0E9o\naGh8fHyMjIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enw8PD///9W+WhdAAAACXBIWXMA\nABJ0AAASdAHeZh94AAAehklEQVR4nO3di5aiuAJG4TSFFFreeP+XbYHSQkUI5CckYX/rrB6n\nLobTuicKAU0FwJlZewOAFBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQE\nCBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQ\nEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIg\nQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBI\ngAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAA\nIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEB\nAvNDOu0LUyvKk3B7gCjNDemamz+74SGAyPgLqTTZz7m5dTlmplxiCGAlHkPKzPlx+2yyJYYA\nVuIxpKfZb3gqJCREhhkJEPD7Hul4aW7xHgmp8RhStevs48iviwwBrMNnSNWpbI4jZcV+5DgS\nISEyXkMKaQhAiZAAAb8v7WyXCBESIuMxpClLhGYOAayEJUKAAAdkAQGWCAECzEiAAEuEAAGW\nCAECLBECBFjZAIz6evX2E4QEjHnr6L0k/yEdcmOK46JDAEo9Hb2V5P040u8eh8GddoSERTUZ\n3Guo/+x5rdb9aYuSfIdUmvJaVZfSHN6/7XRtI8DOhy76U/r0w6uHlJlmv/fV5EsMAYwZKONT\nLTbd+Q7pPtmwRAirGJhiPr+8s/hZ3yF930NiiRDWMCskC15DKvaHo/m53byWLBHCOgZe2Lnc\nrdeQHjsSjMlYIoR1LDIheT2OdD4fDkXR7HIoBzsiJCzmq7NX4b4L3L0jVjZgMwTzzmeEhPQt\nmlCLkJAyDwm11gqJ40hYlreEWoSE1HhOqMVLO6RjlYRahIQ49B9DFR4JckNIiMLHw6hbDIlr\nf2Mmq47WLIlrfyMGva/kNhoS1/7GbKFPSFxpFXEIvCOu/Y1IPCcTWkfMSIjJ+sF8wrW/EYdw\nG2pw7W8EL/CGGlz7GyGLoaEGKxsQpgB2IExBSAhOXAm1CAkhibGhBiEhENE21CAkrC/uhhqE\nhFUl0FCDkLCWVBpqEBJWkFRDDUKCX+k11CAkeJNoQw1CggeRLVOYgZCwrOQTahESFrORhhqE\nhCVsqaEGIcHO4/zu9l9+b7anff+d+f1yRnjne2tv/8IICVb6LpLwSdV7e+3/B8siJNgY62W0\nKcePaA0eIcHGlAmpNyZCUvxKgENgEpeQHkGljJBg5WWCsZmEXn527f8HyyIk2HnU0NvIy8zz\n9+Xq/sNrb//CCAmTJV/FDISESdJ/kTYPIcEeEX1ESLDDVDSIkGCBiMYQEkYwFdkgJAwhIkuE\nhE+YiiYgJPQhookICW+IaDpCwhOmonkICX+IaDZCQoupyAkhoWIqckdIm8dUpOD3w5j3zYcx\nm6Lkw5gDQUQiHkO65ubPbpEhMAVTkZDHkEqT/ZybW5djZsolhoA1ItLyGFJmzo/bZ5MtMQSs\nMBXpeQzJmE//IhsC44hoEcxIW8JUtBi/75GOl+YW75FWQESL8rn7e9fZa5dfFxkC/YhoaX6P\nI5XNcaSs2HMcyR+mIh9Y2ZA2IvKEkNLFVOQRS4QSRUR+sUQoQUxF/rFEKDVEtAoOyKaEqWg1\nLBFKBhGtiRkpCUxFa2OJUPyIKAAsEYobU1EgWCIULyIKCCsbIkVEYSGkCDEVhcdnSNdvY3bH\n3zth9/dMRBQkn0uEsnahXXsnhDQDU1GwvO7+PtxqOmTNMrv3kEzXzCGSRkQh83pAtvnHJcsv\nzEgTMRWFboUlQtfdjpCmIKIIeAwpN/eDsPmOkCwxFUXCY0gH8/1762J2hGSBiOLhc/d3+ajn\nOLI/gZCYiiLj9YDsubjfunwT0gAiig4rG0LDVBQlQgoJEUWLkIJBRDFbKyR2NjxhKoodIa2P\niBLAS7t1MRUlgpBWRETpIKSVMBWlhWt/r4GIksO1v31jKkoS1/72iohSxZVWvWEqShnX/vaD\niBLHjLQ8pqIN4NrfyyKijeDa3wsiou3g2t8LYSraFlY2LIGINoeQ1JiKNomQpIhoqwhJhqlo\nywhJg4g2jpDcMRWBkFwREWqE5ICpCHeENBcRoYOQ5mAqwgtCmoqI0IOQJiEi9CMka0xF+IyQ\n7BARBhHSOKYijCKkEUQEG4Q0gKkItgjpEyLCBITUh6kIExHSGyLCdIT0hKkI8xDSHyLCbITU\nYiqCE0IiIghsPiQigsKmQ2Iqgsp2QyIiCG0zJKYiiG0wJCKC3sZCYirCMrYUEhFhMVsJiakI\ni9pESESEpSUfElMRfEg7JCKCJ+mGxFQEj9IMiYjgWYIhERH8SywkpiKsI6WQiAirCSck0zX5\nt5mKsKpwQnIYgoiwtuhDYipCCOIOiYgQiHhDYipCQCINiYgQlghDYipCeGILiYgQpJhCYipC\nsGIJiYgQtChCIiKELviQmIoQg7BDIiJEItyQmIoQkUBDIiLEJcyQqAiRCTOktT/VHJiIkAAB\nQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUIC\nBAgJECAkQICQAAFCAgQICRDwGtJpX5haUZ6WGgJYhceQrrn5s1tkCGAlHkMqTfZzbm5djpkp\nlxgCWInHkDJzftw+m2yJIYCVeAzJmE//IhsCWAkzEiDg9z3S8dLc4j0SUuNz9/eus9cuvy4y\nBLAOv8eRyuY4UlbsOY6EtLCyARBwDulY1Dvgiotoe/qGAILnGlLzvuf2tcy6pENuTHGUbxWw\nJseQDmZ3rUM6mO/x32t+8XePw+BOO0JCbBxDysy1DWT4AGv7e/WPlKa8VtWlNAfxVgFrcgyp\neVk3JaS6vJurycVbBazJMaT8d0Y6D4fR/l43OJYIISma90jHbPilWvt79S9+30NiiRBS4rrX\nrrA6v6j9PVPsD0fzc7t5LVkihKRIjiOZ4sfm9341NzOWCCElPlc2nM+HQ1E0uxzKwY4ICbFh\niRAgINj93cgGdx64DAFEQBTSxeI4ElcRQrocQjqarvHjSFxFCOlymZG6YeQjc0w1ehWhpyyn\nbxWwJtV7JAtcswHp4ipCgIAqpFMx+nvMSEiXa0il/dsariKEdDmG9NfRyDmvNa4ihGQ5n9j3\nc+vjctmZ8b12XEUI6RLstdvfZqOzzfLveUMAERCEdKzPRdIe+iEkRMYxpOL20u5i8upESNg0\nx5COdUDNToTxqwg93wnHkZAS193f+/b88ZHLa/XcCSEhJZyPBAi4vkeaOhNNHwKIgMdFqzOH\nACIguK6dPU7sQ6ocQ7oWO5slDe3PcmIfkuX80m7SotWhE/sctwpYk8eQOI0C6eLEPkDAY0jM\nSEiXx5A4sQ/p8rmygRP7kCyvS4Q4sQ+pYq0dIEBIgIDrEqH9RbYpH4YAIuB+QHaJlggJkXFd\na/fzvURLhITICN4jnfa5uiVCQmQ0OxvO2W1eGv9cc5chgJBJQjrubD/ZfPYQQNDcQ7rub9NR\nfrzeahq/kP68IYDQuYZ0qnc2lO1qVN1554SEyDifam7yw33Z3PCK7rlDABFwPY5UWHwKhdsQ\nQAS4HBcgwOW4AAGvl+OaNQQQAY+X45o5BBABj1cRmjkEEAFCAgQ4sQ8QICRAwDmkY1G/qiu0\nJyQREiLjGtKufXtkMmlJhITIOIZ0MLtrHdJh6mfI2g8BRMAxpMxc29UN7LXDpgmWCBESIFgi\nVDd0NrlskypCQnQ075GOmfKKDYSE6LjutSusPsrSaQggfJLjSKb4EW1O7xBA8FjZAAi4nkYh\n25CPQwARcN39veOaDYDiKkKm1J/aR0iIjOt7pEt74W/xSzxCQmQEOxsuZWbEL/EICZHR7LU7\ncIYstk0xIzWv7qRHkggJkZG8R8pKu7ORTvt2IUQxtn+CkBAZwV67b8u9dte8c6mU4SVFhITI\nOB9Hsn9JV5rsp/3YissxM4PXOiYkRMbjyobMnB+3z8OfXEFIiIzzzoafne2i1acde8N7+QgJ\nkZFc/MTuNApmJKTL+cS+rD4Sa3Vi3+090rHdu8d7JKTGea9dO8tYnWq+6+y1ywffXRESIqP6\nfCSrlQ2nsjmOlBV7jiMhLbIZSfb5sa9DABHw+B5p5hBABDzutauu33+rxNn9jaS4H0eyvvjJ\nNWsX2rV3QkhIiceLn5T1y7/rIWsmL0JCUjyGlLW/eMnyCyEhMa4hXct6d11WWiy6u7dz3e0I\nCYlxDOmS/V5C3+LzkerrhP/e2hES0uIY0s5813Vcy999CEP+PkPpYnaEhKT4XNlQPn7oOHKN\nB0JCZAQfNFa7Wi0ROj+mrcs3ISEljiGVZlcvmzvthldzuwwBRMDnyoaZQwDh06xs2NmttOMq\nQkiVxwOyXEUI6fK6RGjwKkKma+YQwEpcQzrkty5yk1tc245rNiBdjiEd68mjWdU9XhJXEUK6\nnFc2/DTXa/ix2G3HjIR0CVY2nOv3OxZva7iKENIlCKkwR7slQlxFCMlyfml3Ptav0mxe2nEV\nIaTLfWeDMft6QuIT+7Blzru/23c72s8ZIyTExuMB2bCGAJTWConjSEiKQ0h1C7NX9RASkrJW\nSPKtAtbEeyRAgJAAAaeXdlPPfODEPqTKY0ic2Id0ub60K5qPdTll3x9+uGPkxD7HrQLW5HwV\nofsHjY1fRYjTKJAujxeI5MQ+pMv5ApH2H33JjIR0Ob+0y+odcLf3PPvR3+PEPqRLdYHI8Wvo\nc2IfEqa5QGRhdzYSJ/YhVaxsAAQICRAgJECAkAABQgIECAkQICRAgJAAAeeQjkVz2eKLaHv6\nhgCCJ1kidPtaJi2JkBAZx5AOZnetQzoYizP75g0BRMD5NIpre2oRl+PCpglO7CMkwDGk/HdG\nqj+1T4iQEBnNe6RjZg6yTaoICdFxvoqQ1eW1nIYAwic5jmQK7ccjERJiw8oGQICQAAHnj77M\nq+qSm3zkIgwuQwDhcwzpWO/7zuq9DdKSCAmRcQxpZ36aY0g/2t12hITICFY2NNf9ZmUDNk0Q\nUmGOhISNc35pdz7Wl/HmpR22zX1ng6kv+22M3bVWZwwBRMB593d7Ofxcu7SBkBAZDsgCAoQE\nCDiENP1TzZfcKmBNhAQI8NIOECAkQEAV0snmsy/dhgDC5RpSyXskQPCp5nesbMCWOV8g8qfa\nmctlx/lI2DTB6u/9bTY6s2gVmyYI6Vhf0473SNg0x5CK20u7i8mrEyFh0xTXbGg+2sX50ygW\nWyYBLM919/e+/rdv05xLoUNIiAwrGwABQgIEnFZ/S7ekbwggEs4hLZITISEyhAQIEBIgQEiA\nACEBAoQECHDxE0CAkAABVjYAAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiA\nACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIOA/pENu\nTHFcdAjAN48htdfZ37WX3C8XGQJYie+QSlNeq+pSmsMSQwAr8R1SZq717avJlxgCWInvkO6f\nozT8eUqEhMj4Dun7HlK2xBDASryGVOwPR/Nzu3kth/c2EBIi4zWkx2dkGpNdlxgCWInP40jn\n8+FQFM0uh3KwI0JCbFjZAAgQEiDgNaTTvmjeJRXlaakhgFV4DOmamz+7RYYAVuIxpNJkP+fm\n1uWYsfsbSfEYUmbOj9tnDsgiKd5Xf/f9i2wIYCXMSICA3/dIx0tzi/dISI3P3d+7zl67nCVC\nSInf40hlcxwpK/YcR0JaWNkACBASIMASIUCAJUKAQDhLhEzXzCGAlXBAFhBgiRAgwIwECLBE\nCBBgiRAgwBIhQICVDYAAIQEChAQIrBUSx5GQFEICBHhpBwgQEiBASIAAJ/YBApzYBwiEc2Kf\nYghgJZxGAQhwYh8gEOaM9NWaORDgXdAn9n11zRwV8CGeE/uoCgGL9MQ+okJYEljZwFSF9SUQ\n0hOiwipSC6mLqQrepBzSE/uq5kRHqFu3mZCeDEX11fxv4v3N+B0kZZshdb309HX/Y8Ik8/gd\nbBYhvbhHMWWSISTEFtLyb0Y6EU0piY62LbKQfDxjm1YnTjLsbNi6uELy+BqKSQZTENInTDKY\nIK6QmCcQqMhCYp5AmGILCQhSZCEtPCEx32GmuEJa+C3S+N33rigC4gpp4Z1243f/XhodoUFI\nE+7+/QcmLspDsqIKydNLu69Xne8/b8DX8tuEOMQV0uO//k6zwFsoH7rp+c33CYkFq6hFFtKv\n4VlgfijdAbp313PP958iJNSiDOnruaQZoYz56q/x9Yd+/6QjpBCSs3lz19fjt3VbglhFGVIz\nYcy7Z5uJxm5WIx90xBmSVUcz3h/df4SOMFGkIb0+kWfuU3i9h/47Hx8emxdfSDMmmmHv98CE\nhKmCDmmwGcnOuf4ImZAwVZghTdlp5jLE3LumI7wIM6T+IV6e+LOfzcORftERposopJ71OVPU\n7VjtuRu9YzrCm3hC6l96bc1+n8R4acCrLYT0aIiQsJRgQ3rfsz3nrLrnhixDcvo2tinMkPpf\nhb19cfAkvJfXcvar8xbdWYhEhRmS7RD9z+r+t0PWR20Hf5CO0CuxkJyXOVTDIdER+sUdUveJ\nrWiovcuhY0xArxRCcl9u93yXH++KjvBJ7CEpE/q9y8/BEBI+iTektiH5c3vgdCQ6wkdRhtSd\nh9TP7s8h0RE+iyykWecOTfIxJDrCgHhC+vR2SP8Wqfc+6QhDoghpcJeC9hn+1flzuVGQnNBD\nsjrvQehDSHSEYQGHZLtrm5CwvjBDmnZ4SPgs/7AvkI4wIsyQpg2xeEh0hDEphCR8oveGREcY\nRUi99/TV8zXgsyRCkj3X/y6a9/X+NeAzQuq7n85F+ukINtIISfV0/3r846tzGxhFSD1306y3\n+7TGAeiRSEiaJ/xX5wYdYQpC6r2TSRfCA5IJSfPhFC+36Qi2COnTXSxw9i3SlUxIgqf9S0h0\nBHvhhGS6Zvw+IWFF4YTkPITzE5+VqpiNkD78Pm+RMEVCIbk+819XfBMS7BFS76+z8xvTpBSS\n21P/7cwJQoI9Qur7ZdYHYaKkQnJ66n+93SQkWCOk99/lVCRMllZIDs/9nlNiCQnWCOn1N99f\n4gGjEgtp/nOfy57ABSE9/yLX4cIsqYU098nPBe3ghJCefo1LFWOe5EJSHUtyvi9sCiEN/Soh\nwVJ6IUmf/ZQEO4Tk7b6QsgRDWuJjXoBhhOTrrpC0FEOiJHhHSMPXCyIkWEkypGlP/8HLMxAS\nrBDS8GnlhAQrYYbkfHHGCb8/cn0GSoKNMEMy94+DmGvilMSbJDgKNaSqcmtpUklDP0xIsBFw\nSFXTkuNpEc4ICTbCDqk2syVKgk/hh1TNa4mQ4FMUIVVzWlIFQEiwEEtI1eQ3TIQEjyIKqTal\nJUqCP5GFVE1oiZDgT3whVdYtiQogJIyLMqTKriWmJHgTa0iVzc4HpiT4EnFIteGWCAm+RB5S\nNdySJgFCwqj4Q6oGWmJKgidJhFR9bEmzCJyQMCaVkKr+nQ/TTvD79NOEhDEeQzLPlhjifWKy\nbmDoRFlCwhiPIR08hFS9tiQJiZIwxudLu3O2W3qIVrelKSXxJglzeX2PdDbl0kPcPd4wTVjk\nyhVQMJvfnQ0Hc7b6OdHxH+eLET3uSXQ/SFaYe+10p7eKWqIkDEs8pErUEiFhWJghGe0T170l\nQsIwryGd9kWz57soTyNDyJ+48y/s1f66bEOQJo8hXfPOUaThHeH6kGouLVESBnkMqTTZT7vT\n7nLMhneEG/ET91HQ+hecRJo8hpR19n2fTTY8hPSJ+3SsdV5LhIRBXtfaffqXniHkn6j8tG5o\nvavkIVGhzkjyj698vb+pLVEShvh9j3S8NLcs3iOpp6S+u5vUEiFhiM/d37vOXrv8OjKE9k3S\nx2Xd1i0REob4PY5UNseRsmI/dhyp8vfMtW2JkjAg0JUNVc8T998v/XBWOx8ICQPiCeme0RIl\nVRYTEyFhQKBLhOo/np+5//4tXJL+InmyszgQvkCXCNV/fAhpmVd3v5QXyRs65RapCWeJ0NsF\nHdYIqdJdJG/wIhBITbAHZKuXZ6G3kKpPOx8ICZ+FukSo9hxS5S+kZvC3liyi+Hq57AodbUY0\nM9KtpAV3gPeacI28r9bw7yNhwS4Rqr2+S6q8dvTk6/MyI8dzBpGEYJcI1cJ5fjYZPW0OBaEr\n3CVCVUAhtTsOHptDQXgV7sqGKriQwtkeBCfokMJ55n6xEw6DCMlO81ounM1BaNYKye7TKAJ7\n5ga2OQhI2CGF9tQNbHMQjrBf2oX2zA1scxAOQpoisM1BOAIPKbCnblhbg4AEfGJfI7CnbmCb\ng2AEfGJfI7BnbmCbg2CEc2LfhyHCeuqGtTUIR8inUTQCe+oGtjkIRcgn9jUCe+YGtjkIRfAz\nUmBP3bC2BsEI+sS+RlhP3bC2BsEI+sS+RmBP3cA2B4EI9MS+ri8TkrC2BsGYnoOHlQ3Bjb3m\n/+dxK20dD4YbQgoNIUWJkEJDSFEipNAQUpQIKTSEFCVCCg0hRYmQQkNIUSKk0BBSlAgpNIQU\nJUIKDSFFiZBCQ0hRIqTQEFKUUvn/AayKkAABQgIECAkQICRAgJAAAUICBAgJECAkQICQAAFC\nAgQICRAgJECAkAABQgIECAkQ8BNSmZmsvPZ/4el7B+32WI57yF9+zA+7rZt9Wfc5w3Yegbcf\nW3LcziMgH9cPLyG1HwGT937h6Xtn5RPGetyyuZn5fvDstu6sDmlo2M4j8PZjS47beQTk43ri\nI6STyc7VOTOnni88fe/2T+X2WI57Nt/X+r/E38KhlVtXeBu28wi8/diS43YeAfm4vvgIqTTH\n258/Zt/zhe73DmYnDcly3KIdUzsZyrbu8PcDiw/beQTefmzJcTuPgHxcX3w8ewpTf0hm5z+t\nnS90v2dK7bPZdtyW75Ast+5gDt6G7TwC738/S47bqoeWj+uLj2ePef1PfucL3e+dxc9m23Eb\nV7MTDm3BcusKc/y+vfv2MmznEXj7sUXHbTSPgHxcX0IKqRL/BU4K6dC8pvDIOqSGrPKxJ/R6\nITWPACENjRFBSJfM94sJy60z5uf2H+tS9gIv2JDaR4CQhsYIP6Rr5vmF3eQXnqodwqGG9PsI\nENKA7PUvp/OFrPdh9D7uzv+BiwlbVwn/YoaG7fzzfQsWHvf3EZCP64uPDW73xFxed9hc/vZP\nPb4n/Qu0HveS7y7CccVb15D9xQwN2xnofQsWHffxCMjH9cVHSPvmffzRlD1fePmeNCTbcY++\nd9hN2brM1Mf7dc+soWFrv4/A248tOu7fIyAf1xcfIVmvbBCHZDnuZZWObLeurJ9T11K2T3F4\nZcPjEfC6sqHzCLCyYUj+twe3fZw6X8if9u5qXxvbjftt5OtChVt3zZqbuv9CDw1b/T0Cz19d\ndtzuIyAf1xMvz55rs6K3Hc+8fKFzs1KHZDeu6TyMPln+rdQ3c+HqhqFhq79H4Pmry47bfQTk\n43oS3d4RIESEBAgQEiBASIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBA\nSIAAIQEChAQIEBIgQEiAACEBAoQECBASIEBIgAAhAQKEBAgQEiBASIAAIQEChAQIEBIgQEiA\nACEBAoQUPe8fNYgePAjRI6QQ8CBEj5BCwIMQPUIKAQ9CwG6JlN2P+L6avPlnbq7VsTCPzwS/\nx9T+echNJvwUdFghpIAZszc3u8cXduZy+/Ny+0rzDWPK6i2kwjz/DrwgpIDd5pxzdc7Mz/0L\nP2Z/+3Nvjrdv/dT/aqrXkI5md62uu9uPwCdCCphpcjia4u8r9Wu73Pz9QPUaUnF72Ve/CCwq\n+ERIAfvdjdDZm/B9e213aV7QVZfjftcTkrlbYXu3jL/vgP2FdG/jdHttV5pTVb9duudCSCHg\n7ztg7yFVWV7/r56a8sPx0hvSetu7Zfy1B8w0c8/RfP99qTSHZodD08tLSKf2PRK7GdZASAG7\n77XrpHFrp9mbUDd2/nuPlJtDvavO1Hvybr9THdjZ4BkhBcyY5p3QUxN5e4io/H0ndGpDOjQ/\n13TVvnnKLqts8XYRUsBuYRS390JPX/v5nZ++b5Wdmj3jTT777PYC8LGywXzTkWeEFDB2HMSD\nhypghBQPHqqAEVI8eKgCRkjx4KECBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRA\ngJAAAUICBAgJECAkQICQAAFCAgQICRAgJECAkAABQgIECAkQICRAgJAAAUICBAgJECAkQOA/\nqYu8kdFYTwIAAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plot the result, this function makes a plot of the significant gene sets, based on a call to the GSA (Gene set analysis) function.\n",
    "suppressWarnings(GSA.plot(GSA.obj, fac = 1, FDRcut = 0.5))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2c644a4-e3c9-4ced-b343-78c64d6e9e16",
   "metadata": {},
   "source": [
    "To use the `GSA.plot` function, we need to pass the following arguments:\n",
    "\n",
    "- `GSA.obj`: A GSA object returned by the GSA function.\n",
    "- `fac`: A numeric value for jittering points in plot (\"factor\" in called to jitter()).\n",
    "- `FDRcut`: A numeric value specifying the false discovery rate cutpoint for sets to be plotted. A value of 1 (the default) will cause all sets to be plotted.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7107ef92",
   "metadata": {},
   "source": [
    "### Enrichment analysis using GSA and KEGG pathways\n",
    "We can use the same procedure to per enrichment analysis with the KEGG pathway. All the codes are similar but `genesets` are assigned from `KEGG_hallmark`. The code is shown below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "361aff0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "perm= 10 / 100 \n",
      "perm= 20 / 100 \n",
      "perm= 30 / 100 \n",
      "perm= 40 / 100 \n",
      "perm= 50 / 100 \n",
      "perm= 60 / 100 \n",
      "perm= 70 / 100 \n",
      "perm= 80 / 100 \n",
      "perm= 90 / 100 \n",
      "perm= 100 / 100 \n"
     ]
    }
   ],
   "source": [
    "genesets <- KEGG_hallmark\n",
    "GSA.obj <- GSA(x = as.matrix(new_expression_data), y = as.numeric(groups$groups), \n",
    "               genenames = genenames, genesets = genesets, \n",
    "               resp.type = \"Two class unpaired\", nperms = 100, random.seed = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "00afb710",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List the results from a GSA analysis\n",
    "res <- GSA.listsets(GSA.obj = GSA.obj, geneset.names = names(genesets), FDRcut = .5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e15cdae3",
   "metadata": {},
   "source": [
    "A table of the negative gene sets. \"Negative\" means that lower expression of most genes in the gene set correlates with higher values of the phenotype y. E.g., for two classes coded 1,2, lower expression correlates with class 2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "c9e7aad4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 × 5 of type chr</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Gene_set</th><th scope=col>Gene_set_name</th><th scope=col>Score</th><th scope=col>p-value</th><th scope=col>FDR</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>214</td><td>hsa04260</td><td>-0.6732</td><td>0   </td><td>0     </td></tr>\n",
       "\t<tr><td>239</td><td>hsa04723</td><td>-0.7029</td><td>0   </td><td>0     </td></tr>\n",
       "\t<tr><td>15 </td><td>hsa00190</td><td>-1.0477</td><td>0.01</td><td>0.2485</td></tr>\n",
       "\t<tr><td>89 </td><td>hsa00970</td><td>-0.8305</td><td>0.01</td><td>0.2485</td></tr>\n",
       "\t<tr><td>252</td><td>hsa04714</td><td>-0.5229</td><td>0.01</td><td>0.2485</td></tr>\n",
       "\t<tr><td>318</td><td>hsa05010</td><td>-0.497 </td><td>0.01</td><td>0.2485</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 5 of type chr\n",
       "\\begin{tabular}{lllll}\n",
       " Gene\\_set & Gene\\_set\\_name & Score & p-value & FDR\\\\\n",
       "\\hline\n",
       "\t 214 & hsa04260 & -0.6732 & 0    & 0     \\\\\n",
       "\t 239 & hsa04723 & -0.7029 & 0    & 0     \\\\\n",
       "\t 15  & hsa00190 & -1.0477 & 0.01 & 0.2485\\\\\n",
       "\t 89  & hsa00970 & -0.8305 & 0.01 & 0.2485\\\\\n",
       "\t 252 & hsa04714 & -0.5229 & 0.01 & 0.2485\\\\\n",
       "\t 318 & hsa05010 & -0.497  & 0.01 & 0.2485\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 5 of type chr\n",
       "\n",
       "| Gene_set | Gene_set_name | Score | p-value | FDR |\n",
       "|---|---|---|---|---|\n",
       "| 214 | hsa04260 | -0.6732 | 0    | 0      |\n",
       "| 239 | hsa04723 | -0.7029 | 0    | 0      |\n",
       "| 15  | hsa00190 | -1.0477 | 0.01 | 0.2485 |\n",
       "| 89  | hsa00970 | -0.8305 | 0.01 | 0.2485 |\n",
       "| 252 | hsa04714 | -0.5229 | 0.01 | 0.2485 |\n",
       "| 318 | hsa05010 | -0.497  | 0.01 | 0.2485 |\n",
       "\n"
      ],
      "text/plain": [
       "     Gene_set Gene_set_name Score   p-value FDR   \n",
       "[1,] 214      hsa04260      -0.6732 0       0     \n",
       "[2,] 239      hsa04723      -0.7029 0       0     \n",
       "[3,] 15       hsa00190      -1.0477 0.01    0.2485\n",
       "[4,] 89       hsa00970      -0.8305 0.01    0.2485\n",
       "[5,] 252      hsa04714      -0.5229 0.01    0.2485\n",
       "[6,] 318      hsa05010      -0.497  0.01    0.2485"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "neg.table <- res$negative\n",
    "head(neg.table)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e226ae92",
   "metadata": {},
   "source": [
    "A table of the positive gene sets. \"Positive\" means that higher expression of most genes in the gene set correlates with higher values of the phenotype y. See \"negative\" above for more info."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "bcd7ff3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 3 × 5 of type chr</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Gene_set</th><th scope=col>Gene_set_name</th><th scope=col>Score</th><th scope=col>p-value</th><th scope=col>FDR</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>7  </td><td>hsa00053</td><td>0.7726</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>164</td><td>hsa04520</td><td>0.474 </td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>169</td><td>hsa04810</td><td>0.2232</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 3 × 5 of type chr\n",
       "\\begin{tabular}{lllll}\n",
       " Gene\\_set & Gene\\_set\\_name & Score & p-value & FDR\\\\\n",
       "\\hline\n",
       "\t 7   & hsa00053 & 0.7726 & 0 & 0\\\\\n",
       "\t 164 & hsa04520 & 0.474  & 0 & 0\\\\\n",
       "\t 169 & hsa04810 & 0.2232 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 3 × 5 of type chr\n",
       "\n",
       "| Gene_set | Gene_set_name | Score | p-value | FDR |\n",
       "|---|---|---|---|---|\n",
       "| 7   | hsa00053 | 0.7726 | 0 | 0 |\n",
       "| 164 | hsa04520 | 0.474  | 0 | 0 |\n",
       "| 169 | hsa04810 | 0.2232 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "     Gene_set Gene_set_name Score  p-value FDR\n",
       "[1,] 7        hsa00053      0.7726 0       0  \n",
       "[2,] 164      hsa04520      0.474  0       0  \n",
       "[3,] 169      hsa04810      0.2232 0       0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pos.table <- res$positive\n",
    "head(pos.table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "0d676c13",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 47 × 2 of type chr</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Gene</th><th scope=col>Score</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>ALDH9A1</td><td>2.707 </td></tr>\n",
       "\t<tr><td>ACSS1  </td><td>2.612 </td></tr>\n",
       "\t<tr><td>ADH1B  </td><td>1.513 </td></tr>\n",
       "\t<tr><td>GRHPR  </td><td>0.854 </td></tr>\n",
       "\t<tr><td>ALDH2  </td><td>0.768 </td></tr>\n",
       "\t<tr><td>ALDH3A2</td><td>0.726 </td></tr>\n",
       "\t<tr><td>ALDH7A1</td><td>0.64  </td></tr>\n",
       "\t<tr><td>ACACB  </td><td>0.455 </td></tr>\n",
       "\t<tr><td>ADH6   </td><td>0.346 </td></tr>\n",
       "\t<tr><td>LDHAL6B</td><td>0.22  </td></tr>\n",
       "\t<tr><td>LDHD   </td><td>0.22  </td></tr>\n",
       "\t<tr><td>ADH4   </td><td>0.186 </td></tr>\n",
       "\t<tr><td>ADH1C  </td><td>0.146 </td></tr>\n",
       "\t<tr><td>ADH5   </td><td>0.136 </td></tr>\n",
       "\t<tr><td>ACSS2  </td><td>0.053 </td></tr>\n",
       "\t<tr><td>PKLR   </td><td>0.045 </td></tr>\n",
       "\t<tr><td>PCK1   </td><td>-0.022</td></tr>\n",
       "\t<tr><td>ADH1A  </td><td>-0.123</td></tr>\n",
       "\t<tr><td>PCK2   </td><td>-0.16 </td></tr>\n",
       "\t<tr><td>PC     </td><td>-0.213</td></tr>\n",
       "\t<tr><td>PDHA2  </td><td>-0.239</td></tr>\n",
       "\t<tr><td>ADH7   </td><td>-0.378</td></tr>\n",
       "\t<tr><td>ACOT12 </td><td>-0.533</td></tr>\n",
       "\t<tr><td>AKR1A1 </td><td>-0.559</td></tr>\n",
       "\t<tr><td>ALDH1B1</td><td>-0.623</td></tr>\n",
       "\t<tr><td>ACAT1  </td><td>-0.689</td></tr>\n",
       "\t<tr><td>LDHC   </td><td>-0.882</td></tr>\n",
       "\t<tr><td>LDHAL6A</td><td>-0.939</td></tr>\n",
       "\t<tr><td>PDHB   </td><td>-1.449</td></tr>\n",
       "\t<tr><td>DLD    </td><td>-1.453</td></tr>\n",
       "\t<tr><td>LDHB   </td><td>-1.481</td></tr>\n",
       "\t<tr><td>ME2    </td><td>-1.546</td></tr>\n",
       "\t<tr><td>ACAT2  </td><td>-1.793</td></tr>\n",
       "\t<tr><td>PKM    </td><td>-1.877</td></tr>\n",
       "\t<tr><td>HAGH   </td><td>-1.899</td></tr>\n",
       "\t<tr><td>DLAT   </td><td>-1.904</td></tr>\n",
       "\t<tr><td>ME3    </td><td>-2.031</td></tr>\n",
       "\t<tr><td>MDH1   </td><td>-2.045</td></tr>\n",
       "\t<tr><td>ACYP2  </td><td>-2.244</td></tr>\n",
       "\t<tr><td>GLO1   </td><td>-2.309</td></tr>\n",
       "\t<tr><td>FH     </td><td>-2.453</td></tr>\n",
       "\t<tr><td>MDH2   </td><td>-2.547</td></tr>\n",
       "\t<tr><td>LDHA   </td><td>-2.576</td></tr>\n",
       "\t<tr><td>ACYP1  </td><td>-2.677</td></tr>\n",
       "\t<tr><td>ME1    </td><td>-2.688</td></tr>\n",
       "\t<tr><td>PDHA1  </td><td>-3.078</td></tr>\n",
       "\t<tr><td>ACACA  </td><td>-3.324</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 47 × 2 of type chr\n",
       "\\begin{tabular}{ll}\n",
       " Gene & Score\\\\\n",
       "\\hline\n",
       "\t ALDH9A1 & 2.707 \\\\\n",
       "\t ACSS1   & 2.612 \\\\\n",
       "\t ADH1B   & 1.513 \\\\\n",
       "\t GRHPR   & 0.854 \\\\\n",
       "\t ALDH2   & 0.768 \\\\\n",
       "\t ALDH3A2 & 0.726 \\\\\n",
       "\t ALDH7A1 & 0.64  \\\\\n",
       "\t ACACB   & 0.455 \\\\\n",
       "\t ADH6    & 0.346 \\\\\n",
       "\t LDHAL6B & 0.22  \\\\\n",
       "\t LDHD    & 0.22  \\\\\n",
       "\t ADH4    & 0.186 \\\\\n",
       "\t ADH1C   & 0.146 \\\\\n",
       "\t ADH5    & 0.136 \\\\\n",
       "\t ACSS2   & 0.053 \\\\\n",
       "\t PKLR    & 0.045 \\\\\n",
       "\t PCK1    & -0.022\\\\\n",
       "\t ADH1A   & -0.123\\\\\n",
       "\t PCK2    & -0.16 \\\\\n",
       "\t PC      & -0.213\\\\\n",
       "\t PDHA2   & -0.239\\\\\n",
       "\t ADH7    & -0.378\\\\\n",
       "\t ACOT12  & -0.533\\\\\n",
       "\t AKR1A1  & -0.559\\\\\n",
       "\t ALDH1B1 & -0.623\\\\\n",
       "\t ACAT1   & -0.689\\\\\n",
       "\t LDHC    & -0.882\\\\\n",
       "\t LDHAL6A & -0.939\\\\\n",
       "\t PDHB    & -1.449\\\\\n",
       "\t DLD     & -1.453\\\\\n",
       "\t LDHB    & -1.481\\\\\n",
       "\t ME2     & -1.546\\\\\n",
       "\t ACAT2   & -1.793\\\\\n",
       "\t PKM     & -1.877\\\\\n",
       "\t HAGH    & -1.899\\\\\n",
       "\t DLAT    & -1.904\\\\\n",
       "\t ME3     & -2.031\\\\\n",
       "\t MDH1    & -2.045\\\\\n",
       "\t ACYP2   & -2.244\\\\\n",
       "\t GLO1    & -2.309\\\\\n",
       "\t FH      & -2.453\\\\\n",
       "\t MDH2    & -2.547\\\\\n",
       "\t LDHA    & -2.576\\\\\n",
       "\t ACYP1   & -2.677\\\\\n",
       "\t ME1     & -2.688\\\\\n",
       "\t PDHA1   & -3.078\\\\\n",
       "\t ACACA   & -3.324\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 47 × 2 of type chr\n",
       "\n",
       "| Gene | Score |\n",
       "|---|---|\n",
       "| ALDH9A1 | 2.707  |\n",
       "| ACSS1   | 2.612  |\n",
       "| ADH1B   | 1.513  |\n",
       "| GRHPR   | 0.854  |\n",
       "| ALDH2   | 0.768  |\n",
       "| ALDH3A2 | 0.726  |\n",
       "| ALDH7A1 | 0.64   |\n",
       "| ACACB   | 0.455  |\n",
       "| ADH6    | 0.346  |\n",
       "| LDHAL6B | 0.22   |\n",
       "| LDHD    | 0.22   |\n",
       "| ADH4    | 0.186  |\n",
       "| ADH1C   | 0.146  |\n",
       "| ADH5    | 0.136  |\n",
       "| ACSS2   | 0.053  |\n",
       "| PKLR    | 0.045  |\n",
       "| PCK1    | -0.022 |\n",
       "| ADH1A   | -0.123 |\n",
       "| PCK2    | -0.16  |\n",
       "| PC      | -0.213 |\n",
       "| PDHA2   | -0.239 |\n",
       "| ADH7    | -0.378 |\n",
       "| ACOT12  | -0.533 |\n",
       "| AKR1A1  | -0.559 |\n",
       "| ALDH1B1 | -0.623 |\n",
       "| ACAT1   | -0.689 |\n",
       "| LDHC    | -0.882 |\n",
       "| LDHAL6A | -0.939 |\n",
       "| PDHB    | -1.449 |\n",
       "| DLD     | -1.453 |\n",
       "| LDHB    | -1.481 |\n",
       "| ME2     | -1.546 |\n",
       "| ACAT2   | -1.793 |\n",
       "| PKM     | -1.877 |\n",
       "| HAGH    | -1.899 |\n",
       "| DLAT    | -1.904 |\n",
       "| ME3     | -2.031 |\n",
       "| MDH1    | -2.045 |\n",
       "| ACYP2   | -2.244 |\n",
       "| GLO1    | -2.309 |\n",
       "| FH      | -2.453 |\n",
       "| MDH2    | -2.547 |\n",
       "| LDHA    | -2.576 |\n",
       "| ACYP1   | -2.677 |\n",
       "| ME1     | -2.688 |\n",
       "| PDHA1   | -3.078 |\n",
       "| ACACA   | -3.324 |\n",
       "\n"
      ],
      "text/plain": [
       "      Gene    Score \n",
       " [1,] ALDH9A1 2.707 \n",
       " [2,] ACSS1   2.612 \n",
       " [3,] ADH1B   1.513 \n",
       " [4,] GRHPR   0.854 \n",
       " [5,] ALDH2   0.768 \n",
       " [6,] ALDH3A2 0.726 \n",
       " [7,] ALDH7A1 0.64  \n",
       " [8,] ACACB   0.455 \n",
       " [9,] ADH6    0.346 \n",
       "[10,] LDHAL6B 0.22  \n",
       "[11,] LDHD    0.22  \n",
       "[12,] ADH4    0.186 \n",
       "[13,] ADH1C   0.146 \n",
       "[14,] ADH5    0.136 \n",
       "[15,] ACSS2   0.053 \n",
       "[16,] PKLR    0.045 \n",
       "[17,] PCK1    -0.022\n",
       "[18,] ADH1A   -0.123\n",
       "[19,] PCK2    -0.16 \n",
       "[20,] PC      -0.213\n",
       "[21,] PDHA2   -0.239\n",
       "[22,] ADH7    -0.378\n",
       "[23,] ACOT12  -0.533\n",
       "[24,] AKR1A1  -0.559\n",
       "[25,] ALDH1B1 -0.623\n",
       "[26,] ACAT1   -0.689\n",
       "[27,] LDHC    -0.882\n",
       "[28,] LDHAL6A -0.939\n",
       "[29,] PDHB    -1.449\n",
       "[30,] DLD     -1.453\n",
       "[31,] LDHB    -1.481\n",
       "[32,] ME2     -1.546\n",
       "[33,] ACAT2   -1.793\n",
       "[34,] PKM     -1.877\n",
       "[35,] HAGH    -1.899\n",
       "[36,] DLAT    -1.904\n",
       "[37,] ME3     -2.031\n",
       "[38,] MDH1    -2.045\n",
       "[39,] ACYP2   -2.244\n",
       "[40,] GLO1    -2.309\n",
       "[41,] FH      -2.453\n",
       "[42,] MDH2    -2.547\n",
       "[43,] LDHA    -2.576\n",
       "[44,] ACYP1   -2.677\n",
       "[45,] ME1     -2.688\n",
       "[46,] PDHA1   -3.078\n",
       "[47,] ACACA   -3.324"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Individual gene scores from a gene set analysis\n",
    "# look at 10th gene set\n",
    "GSA.genescores(geneset.number = 10, genesets = genesets, \n",
    "               GSA.obj = GSA.obj, genenames = genenames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "2284bc1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAANlBMVEUAAAAil+ZNTU1h0E9o\naGh8fHyMjIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enw8PD///9W+WhdAAAACXBIWXMA\nABJ0AAASdAHeZh94AAAZgElEQVR4nO3diVbiSACG0ZoIiIgs7/+yI9Aq7sH8VBbuPXN6aJdJ\nIX4TqFRC2QOdlb4HAFMgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFB\ngJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAk\nCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKE\nBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGA\nkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCPh7SE8Pi3KwWD4FxwOj9NeQdrPyZh4d\nEozPX0NaluZxc7y1XTdlmRsQjNFfQ2rK5vX2pjSZwcBY/TWkUr77C9wgeyQI6PAaab093vIa\nCf4+/T0/m7Wb7ZJDgvHpcBxpeTyO1CweHEfi5pkmgAAhQYAlQhBgiRAEWCIEAQ7IQkCFJUIF\nRuYPQVz+LUcX7JFMDDIyFUO6YImQkBiZiiFdsERISIxMzZDaLxESEiNTNaQhbQKShAQB9UNa\nzUpZrK+6CaitYkinqfZ/Mw4/L2wQEiNTO6RlWe72++2yrD5/utPRLehT7ZCacpz33pXZNTYB\nPakd0svO5pclQn/cBPSkdkj3LyFZIsSUVA1p8bBal8fnm7ulJUJMStWQXicSSmksEWJKah5H\n2mxWq8XiOOWw/PlqXEJiaO7ufvy0lQ3Qwt3xn+8JCX539/rHN4QEvxtsSI4jMSofntp9esUk\nJGjjXTqfXzF5ageX+uKJnpDgUkKChJ6f2rW+9reQGIbvJur6nGy44NrfQmIIflnNcK7ude3a\nXvtbSPTvgoyqhuRKq4zJJRn1cM2Gr/4S2wSEXLQ72tsjwRcuzci1v+GTyzNy7W/46A8ZufY3\nvPeX3dHeygY498eMhARn/pqRkODVn3dHeyHBP10yEhKcdMpISHDQbXe0FxIEMhISBDISEjcv\nkZGQuHGR3dFeSNy0VEZC4pbFMhIStyu3O9oLiVsVzUhI3KhsRkLiJoV3R3shcYPyGQmJm3ON\njITErblKRkLitlxnd7QXErfkahkJiRtyvYyExM244u5oLyRuxHUzEhK34coZCYlbcO3d0V5I\nTF+FjITE0MR/7WtkJCQG5vP7hbf8vm++q8ruaC8khuXu9MfFv/7f9FcrIyExLKeQLt4t3b3+\n8e6j1TISEgNz9/L7f1EEX35HxYyExNA8Z3R5SF88tau5O9oLiSFq89Tu7uD8rx8+Gx/Uz4TE\nAP3Qwd0/P39/ekC/EhLD9HHGoVVAb99am5AYpLuzeYf9dwF9+dE+MhISw/Q63/Djk7xvv7M+\nITFEZxN336R0wT6qhqohPT0sysFi+XStTTARr6+Q7r6sY2AZVQ1pNytv5lfZBNPxGsXnJ3iX\nvGKqpWJIy9I8bo63tuumLK+xCabo7n1K3y5PrTScr1UMqSmb19ub0lxjE4zMxz3Nd1/29tm+\nV3l/p2JIpXz3l9gmGJf3h4p+Xc/w/WGkvjOyR6JH79fUfVph92kR6renV/SeUe3XSOvt8ZbX\nSBz8EtL+q+d9A2jmSzWnv+dns3az3VU2wai0f2r3usTh6mP6o7rHkZbH40jN4sFxJA7aTTa8\nvjR6l9qworKygUF6m1e4+/rJ318v7nAlQmKoPi33Pg/pDyf/XZUlQgzYh5bu3u2QbjUkS4S4\nzNmzu8+TDbf71M4SITr4dFbfzU42OCDL737O4/M5skPJyRIhhuTjE7avOrkb4hM8eyQG5OMU\nwm+dDGfKwRIhBuRDGL92cpMhWSLErz4tGvp9lzSIjiwRYlg+zMz92skNTjYMaxOMwlA6+ZWQ\nIKBmSLv7Uubrf/8R099MSc0lQs1pod3pPyIkpqTq9PfquaZVc1xm9zmkcu6Pm4CeVD0ge/zX\ntplt7ZGYmB6WCO3mcyExMRVDmpWXg7CzuZCYloohrcr9v1vbMhcSk1Jz+nv5Ws/6l/kEITEy\nVQ/IbhYvt7b3QmJKrGyAACFBgJAgoK+QTDYwKUKCAE/tIEBIECAkCHDtbwhw7W8IcO1vCHCl\nVQhw7W8IsEeCANf+hgDX/oYA1/6GACsbIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQI\nEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQE\nAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBgOCGVc9fZBFzLcEKqvAlIEhIECAkC\nhAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFB\ngJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAk\nCKga0tPDohwslk/X2gT0omJIu1l5M7/KJqAnFUNaluZxc7y1XTdleY1NQE8qhtSUzevtTWmu\nsQnoScWQSvnuL7FNQE/skSCg7muk9fZ4y2skpqbm9Pf8bNZutrvKJqAfdY8jLY/HkZrFg+NI\nTIuVDRAgJAioH9JqVspifdVNQG3VjyP9m3H4cdJOSIxN7ZCWZbnb77fLsrrGJqAntUNqynHe\ne1dm19gE9KR2SC9LgywRYlJqh3T/EpIlQkxJ1ZAWD6t1eXy+uVtaIsSkdA5pvTjsaRbbFt/3\nz/FmY4kQU9I1pPkpjdK0KGmzWa0Wi+OUw/LHjoTE2HQMaVXmu0NIq3IfG9JeSIxOx5AOs9nv\nZuMyhMTIdAzp+LSudUiuIsRUdQxp9m+PtPn5AOuRqwgxXZnXSOvm5yU/R79cRaicu3xU0Keu\ns3aLVnuYI9dsYLoix5HK4rHN97mKEJPlKkIQ4CpCEBCY/j5qftzDnLiKEJMVCmnb7jiSqwgx\nUR1CWr+bsP79ONJ1RwV96rJHOj/AOvtlH3P1UUGfUq+RsoTEyPR1XTvHkZiUVEhPiwv/I0Ji\nSrqGtLzK8jghMTIdQ3rr6Jdrp/59EzACnU/se9zPy3Y7L2btuGWBWbuH573Rps3ybyf2MVmB\nkNaHc5FavEZyYh/T1TGkxfNTu22Z7Z9ahPTLiX0dRwV96hjS+hDQcTHq71cRchoF09V1+vvh\ndB3iX96m5fR9TuxjspzYBwFdXyO12BO9cGIf01Vz0aoT+5iswHXt2nNiH1PVMaTdYh5d0vDF\nJmAEOj+1s2gVhAQRfZ3Y1/smIElIECAkCBASBAgJAoQEAUKCgK5LhB62saF8swkYge4HZK/R\nkpAYma5r7R7vr9GSkBiZwGukp4dZuiUhMTKZyYZN87xf+v19zbtsAoYsEtJ63vadzf+8CRi0\n7iHtHp53R7P17rmmCy+k33oTMHRdQ3o6TDYsT1c1yZ1KISRGpvOp5mW2ejnb/OcrA/11EzAC\nXY8jLaLvQvHVJmAEKl6O64+bgBHwHrIQUPVyXH/aBIyAy3FBgKsIQYCQIMCJfRAgJAjoHNJ6\ncXhWt8iekCQkRqZrSPPTy6PSREsSEiPTMaRVme8OIa1avIfsHzcBI9AxpKbsTqsbzNpx0wJL\nhIQEgSVCh4Y2ZRYb0l5IjE7mNdK6SV6xQUiMTtdZu8W/dQ256zV82gQMX+Q4Ulk8hobz5SZg\n8KxsgICup1Fc9J1PD6dngovlL+deCImR6Tr9PW9/zYbd7Gyt+M+vqYTEyHS/ilD5bffyYlma\nx9N1u7brpvx4sQchMTJdXyNtTxf+bvMUrymb19ubny/dJSRGJjDZsF02pc1TvHeLH35eCSEk\nRiYza7dqc4asPRLTldgjHZ/d/X4k6fk10vp0soXXSExN5DVSs2x1NtL8bNZu9uOrKiExMoFZ\nu/vWF+R6Wh6PIzWLB8eRmJbOx5HCi4M+bwJGoOrKhj9tAkag82TD47z1otXd/ds0uelvJiVy\n8ZN2p1HsmtNCu9N/REhMSecT+5rDLqbViX3Lw9fsVs2xOSExKZ1n7U4HWducat6cvnHbzLZC\nYmJS74/UYmXDy5fs5nMhMTGxPdLv7x/79l5Ks7mQmJaKr5HeLiK5LXMhMSkVZ+32y9d61r8s\nchUSI9P9ONIFFz/ZLF5ube+FxJS4+AkECAkCuoa0Wx6m65plq0V3riLEVHUMadv8u4R+i/dH\nchUhpqtjSPNyf9gX7ZZl8d2Xv/rlKkLl3OWjgj5VXNngmg1MV+CNxg52FywR+vyXxKigTx1D\nWpb5Yd7gaf7zxUyO7JGYroorG1xFiOnKrGyYt3qbMVcRYrKqHpB1FSGmysoGCOga0mr2/JJn\nVmatr213+SZg+DqGtD5MYx8vahItSUiMTOeVDY/H6zU8XvpuzI4jMSmBlQ2bw1T2pat6bjWk\nu7u+R8BVBEJalPXlIbXfxKTcHf9hejo/tdusD4sULn5q134TU3L3+gcT032yoZSHww6p/Zsy\nX7iJKRHSZHWe/j4t9mnxPmN7J/Z5ajddFQ/IOrFvb7JhsiqG9MuJfYlNQE86hHSYqLvkpFan\nUTBdFUNyYh/TVfGpnT0S01X3NZIT+5ioTk/tLrzwjxP7mKyaITmxj8nq+tRucXxbl6fm/psv\n/hshMTKdryL08kZjv19F6I+bgBGoeIHIP24CRqDzBSLbv/XlHzcBI9D5qV1zmDdYN4cl4DlC\nYmRSF4j8/Rr6f94EDF/mApGL6NlIQmJ0XNcOAoQEAUKCACFBgJAgQEgQICQIEBIEdA5pvThe\ntngbGs9Xm4DBiywRev5YEy1JSIxMx5BWZb47hLQq0TP7hMTIdD6NYnc6Fcn5SNy0wIl9QoKO\nIc3+7ZEO79oXJCRGJvMaad2UVWxIeyExOp2vItTq3SU6bQKGL3IcqSxavT3SXzcBg2dlAwQI\nCQI6v/XlbL/fzsrsl2sQd9kEDF/HkNaHue/mMNsQLUlIjEzHkObl8XgM6TE7bSckRiawsuF4\n3W8rG7hpgZAWZS0kblznp3ab9eGy357acdu6TzaUw2W/S4lea1VIjEzn6e/Tu8HOsksbhMTI\nOCALAUKCgKpvxnzNUUGfhAQBntpBgJAgIBXSU/S9L4XEyHQNaek1EgTe1fxF55UNV5u5gOvr\nfIHIx/28bLdz5yNx0wKrvx+e90Ybi1a5aYGQ1odr2nmNxE3rGNLi+andtsz2T0LipiWu2XB8\naxfvRsEt6zr9/XD42305nkuRIyRGxsoGCBASBHRa/R0dyVebgJHoHNJVchISIyMkCBASBAgJ\nAoQEAUKCABc/gQAhQYCVDRAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKC\nACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUFA/ZBWs1IW66tuAmqrGNLpGpLz0+Ukf37PWSEx\nMrVDWpblbr/fLsvqGpuAntQOqSm7w+1dmV1jE9CT2iG9XCP852uFC4mRqR3S/UtIzTU2AT2p\nGtLiYbUuj883d8ufZxuExMhUDen1/V9KaXbX2AT0pOZxpM1mtVosjlMOyx87EhJjY2UDBAgJ\nAqqG9PSwOL5KWiyfrrUJ6EXFkHazs3ecnV9lE9CTiiEtS/O4Od7arhvT30xKxZCasnm9vXFA\nlkmpvvr7q7/ENgE9sUeCgLqvkdbb4y2vkZiamtPf87NZu5klQkxJ3eNIy+NxpGbx4DgS02Jl\nAwQICQIsEYIAS4QgYDhLhMq5P24CeuKALARYIgQB9kgQYIkQBFgiBAGWCEGAlQ0QICQIEBIE\n9BWS40hMipAgwFM7CBASBAgJApzYBwFO7IOA4ZzYl9gE9MRpFBDgxD4IsEeCACf2QYAT+yDA\niX0QYGUDBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQI\nCQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIA\nIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQ\nICQIEBIECAkChAQBQoIAIUGAkCBASBAwnJDKuetsAq5lOCFV3gQkCQkChAQBQoIAIUGAkCBA\nSBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIE\nCAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEARVD\nKu9dYxPQk4ohrYTEZNV8ardp5tfeBPSj6mukTVleexPQi7qTDauyufYmoA9m7SBASBAgJAio\nGtLTw+I4871YPl1rE9CLiiHtZmdHkX6eCBcSI1MxpGVpHk+Tdtt18/NEuJAYmYohNWdz35vS\nXGMT0JOqa+2++0tsE9ATeyQIqPsaab093vIaiampOf09P5u1m+2usgnoR93jSMvjcaRm8eA4\nEtMyspUN//13/Y3D5cYV0n//KYlBGtMSoWNFSmKIRrRE6L9TREJigIazROi3Czr8d9ol2SMx\nROM5IHvcG3mNxDCNaInQv10SDNB49kjmGRgwS4QgwBIhCLBECALGtbIBBkpIECAkCOgrJKea\nMylCggBP7SBASBAgJAgY04l9MFgjOrEPhms4J/YlNgE9GdNpFDBYIzqxD4bLHgkCnNgHAU7s\ng4CBntgHI3N5Dn3uLoa3qxreiC40vDswuBFdaUBCOje8EV1oeHdgcCMSUgXDG9GFhncHBjci\nIVUwvBFdaHh3YHAjElIFwxvRhYZ3BwY3IiFVMLwRXWh4d2BwIxJSBcMb0YWGdwcGNyIhVTC8\nEV1oeHdgcCMSUgXDG9GFhncHBjciIVUwvBFdaHh3YHAjElIFwxvRhYZ3BwY3IiFVMLwRXWh4\nd2BwI5pgSDAZQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAuqE\ntGxKs9x9/YF3n1vVCrvliFazD182GO3uwJ+vCR8e0dkj++nLehnQ2QObGVCVn/HpLWBmX37g\n3ec2lR7ztiNaHm82Ayyp3R3YVAzppxGdPbKfvqyXAZ09sKEB1fgZP5Vms9805emLD7z73PO/\n64TUckSbcr87/L/0vsqgLtH6DiyGMKKzR/bTl/UyoLMHNjWgGr+4y7J+/vOxPHzxgfPPrcq8\n1rOQdiNanEZTazd5gZZ3YPX2BX2O6OyR/fRlvQzo7IFNDajG78iiHN4k8+z/jmcfOP9cWdb6\nnW07opMBhtTyDqzKaggjOntkP/90exnQyWFMqQHV+B0pH//HfvaB889tqv3Oth3R0a7Mqwzq\nEi3vwKKs759fSfc9orNH9tOX9TOgo+MDmxrQkELaV/kBXziiw//W11UGdYnWIR3V+B/Bb7+3\nAwzp+MAKqdKI9tum2gv29lregVIen//Hu6zxBG98IZ0eWCFVGtGuGd4Tu4ufm1aYbR5dSP8e\n2DGF1Hwc69kHmi9/3AMa0bzGIY+LXXAH9nV+qj+N6OzfnwfX14D+PbCpAdX4xT1NjGw/zp9s\n36aYXj9XKaTWI9rO5tsqI7rQBT/SfZ2f6k8jOhvD58H1M6DXBzY1oBq/uA/HV+vrsvziAx8+\nVymktiNaD3DC7qjlHWjK4dh9jV/bH0d08O+R/fRl/Qzo7YFNDajGL27rlQ3VQmo5ou1QO2p7\nB5aH34/dssa0488rG14f2WGsbDh7YMe0smE/e5uEPf08zz4wezdBWymkliO6LzUXfV6k3R3Y\nNcebVQ4k/TSi/dsj+/6jPQ3o/IENDajK78juuMD2tL3y4QNnN/f1Qmo3onL28x6Ylj/Sw81Z\nndUNP41o//bIvv9oTwM6f2BDAxrg7wiMj5AgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAg\nJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkC\nhAQBQoIAIUGAkEZviG8oeHs8CKMnpCHwIIyekIbAgzB6QhoCD8KAPSeyPH/H7V2ZHf89K7v9\nelFe36L7JabTn6tZaeq8kTlvhDRgpTwc3sN+/vqBedk+/7l9/sjD6e3tl/tPIS3K+++hCiEN\n2PM+Z7PfNOXx5QOP5eH5z4eyfv7U4+GvZf8xpHWZ7/a7+fOXUJOQBqwcc1iXxdtHDs/tZuXt\nC/YfQ1o8P+07PAlc7KlJSAP2bxrhbDbh/vm53fb4hG6/XT/MvwipvOhhvLfMz3vA3kJ6aePp\n+bndsjztDy+XXnIR0hD4eQ/Y55D2zezwz2HXNFutt1+G1N94b5kf+4CV475nXe7fPrQsq+OE\nw7GXDyE9nV4jmWbog5AG7GXW7iyN53aOswmHxjZvr5FmZXWYqiuHmbzn79mvTDZUJqQBK+X4\nSuhdE7PTIaLlv1dCT6eQVsevO3Z1evHUbHsZ8e0S0oA9h7F4fi307mOP//ZP98+VPR1nxo/5\nPDTPTwBfVzaUex1VJqQBM3EwHh6qARPSeHioBkxI4+GhGjAhjYeHCgKEBAFCggAhQYCQIEBI\nECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQI\nCQKEBAFCgoD/ASdtRbeB/S6NAAAAAElFTkSuQmCC",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plot the result, this function makes a plot of the significant gene sets, based on a call to the GSA (Gene set analysis) function.\n",
    "suppressWarnings(GSA.plot(GSA.obj, fac = 1, FDRcut = 0.5))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57c8b700-fe1f-40f0-b96e-97d3eedda337",
   "metadata": {},
   "source": [
    "### Enrichment analysis using GSA and REACTOME pathways\n",
    "We can use the same procedure to per enrichment analysis with the REACTOME pathway. All the codes are similar but `genesets` are assigned from `REACTOME_hallmark`. The code is shown below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "41b72bcd-453b-45ae-81b2-e99ac045aedc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the probe IDs\n",
    "expression_data$PROBEID <- rownames(expression_data)\n",
    "probeIDs <- rownames(expression_data)\n",
    "# Perform gene mapping\n",
    "suppressMessages({\n",
    "    annotLookup <- AnnotationDbi::select(hgu133plus2.db, keys = probeIDs, columns = c('PROBEID', 'GENENAME', 'ENTREZID'))\n",
    "})\n",
    "# Merge DE result  data frame with annotation table\n",
    "new_expression_data = merge(annotLookup, expression_data, by = \"PROBEID\")\n",
    "# Remove NA value\n",
    "new_expression_data <- new_expression_data[!is.na(new_expression_data$ENTREZID), ]\n",
    "# Remove duplicated genes ENTREZID\n",
    "new_expression_data <- new_expression_data[!duplicated(new_expression_data$ENTREZID, fromLast = FALSE), ]\n",
    "rownames(new_expression_data) <- new_expression_data$ENTREZID\n",
    "# Drop PROBEID, GENENAME, and ENTREZID columns\n",
    "new_expression_data <- new_expression_data[, -c(1:3)]\n",
    "genenames <- rownames(new_expression_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "9f64fdfa-b967-4290-abe3-64e3d4487f58",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "perm= 10 / 100 \n",
      "perm= 20 / 100 \n",
      "perm= 30 / 100 \n",
      "perm= 40 / 100 \n",
      "perm= 50 / 100 \n",
      "perm= 60 / 100 \n",
      "perm= 70 / 100 \n",
      "perm= 80 / 100 \n",
      "perm= 90 / 100 \n",
      "perm= 100 / 100 \n"
     ]
    }
   ],
   "source": [
    "genesets <- REACTOME_hallmark\n",
    "GSA.obj <- GSA(x = as.matrix(new_expression_data), y = as.numeric(groups$groups), \n",
    "               genenames = genenames, genesets = genesets, \n",
    "               resp.type = \"Two class unpaired\", nperms = 100, random.seed = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "f47bfccc-ee95-4c5c-80df-a5233c3b82db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List the results from a GSA analysis\n",
    "res <- GSA.listsets(GSA.obj = GSA.obj, geneset.names = names(genesets), FDRcut = .5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "af79ff82-2352-4a01-9146-77318039a5bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 × 5 of type chr</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Gene_set</th><th scope=col>Gene_set_name</th><th scope=col>Score</th><th scope=col>p-value</th><th scope=col>FDR</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>149</td><td>R-HSA-983168 </td><td>-0.3143</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>187</td><td>R-HSA-9612973</td><td>-0.3   </td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>266</td><td>R-HSA-5607764</td><td>-0.4568</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>335</td><td>R-HSA-390466 </td><td>-0.5509</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>393</td><td>R-HSA-6814122</td><td>-0.6076</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>394</td><td>R-HSA-389958 </td><td>-1.1833</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 5 of type chr\n",
       "\\begin{tabular}{lllll}\n",
       " Gene\\_set & Gene\\_set\\_name & Score & p-value & FDR\\\\\n",
       "\\hline\n",
       "\t 149 & R-HSA-983168  & -0.3143 & 0 & 0\\\\\n",
       "\t 187 & R-HSA-9612973 & -0.3    & 0 & 0\\\\\n",
       "\t 266 & R-HSA-5607764 & -0.4568 & 0 & 0\\\\\n",
       "\t 335 & R-HSA-390466  & -0.5509 & 0 & 0\\\\\n",
       "\t 393 & R-HSA-6814122 & -0.6076 & 0 & 0\\\\\n",
       "\t 394 & R-HSA-389958  & -1.1833 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 5 of type chr\n",
       "\n",
       "| Gene_set | Gene_set_name | Score | p-value | FDR |\n",
       "|---|---|---|---|---|\n",
       "| 149 | R-HSA-983168  | -0.3143 | 0 | 0 |\n",
       "| 187 | R-HSA-9612973 | -0.3    | 0 | 0 |\n",
       "| 266 | R-HSA-5607764 | -0.4568 | 0 | 0 |\n",
       "| 335 | R-HSA-390466  | -0.5509 | 0 | 0 |\n",
       "| 393 | R-HSA-6814122 | -0.6076 | 0 | 0 |\n",
       "| 394 | R-HSA-389958  | -1.1833 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "     Gene_set Gene_set_name Score   p-value FDR\n",
       "[1,] 149      R-HSA-983168  -0.3143 0       0  \n",
       "[2,] 187      R-HSA-9612973 -0.3    0       0  \n",
       "[3,] 266      R-HSA-5607764 -0.4568 0       0  \n",
       "[4,] 335      R-HSA-390466  -0.5509 0       0  \n",
       "[5,] 393      R-HSA-6814122 -0.6076 0       0  \n",
       "[6,] 394      R-HSA-389958  -1.1833 0       0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "neg.table <- res$negative\n",
    "head(neg.table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "e84d66dd-c9c9-48a8-b2cb-80b373e43648",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 × 5 of type chr</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Gene_set</th><th scope=col>Gene_set_name</th><th scope=col>Score</th><th scope=col>p-value</th><th scope=col>FDR</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>118 </td><td>R-HSA-418990 </td><td>0.5567</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>316 </td><td>R-HSA-446728 </td><td>0.4345</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>320 </td><td>R-HSA-1500931</td><td>0.3282</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>981 </td><td>R-HSA-354194 </td><td>0.685 </td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>1689</td><td>R-HSA-174824 </td><td>0.2829</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>1942</td><td>R-HSA-186712 </td><td>1.4641</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 5 of type chr\n",
       "\\begin{tabular}{lllll}\n",
       " Gene\\_set & Gene\\_set\\_name & Score & p-value & FDR\\\\\n",
       "\\hline\n",
       "\t 118  & R-HSA-418990  & 0.5567 & 0 & 0\\\\\n",
       "\t 316  & R-HSA-446728  & 0.4345 & 0 & 0\\\\\n",
       "\t 320  & R-HSA-1500931 & 0.3282 & 0 & 0\\\\\n",
       "\t 981  & R-HSA-354194  & 0.685  & 0 & 0\\\\\n",
       "\t 1689 & R-HSA-174824  & 0.2829 & 0 & 0\\\\\n",
       "\t 1942 & R-HSA-186712  & 1.4641 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 5 of type chr\n",
       "\n",
       "| Gene_set | Gene_set_name | Score | p-value | FDR |\n",
       "|---|---|---|---|---|\n",
       "| 118  | R-HSA-418990  | 0.5567 | 0 | 0 |\n",
       "| 316  | R-HSA-446728  | 0.4345 | 0 | 0 |\n",
       "| 320  | R-HSA-1500931 | 0.3282 | 0 | 0 |\n",
       "| 981  | R-HSA-354194  | 0.685  | 0 | 0 |\n",
       "| 1689 | R-HSA-174824  | 0.2829 | 0 | 0 |\n",
       "| 1942 | R-HSA-186712  | 1.4641 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "     Gene_set Gene_set_name Score  p-value FDR\n",
       "[1,] 118      R-HSA-418990  0.5567 0       0  \n",
       "[2,] 316      R-HSA-446728  0.4345 0       0  \n",
       "[3,] 320      R-HSA-1500931 0.3282 0       0  \n",
       "[4,] 981      R-HSA-354194  0.685  0       0  \n",
       "[5,] 1689     R-HSA-174824  0.2829 0       0  \n",
       "[6,] 1942     R-HSA-186712  1.4641 0       0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pos.table <- res$positive\n",
    "head(pos.table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "3700db02-950e-4f26-9682-233edf803d5b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 22 × 2 of type chr</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Gene</th><th scope=col>Score</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>55970 </td><td>2.413 </td></tr>\n",
       "\t<tr><td>2773  </td><td>1.846 </td></tr>\n",
       "\t<tr><td>2771  </td><td>1.415 </td></tr>\n",
       "\t<tr><td>2787  </td><td>1.226 </td></tr>\n",
       "\t<tr><td>59345 </td><td>1.134 </td></tr>\n",
       "\t<tr><td>64805 </td><td>0.502 </td></tr>\n",
       "\t<tr><td>2788  </td><td>0.021 </td></tr>\n",
       "\t<tr><td>346562</td><td>0.012 </td></tr>\n",
       "\t<tr><td>51764 </td><td>-0.008</td></tr>\n",
       "\t<tr><td>2791  </td><td>-0.078</td></tr>\n",
       "\t<tr><td>94235 </td><td>-0.232</td></tr>\n",
       "\t<tr><td>2770  </td><td>-0.365</td></tr>\n",
       "\t<tr><td>2792  </td><td>-0.603</td></tr>\n",
       "\t<tr><td>2786  </td><td>-0.823</td></tr>\n",
       "\t<tr><td>2793  </td><td>-1.031</td></tr>\n",
       "\t<tr><td>2783  </td><td>-1.062</td></tr>\n",
       "\t<tr><td>2790  </td><td>-1.499</td></tr>\n",
       "\t<tr><td>10681 </td><td>-1.674</td></tr>\n",
       "\t<tr><td>2782  </td><td>-1.76 </td></tr>\n",
       "\t<tr><td>54331 </td><td>-2.541</td></tr>\n",
       "\t<tr><td>2785  </td><td>-2.754</td></tr>\n",
       "\t<tr><td>2784  </td><td>-2.932</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 22 × 2 of type chr\n",
       "\\begin{tabular}{ll}\n",
       " Gene & Score\\\\\n",
       "\\hline\n",
       "\t 55970  & 2.413 \\\\\n",
       "\t 2773   & 1.846 \\\\\n",
       "\t 2771   & 1.415 \\\\\n",
       "\t 2787   & 1.226 \\\\\n",
       "\t 59345  & 1.134 \\\\\n",
       "\t 64805  & 0.502 \\\\\n",
       "\t 2788   & 0.021 \\\\\n",
       "\t 346562 & 0.012 \\\\\n",
       "\t 51764  & -0.008\\\\\n",
       "\t 2791   & -0.078\\\\\n",
       "\t 94235  & -0.232\\\\\n",
       "\t 2770   & -0.365\\\\\n",
       "\t 2792   & -0.603\\\\\n",
       "\t 2786   & -0.823\\\\\n",
       "\t 2793   & -1.031\\\\\n",
       "\t 2783   & -1.062\\\\\n",
       "\t 2790   & -1.499\\\\\n",
       "\t 10681  & -1.674\\\\\n",
       "\t 2782   & -1.76 \\\\\n",
       "\t 54331  & -2.541\\\\\n",
       "\t 2785   & -2.754\\\\\n",
       "\t 2784   & -2.932\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 22 × 2 of type chr\n",
       "\n",
       "| Gene | Score |\n",
       "|---|---|\n",
       "| 55970  | 2.413  |\n",
       "| 2773   | 1.846  |\n",
       "| 2771   | 1.415  |\n",
       "| 2787   | 1.226  |\n",
       "| 59345  | 1.134  |\n",
       "| 64805  | 0.502  |\n",
       "| 2788   | 0.021  |\n",
       "| 346562 | 0.012  |\n",
       "| 51764  | -0.008 |\n",
       "| 2791   | -0.078 |\n",
       "| 94235  | -0.232 |\n",
       "| 2770   | -0.365 |\n",
       "| 2792   | -0.603 |\n",
       "| 2786   | -0.823 |\n",
       "| 2793   | -1.031 |\n",
       "| 2783   | -1.062 |\n",
       "| 2790   | -1.499 |\n",
       "| 10681  | -1.674 |\n",
       "| 2782   | -1.76  |\n",
       "| 54331  | -2.541 |\n",
       "| 2785   | -2.754 |\n",
       "| 2784   | -2.932 |\n",
       "\n"
      ],
      "text/plain": [
       "      Gene   Score \n",
       " [1,] 55970  2.413 \n",
       " [2,] 2773   1.846 \n",
       " [3,] 2771   1.415 \n",
       " [4,] 2787   1.226 \n",
       " [5,] 59345  1.134 \n",
       " [6,] 64805  0.502 \n",
       " [7,] 2788   0.021 \n",
       " [8,] 346562 0.012 \n",
       " [9,] 51764  -0.008\n",
       "[10,] 2791   -0.078\n",
       "[11,] 94235  -0.232\n",
       "[12,] 2770   -0.365\n",
       "[13,] 2792   -0.603\n",
       "[14,] 2786   -0.823\n",
       "[15,] 2793   -1.031\n",
       "[16,] 2783   -1.062\n",
       "[17,] 2790   -1.499\n",
       "[18,] 10681  -1.674\n",
       "[19,] 2782   -1.76 \n",
       "[20,] 54331  -2.541\n",
       "[21,] 2785   -2.754\n",
       "[22,] 2784   -2.932"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Individual gene scores from a gene set analysis\n",
    "# look at 10th gene set\n",
    "GSA.genescores(geneset.number = 10, genesets = genesets, \n",
    "               GSA.obj = GSA.obj, genenames = genenames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "dfcb5446-3ef1-4825-aad1-5cd331945c76",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAANlBMVEUAAAAil+ZNTU1h0E9o\naGh8fHyMjIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enw8PD///9W+WhdAAAACXBIWXMA\nABJ0AAASdAHeZh94AAAYNUlEQVR4nO3diXaiSACG0Rqixhjj8v4vO4pZTDoxKr9AmXvPnB47\nS4NtvgaKAssW6KwMvQJwD4QEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAg\nJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkC\nhAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFB\ngJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAk\nCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKE\nBAHXh/TyNCt7s/lLcH2gSteGtJmUD9PoKkF9rg1pXprnVftovWzKPLdCUKNrQ2rK6v3xqjSZ\nlYFaXRtSKT/9Bv4gWyQI6HCMtFy3jxwjwfXD39OjUbvJ5uQioDI9hrR9mbfnkZrZ0y/nkRxA\nUZleQxrTIiBJSBDQ767duVOEhERlegzpgilCQqJ3Dw9dvrvHkC6YIiQk+vbQ/ne1HkO64ISs\nkOjZw/sv1+kxpAumCAmJnlUUki0SI1bPrt0FU4SERO+qGWy4ZIrQtYuAjq7syRQhOHLtHp6Z\nDfDh6jEHIcGHLyE9nO/yZXX8KV9MSpktb7oIuFYFu3aHU0evIw6nr+sTEkMZ/2BDG9K8zDfb\n7XpeFv9+utNVUjCkvkNqSjvuvSmTWywCurrybFLfIb1tbEwRYqTq2LV7fAvJFCHGqYKQZk+L\nZXnePdzMTRFirMY/avc+kFBKY4oQIzX2kLar1WIxm7VDDvOTHQmJIZnZAAFCgoTLSxpgitBI\nFgE/unjq3MNgWyTnkRivXUcXlfQw3K6dkBith/Omrn5stgYMafBFwE8eXls646veHwsJvnr4\nNHjw/XVHn69cMtgA/zpjtOHLJYDu/Q3f+CakL1fCfj6Ocu9v/rrvtz4Pb5/88WLyTx9x72/+\nuO9G6C6/FYM7rfK3fXu7k8v/GPf+5m/retPvV7ZI/HEdb/r9yr2/+eu63fT7lXt/w5Frq3Lv\nb/ikHWt47WmUo3bjWgT87PW46YLDJyHBV68jeZcM6AkJvhISJNi1gwSDDTAEIUGAkCBASBAg\nJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkC\nhAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAH9vqv5U/uu5mU2967m3JceQ9pMyofpTRYBA+kx\npHlpnlfto/WyKfNbLAIG0mNITVm9P16V5haLgIH0GFIpP/0mtggYiC0SBPR7jLRct48cI3Fv\n+hz+nh6N2k02N1kEDKPf80jz9jxSM3tyHon7YmYDBAgJAkwRggBThCDAFCEIcEIWAkwRggBb\nJAgwRQgCTBGCAFOEIMDMBggQEgT0GdLmsZTp8vUPMfzNPelzilBzmGh3+EOExD3pdfh7satp\n0bTT7P4NqRy7chEwkF5PyLb/WzeTtS0Sd2aAKUKb6VRI3JkeQ5qUt5Owk6mQuC89hrQoj6+P\n1mUqJO5Kn8Pf8/d6lr+MJwiJyvR6QnY1e3u0fhQS98TMBggQEgQICQKGCslgA3dFSBBg1w4C\nhAQBQoIA9/6GAPf+hgD3/oYAd1qFAPf+hgBbJAhw728IcO9vCHDvbwgwswEChAQBQoIAIUGA\nkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQI\nEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAsYT\nUjl2m0XArYwnpJ4XAUlCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIE\nCAkC/kRIDw/ZPw+++gshPbT/wQ39gZAe3n+BWxESBPyBkOzacXt/ISSDDdzcnwgJbk1IECAk\nCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChFQz09pHQ0gVc6HVeAipXi79HREh1UtI\nIyKkitm1G49eQ3p5mpW92fzlVov4Www2jEaPIW0m5cP0JouAgfQY0rw0z6v20XrZlPktFgED\n6TGkpqzeH69Kc4tFwEB6DKmUn34TWwQMxBYJAvo9Rlqu20eOkbg3fQ5/T49G7SabmywChtHv\neaR5ex6pmT05j8R9MbMBAoQEAf2HtJiUMlvedBHQt97PI72OOJwctBMStekc0nK2D2S2PuP7\n9t84L/PNdruel0V4rWBIXUNqNzC7jzW/l9SG1JR23HtTJuG1giF1DGlRppt9IIvy+Pv3lbdf\njv4fWysYUseQ9huYT4Gc+r79lzy+hWSKEPekY0jtbt3ZIc2eFsvyvHu4mZsixF3pGNLkdYu0\nOn3Mc/i+V+3DxhQh7knmGGnZnB6FO1itFovZrB1ymJ/sSEjUpuuo3eysS8c7LQLGL3Ieqcye\nQ6vz7SJg9NxFCALcRQgCAsPfrebkeaHWL3cRKscuXysYUiik9Rk/++7ZwP3qENLy0zbknPNI\nP/0msVYwpC5bpONjnskvwwdbWyTuWeoY6QzuIsT9chchCEiF9DI74zvdRYh71TWk+U1GrIVE\nZTqG9NHRL7czuX4RUIHOF/Y97w591utp+X3U7spFQAUCo3ZPu63R6tLp384jcVcCIS331yJd\neowkJO5Kx5Bmu127dZlsXww28Kd1DGm5D6g9P/T7XYSuXARUoOvw99Ph1kC/3Dm10yJg/FzY\nBwFdj5Eu2BK5sI/71e+k1VMX9nVcKxhS4L5253IZBferY0ib2fTsKQ0u7ON+dd61O3/Sah1b\npIeH4ZZNvXoMqYoL+x7a/+BCLuz75OH9F7hEv+eRRn9hn5C4Tq8hjWkRP7Brx1WE9IXBBq4h\nJAgQEgQICQK6ThF6WsdW5YdFQAW6n5C9RUtCojJd59o9P96iJSFRmcAx0svTJN2SkKhMZrBh\n1ey2S2e8r3mHRcCYRUJaTsPvbC4kKtM9pM3TbnM0WW52NZ1zI/1rFgFj1zWkl/1gw/xwoVHu\n3nZCojKdLzUvk8XbFRGnL9a7dhFQga7nkWbRd6H4bhFQgR5vx3XlIqACPd6O68pFQAV6vB3X\nlYuACvR4O64rFwEV6PEuQlcuAiogJAhwYR8ECAkCOoe0nO336mbZC5KERGW6hjQ9HB6VJlqS\nkKhMx5AWZbrZh7TwHrL8aR1DasrmMLvBqB1/WmCKkJAgMEVo39CqTGKrtBUS1ckcIy2b5B0b\nhER1uo7azc56l/JOi4Dxi5xHKrPn0Op8uwgYPTMbIKDrZRSxFflxEVCBrsPfU/dsgMRdhMo8\nf2mfkKhM12Ok9eHG3+FdPCFRmcBgw3relPN28V6eDqPls982YkKiMplRu8U5V8huJkfX054+\n7yQkKpPYIrV7d7+fSZqX5vlwb+P1siknb4gnJCoTOUZq5udcjdSU1fvj1enbGwuJygRG7R7P\nHLX7tPd3eldQSFSm83mk8ycH2SJxv3qc2bA7RloedgEdI3FvOg82PE/PnrQ6PRq1m5xMUEhU\nJnLzk3Mvo3iZt+eRmtmT80jcl84X9jX7M7Eu7Ovs4WHoNaCLzqN2hwEEl5p39ND+R7VS7490\nzs1PNo8fU4kMf3/y8P4LdYptkX5//9hNc5hod/hDhHRMSLXr8Rhpvv+azaJpxyWE9Jldu8r1\nOGrXHL5x3UzWQvqHwYa6dT+PdPbNT97a2UynQuLO9Hjzk4/3m51MhcR96TGkjxvtr8tUSNyV\nriFt5vvhumZ+zqS7+Xs9y18uBBw4JMcrXKpjSOvm9Rb6Z70/0mr2/n2PIw7JCBoX6xjStDzu\nt0WbeZn99OXXGDQk53S4XJ8zG65bRO+ExOUCbzS2tzkrpEruImTXjot1DGlepvsmXqanL9Rr\n1XMXIYMNXKrHmQ2/3EWoHLt8rWBImZkN03OuRnLPBu5Xjydk3UWI+9VjSLZI3K+uIS0mu0Oe\nSZmccW87dxHifnUMabnfRWsv2DujJHcR4m51ntnw3N6v4fms2wi5ixD3KjCzYbXfTbujmQ1w\nuUBIs7IUEn9c51271XI/AHfert1Vi4AKdB9sKOVpv0G68E2ZnUfirnQe/j4MZJ/xPmNf/hAh\ncU96PCE7rkVAkpAgoENI+72zG03YFhKV6TWkSi7sg4v1uGtXz4V9cKkeQ/rlwr7EImAgnXbt\nLruo1WUU3K8eQ3JhH/er667drH1bl5fm8YcvPmKLxP3qfBehtzca+/0uQi7s4371eYNIF/Zx\ntzrfIPL8t750YR/3q/OuXbNvYrer9pRao6+LgAqkbhAZvYe+kKhN5gaRswuvRrpsETB6Zn9D\ngJAgQEgQICQIEBIECAkChAQBQoKAziEtZ+1ti9eh9fluETB6kSlCu4810ZKERGU6hrQo080+\npEU548q+6xYBFeh8GcXmcCmS+9rxpwUu7BMSdAxp8rpF2r9rX5CQqEzmGGnZlEVslbZCojqd\n7yJ01p1TOy0Cxi9yHqnMLn17pIsWAaNnZgMECAkCOr/15WS7XU/K5Jf7a3VZBIxfx5CW+7Hv\nZj/aEC1JSFSmY0jT8tyeQ3rODtsJicoEZja09/02s4E/LRDSrCyFxB/Xeddutdzf9tuuHX9b\n98GGsr/tdynRe60Kicp0Hv4+vNPRJDu1QUhUxglZCBASBHQI6dI3Y77tWsGQhAQBdu0gQEgQ\nkArpJfrel0KiMl1DmjtGgsC7mr8xs4G/rPMNIp+307JeT12PxJ8WmP39tNsarUxa5U8LhLTc\n39Ou+zHSzU5Kwe11DGm227Vbl8n2xWADf1ring3tW7t4Nwr+sq7D30/73z2W9lqKHCFRGTMb\nIEBIENBp9nd0Tb5bBFSic0g3yUlIVEZIECAkCBASBAgJAoQEAW5+AgFCggAzGyBASBAgJAgQ\nEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQB\nQoIAIUGAkCCg/5AWk1Jmy5suAvrWY0iHuxpPDzc4Pv0u6EKiMn2HNC/zzXa7npfFLRYBA+k7\npKZs9o83ZXKLRcBA+g7p7V0rTr97hZCoTN8hPb6F1NxiETCQXkOaPS2W5Xn3cDM/PdogJCrT\na0jv70hWSrO5xSJgIH2eR1qtFovZrB1ymJ/sSEjUxswGCBASBPQa0svTrD1Kms1fbrUIGESP\nIW0mR++BPr3JImAgPYY0L83zqn20XjaGv7krPYbUlNX745UTstyV3md/f/eb2CJgILZIENDv\nMdJy3T5yjMS96XP4e3o0ajcxRYh70u95pHl7HqmZPTmPxH0xswEChAQBpghBgClCEDCeKULl\n2JWLgIE4IQsBpghBgC0SBJgiBAGmCEGAKUIQYGYDBAgJAoQEAUOF5DwSd0VIEGDXDgKEBAFC\nggAX9kGAC/sgYDwX9iUWAQNxGQUEuLAPAmyRIMCFfRDgwj4IcGEfBJjZAAFCggAhQYCQIEBI\nECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQI\nCQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIA\nIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQ\nICQIGE9I5dhtFgG3Mp6Qel4EJAkJAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGA\nkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFBgJAgQEgQICQI\nEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQJ6DKl8dotFwEB6DGkhJO5Wn7t2q2Z660XA\nMHo9RlqV+a0XAYPod7BhUVa3XgQMwagdBAgJAoQEAb2G9PI0a0e+Z/OXWy0CBtFjSJvJ0Vmk\n0wPhQqIyPYY0L83zYdBuvWxOD4QLicr0GFJzNPa9Ks0tFgED6XWu3U+/iS0CBmKLBAH9HiMt\n1+0jx0jcmz6Hv6dHo3aTzU0WAcPo9zzSvD2P1MyenEfivlQ1s+G//26/aLhGPSH9t7NVEuNU\nzRSh/9qSbJMYp1qmCO072iqJsRrPFKHTN3T477Ujx0mMUjUnZP973bNznMQY1TNF6K0jITFC\n1WyR3krSEWNU2xQhh0iMkilCEGCKEATUM7MBRkxIECAkCBgqJJeac1eEBAF27SBASBAgJAio\n5sI+GLNaLuyDURvPhX2JRcBAKrqMAsarngv7YMRskSCgtgv7YJRc2AcBI72wDypzeQ5Dbi7q\n3VTVu+bXq/Y597PiQrpGvWt+vWqfs5DGq941v161z1lI41Xvml+v2ucspPGqd82vV+1zFtJ4\n1bvm16v2OQtpvOpd8+tV+5yFNF71rvn1qn3OQhqvetf8etU+ZyGNV71rfr1qn7OQxqveNb9e\ntc9ZSONV75pfr9rnfP8hwd0QEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKCACFB\ngJAgQEgQ0E9I86Y08833H/j0ucXYwj5zzReTL19Ws/Oe89W3m7+dUyt+9LP1z5cl9PIXcXgL\nmMm3H/j0udW4Xpiz13zePmzuo6TznvNqfCGdWvGjn61/viyij7+Il9KstqumvHzzgU+f2/1/\nTC/M2Wu+Ko+b/b94j0Oua8rZz3k25Fp+49SKH/1s/fNlGX384M7Lcvfrc3n65gPHn1uU6chC\nOnPNZ4e1HtnKX+nM57z4+IKROLXiRz9b/3xZRh+v/azs3yTz6J+wow8cf67Mx/azeO6aH4xs\n5a905nNelMVgq/i9Uyt+9LP17+sW0cdrX77+g330gePPrUb3s3jumrc2Zdrryt3Imc95VpaP\nu2P2AVbwJ6dW/Ohn658vCy09+8d9v4zzfxxrDmnR7jNU7+yQWiP6t+NkSP/8Vkh9uiSkdTO2\no+/rnPmcS3nebYTnI9rBE9L2HkLaNCP6x7mLC3dn08PI17v/kJqvq370gWbUIV2w5tPx/ER1\nc8Fz3o7qBTu14kf///c5RPTx93AYJ1l/HU5Zf4wDvX9uRK/L3tlrvp5M1wOtY9oFr9Z2VC/Y\nqRXf+zRqt65x1O6pPQpflvk3H/jyuRG9LnvnrvlyTAfdHZ35nJuyn8cR/3ns4NSK773+bP3z\nZRl9/OCePbNhdCGduebrO+ro3Oc83/8kbuYjGqk8PbPh/Wer4pkN28nHSOnh2Rx9YPJpFHVk\nIZ255o9lfBPPrnfec9407cMxnUg6teLbj5+tzx9N6eW137TzbQ/LK18+cPRwO76Qzlvzclch\nnflq7R9OxjP4vT294tuPn63PH025j9ceBiYkCBASBAgJAoQEAUKCACFBgJAgQEgQICQIEBIE\nCAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAhQYCQIEBIECAkCBASBAgJAoQEAUKC\nACFBgJAgQEgQIKTq3ckbBVbOi1A9IY2BF6F6QhoDL0L1hDQGXoQR2yUyP34D7k2ZtP+flM12\nOSvv76n+FtPh18WkNKN6t/E/QUgjVspT2Zm+f2Ba1rtf17uPtJ8oZb79J6RZ+fw99EJII7bb\n5qy2q6Y8v33guTztfn0qy92nnve/LduvIS3LdLPdTHdfQp+ENGKlzWFZZh8f2e/bTcrHF2y/\nhjTb7fbtdwJnW/okpBF7HUY4Gk143O3brdsduu16+TT9JqTyZoD1/cv8fY/YR0hvbbzs9u3m\n5WW7P1x6y0VIY+Dve8T+DWnbTPb/7TdNk8Vy/W1Iw63vX+avfcRKu+1ZlsePD83Loh1waHv5\nEtLL4RjJMMMQhDRib6N2R2ns2mlHE/aNrT6OkSZlsR+qK/uRvN33bBcGG3ompBErpT0S+tTE\n5HCKaP56JPRyCGnRfl3b1eHgqVkPssZ/l5BGbBfGbHcs9Oljz6/bp8ddZS/tyHibz1Oz2wF8\nn9lQHnXUMyGNmIGDenipRkxI9fBSjZiQ6uGlGjEh1cNLBQFCggAhQYCQIEBIECAkCBASBAgJ\nAoQEAUKCACFBgJAgQEgQICQIEBIECAkChAQBQoIAIUGAkCBASBAgJAgQEgQICQKEBAFCggAh\nQcD/IV9b0mbJvosAAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plot the result, this function makes a plot of the significant gene sets, based on a call to the GSA (Gene set analysis) function.\n",
    "suppressWarnings(GSA.plot(GSA.obj, fac = 1, FDRcut = 0.5))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36b0036d-b465-4b52-bef0-fa7f39bd40b3",
   "metadata": {},
   "source": [
    "# Enrichment analysis using PADOG\n",
    "\n",
    "Pathway Analysis with Down-weighting of Overlapping Genes (PADOG) is a specialized method for enrichment analysis. Focused on deciphering the functional implications of differentially expressed genes in biological pathways, PADOG uniquely emphasizes the uniqueness of genes. Its down-weighting mechanism enhances precision by addressing challenges posed by overlapping genes, offering valuable insights into specific pathways influenced by gene expression changes.\n",
    "\n",
    "To perform PADOG enrichment analysis, we will use the \"PADOG\" package. We first need to install and load the package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "b09bba68-4e7e-4259-aa48-52a2c36422b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressWarnings({\n",
    "    if (!require(\"BiocManager\", quietly = TRUE)) {\n",
    "        install.packages(\"BiocManager\")\n",
    "    }\n",
    "})\n",
    "\n",
    "suppressWarnings({suppressMessages({BiocManager::install(\"PADOG\")})})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "c14e57fa-057a-4191-ad5c-9c5657476206",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressMessages({library(PADOG)})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbae06eb-3102-4ff8-8bf4-ad718f0a1d37",
   "metadata": {},
   "source": [
    "## Data Preparation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ec13aac-5cea-4ce4-a469-714d94380a5d",
   "metadata": {},
   "source": [
    "The PADOG method requires the inputs that is quite similar to GSA method. Therefore, to perform analysis using PADOG, we will use the same prepartion steps as for the GSA method above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "8340b750-91ec-44a8-909a-160465778a3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 54</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>GSM300173</th><th scope=col>GSM300177</th><th scope=col>GSM300181</th><th scope=col>GSM300186</th><th scope=col>GSM300189</th><th scope=col>GSM300192</th><th scope=col>GSM300196</th><th scope=col>GSM300204</th><th scope=col>GSM300208</th><th scope=col>GSM300214</th><th scope=col>⋯</th><th scope=col>GSM1176201</th><th scope=col>GSM1176202</th><th scope=col>GSM1176203</th><th scope=col>GSM1176204</th><th scope=col>GSM1176205</th><th scope=col>GSM1176206</th><th scope=col>GSM1176207</th><th scope=col>GSM1176208</th><th scope=col>GSM1176209</th><th scope=col>GSM1176210</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>DDR1</th><td>0.6085351</td><td>0.8513945</td><td>1.1591995</td><td>2.0921063</td><td>1.2385797</td><td>0.8326798</td><td>0.8034588</td><td>0.6996407</td><td>1.3955209</td><td>1.1198790</td><td>⋯</td><td>1.1563261</td><td>1.6152382</td><td>1.1901788</td><td>1.4972300</td><td>0.6329609</td><td>1.4160284</td><td>1.2335104</td><td>1.2999401</td><td>0.8595833</td><td>1.2998189</td></tr>\n",
       "\t<tr><th scope=row>RFC2</th><td>0.9081397</td><td>0.9336113</td><td>0.8563885</td><td>1.0015364</td><td>0.7332213</td><td>0.8570584</td><td>1.0210657</td><td>0.9641521</td><td>0.8585522</td><td>0.8584299</td><td>⋯</td><td>0.8848516</td><td>0.9385835</td><td>1.0359085</td><td>0.9225201</td><td>1.0405487</td><td>0.8187565</td><td>0.9541863</td><td>0.9855680</td><td>1.1704986</td><td>0.9924505</td></tr>\n",
       "\t<tr><th scope=row>HSPA6</th><td>0.8674964</td><td>0.8385732</td><td>0.9956165</td><td>0.9797408</td><td>3.5688195</td><td>1.0145685</td><td>1.1729818</td><td>1.0208826</td><td>0.8163447</td><td>0.8572931</td><td>⋯</td><td>0.9469172</td><td>1.3414407</td><td>0.9509885</td><td>1.3277143</td><td>1.0712911</td><td>0.9801153</td><td>0.9847162</td><td>1.4585433</td><td>1.2102244</td><td>0.9023996</td></tr>\n",
       "\t<tr><th scope=row>PAX8</th><td>1.1053528</td><td>1.0432043</td><td>1.0137772</td><td>1.0931698</td><td>1.0356631</td><td>0.9435998</td><td>1.1165266</td><td>1.0960624</td><td>1.0099041</td><td>1.0553276</td><td>⋯</td><td>1.0055602</td><td>0.8999849</td><td>0.9461904</td><td>1.2250262</td><td>1.0535747</td><td>0.9321033</td><td>1.0115234</td><td>1.0048561</td><td>1.1175232</td><td>1.0730404</td></tr>\n",
       "\t<tr><th scope=row>GUCA1A</th><td>1.2764652</td><td>0.8625501</td><td>1.0411838</td><td>1.0323309</td><td>0.8504494</td><td>1.3334491</td><td>1.1356378</td><td>1.0168769</td><td>0.9045680</td><td>0.8634836</td><td>⋯</td><td>1.3178803</td><td>0.8809971</td><td>0.9128050</td><td>1.0231747</td><td>1.0368457</td><td>0.9140167</td><td>1.0870208</td><td>0.9601511</td><td>3.4069852</td><td>1.0806457</td></tr>\n",
       "\t<tr><th scope=row>UBA7</th><td>0.8337133</td><td>1.0081274</td><td>1.0330204</td><td>1.5409547</td><td>0.8751729</td><td>1.0063332</td><td>0.7226255</td><td>0.9519509</td><td>0.8132348</td><td>1.0214182</td><td>⋯</td><td>1.2466778</td><td>1.8324052</td><td>0.8732074</td><td>1.4028883</td><td>0.9107665</td><td>0.9586981</td><td>1.1221024</td><td>1.1861940</td><td>1.0770138</td><td>1.2797908</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 54\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & GSM300173 & GSM300177 & GSM300181 & GSM300186 & GSM300189 & GSM300192 & GSM300196 & GSM300204 & GSM300208 & GSM300214 & ⋯ & GSM1176201 & GSM1176202 & GSM1176203 & GSM1176204 & GSM1176205 & GSM1176206 & GSM1176207 & GSM1176208 & GSM1176209 & GSM1176210\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\tDDR1 & 0.6085351 & 0.8513945 & 1.1591995 & 2.0921063 & 1.2385797 & 0.8326798 & 0.8034588 & 0.6996407 & 1.3955209 & 1.1198790 & ⋯ & 1.1563261 & 1.6152382 & 1.1901788 & 1.4972300 & 0.6329609 & 1.4160284 & 1.2335104 & 1.2999401 & 0.8595833 & 1.2998189\\\\\n",
       "\tRFC2 & 0.9081397 & 0.9336113 & 0.8563885 & 1.0015364 & 0.7332213 & 0.8570584 & 1.0210657 & 0.9641521 & 0.8585522 & 0.8584299 & ⋯ & 0.8848516 & 0.9385835 & 1.0359085 & 0.9225201 & 1.0405487 & 0.8187565 & 0.9541863 & 0.9855680 & 1.1704986 & 0.9924505\\\\\n",
       "\tHSPA6 & 0.8674964 & 0.8385732 & 0.9956165 & 0.9797408 & 3.5688195 & 1.0145685 & 1.1729818 & 1.0208826 & 0.8163447 & 0.8572931 & ⋯ & 0.9469172 & 1.3414407 & 0.9509885 & 1.3277143 & 1.0712911 & 0.9801153 & 0.9847162 & 1.4585433 & 1.2102244 & 0.9023996\\\\\n",
       "\tPAX8 & 1.1053528 & 1.0432043 & 1.0137772 & 1.0931698 & 1.0356631 & 0.9435998 & 1.1165266 & 1.0960624 & 1.0099041 & 1.0553276 & ⋯ & 1.0055602 & 0.8999849 & 0.9461904 & 1.2250262 & 1.0535747 & 0.9321033 & 1.0115234 & 1.0048561 & 1.1175232 & 1.0730404\\\\\n",
       "\tGUCA1A & 1.2764652 & 0.8625501 & 1.0411838 & 1.0323309 & 0.8504494 & 1.3334491 & 1.1356378 & 1.0168769 & 0.9045680 & 0.8634836 & ⋯ & 1.3178803 & 0.8809971 & 0.9128050 & 1.0231747 & 1.0368457 & 0.9140167 & 1.0870208 & 0.9601511 & 3.4069852 & 1.0806457\\\\\n",
       "\tUBA7 & 0.8337133 & 1.0081274 & 1.0330204 & 1.5409547 & 0.8751729 & 1.0063332 & 0.7226255 & 0.9519509 & 0.8132348 & 1.0214182 & ⋯ & 1.2466778 & 1.8324052 & 0.8732074 & 1.4028883 & 0.9107665 & 0.9586981 & 1.1221024 & 1.1861940 & 1.0770138 & 1.2797908\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 54\n",
       "\n",
       "| <!--/--> | GSM300173 &lt;dbl&gt; | GSM300177 &lt;dbl&gt; | GSM300181 &lt;dbl&gt; | GSM300186 &lt;dbl&gt; | GSM300189 &lt;dbl&gt; | GSM300192 &lt;dbl&gt; | GSM300196 &lt;dbl&gt; | GSM300204 &lt;dbl&gt; | GSM300208 &lt;dbl&gt; | GSM300214 &lt;dbl&gt; | ⋯ ⋯ | GSM1176201 &lt;dbl&gt; | GSM1176202 &lt;dbl&gt; | GSM1176203 &lt;dbl&gt; | GSM1176204 &lt;dbl&gt; | GSM1176205 &lt;dbl&gt; | GSM1176206 &lt;dbl&gt; | GSM1176207 &lt;dbl&gt; | GSM1176208 &lt;dbl&gt; | GSM1176209 &lt;dbl&gt; | GSM1176210 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| DDR1 | 0.6085351 | 0.8513945 | 1.1591995 | 2.0921063 | 1.2385797 | 0.8326798 | 0.8034588 | 0.6996407 | 1.3955209 | 1.1198790 | ⋯ | 1.1563261 | 1.6152382 | 1.1901788 | 1.4972300 | 0.6329609 | 1.4160284 | 1.2335104 | 1.2999401 | 0.8595833 | 1.2998189 |\n",
       "| RFC2 | 0.9081397 | 0.9336113 | 0.8563885 | 1.0015364 | 0.7332213 | 0.8570584 | 1.0210657 | 0.9641521 | 0.8585522 | 0.8584299 | ⋯ | 0.8848516 | 0.9385835 | 1.0359085 | 0.9225201 | 1.0405487 | 0.8187565 | 0.9541863 | 0.9855680 | 1.1704986 | 0.9924505 |\n",
       "| HSPA6 | 0.8674964 | 0.8385732 | 0.9956165 | 0.9797408 | 3.5688195 | 1.0145685 | 1.1729818 | 1.0208826 | 0.8163447 | 0.8572931 | ⋯ | 0.9469172 | 1.3414407 | 0.9509885 | 1.3277143 | 1.0712911 | 0.9801153 | 0.9847162 | 1.4585433 | 1.2102244 | 0.9023996 |\n",
       "| PAX8 | 1.1053528 | 1.0432043 | 1.0137772 | 1.0931698 | 1.0356631 | 0.9435998 | 1.1165266 | 1.0960624 | 1.0099041 | 1.0553276 | ⋯ | 1.0055602 | 0.8999849 | 0.9461904 | 1.2250262 | 1.0535747 | 0.9321033 | 1.0115234 | 1.0048561 | 1.1175232 | 1.0730404 |\n",
       "| GUCA1A | 1.2764652 | 0.8625501 | 1.0411838 | 1.0323309 | 0.8504494 | 1.3334491 | 1.1356378 | 1.0168769 | 0.9045680 | 0.8634836 | ⋯ | 1.3178803 | 0.8809971 | 0.9128050 | 1.0231747 | 1.0368457 | 0.9140167 | 1.0870208 | 0.9601511 | 3.4069852 | 1.0806457 |\n",
       "| UBA7 | 0.8337133 | 1.0081274 | 1.0330204 | 1.5409547 | 0.8751729 | 1.0063332 | 0.7226255 | 0.9519509 | 0.8132348 | 1.0214182 | ⋯ | 1.2466778 | 1.8324052 | 0.8732074 | 1.4028883 | 0.9107665 | 0.9586981 | 1.1221024 | 1.1861940 | 1.0770138 | 1.2797908 |\n",
       "\n"
      ],
      "text/plain": [
       "       GSM300173 GSM300177 GSM300181 GSM300186 GSM300189 GSM300192 GSM300196\n",
       "DDR1   0.6085351 0.8513945 1.1591995 2.0921063 1.2385797 0.8326798 0.8034588\n",
       "RFC2   0.9081397 0.9336113 0.8563885 1.0015364 0.7332213 0.8570584 1.0210657\n",
       "HSPA6  0.8674964 0.8385732 0.9956165 0.9797408 3.5688195 1.0145685 1.1729818\n",
       "PAX8   1.1053528 1.0432043 1.0137772 1.0931698 1.0356631 0.9435998 1.1165266\n",
       "GUCA1A 1.2764652 0.8625501 1.0411838 1.0323309 0.8504494 1.3334491 1.1356378\n",
       "UBA7   0.8337133 1.0081274 1.0330204 1.5409547 0.8751729 1.0063332 0.7226255\n",
       "       GSM300204 GSM300208 GSM300214 ⋯ GSM1176201 GSM1176202 GSM1176203\n",
       "DDR1   0.6996407 1.3955209 1.1198790 ⋯ 1.1563261  1.6152382  1.1901788 \n",
       "RFC2   0.9641521 0.8585522 0.8584299 ⋯ 0.8848516  0.9385835  1.0359085 \n",
       "HSPA6  1.0208826 0.8163447 0.8572931 ⋯ 0.9469172  1.3414407  0.9509885 \n",
       "PAX8   1.0960624 1.0099041 1.0553276 ⋯ 1.0055602  0.8999849  0.9461904 \n",
       "GUCA1A 1.0168769 0.9045680 0.8634836 ⋯ 1.3178803  0.8809971  0.9128050 \n",
       "UBA7   0.9519509 0.8132348 1.0214182 ⋯ 1.2466778  1.8324052  0.8732074 \n",
       "       GSM1176204 GSM1176205 GSM1176206 GSM1176207 GSM1176208 GSM1176209\n",
       "DDR1   1.4972300  0.6329609  1.4160284  1.2335104  1.2999401  0.8595833 \n",
       "RFC2   0.9225201  1.0405487  0.8187565  0.9541863  0.9855680  1.1704986 \n",
       "HSPA6  1.3277143  1.0712911  0.9801153  0.9847162  1.4585433  1.2102244 \n",
       "PAX8   1.2250262  1.0535747  0.9321033  1.0115234  1.0048561  1.1175232 \n",
       "GUCA1A 1.0231747  1.0368457  0.9140167  1.0870208  0.9601511  3.4069852 \n",
       "UBA7   1.4028883  0.9107665  0.9586981  1.1221024  1.1861940  1.0770138 \n",
       "       GSM1176210\n",
       "DDR1   1.2998189 \n",
       "RFC2   0.9924505 \n",
       "HSPA6  0.9023996 \n",
       "PAX8   1.0730404 \n",
       "GUCA1A 1.0806457 \n",
       "UBA7   1.2797908 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load the expression data, the normalized expression data and the groups\n",
    "data <- readRDS(\"./data/GSE48350.rds\")\n",
    "expression_data <- data$expression_data\n",
    "norm_expression_data <- data$norm_expression_data\n",
    "groups <- data$groups\n",
    "\n",
    "# Get the probe IDs\n",
    "norm_expression_data$PROBEID <- rownames(norm_expression_data)\n",
    "probeIDs <- rownames(norm_expression_data)\n",
    "# Perform gene mapping\n",
    "suppressMessages({\n",
    "    annotLookup <- AnnotationDbi::select(hgu133plus2.db, keys = probeIDs, columns = c('PROBEID', 'GENENAME', 'SYMBOL'))\n",
    "})\n",
    "# Merge DE result  data frame with annotation table\n",
    "new_expression_data = merge(annotLookup, norm_expression_data, by = \"PROBEID\")\n",
    "# Remove NA value\n",
    "new_expression_data <- new_expression_data[!is.na(new_expression_data$SYMBOL), ]\n",
    "# Remove duplicated genes symbol\n",
    "new_expression_data <- new_expression_data[!duplicated(new_expression_data$SYMBOL, fromLast = FALSE), ]\n",
    "rownames(new_expression_data) <- new_expression_data$SYMBOL\n",
    "# Drop PROBEID, GENENAME, and SYMBOL columns\n",
    "new_expression_data <- new_expression_data[, -c(1:3)]\n",
    "\n",
    "# Show some first rows of the new expression data\n",
    "head(new_expression_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fdfcffe-e916-4803-93ac-16b440778668",
   "metadata": {},
   "source": [
    "In order to perform PADOG enrichment analysis, we will use the `.padog` function bellow. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d28b96e6-84ac-460c-9ff2-8c04f8f45235",
   "metadata": {},
   "source": [
    "### Enrichment analysis using PADOG and GO terms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "fc66fcc6-1696-4572-826c-1462ba7baf40",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting with 6909 gene sets!\n",
      "Analyzing 6909 gene sets with 3 or more genes!\n",
      "# of permutations used: 100 \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>pathway</th><th scope=col>p.value</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>GO:0065003</th><td>GO:0065003</td><td>1e-04</td></tr>\n",
       "\t<tr><th scope=row>GO:0007005</th><td>GO:0007005</td><td>1e-04</td></tr>\n",
       "\t<tr><th scope=row>GO:0042773</th><td>GO:0042773</td><td>1e-04</td></tr>\n",
       "\t<tr><th scope=row>GO:0042775</th><td>GO:0042775</td><td>1e-04</td></tr>\n",
       "\t<tr><th scope=row>GO:0007007</th><td>GO:0007007</td><td>1e-04</td></tr>\n",
       "\t<tr><th scope=row>GO:0042407</th><td>GO:0042407</td><td>1e-04</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & pathway & p.value\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\tGO:0065003 & GO:0065003 & 1e-04\\\\\n",
       "\tGO:0007005 & GO:0007005 & 1e-04\\\\\n",
       "\tGO:0042773 & GO:0042773 & 1e-04\\\\\n",
       "\tGO:0042775 & GO:0042775 & 1e-04\\\\\n",
       "\tGO:0007007 & GO:0007007 & 1e-04\\\\\n",
       "\tGO:0042407 & GO:0042407 & 1e-04\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | pathway &lt;chr&gt; | p.value &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| GO:0065003 | GO:0065003 | 1e-04 |\n",
       "| GO:0007005 | GO:0007005 | 1e-04 |\n",
       "| GO:0042773 | GO:0042773 | 1e-04 |\n",
       "| GO:0042775 | GO:0042775 | 1e-04 |\n",
       "| GO:0007007 | GO:0007007 | 1e-04 |\n",
       "| GO:0042407 | GO:0042407 | 1e-04 |\n",
       "\n"
      ],
      "text/plain": [
       "           pathway    p.value\n",
       "GO:0065003 GO:0065003 1e-04  \n",
       "GO:0007005 GO:0007005 1e-04  \n",
       "GO:0042773 GO:0042773 1e-04  \n",
       "GO:0042775 GO:0042775 1e-04  \n",
       "GO:0007007 GO:0007007 1e-04  \n",
       "GO:0042407 GO:0042407 1e-04  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Assign the geneset variable to GO_term_hallmark loaded in previous code\n",
    "geneset <- GO_term_hallmark\n",
    "# Perform PADOG enrichment analysis using the .padog function\n",
    "res <- PADOG::padog(esetm = as.matrix(new_expression_data), group = as.character(groups$groups), \n",
    "                    gslist = geneset, gs.names = names(geneset), NI = 100, plots = FALSE, \n",
    "                    dseed = 1, parallel = TRUE, ncr = 1)\n",
    "res <- res[, c('ID', 'Ppadog')]\n",
    "colnames(res) <- c(\"pathway\", \"p.value\")\n",
    "head(res)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "776df6b0-a160-4fa5-89a6-d8034d670b6d",
   "metadata": {},
   "source": [
    "To use the `padog` function, we need to pass the following arguments:\n",
    "\n",
    "- `esetm`: A matrix containing log transfomed and normalized gene expression data. Rows correspond to genes and columns to samples.\n",
    "- `group`: A character vector with the class labels of the samples. It can only contain \"c\" for control samples or \"d\" for disease samples.\n",
    "- `gslist`: Either the value \"KEGGRESTpathway\" or a list with the gene sets.\n",
    "- `gs.names`: A character vector with the names of the gene sets. If specified, must have the same length as gslist.\r\n",
    "\n",
    "- `NI`: An integer specifying the number of iterations to determine the gene set score significance p-value\n",
    "- `plots`: A boolean, if set to TRUE then the distribution of the PADOG scores with and without weighting the genes in raw and standardized form are shown using boxplots. A pdf file will be created in the current directory having the name provided in the targetgs field. The scores for the targetgs gene set will be shown in red.\n",
    "- `dseed`: An integer specifying optional initial seed for random number generator.\r\n",
    "\n",
    "- `parallel`: A boolean, if set to TRUE, the NI iterations will be executed in parallel if multiple CPU cores are available and foreach and doRNG packages are installed.\r\n",
    "s.\n",
    "- `ncr`: An integer specifying the number of CPU cores used when parallel set to TRUE. Default is to use all CPU cores detected.\n",
    "\n",
    "The `padog` function returns a dataframe containing the analysis result."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60a039b9-b62e-4ab7-be15-656e36118110",
   "metadata": {},
   "source": [
    "### Enrichment analysis using PADOG and KEGG pathways\n",
    "\n",
    "We can use the same procedure to per enrichment analysis with the KEGG pathway. All the codes are similar but `genesets` are loaded from `KEGG_hallmark`. The code is shown below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "7ff3b556-d4d5-4384-888c-f69f98b79537",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting with 344 gene sets!\n",
      "Analyzing 344 gene sets with 3 or more genes!\n",
      "# of permutations used: 100 \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>pathway</th><th scope=col>p.value</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>hsa05016</th><td>hsa05016</td><td>1e-04</td></tr>\n",
       "\t<tr><th scope=row>hsa05022</th><td>hsa05022</td><td>1e-04</td></tr>\n",
       "\t<tr><th scope=row>hsa05012</th><td>hsa05012</td><td>1e-04</td></tr>\n",
       "\t<tr><th scope=row>hsa00190</th><td>hsa00190</td><td>1e-04</td></tr>\n",
       "\t<tr><th scope=row>hsa05020</th><td>hsa05020</td><td>1e-04</td></tr>\n",
       "\t<tr><th scope=row>hsa05010</th><td>hsa05010</td><td>1e-04</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & pathway & p.value\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\thsa05016 & hsa05016 & 1e-04\\\\\n",
       "\thsa05022 & hsa05022 & 1e-04\\\\\n",
       "\thsa05012 & hsa05012 & 1e-04\\\\\n",
       "\thsa00190 & hsa00190 & 1e-04\\\\\n",
       "\thsa05020 & hsa05020 & 1e-04\\\\\n",
       "\thsa05010 & hsa05010 & 1e-04\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | pathway &lt;chr&gt; | p.value &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| hsa05016 | hsa05016 | 1e-04 |\n",
       "| hsa05022 | hsa05022 | 1e-04 |\n",
       "| hsa05012 | hsa05012 | 1e-04 |\n",
       "| hsa00190 | hsa00190 | 1e-04 |\n",
       "| hsa05020 | hsa05020 | 1e-04 |\n",
       "| hsa05010 | hsa05010 | 1e-04 |\n",
       "\n"
      ],
      "text/plain": [
       "         pathway  p.value\n",
       "hsa05016 hsa05016 1e-04  \n",
       "hsa05022 hsa05022 1e-04  \n",
       "hsa05012 hsa05012 1e-04  \n",
       "hsa00190 hsa00190 1e-04  \n",
       "hsa05020 hsa05020 1e-04  \n",
       "hsa05010 hsa05010 1e-04  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Assign the geneset variable to KEGG_hallmark loaded in previous code\n",
    "geneset <- KEGG_hallmark\n",
    "# Perform PADOG enrichment analysis using the .padog function\n",
    "res <- PADOG::padog(esetm = as.matrix(new_expression_data), group = as.character(groups$groups), \n",
    "                    gslist = geneset, gs.names = names(geneset), NI = 100, plots = FALSE, \n",
    "                    dseed = 1, parallel = TRUE, ncr = 1)\n",
    "res <- res[, c('ID', 'Ppadog')]\n",
    "colnames(res) <- c(\"pathway\", \"p.value\")\n",
    "head(res)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c29ad340-33ee-4a39-a8ae-cd7aa05926bc",
   "metadata": {},
   "source": [
    "### Enrichment analysis using PADOG and REACTOME pathways\n",
    "\n",
    "We can use the same procedure to per enrichment analysis with the KEGG pathway. All the codes are similar but `genesets` are loaded from `KEGG_hallmark`, we also need to replace SYMBOL with ENTREZID. The code is shown below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "9fcf6167-be75-4150-b3f0-4cb6b82a182f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 54</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>GSM300173</th><th scope=col>GSM300177</th><th scope=col>GSM300181</th><th scope=col>GSM300186</th><th scope=col>GSM300189</th><th scope=col>GSM300192</th><th scope=col>GSM300196</th><th scope=col>GSM300204</th><th scope=col>GSM300208</th><th scope=col>GSM300214</th><th scope=col>⋯</th><th scope=col>GSM1176201</th><th scope=col>GSM1176202</th><th scope=col>GSM1176203</th><th scope=col>GSM1176204</th><th scope=col>GSM1176205</th><th scope=col>GSM1176206</th><th scope=col>GSM1176207</th><th scope=col>GSM1176208</th><th scope=col>GSM1176209</th><th scope=col>GSM1176210</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>780</th><td>0.6085351</td><td>0.8513945</td><td>1.1591995</td><td>2.0921063</td><td>1.2385797</td><td>0.8326798</td><td>0.8034588</td><td>0.6996407</td><td>1.3955209</td><td>1.1198790</td><td>⋯</td><td>1.1563261</td><td>1.6152382</td><td>1.1901788</td><td>1.4972300</td><td>0.6329609</td><td>1.4160284</td><td>1.2335104</td><td>1.2999401</td><td>0.8595833</td><td>1.2998189</td></tr>\n",
       "\t<tr><th scope=row>5982</th><td>0.9081397</td><td>0.9336113</td><td>0.8563885</td><td>1.0015364</td><td>0.7332213</td><td>0.8570584</td><td>1.0210657</td><td>0.9641521</td><td>0.8585522</td><td>0.8584299</td><td>⋯</td><td>0.8848516</td><td>0.9385835</td><td>1.0359085</td><td>0.9225201</td><td>1.0405487</td><td>0.8187565</td><td>0.9541863</td><td>0.9855680</td><td>1.1704986</td><td>0.9924505</td></tr>\n",
       "\t<tr><th scope=row>3310</th><td>0.8674964</td><td>0.8385732</td><td>0.9956165</td><td>0.9797408</td><td>3.5688195</td><td>1.0145685</td><td>1.1729818</td><td>1.0208826</td><td>0.8163447</td><td>0.8572931</td><td>⋯</td><td>0.9469172</td><td>1.3414407</td><td>0.9509885</td><td>1.3277143</td><td>1.0712911</td><td>0.9801153</td><td>0.9847162</td><td>1.4585433</td><td>1.2102244</td><td>0.9023996</td></tr>\n",
       "\t<tr><th scope=row>7849</th><td>1.1053528</td><td>1.0432043</td><td>1.0137772</td><td>1.0931698</td><td>1.0356631</td><td>0.9435998</td><td>1.1165266</td><td>1.0960624</td><td>1.0099041</td><td>1.0553276</td><td>⋯</td><td>1.0055602</td><td>0.8999849</td><td>0.9461904</td><td>1.2250262</td><td>1.0535747</td><td>0.9321033</td><td>1.0115234</td><td>1.0048561</td><td>1.1175232</td><td>1.0730404</td></tr>\n",
       "\t<tr><th scope=row>2978</th><td>1.2764652</td><td>0.8625501</td><td>1.0411838</td><td>1.0323309</td><td>0.8504494</td><td>1.3334491</td><td>1.1356378</td><td>1.0168769</td><td>0.9045680</td><td>0.8634836</td><td>⋯</td><td>1.3178803</td><td>0.8809971</td><td>0.9128050</td><td>1.0231747</td><td>1.0368457</td><td>0.9140167</td><td>1.0870208</td><td>0.9601511</td><td>3.4069852</td><td>1.0806457</td></tr>\n",
       "\t<tr><th scope=row>7318</th><td>0.8337133</td><td>1.0081274</td><td>1.0330204</td><td>1.5409547</td><td>0.8751729</td><td>1.0063332</td><td>0.7226255</td><td>0.9519509</td><td>0.8132348</td><td>1.0214182</td><td>⋯</td><td>1.2466778</td><td>1.8324052</td><td>0.8732074</td><td>1.4028883</td><td>0.9107665</td><td>0.9586981</td><td>1.1221024</td><td>1.1861940</td><td>1.0770138</td><td>1.2797908</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 54\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & GSM300173 & GSM300177 & GSM300181 & GSM300186 & GSM300189 & GSM300192 & GSM300196 & GSM300204 & GSM300208 & GSM300214 & ⋯ & GSM1176201 & GSM1176202 & GSM1176203 & GSM1176204 & GSM1176205 & GSM1176206 & GSM1176207 & GSM1176208 & GSM1176209 & GSM1176210\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t780 & 0.6085351 & 0.8513945 & 1.1591995 & 2.0921063 & 1.2385797 & 0.8326798 & 0.8034588 & 0.6996407 & 1.3955209 & 1.1198790 & ⋯ & 1.1563261 & 1.6152382 & 1.1901788 & 1.4972300 & 0.6329609 & 1.4160284 & 1.2335104 & 1.2999401 & 0.8595833 & 1.2998189\\\\\n",
       "\t5982 & 0.9081397 & 0.9336113 & 0.8563885 & 1.0015364 & 0.7332213 & 0.8570584 & 1.0210657 & 0.9641521 & 0.8585522 & 0.8584299 & ⋯ & 0.8848516 & 0.9385835 & 1.0359085 & 0.9225201 & 1.0405487 & 0.8187565 & 0.9541863 & 0.9855680 & 1.1704986 & 0.9924505\\\\\n",
       "\t3310 & 0.8674964 & 0.8385732 & 0.9956165 & 0.9797408 & 3.5688195 & 1.0145685 & 1.1729818 & 1.0208826 & 0.8163447 & 0.8572931 & ⋯ & 0.9469172 & 1.3414407 & 0.9509885 & 1.3277143 & 1.0712911 & 0.9801153 & 0.9847162 & 1.4585433 & 1.2102244 & 0.9023996\\\\\n",
       "\t7849 & 1.1053528 & 1.0432043 & 1.0137772 & 1.0931698 & 1.0356631 & 0.9435998 & 1.1165266 & 1.0960624 & 1.0099041 & 1.0553276 & ⋯ & 1.0055602 & 0.8999849 & 0.9461904 & 1.2250262 & 1.0535747 & 0.9321033 & 1.0115234 & 1.0048561 & 1.1175232 & 1.0730404\\\\\n",
       "\t2978 & 1.2764652 & 0.8625501 & 1.0411838 & 1.0323309 & 0.8504494 & 1.3334491 & 1.1356378 & 1.0168769 & 0.9045680 & 0.8634836 & ⋯ & 1.3178803 & 0.8809971 & 0.9128050 & 1.0231747 & 1.0368457 & 0.9140167 & 1.0870208 & 0.9601511 & 3.4069852 & 1.0806457\\\\\n",
       "\t7318 & 0.8337133 & 1.0081274 & 1.0330204 & 1.5409547 & 0.8751729 & 1.0063332 & 0.7226255 & 0.9519509 & 0.8132348 & 1.0214182 & ⋯ & 1.2466778 & 1.8324052 & 0.8732074 & 1.4028883 & 0.9107665 & 0.9586981 & 1.1221024 & 1.1861940 & 1.0770138 & 1.2797908\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 54\n",
       "\n",
       "| <!--/--> | GSM300173 &lt;dbl&gt; | GSM300177 &lt;dbl&gt; | GSM300181 &lt;dbl&gt; | GSM300186 &lt;dbl&gt; | GSM300189 &lt;dbl&gt; | GSM300192 &lt;dbl&gt; | GSM300196 &lt;dbl&gt; | GSM300204 &lt;dbl&gt; | GSM300208 &lt;dbl&gt; | GSM300214 &lt;dbl&gt; | ⋯ ⋯ | GSM1176201 &lt;dbl&gt; | GSM1176202 &lt;dbl&gt; | GSM1176203 &lt;dbl&gt; | GSM1176204 &lt;dbl&gt; | GSM1176205 &lt;dbl&gt; | GSM1176206 &lt;dbl&gt; | GSM1176207 &lt;dbl&gt; | GSM1176208 &lt;dbl&gt; | GSM1176209 &lt;dbl&gt; | GSM1176210 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 780 | 0.6085351 | 0.8513945 | 1.1591995 | 2.0921063 | 1.2385797 | 0.8326798 | 0.8034588 | 0.6996407 | 1.3955209 | 1.1198790 | ⋯ | 1.1563261 | 1.6152382 | 1.1901788 | 1.4972300 | 0.6329609 | 1.4160284 | 1.2335104 | 1.2999401 | 0.8595833 | 1.2998189 |\n",
       "| 5982 | 0.9081397 | 0.9336113 | 0.8563885 | 1.0015364 | 0.7332213 | 0.8570584 | 1.0210657 | 0.9641521 | 0.8585522 | 0.8584299 | ⋯ | 0.8848516 | 0.9385835 | 1.0359085 | 0.9225201 | 1.0405487 | 0.8187565 | 0.9541863 | 0.9855680 | 1.1704986 | 0.9924505 |\n",
       "| 3310 | 0.8674964 | 0.8385732 | 0.9956165 | 0.9797408 | 3.5688195 | 1.0145685 | 1.1729818 | 1.0208826 | 0.8163447 | 0.8572931 | ⋯ | 0.9469172 | 1.3414407 | 0.9509885 | 1.3277143 | 1.0712911 | 0.9801153 | 0.9847162 | 1.4585433 | 1.2102244 | 0.9023996 |\n",
       "| 7849 | 1.1053528 | 1.0432043 | 1.0137772 | 1.0931698 | 1.0356631 | 0.9435998 | 1.1165266 | 1.0960624 | 1.0099041 | 1.0553276 | ⋯ | 1.0055602 | 0.8999849 | 0.9461904 | 1.2250262 | 1.0535747 | 0.9321033 | 1.0115234 | 1.0048561 | 1.1175232 | 1.0730404 |\n",
       "| 2978 | 1.2764652 | 0.8625501 | 1.0411838 | 1.0323309 | 0.8504494 | 1.3334491 | 1.1356378 | 1.0168769 | 0.9045680 | 0.8634836 | ⋯ | 1.3178803 | 0.8809971 | 0.9128050 | 1.0231747 | 1.0368457 | 0.9140167 | 1.0870208 | 0.9601511 | 3.4069852 | 1.0806457 |\n",
       "| 7318 | 0.8337133 | 1.0081274 | 1.0330204 | 1.5409547 | 0.8751729 | 1.0063332 | 0.7226255 | 0.9519509 | 0.8132348 | 1.0214182 | ⋯ | 1.2466778 | 1.8324052 | 0.8732074 | 1.4028883 | 0.9107665 | 0.9586981 | 1.1221024 | 1.1861940 | 1.0770138 | 1.2797908 |\n",
       "\n"
      ],
      "text/plain": [
       "     GSM300173 GSM300177 GSM300181 GSM300186 GSM300189 GSM300192 GSM300196\n",
       "780  0.6085351 0.8513945 1.1591995 2.0921063 1.2385797 0.8326798 0.8034588\n",
       "5982 0.9081397 0.9336113 0.8563885 1.0015364 0.7332213 0.8570584 1.0210657\n",
       "3310 0.8674964 0.8385732 0.9956165 0.9797408 3.5688195 1.0145685 1.1729818\n",
       "7849 1.1053528 1.0432043 1.0137772 1.0931698 1.0356631 0.9435998 1.1165266\n",
       "2978 1.2764652 0.8625501 1.0411838 1.0323309 0.8504494 1.3334491 1.1356378\n",
       "7318 0.8337133 1.0081274 1.0330204 1.5409547 0.8751729 1.0063332 0.7226255\n",
       "     GSM300204 GSM300208 GSM300214 ⋯ GSM1176201 GSM1176202 GSM1176203\n",
       "780  0.6996407 1.3955209 1.1198790 ⋯ 1.1563261  1.6152382  1.1901788 \n",
       "5982 0.9641521 0.8585522 0.8584299 ⋯ 0.8848516  0.9385835  1.0359085 \n",
       "3310 1.0208826 0.8163447 0.8572931 ⋯ 0.9469172  1.3414407  0.9509885 \n",
       "7849 1.0960624 1.0099041 1.0553276 ⋯ 1.0055602  0.8999849  0.9461904 \n",
       "2978 1.0168769 0.9045680 0.8634836 ⋯ 1.3178803  0.8809971  0.9128050 \n",
       "7318 0.9519509 0.8132348 1.0214182 ⋯ 1.2466778  1.8324052  0.8732074 \n",
       "     GSM1176204 GSM1176205 GSM1176206 GSM1176207 GSM1176208 GSM1176209\n",
       "780  1.4972300  0.6329609  1.4160284  1.2335104  1.2999401  0.8595833 \n",
       "5982 0.9225201  1.0405487  0.8187565  0.9541863  0.9855680  1.1704986 \n",
       "3310 1.3277143  1.0712911  0.9801153  0.9847162  1.4585433  1.2102244 \n",
       "7849 1.2250262  1.0535747  0.9321033  1.0115234  1.0048561  1.1175232 \n",
       "2978 1.0231747  1.0368457  0.9140167  1.0870208  0.9601511  3.4069852 \n",
       "7318 1.4028883  0.9107665  0.9586981  1.1221024  1.1861940  1.0770138 \n",
       "     GSM1176210\n",
       "780  1.2998189 \n",
       "5982 0.9924505 \n",
       "3310 0.9023996 \n",
       "7849 1.0730404 \n",
       "2978 1.0806457 \n",
       "7318 1.2797908 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load the expression data, the normalized expression data and the groups\n",
    "data <- readRDS(\"./data/GSE48350.rds\")\n",
    "expression_data <- data$expression_data\n",
    "norm_expression_data <- data$norm_expression_data\n",
    "groups <- data$groups\n",
    "\n",
    "# Get the probe IDs\n",
    "norm_expression_data$PROBEID <- rownames(norm_expression_data)\n",
    "probeIDs <- rownames(norm_expression_data)\n",
    "# Perform gene mapping\n",
    "suppressMessages({\n",
    "    annotLookup <- AnnotationDbi::select(hgu133plus2.db, keys = probeIDs, columns = c('PROBEID', 'GENENAME', 'ENTREZID'))\n",
    "})\n",
    "# Merge DE result  data frame with annotation table\n",
    "new_expression_data = merge(annotLookup, norm_expression_data, by = \"PROBEID\")\n",
    "# Remove NA value\n",
    "new_expression_data <- new_expression_data[!is.na(new_expression_data$ENTREZID), ]\n",
    "# Remove duplicated genes ENTREZID\n",
    "new_expression_data <- new_expression_data[!duplicated(new_expression_data$ENTREZID, fromLast = FALSE), ]\n",
    "rownames(new_expression_data) <- new_expression_data$ENTREZID\n",
    "# Drop PROBEID, GENENAME, and ENTREZID columns\n",
    "new_expression_data <- new_expression_data[, -c(1:3)]\n",
    "\n",
    "# show some first rows of the new expression data\n",
    "head(new_expression_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "c2b0002b-3c8e-4b7f-91f8-d2198e785100",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting with 2647 gene sets!\n",
      "Analyzing 2198 gene sets with 3 or more genes!\n",
      "# of permutations used: 100 \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>pathway</th><th scope=col>p.value</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>R-HSA-1428517</th><td>R-HSA-1428517</td><td>1e-04</td></tr>\n",
       "\t<tr><th scope=row>R-HSA-444411</th><td>R-HSA-444411 </td><td>1e-04</td></tr>\n",
       "\t<tr><th scope=row>R-HSA-210747</th><td>R-HSA-210747 </td><td>1e-04</td></tr>\n",
       "\t<tr><th scope=row>R-HSA-1852241</th><td>R-HSA-1852241</td><td>1e-04</td></tr>\n",
       "\t<tr><th scope=row>R-HSA-210744</th><td>R-HSA-210744 </td><td>1e-04</td></tr>\n",
       "\t<tr><th scope=row>R-HSA-8949613</th><td>R-HSA-8949613</td><td>1e-04</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & pathway & p.value\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\tR-HSA-1428517 & R-HSA-1428517 & 1e-04\\\\\n",
       "\tR-HSA-444411 & R-HSA-444411  & 1e-04\\\\\n",
       "\tR-HSA-210747 & R-HSA-210747  & 1e-04\\\\\n",
       "\tR-HSA-1852241 & R-HSA-1852241 & 1e-04\\\\\n",
       "\tR-HSA-210744 & R-HSA-210744  & 1e-04\\\\\n",
       "\tR-HSA-8949613 & R-HSA-8949613 & 1e-04\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | pathway &lt;chr&gt; | p.value &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| R-HSA-1428517 | R-HSA-1428517 | 1e-04 |\n",
       "| R-HSA-444411 | R-HSA-444411  | 1e-04 |\n",
       "| R-HSA-210747 | R-HSA-210747  | 1e-04 |\n",
       "| R-HSA-1852241 | R-HSA-1852241 | 1e-04 |\n",
       "| R-HSA-210744 | R-HSA-210744  | 1e-04 |\n",
       "| R-HSA-8949613 | R-HSA-8949613 | 1e-04 |\n",
       "\n"
      ],
      "text/plain": [
       "              pathway       p.value\n",
       "R-HSA-1428517 R-HSA-1428517 1e-04  \n",
       "R-HSA-444411  R-HSA-444411  1e-04  \n",
       "R-HSA-210747  R-HSA-210747  1e-04  \n",
       "R-HSA-1852241 R-HSA-1852241 1e-04  \n",
       "R-HSA-210744  R-HSA-210744  1e-04  \n",
       "R-HSA-8949613 R-HSA-8949613 1e-04  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Assign the geneset variable to KEGG_hallmark loaded in previous code\n",
    "geneset <- REACTOME_hallmark\n",
    "# Perform PADOG enrichment analysis using the .padog function\n",
    "res <- PADOG::padog(esetm = as.matrix(new_expression_data), group = as.character(groups$groups), gslist = geneset,\n",
    "                        gs.names = names(geneset), NI = 100, plots = FALSE, dseed = 1,\n",
    "                         parallel = TRUE, ncr = 1)\n",
    "res <- res[, c('ID', 'Ppadog')]\n",
    "colnames(res) <- c(\"pathway\", \"p.value\")\n",
    "head(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbf4d680-3478-4c7a-af18-3746faefcc99",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "49fd3929-7d2e-401d-93c6-24572bec35f6",
   "metadata": {},
   "source": [
    "The next submodule will go indepth about meta analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "e870e8eb-ee47-439b-8154-0a720927bd79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.2 (2023-10-31 ucrt)\n",
       "Platform: x86_64-w64-mingw32/x64 (64-bit)\n",
       "Running under: Windows 10 x64 (build 19045)\n",
       "\n",
       "Matrix products: default\n",
       "\n",
       "\n",
       "locale:\n",
       "[1] LC_COLLATE=English_United States.utf8 \n",
       "[2] LC_CTYPE=English_United States.utf8   \n",
       "[3] LC_MONETARY=English_United States.utf8\n",
       "[4] LC_NUMERIC=C                          \n",
       "[5] LC_TIME=English_United States.utf8    \n",
       "\n",
       "time zone: America/Chicago\n",
       "tzcode source: internal\n",
       "\n",
       "attached base packages:\n",
       "[1] stats4    stats     graphics  grDevices utils     datasets  methods  \n",
       "[8] base     \n",
       "\n",
       "other attached packages:\n",
       " [1] PADOG_1.44.0             KEGGdzPathwaysGEO_1.40.0 GSA_1.03.2              \n",
       " [4] fgsea_1.28.0             BiocManager_1.30.22      hgu133plus2.db_3.13.0   \n",
       " [7] org.Hs.eg.db_3.18.0      AnnotationDbi_1.64.1     IRanges_2.36.0          \n",
       "[10] S4Vectors_0.40.2         Biobase_2.62.0           BiocGenerics_0.48.1     \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] tidyselect_1.2.0        IRdisplay_1.1           dplyr_1.1.4            \n",
       " [4] farver_2.1.1            blob_1.2.4              Biostrings_2.70.1      \n",
       " [7] bitops_1.0-7            fastmap_1.1.1           RCurl_1.98-1.13        \n",
       "[10] digest_0.6.33           lifecycle_1.0.4         statmod_1.5.0          \n",
       "[13] KEGGREST_1.42.0         RSQLite_2.3.4           magrittr_2.0.3         \n",
       "[16] compiler_4.3.2          rngtools_1.5.2          rlang_1.1.2            \n",
       "[19] tools_4.3.2             utf8_1.2.4              data.table_1.14.10     \n",
       "[22] labeling_0.4.3          doRNG_1.8.6             bit_4.0.5              \n",
       "[25] repr_1.1.6              BiocParallel_1.36.0     pbdZMQ_0.3-10          \n",
       "[28] withr_2.5.2             hgu133a.db_3.13.0       grid_4.3.2             \n",
       "[31] fansi_1.0.6             colorspace_2.1-0        ggplot2_3.4.4          \n",
       "[34] iterators_1.0.14        scales_1.3.0            cli_3.6.2              \n",
       "[37] crayon_1.5.2            generics_0.1.3          httr_1.4.7             \n",
       "[40] DBI_1.2.0               cachem_1.0.8            zlibbioc_1.48.0        \n",
       "[43] parallel_4.3.2          XVector_0.42.0          base64enc_0.1-3        \n",
       "[46] vctrs_0.6.5             Matrix_1.6-1.1          jsonlite_1.8.8         \n",
       "[49] bit64_4.0.5             foreach_1.5.2           limma_3.58.1           \n",
       "[52] snow_0.4-4              glue_1.6.2              codetools_0.2-19       \n",
       "[55] cowplot_1.1.2           gtable_0.3.4            GenomeInfoDb_1.38.2    \n",
       "[58] munsell_0.5.0           tibble_3.2.1            pillar_1.9.0           \n",
       "[61] htmltools_0.5.7         IRkernel_1.3.2          GenomeInfoDbData_1.2.11\n",
       "[64] R6_2.5.1                doParallel_1.0.17       evaluate_0.23          \n",
       "[67] lattice_0.21-9          png_0.1-8               memoise_2.0.1          \n",
       "[70] Rcpp_1.0.11             uuid_1.1-1              fastmatch_1.1-4        \n",
       "[73] nlme_3.1-163            pkgconfig_2.0.3        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca1ba4ed-3e1c-4d5a-8554-40703527a4d5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "environment": {
   "kernel": "ir",
   "name": "r-cpu.4-2.m107",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/r-cpu.4-2:m107"
  },
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  },
  "toc-showcode": true,
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
